{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "621bbcd9",
   "metadata": {},
   "source": [
    "be really clear about what's going into repaired tables: \n",
    "- repaired\n",
    "- already OK\n",
    "- unrepairable\n",
    "\n",
    "Only ~ 215k repaired `env_package`s ?!\n",
    "- yeah, there just aren't that many annotated in the biosample metadata\n",
    "\n",
    "\n",
    "- some of this code should probably be moved back into the package\n",
    "    - splitting and repiping \n",
    "\n",
    "- read in previous SSSOMs and don't re-map? or at least double check?\n",
    "\n",
    "- don't forget to consult \"model\" field in biosample table too\n",
    "    - for scoping, in addition to en_package and taxonomy_id?\n",
    "\n",
    "- what to do about hits against obsolete classes... does OLS provide a mechanism to avoid?\n",
    "- be careful with GAZ matches\n",
    "- avoid IDless, OMIT, NCIT, MICRO, BTO\n",
    "\n",
    "- is it worth trimming and concatenating \"x1\", \"x2\", \"x3\", where x is a mappable term? numbers may be low.\n",
    "    - example = XXX\n",
    "\n",
    "- reconcile \"x biome\" and \"x\" mappings\n",
    "- take note of non-local salvage hits\n",
    "    - like XXX\n",
    "    - small/large (freshwater) lake?\n",
    "    - include native labels in the review frame, in addition to local?\n",
    "    - should use rdftab, not additional OLS calls\n",
    "    \n",
    "\n",
    "- switch from hard-coded `env_package_overrides` to something like SSSOM?\n",
    "- mappings can come from merging strategies, first pass OLS or salvage OLS\n",
    "- reflect those methods in SSSOM\n",
    "- add string distance/confidence\n",
    "    - multiple columns for multiple methods?\n",
    "\n",
    "review these SSSOM columns\n",
    "\n",
    "```\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "```\n",
    "\n",
    "- are there samples with a host_taxid and a (sample) taxonomy_id from \"unclassified\"?\n",
    "- start at uncultured or metagenome taxonomy_id?\n",
    "\n",
    "- what scoping on or mapping of  \"gold\" or \"path\"? WON'T FIND THAT IN NCBI BIOSAMPLE\n",
    "\n",
    "\n",
    "What to save where? \n",
    "- new columns in biosample sqlite? \n",
    "- delimited file dumps\n",
    "- move content from original column and insert the new values into that original column?\n",
    "- pandas clipboard integration handy for review step\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fac2962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "\n",
    "import scoped_mapping\n",
    "\n",
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3300da64",
   "metadata": {},
   "source": [
    "# User-provided data\n",
    "See repo README for notes on setting up SQLite databases of OBO ontologies with semantic-sql, relation-graph and rdftab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5684af0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://www.ncbi.nlm.nih.gov/biosample/docs/packages/?format=xml\n",
    "# see also https://www.ncbi.nlm.nih.gov/biosample/docs/packages/\n",
    "biosample_packages_file = \"../target/biosample_packages.xml\"\n",
    "\n",
    "# from ftp://ftp.ncbi.nlm.nih.gov//biosample/biosample_set.xml.gz\n",
    "# via harmonized_table.db.gz\n",
    "# in https://drive.google.com/drive/u/0/folders/1eL0v0stoduahjDpoDJIk3z2pJBAU4b2Y\n",
    "biosample_sqlite_file = \"../target/harmonized_table.db\"\n",
    "ncbitaxon_sqlite_file = \"../semantic-sql/db/ncbitaxon.db\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c4494ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)\n",
    "ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80370e30",
   "metadata": {},
   "source": [
    "# Utilizing taxonomy for broad subsetting\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46e3ee59",
   "metadata": {},
   "source": [
    "**This uses an SQLite database in which the transitive closure over subClassOf has already been materialized. See the README and Makefile.**\n",
    "\n",
    "Specifically, flag the Biosamples whose `taxon_id` indicates they are an unclassified entity. Ignoring the others will throw out samples of multicellular organisms, like fruit flies.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bfd8416",
   "metadata": {},
   "source": [
    "## Get a listing of all taxa that are transitive subclasses of `NCBITaxon:2787823`\n",
    "\n",
    "I.e. 'unclassified entities'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c40903d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:10.132506\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:1006967</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:1041057</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:1046002</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:1046003</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:1046004</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>NCBITaxon:939928</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>NCBITaxon:941420</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>NCBITaxon:941421</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>NCBITaxon:941422</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>NCBITaxon:942017</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>994 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               subject  unclassified\n",
       "0    NCBITaxon:1006967          True\n",
       "1    NCBITaxon:1041057          True\n",
       "2    NCBITaxon:1046002          True\n",
       "3    NCBITaxon:1046003          True\n",
       "4    NCBITaxon:1046004          True\n",
       "..                 ...           ...\n",
       "989   NCBITaxon:939928          True\n",
       "990   NCBITaxon:941420          True\n",
       "991   NCBITaxon:941421          True\n",
       "992   NCBITaxon:941422          True\n",
       "993   NCBITaxon:942017          True\n",
       "\n",
       "[994 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    distinct s.subject\n",
    "from\n",
    "    entailed_edge ee\n",
    "join statements s on\n",
    "    ee.subject = s.subject\n",
    "where\n",
    "    ee.predicate = 'rdfs:subClassOf'\n",
    "    and ee.object = 'NCBITaxon:2787823'\n",
    "    and s.predicate = 'rdfs:label'\n",
    "\"\"\"\n",
    "[unclassified_taxa, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "unclassified_taxa[\"unclassified\"] = True\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "unclassified_taxa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cad3ffdb",
   "metadata": {},
   "source": [
    "## Get taxon counts from the Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a5d8f028",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:01.338698\n"
     ]
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    taxonomy_id biosample_taxid,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    taxonomy_id\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[biosample_sample_taxon_summary, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx\n",
    ")\n",
    "biosample_sample_taxon_summary[\"curie\"] = \"NCBITaxon:\" + biosample_sample_taxon_summary[\n",
    "    \"biosample_taxid\"\n",
    "].astype(str)\n",
    "\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cf3a275",
   "metadata": {},
   "source": [
    "## Merge the two taxonomy dataframes\n",
    "\n",
    "I.e. flag the the Biosample records whose `taxonomy_id` field belongs to a subclass of 'unclassified entries'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7ad29a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>curie</th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       biosample_taxid    count             curie           subject  \\\n",
       "0                 9606  6819707    NCBITaxon:9606               NaN   \n",
       "1                10090   964219   NCBITaxon:10090               NaN   \n",
       "2               408170   290862  NCBITaxon:408170  NCBITaxon:408170   \n",
       "3               410658   280666  NCBITaxon:410658  NCBITaxon:410658   \n",
       "4               646099   208741  NCBITaxon:646099  NCBITaxon:646099   \n",
       "...                ...      ...               ...               ...   \n",
       "163372          999891        1  NCBITaxon:999891               NaN   \n",
       "163373          999892        1  NCBITaxon:999892               NaN   \n",
       "163374          999898        1  NCBITaxon:999898               NaN   \n",
       "163375          999931        1  NCBITaxon:999931               NaN   \n",
       "163376          999947        1  NCBITaxon:999947               NaN   \n",
       "\n",
       "        unclassified  \n",
       "0              False  \n",
       "1              False  \n",
       "2               True  \n",
       "3               True  \n",
       "4               True  \n",
       "...              ...  \n",
       "163372         False  \n",
       "163373         False  \n",
       "163374         False  \n",
       "163375         False  \n",
       "163376         False  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biosample_sample_taxon_summary = biosample_sample_taxon_summary.merge(\n",
    "    unclassified_taxa, left_on=\"curie\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "biosample_sample_taxon_summary.unclassified.fillna(False, inplace=True)\n",
    "\n",
    "biosample_sample_taxon_summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f7aff98",
   "metadata": {},
   "source": [
    "## Add labels to all taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a6acc349",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:06.125497\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>unclassified</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>False</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>False</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>True</td>\n",
       "      <td>human gut metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>True</td>\n",
       "      <td>soil metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>True</td>\n",
       "      <td>human metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Bacillus amyloliquefaciens TA208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>[Propionibacterium] humerusii P08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Peptococcaceae bacterium CEB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Barrientosiimonas humi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Angomonas ambiguus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   curie biosample_taxid    count  unclassified  \\\n",
       "0         NCBITaxon:9606            9606  6819707         False   \n",
       "1        NCBITaxon:10090           10090   964219         False   \n",
       "2       NCBITaxon:408170          408170   290862          True   \n",
       "3       NCBITaxon:410658          410658   280666          True   \n",
       "4       NCBITaxon:646099          646099   208741          True   \n",
       "...                  ...             ...      ...           ...   \n",
       "163372  NCBITaxon:999891          999891        1         False   \n",
       "163373  NCBITaxon:999892          999892        1         False   \n",
       "163374  NCBITaxon:999898          999898        1         False   \n",
       "163375  NCBITaxon:999931          999931        1         False   \n",
       "163376  NCBITaxon:999947          999947        1         False   \n",
       "\n",
       "                                    label  \n",
       "0                            Homo sapiens  \n",
       "1                            Mus musculus  \n",
       "2                    human gut metagenome  \n",
       "3                         soil metagenome  \n",
       "4                        human metagenome  \n",
       "...                                   ...  \n",
       "163372   Bacillus amyloliquefaciens TA208  \n",
       "163373  [Propionibacterium] humerusii P08  \n",
       "163374      Peptococcaceae bacterium CEB3  \n",
       "163375             Barrientosiimonas humi  \n",
       "163376                 Angomonas ambiguus  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    subject ,\n",
    "    value\n",
    "from statements\n",
    "where\n",
    "    predicate = 'rdfs:label' and subject = stanza\n",
    "\"\"\"\n",
    "[all_tax_labels, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "\n",
    "biosample_sample_taxon_summary = biosample_sample_taxon_summary.merge(\n",
    "    all_tax_labels, left_on=\"curie\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "\n",
    "biosample_sample_taxon_summary = biosample_sample_taxon_summary[\n",
    "    [\"curie\", \"biosample_taxid\", \"count\", \"unclassified\", \"value\"]\n",
    "]\n",
    "biosample_sample_taxon_summary.columns = [\n",
    "    \"curie\",\n",
    "    \"biosample_taxid\",\n",
    "    \"count\",\n",
    "    \"unclassified\",\n",
    "    \"label\",\n",
    "]\n",
    "\n",
    "print(query_duration)\n",
    "biosample_sample_taxon_summary.to_sql(\n",
    "    \"biosample_sample_taxon_summary\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "\n",
    "all_tax_labels.to_sql(\"all_tax_labels\", biosample_cnx, if_exists=\"replace\", index=False)\n",
    "\n",
    "biosample_sample_taxon_summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "554a82c0",
   "metadata": {},
   "source": [
    "**Almost all of the taxa that are common in the biosample collection are either unclassified/metagenomes or easily recognized cellular organisms. Cellular organism samples are de-prioritized in this exercise**\n",
    "\n",
    "Exceptions include:\n",
    "- 32630 = synthetic construct (other entries; other sequences; artificial sequences)\n",
    "    - 'other entries' would add 16k rows on top of the 1k 'unclassified entities'\n",
    "    - metagenomes account for 331 of the 'unclassified entities'\n",
    "    - there are also a small number of uncultured/unclassified microorganisms in the biosample dataset\n",
    "- 77133 = uncultured bacterium (cellular organisms; Bacteria; environmental samples)\n",
    "    - 'cellular organisms' would add 2M rows on top of the 1k 'unclassified entities'\n",
    "    - 'cellular organisms; Bacteria; environmental samples' adds 26k\n",
    "    \n",
    "----"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
