{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fac2962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy\n",
    "import scoped_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5684af0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_sqlite_file = \"/Users/MAM/Documents/gitrepos/biosample-analysis/target/harmonized_table.db\"\n",
    "# TODO process these as a list?\n",
    "ncbitaxon_sqlite_file = \"/Users/MAM/Documents/gitrepos/semantic-sql/db/ncbitaxon.db\"\n",
    "envo_sqlite_file = \"/Users/MAM/Documents/gitrepos/semantic-sql/db/envo.db\"\n",
    "ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)\n",
    "envo_cnx = sqlite3.connect(envo_sqlite_file)\n",
    "target_onto_prefix = 'ENVO'\n",
    "chars_to_whiteout = '._-'\n",
    "my_query_fields = ''\n",
    "my_row_req = 3\n",
    "\n",
    "env_package_overrides = {\n",
    "    'built environment': 'built',\n",
    "    'misc environment': 'miscellaneous',\n",
    "    'missing': 'no environmental package',\n",
    "    'unknown': 'no environmental package',\n",
    "    'default': 'no environmental package',\n",
    "    'unspecified': 'no environmental package',\n",
    "    'not available': 'no environmental package',\n",
    "    'not collected': 'no environmental package'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9282126f",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "caff574e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       id env_package      package  \\\n",
      "0  BIOSAMPLE:SAMN00000002     missing  MIGS.ba.5.0   \n",
      "1  BIOSAMPLE:SAMN00000003     missing  MIGS.ba.5.0   \n",
      "2  BIOSAMPLE:SAMN00000004     missing  MIGS.ba.5.0   \n",
      "3  BIOSAMPLE:SAMN00000005        None  Generic.1.0   \n",
      "4  BIOSAMPLE:SAMN00000006        None  Generic.1.0   \n",
      "5  BIOSAMPLE:SAMN00000007     missing  MIGS.ba.5.0   \n",
      "6  BIOSAMPLE:SAMN00000008     missing  MIGS.ba.5.0   \n",
      "7  BIOSAMPLE:SAMN00000009        None  Generic.1.0   \n",
      "8  BIOSAMPLE:SAMN00000010        None  Generic.1.0   \n",
      "9  BIOSAMPLE:SAMN00000011        None  Generic.1.0   \n",
      "\n",
      "                                   package_name host_taxid taxonomy_id  \\\n",
      "0  MIGS: cultured bacteria/archaea; version 5.0       9606      445970   \n",
      "1  MIGS: cultured bacteria/archaea; version 5.0       9606      445972   \n",
      "2  MIGS: cultured bacteria/archaea; version 5.0       9606      449673   \n",
      "3                                       Generic       None        6526   \n",
      "4                                       Generic       None        9483   \n",
      "5  MIGS: cultured bacteria/archaea; version 5.0       9606      445974   \n",
      "6  MIGS: cultured bacteria/archaea; version 5.0       9606      411461   \n",
      "7                                       Generic       None       13616   \n",
      "8                                       Generic       None      451639   \n",
      "9                                       Generic       None      451638   \n",
      "\n",
      "                     env_broad_scale  \\\n",
      "0  terrestrial biome [ENVO:00000446]   \n",
      "1  terrestrial biome [ENVO:00000446]   \n",
      "2  terrestrial biome [ENVO:00000446]   \n",
      "3                               None   \n",
      "4                               None   \n",
      "5  terrestrial biome [ENVO:00000446]   \n",
      "6  terrestrial biome [ENVO:00000446]   \n",
      "7                               None   \n",
      "8                               None   \n",
      "9                               None   \n",
      "\n",
      "                            env_local_scale  \\\n",
      "0  human-associated habitat [ENVO:00009003]   \n",
      "1  human-associated habitat [ENVO:00009003]   \n",
      "2  human-associated habitat [ENVO:00009003]   \n",
      "3                                      None   \n",
      "4                                      None   \n",
      "5  human-associated habitat [ENVO:00009003]   \n",
      "6  human-associated habitat [ENVO:00009003]   \n",
      "7                                      None   \n",
      "8                                      None   \n",
      "9                                      None   \n",
      "\n",
      "                           env_medium  \n",
      "0  biological product [ENVO:02000043]  \n",
      "1  biological product [ENVO:02000043]  \n",
      "2  biological product [ENVO:02000043]  \n",
      "3                                None  \n",
      "4                                None  \n",
      "5  biological product [ENVO:02000043]  \n",
      "6  biological product [ENVO:02000043]  \n",
      "7                                None  \n",
      "8                                None  \n",
      "9                                None  \n"
     ]
    }
   ],
   "source": [
    "# Sample of the data we're working with\n",
    "q = \"\"\"\n",
    "select\n",
    "    id,\n",
    "    env_package,\n",
    "    package,\n",
    "    package_name,\n",
    "    host_taxid,\n",
    "    taxonomy_id,\n",
    "    env_broad_scale,\n",
    "    env_local_scale,\n",
    "    env_medium\n",
    "    from biosample b\n",
    "limit 10\n",
    "\"\"\"\n",
    "biosample_first_ten = pd.read_sql(q, biosample_cnx)\n",
    "print(biosample_first_ten)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c6cc9596",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                            Name  \\\n",
      "0                    Generic.1.0   \n",
      "1              SARS-CoV-2.cl.1.0   \n",
      "2                Pathogen.cl.1.0   \n",
      "3               Pathogen.env.1.0   \n",
      "4                    Microbe.1.0   \n",
      "..                           ...   \n",
      "149  MIUVIG.plant-associated.5.0   \n",
      "150          MIUVIG.sediment.5.0   \n",
      "151              MIUVIG.soil.5.0   \n",
      "152        MIUVIG.wastewater.5.0   \n",
      "153             MIUVIG.water.5.0   \n",
      "\n",
      "                                           DisplayName  \\\n",
      "0                                              Generic   \n",
      "1    SARS-CoV-2: clinical or host-associated; versi...   \n",
      "2    Pathogen: clinical or host-associated; version...   \n",
      "3      Pathogen: environmental/food/other; version 1.0   \n",
      "4                                 Microbe; version 1.0   \n",
      "..                                                 ...   \n",
      "149  MIUVIG: uncultivated virus genome, plant-assoc...   \n",
      "150  MIUVIG: uncultivated virus genome, sediment; v...   \n",
      "151  MIUVIG: uncultivated virus genome, soil; versi...   \n",
      "152  MIUVIG: uncultivated virus genome, wastewater;...   \n",
      "153  MIUVIG: uncultivated virus genome, water; vers...   \n",
      "\n",
      "                                   ShortName        EnvPackage  \\\n",
      "0                                                                \n",
      "1    SARS-CoV-2: clinical or host-associated                     \n",
      "2      Pathogen: clinical or host-associated                     \n",
      "3         Pathogen: environmental/food/other                     \n",
      "4                                    Microbe                     \n",
      "..                                       ...               ...   \n",
      "149         MIUVIG Uncultivated Virus Genome  plant-associated   \n",
      "150         MIUVIG Uncultivated Virus Genome          sediment   \n",
      "151         MIUVIG Uncultivated Virus Genome              soil   \n",
      "152         MIUVIG Uncultivated Virus Genome        wastewater   \n",
      "153         MIUVIG Uncultivated Virus Genome             water   \n",
      "\n",
      "    EnvPackageDisplay                 NotAppropriateFor  \\\n",
      "0                                                         \n",
      "1                      wgs_single;wgs_batch;wgs_diploid   \n",
      "2                                                         \n",
      "3                                                         \n",
      "4                                                         \n",
      "..                ...                               ...   \n",
      "149  plant-associated  wgs_single;wgs_batch;wgs_diploid   \n",
      "150          sediment  wgs_single;wgs_batch;wgs_diploid   \n",
      "151              soil  wgs_single;wgs_batch;wgs_diploid   \n",
      "152        wastewater  wgs_single;wgs_batch;wgs_diploid   \n",
      "153             water  wgs_single;wgs_batch;wgs_diploid   \n",
      "\n",
      "                                           Description       Example  \n",
      "0                                              Generic                \n",
      "1    Use for SARS-CoV-2 samples that are relevant t...                \n",
      "2                 Clinical or host-associated pathogen  SAMN02928182  \n",
      "3                Environmental, food or other pathogen  SAMN02730065  \n",
      "4    Use for bacteria or other unicellular microbes...  SAMN02911891  \n",
      "..                                                 ...           ...  \n",
      "149  Use for uncultivated virus genome identified i...                \n",
      "150  Use for uncultivated virus genome identified i...                \n",
      "151  Use for uncultivated virus genome identified i...                \n",
      "152  Use for uncultivated virus genome identified i...                \n",
      "153  Use for uncultivated virus genome identified i...                \n",
      "\n",
      "[154 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "# Get the canonical checklist and package terms from NCBI\n",
    "# Unfortunately it doesn't do a very good job of differentiating \n",
    "# checklists (MIMAG, MIMARKS, etc.) \n",
    "# from packages (soil, water, etc.)\n",
    "# what about ba , euk, etc?\n",
    "package_dictionary = scoped_mapping.get_package_dictionary()\n",
    "print(package_dictionary)\n",
    "package_dictionary.to_sql('package_dictionary', biosample_cnx, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0cfa7d0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   biosample_row_count\n",
      "0             14300584\n",
      "0:00:00.049769\n"
     ]
    }
   ],
   "source": [
    "# Do the Biosample checklist/package fields match any of the cannonical values?\n",
    "# How many Biosample rows are there?\n",
    "q = \"\"\"\n",
    "select count(*) as biosample_row_count\n",
    "from biosample b\n",
    "\"\"\"\n",
    "[biosample_row_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx, print_timing=False)\n",
    "print(biosample_row_count)\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "03a8b90c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-05-24 14:19:24.990680\n",
      "2021-05-24 14:19:26.123880\n",
      "0:00:01.133200\n",
      "   cannonical_package_name_count\n",
      "0                       14300584\n",
      "0:00:01.133200\n"
     ]
    }
   ],
   "source": [
    "# How many of those rows can be inner-joined with the canonical checklists/packages?\n",
    "# Specifically, joining biosample.package_name = package_dictionary.DisplayName\n",
    "# TODO add indexing to docs and or makefile\n",
    "# create index biosample_package_name_idx on biosample(package_name);\n",
    "# create index package_dictionary_DisplayName_idx on package_dictionary(DisplayName);\n",
    "# create index biosample_package_idx on biosample(package);\n",
    "# create index biosample_p_pn_idx on biosample(package, package_name);\n",
    "q = \"\"\"\n",
    "select\n",
    "    count(*) as cannonical_package_name_count\n",
    "from\n",
    "    biosample b\n",
    "inner join package_dictionary pd on\n",
    "    b.package_name = pd.DisplayName\n",
    "\"\"\"\n",
    "[cannonical_package_name_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx, print_timing=True)\n",
    "print(cannonical_package_name_count)\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "83485b99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-05-24 14:19:26.580389\n",
      "2021-05-24 14:19:28.497415\n",
      "0:00:01.917026\n",
      "                       package  \\\n",
      "0           Beta-lactamase.1.0   \n",
      "1                  Generic.1.0   \n",
      "2                    Human.1.0   \n",
      "3             Invertebrate.1.0   \n",
      "4                  MIGS.ba.5.0   \n",
      "..                         ...   \n",
      "124  Model.organism.animal.1.0   \n",
      "125            Pathogen.cl.1.0   \n",
      "126           Pathogen.env.1.0   \n",
      "127                  Plant.1.0   \n",
      "128                  Virus.1.0   \n",
      "\n",
      "                                          package_name     count  \n",
      "0                          Beta-lactamase; version 1.0       556  \n",
      "1                                              Generic  10186430  \n",
      "2                                   Human; version 1.0    368893  \n",
      "3                            Invertebrate; version 1.0    131085  \n",
      "4         MIGS: cultured bacteria/archaea; version 5.0     28686  \n",
      "..                                                 ...       ...  \n",
      "124              Model organism or animal; version 1.0    382980  \n",
      "125  Pathogen: clinical or host-associated; version...    511040  \n",
      "126    Pathogen: environmental/food/other; version 1.0    246497  \n",
      "127                                 Plant; version 1.0    385939  \n",
      "128                                 Virus; version 1.0     15560  \n",
      "\n",
      "[129 rows x 3 columns]\n",
      "0:00:01.917026\n"
     ]
    }
   ],
   "source": [
    "# What do the combinations of package and package_name look like in the Biosample dataset?\n",
    "q = \"\"\"\n",
    "select\n",
    "    package,\n",
    "    package_name,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    package ,\n",
    "    package_name\n",
    "order by\n",
    "    package ,\n",
    "    package_name\n",
    "\"\"\"\n",
    "[package_name_combos, query_duration] = scoped_mapping.timed_query(q, biosample_cnx, print_timing=True)\n",
    "print(package_name_combos)\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d042df82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        env_package     count\n",
      "0                              None  14083847\n",
      "1                   host-associated     49254\n",
      "2                         human-gut     47921\n",
      "3                             water     16367\n",
      "4                        human-skin     13706\n",
      "..                              ...       ...\n",
      "87                      env_package         1\n",
      "88                   gut microbiome         1\n",
      "89  marine sediment (ENVO:00002113)         1\n",
      "90                           saliva         1\n",
      "91       sea water, [ENVO:00002149]         1\n",
      "\n",
      "[92 rows x 2 columns]\n",
      "0:00:01.018992\n"
     ]
    }
   ],
   "source": [
    "# What about the Biosample env_package values?\n",
    "# Are they also a small, highly regular set?\n",
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    env_package\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[env_package_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "print(env_package_count)\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "82ba2408",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['', 'air', 'built', 'host associated', 'human associated', 'human gut', 'human oral', 'human skin', 'human vaginal', 'microbial', 'microbial mat biofilm', 'miscellaneous', 'miscellaneous or artificial', 'no environmental package', 'plant associated', 'sediment', 'soil', 'wastewater', 'wastewater sludge', 'water']\n"
     ]
    }
   ],
   "source": [
    "# env_package is going to need some cleanup\n",
    "# First, get a set of all canonical env_package values\n",
    "package_dictionary = scoped_mapping.make_tidy_col(package_dictionary, 'EnvPackage', 'eptidy')\n",
    "package_dictionary =scoped_mapping.make_tidy_col(package_dictionary, 'EnvPackageDisplay', 'epdtidy')\n",
    "# update in sqlite\n",
    "package_dictionary.to_sql('package_dictionary', biosample_cnx, if_exists='replace', index=False)\n",
    "valid_combo = []\n",
    "valid_combo = scoped_mapping.add_unique_to_list(valid_combo, package_dictionary['eptidy'])\n",
    "valid_combo = scoped_mapping.add_unique_to_list(valid_combo, package_dictionary['epdtidy'])\n",
    "print(valid_combo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5d157f3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.026371\n",
      "              stanza  prefix\n",
      "0        BFO:0000001     BFO\n",
      "1        BFO:0000002     BFO\n",
      "2        BFO:0000003     BFO\n",
      "3        BFO:0000004     BFO\n",
      "4        BFO:0000006     BFO\n",
      "...              ...     ...\n",
      "6561  UBERON:3000972  UBERON\n",
      "6562  UBERON:3000977  UBERON\n",
      "6563  UBERON:3010200  UBERON\n",
      "6564  UBERON:4100000  UBERON\n",
      "6565  UBERON:8410024  UBERON\n",
      "\n",
      "[6566 rows x 2 columns]\n",
      "{'BFO': 'BFO:\\\\d{7}', 'CARO': 'CARO:\\\\d{7}', 'CHEBI': 'CHEBI:\\\\d{4,6}', 'ENVO': 'ENVO:\\\\d{7,8}', 'FAO': 'FAO:0000001', 'FOODON': 'FOODON:\\\\d{8}', 'GO': 'GO:\\\\d{7}', 'IAO': 'IAO:\\\\d{7}', 'NCBITaxon': 'NCBITaxon:\\\\d+', 'OBI': 'OBI:\\\\d{7}', 'PATO': 'PATO:\\\\d{7}', 'PCO': 'PCO:\\\\d{7}', 'PO': 'PO:\\\\d{7}', 'RO': 'RO:0002577', 'UBERON': 'UBERON:\\\\d{7}'}\n"
     ]
    }
   ],
   "source": [
    "# determine ID patterns\n",
    "q = \"\"\"\n",
    "select\n",
    "    distinct stanza\n",
    "    from statements s\n",
    "where\n",
    "    predicate = 'rdf:type'\n",
    "    and \"object\" = 'owl:Class'\n",
    "    and stanza = subject\"\"\"\n",
    "# include non-envo IDs that come from envo?\n",
    "[ids_from_envo, query_duration] = scoped_mapping.timed_query(q, envo_cnx)\n",
    "print(query_duration)\n",
    "ids_from_envo = scoped_mapping.add_prefix_col(ids_from_envo, 'stanza', 'prefix')\n",
    "print(ids_from_envo)\n",
    "\n",
    "id_patterns = scoped_mapping.get_multi_term_patterns(ids_from_envo, 'stanza', 'prefix')\n",
    "print(id_patterns)\n",
    "\n",
    "def env_package_nomralizastion(dataframe, col_to_normalize, pattern_name, id_replacement_rule):\n",
    "    dataframe[['lhs', 'rhs']] = dataframe[col_to_normalize].str.split('.', expand=True)\n",
    "    flag = dataframe['rhs'].apply(lambda x: x is None)\n",
    "    temp = dataframe['lhs'][flag]\n",
    "    dataframe.loc[flag, 'rhs'] = temp\n",
    "    dataframe.loc[flag, 'lhs'] = ''\n",
    "    series_decomposition = scoped_mapping.decompose_series(dataframe['rhs'], id_replacement_rule)\n",
    "    dataframe = pd.concat([dataframe, series_decomposition], axis=1)\n",
    "    return dataframe\n",
    "\n",
    "env_package_normalized = env_package_nomralizastion(env_package_count, 'env_package',\n",
    "                                                                target_onto_prefix, id_patterns['ENVO'])\n",
    "\n",
    "env_package_normalized = scoped_mapping.add_overrides(env_package_normalized, 'remaining_tidied', 'ep_override',\n",
    "                                                   env_package_overrides)\n",
    "\n",
    "env_package_normalized = scoped_mapping.flag_canonical(env_package_normalized, 'ep_override', 'is_canonical',\n",
    "                                                    valid_combo)\n",
    "\n",
    "env_package_normalized.to_sql('env_package_normalized', biosample_cnx, if_exists='replace', index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c1fbb75",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
