{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fac2962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy\n",
    "import scoped_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3300da64",
   "metadata": {},
   "source": [
    "## User-provided data\n",
    "See repo README for notes on setting up SQLite databases of OBO ontologies with semantic-sql, relation-graph and rdftab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5684af0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_packages_file = '../target/biosample_packages.xml'\n",
    "\n",
    "biosample_sqlite_file   = \"../target/harmonized_table.db\"\n",
    "ncbitaxon_sqlite_file   = \"../semantic-sql/db/ncbitaxon.db\"\n",
    "envo_sqlite_file        = \"../semantic-sql/db/envo.db\"\n",
    "\n",
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)\n",
    "ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)\n",
    "envo_cnx      = sqlite3.connect(envo_sqlite_file)\n",
    "\n",
    "target_onto_prefix = 'ENVO'\n",
    "chars_to_whiteout  = '._-'\n",
    "my_query_fields    = '' # OLS weighted default\n",
    "my_row_req         = 3\n",
    "my_string_dist_arg = 2\n",
    "my_max_string_dist = 0.1\n",
    "\n",
    "env_package_overrides = {\n",
    "    'built environment': 'built',\n",
    "    'misc environment': 'miscellaneous',\n",
    "    'missing': 'no environmental package',\n",
    "    'unknown': 'no environmental package',\n",
    "    'default': 'no environmental package',\n",
    "    'unspecified': 'no environmental package',\n",
    "    'not available': 'no environmental package',\n",
    "    'not collected': 'no environmental package'\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c54707",
   "metadata": {},
   "source": [
    "## Print a sample of the data we're working with\n",
    "Specifically, INSDC/NCBI Biosample Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "caff574e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_package</th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMN00000002</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445970</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BIOSAMPLE:SAMN00000003</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445972</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIOSAMPLE:SAMN00000004</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>449673</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BIOSAMPLE:SAMN00000005</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>6526</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BIOSAMPLE:SAMN00000006</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>9483</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BIOSAMPLE:SAMN00000007</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445974</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BIOSAMPLE:SAMN00000008</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>411461</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BIOSAMPLE:SAMN00000009</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>13616</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BIOSAMPLE:SAMN00000010</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>451639</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BIOSAMPLE:SAMN00000011</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>451638</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id env_package      package  \\\n",
       "0  BIOSAMPLE:SAMN00000002     missing  MIGS.ba.5.0   \n",
       "1  BIOSAMPLE:SAMN00000003     missing  MIGS.ba.5.0   \n",
       "2  BIOSAMPLE:SAMN00000004     missing  MIGS.ba.5.0   \n",
       "3  BIOSAMPLE:SAMN00000005        None  Generic.1.0   \n",
       "4  BIOSAMPLE:SAMN00000006        None  Generic.1.0   \n",
       "5  BIOSAMPLE:SAMN00000007     missing  MIGS.ba.5.0   \n",
       "6  BIOSAMPLE:SAMN00000008     missing  MIGS.ba.5.0   \n",
       "7  BIOSAMPLE:SAMN00000009        None  Generic.1.0   \n",
       "8  BIOSAMPLE:SAMN00000010        None  Generic.1.0   \n",
       "9  BIOSAMPLE:SAMN00000011        None  Generic.1.0   \n",
       "\n",
       "                                   package_name host_taxid taxonomy_id  \\\n",
       "0  MIGS: cultured bacteria/archaea; version 5.0       9606      445970   \n",
       "1  MIGS: cultured bacteria/archaea; version 5.0       9606      445972   \n",
       "2  MIGS: cultured bacteria/archaea; version 5.0       9606      449673   \n",
       "3                                       Generic       None        6526   \n",
       "4                                       Generic       None        9483   \n",
       "5  MIGS: cultured bacteria/archaea; version 5.0       9606      445974   \n",
       "6  MIGS: cultured bacteria/archaea; version 5.0       9606      411461   \n",
       "7                                       Generic       None       13616   \n",
       "8                                       Generic       None      451639   \n",
       "9                                       Generic       None      451638   \n",
       "\n",
       "                     env_broad_scale  \\\n",
       "0  terrestrial biome [ENVO:00000446]   \n",
       "1  terrestrial biome [ENVO:00000446]   \n",
       "2  terrestrial biome [ENVO:00000446]   \n",
       "3                               None   \n",
       "4                               None   \n",
       "5  terrestrial biome [ENVO:00000446]   \n",
       "6  terrestrial biome [ENVO:00000446]   \n",
       "7                               None   \n",
       "8                               None   \n",
       "9                               None   \n",
       "\n",
       "                            env_local_scale  \\\n",
       "0  human-associated habitat [ENVO:00009003]   \n",
       "1  human-associated habitat [ENVO:00009003]   \n",
       "2  human-associated habitat [ENVO:00009003]   \n",
       "3                                      None   \n",
       "4                                      None   \n",
       "5  human-associated habitat [ENVO:00009003]   \n",
       "6  human-associated habitat [ENVO:00009003]   \n",
       "7                                      None   \n",
       "8                                      None   \n",
       "9                                      None   \n",
       "\n",
       "                           env_medium  \n",
       "0  biological product [ENVO:02000043]  \n",
       "1  biological product [ENVO:02000043]  \n",
       "2  biological product [ENVO:02000043]  \n",
       "3                                None  \n",
       "4                                None  \n",
       "5  biological product [ENVO:02000043]  \n",
       "6  biological product [ENVO:02000043]  \n",
       "7                                None  \n",
       "8                                None  \n",
       "9                                None  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    id,\n",
    "    env_package,\n",
    "    package,\n",
    "    package_name,\n",
    "    host_taxid,\n",
    "    taxonomy_id,\n",
    "    env_broad_scale,\n",
    "    env_local_scale,\n",
    "    env_medium\n",
    "    from biosample b\n",
    "limit 10\n",
    "\"\"\"\n",
    "biosample_first_ten = pd.read_sql(q, biosample_cnx)\n",
    "biosample_first_ten"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef577997",
   "metadata": {},
   "source": [
    "## Get the canonical checklist and package terms from NCBI\n",
    "\n",
    "Unfortunately it doesn't do a very good job of differentiating checklists (MIMAG, MIMARKS, etc.) from packages (soil, water, etc.)\n",
    "\n",
    "_What about .ba, .euk, etc?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c6cc9596",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>DisplayName</th>\n",
       "      <th>ShortName</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>EnvPackageDisplay</th>\n",
       "      <th>NotAppropriateFor</th>\n",
       "      <th>Description</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2.cl.1.0</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated; versi...</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for SARS-CoV-2 samples that are relevant t...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>Pathogen: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Clinical or host-associated pathogen</td>\n",
       "      <td>SAMN02928182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>Pathogen: environmental/food/other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Environmental, food or other pathogen</td>\n",
       "      <td>SAMN02730065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Microbe.1.0</td>\n",
       "      <td>Microbe; version 1.0</td>\n",
       "      <td>Microbe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Use for bacteria or other unicellular microbes...</td>\n",
       "      <td>SAMN02911891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>MIUVIG.plant-associated.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, plant-assoc...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>MIUVIG.sediment.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, sediment; v...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>MIUVIG.soil.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, soil; versi...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>MIUVIG.wastewater.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, wastewater;...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>MIUVIG.water.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, water; vers...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Name  \\\n",
       "0                    Generic.1.0   \n",
       "1              SARS-CoV-2.cl.1.0   \n",
       "2                Pathogen.cl.1.0   \n",
       "3               Pathogen.env.1.0   \n",
       "4                    Microbe.1.0   \n",
       "..                           ...   \n",
       "149  MIUVIG.plant-associated.5.0   \n",
       "150          MIUVIG.sediment.5.0   \n",
       "151              MIUVIG.soil.5.0   \n",
       "152        MIUVIG.wastewater.5.0   \n",
       "153             MIUVIG.water.5.0   \n",
       "\n",
       "                                           DisplayName  \\\n",
       "0                                              Generic   \n",
       "1    SARS-CoV-2: clinical or host-associated; versi...   \n",
       "2    Pathogen: clinical or host-associated; version...   \n",
       "3      Pathogen: environmental/food/other; version 1.0   \n",
       "4                                 Microbe; version 1.0   \n",
       "..                                                 ...   \n",
       "149  MIUVIG: uncultivated virus genome, plant-assoc...   \n",
       "150  MIUVIG: uncultivated virus genome, sediment; v...   \n",
       "151  MIUVIG: uncultivated virus genome, soil; versi...   \n",
       "152  MIUVIG: uncultivated virus genome, wastewater;...   \n",
       "153  MIUVIG: uncultivated virus genome, water; vers...   \n",
       "\n",
       "                                   ShortName        EnvPackage  \\\n",
       "0                                                                \n",
       "1    SARS-CoV-2: clinical or host-associated                     \n",
       "2      Pathogen: clinical or host-associated                     \n",
       "3         Pathogen: environmental/food/other                     \n",
       "4                                    Microbe                     \n",
       "..                                       ...               ...   \n",
       "149         MIUVIG Uncultivated Virus Genome  plant-associated   \n",
       "150         MIUVIG Uncultivated Virus Genome          sediment   \n",
       "151         MIUVIG Uncultivated Virus Genome              soil   \n",
       "152         MIUVIG Uncultivated Virus Genome        wastewater   \n",
       "153         MIUVIG Uncultivated Virus Genome             water   \n",
       "\n",
       "    EnvPackageDisplay                 NotAppropriateFor  \\\n",
       "0                                                         \n",
       "1                      wgs_single;wgs_batch;wgs_diploid   \n",
       "2                                                         \n",
       "3                                                         \n",
       "4                                                         \n",
       "..                ...                               ...   \n",
       "149  plant-associated  wgs_single;wgs_batch;wgs_diploid   \n",
       "150          sediment  wgs_single;wgs_batch;wgs_diploid   \n",
       "151              soil  wgs_single;wgs_batch;wgs_diploid   \n",
       "152        wastewater  wgs_single;wgs_batch;wgs_diploid   \n",
       "153             water  wgs_single;wgs_batch;wgs_diploid   \n",
       "\n",
       "                                           Description       Example  \n",
       "0                                              Generic                \n",
       "1    Use for SARS-CoV-2 samples that are relevant t...                \n",
       "2                 Clinical or host-associated pathogen  SAMN02928182  \n",
       "3                Environmental, food or other pathogen  SAMN02730065  \n",
       "4    Use for bacteria or other unicellular microbes...  SAMN02911891  \n",
       "..                                                 ...           ...  \n",
       "149  Use for uncultivated virus genome identified i...                \n",
       "150  Use for uncultivated virus genome identified i...                \n",
       "151  Use for uncultivated virus genome identified i...                \n",
       "152  Use for uncultivated virus genome identified i...                \n",
       "153  Use for uncultivated virus genome identified i...                \n",
       "\n",
       "[154 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "package_dictionary = scoped_mapping.get_package_dictionary(biosample_packages_file)\n",
    "package_dictionary.to_sql('package_dictionary', biosample_cnx, if_exists='replace', index=False)\n",
    "package_dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa900ad8",
   "metadata": {},
   "source": [
    "## Do the Biosample checklist/package fields match any of the cannonical values?\n",
    "\n",
    "Start by counting the Biosample rows/records. See XXX notes on extracting this *harmonized* database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0cfa7d0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.073628\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_row_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14300584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   biosample_row_count\n",
       "0             14300584"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select count(*) as biosample_row_count\n",
    "from biosample b\n",
    "\"\"\"\n",
    "[biosample_row_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx, print_timing=False)\n",
    "\n",
    "print(query_duration)\n",
    "biosample_row_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc896a9c",
   "metadata": {},
   "source": [
    "## How many of those rows can be inner-joined with the canonical checklists/packages?\n",
    "Specifically, joining `biosample.package_name` with `package_dictionary.DisplayName`\n",
    "\n",
    "_Note that indices are built as part of the makefile_\n",
    "\n",
    "\n",
    "- create index biosample_package_name_idx on biosample(package_name);\n",
    "- create index package_dictionary_DisplayName_idx on package_dictionary(DisplayName);\n",
    "- create index biosample_package_idx on biosample(package);\n",
    "- create index biosample_p_pn_idx on biosample(package, package_name);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "03a8b90c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-05-26 12:50:36.366903\n",
      "2021-05-26 12:50:38.109956\n",
      "0:00:01.743053\n",
      "0:00:01.743053\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cannonical_package_name_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14300584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cannonical_package_name_count\n",
       "0                       14300584"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    count(*) as cannonical_package_name_count\n",
    "from\n",
    "    biosample b\n",
    "inner join package_dictionary pd on\n",
    "    b.package_name = pd.DisplayName\n",
    "\"\"\"\n",
    "[cannonical_package_name_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx, print_timing=True)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "cannonical_package_name_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5951b7f1",
   "metadata": {},
   "source": [
    "## Combinations of `package` and `package_name` values in the Biosample dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "83485b99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-05-26 12:50:38.117583\n",
      "2021-05-26 12:50:40.011075\n",
      "0:00:01.893492\n",
      "0:00:01.893492\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Beta-lactamase.1.0</td>\n",
       "      <td>Beta-lactamase; version 1.0</td>\n",
       "      <td>556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>10186430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Human.1.0</td>\n",
       "      <td>Human; version 1.0</td>\n",
       "      <td>368893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Invertebrate.1.0</td>\n",
       "      <td>Invertebrate; version 1.0</td>\n",
       "      <td>131085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>28686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>Model.organism.animal.1.0</td>\n",
       "      <td>Model organism or animal; version 1.0</td>\n",
       "      <td>382980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>511040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>246497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>Plant.1.0</td>\n",
       "      <td>Plant; version 1.0</td>\n",
       "      <td>385939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Virus.1.0</td>\n",
       "      <td>Virus; version 1.0</td>\n",
       "      <td>15560</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       package  \\\n",
       "0           Beta-lactamase.1.0   \n",
       "1                  Generic.1.0   \n",
       "2                    Human.1.0   \n",
       "3             Invertebrate.1.0   \n",
       "4                  MIGS.ba.5.0   \n",
       "..                         ...   \n",
       "124  Model.organism.animal.1.0   \n",
       "125            Pathogen.cl.1.0   \n",
       "126           Pathogen.env.1.0   \n",
       "127                  Plant.1.0   \n",
       "128                  Virus.1.0   \n",
       "\n",
       "                                          package_name     count  \n",
       "0                          Beta-lactamase; version 1.0       556  \n",
       "1                                              Generic  10186430  \n",
       "2                                   Human; version 1.0    368893  \n",
       "3                            Invertebrate; version 1.0    131085  \n",
       "4         MIGS: cultured bacteria/archaea; version 5.0     28686  \n",
       "..                                                 ...       ...  \n",
       "124              Model organism or animal; version 1.0    382980  \n",
       "125  Pathogen: clinical or host-associated; version...    511040  \n",
       "126    Pathogen: environmental/food/other; version 1.0    246497  \n",
       "127                                 Plant; version 1.0    385939  \n",
       "128                                 Virus; version 1.0     15560  \n",
       "\n",
       "[129 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    package,\n",
    "    package_name,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    package ,\n",
    "    package_name\n",
    "order by\n",
    "    package ,\n",
    "    package_name\n",
    "\"\"\"\n",
    "[package_name_combos, query_duration] = scoped_mapping.timed_query(q, biosample_cnx, print_timing=True)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "package_name_combos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e15812f",
   "metadata": {},
   "source": [
    "## What about the Biosample `env_package` values?\n",
    "Are they also a small, highly regular set, like the `package` and `package_name` combinations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d042df82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.906749\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count\n",
       "0                              None  14083847\n",
       "1                   host-associated     49254\n",
       "2                         human-gut     47921\n",
       "3                             water     16367\n",
       "4                        human-skin     13706\n",
       "..                              ...       ...\n",
       "87                      env_package         1\n",
       "88                   gut microbiome         1\n",
       "89  marine sediment (ENVO:00002113)         1\n",
       "90                           saliva         1\n",
       "91       sea water, [ENVO:00002149]         1\n",
       "\n",
       "[92 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    env_package\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[env_package_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "env_package_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ab5ec5d",
   "metadata": {},
   "source": [
    "## `env_package` is a mixture of `ENVO` term ids and strings\n",
    "Those strings may or may not be the term's labels. There are many redundancies due to small spelling and punctiation variations.\n",
    "\n",
    "Start by getting a set of all canonical `env_package` values recognized by INSDC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "82ba2408",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " 'air',\n",
       " 'built',\n",
       " 'host associated',\n",
       " 'human associated',\n",
       " 'human gut',\n",
       " 'human oral',\n",
       " 'human skin',\n",
       " 'human vaginal',\n",
       " 'microbial',\n",
       " 'microbial mat biofilm',\n",
       " 'miscellaneous',\n",
       " 'miscellaneous or artificial',\n",
       " 'no environmental package',\n",
       " 'plant associated',\n",
       " 'sediment',\n",
       " 'soil',\n",
       " 'wastewater',\n",
       " 'wastewater sludge',\n",
       " 'water']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "package_dictionary = scoped_mapping.make_tidy_col(package_dictionary, 'EnvPackage', 'eptidy')\n",
    "package_dictionary =scoped_mapping.make_tidy_col(package_dictionary, 'EnvPackageDisplay', 'epdtidy')\n",
    "# update in sqlite\n",
    "package_dictionary.to_sql('package_dictionary', biosample_cnx, if_exists='replace', index=False)\n",
    "valid_combo = []\n",
    "valid_combo = scoped_mapping.add_unique_to_list(valid_combo, package_dictionary['eptidy'])\n",
    "valid_combo = scoped_mapping.add_unique_to_list(valid_combo, package_dictionary['epdtidy'])\n",
    "\n",
    "valid_combo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70be9eb2",
   "metadata": {},
   "source": [
    "## Determine ID patterns for common ontologies, like `ENVO`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5d157f3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.029921\n"
     ]
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    distinct stanza\n",
    "    from statements s\n",
    "where\n",
    "    predicate = 'rdf:type'\n",
    "    and \"object\" = 'owl:Class'\n",
    "    and stanza = subject\"\"\"\n",
    "# include non-envo IDs that come from envo?\n",
    "[ids_from_envo, query_duration] = scoped_mapping.timed_query(q, envo_cnx)\n",
    "print(query_duration)\n",
    "ids_from_envo = scoped_mapping.add_prefix_col(ids_from_envo, 'stanza', 'prefix')\n",
    "\n",
    "id_patterns = scoped_mapping.get_multi_term_patterns(ids_from_envo, 'stanza', 'prefix')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dcf3bb61",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BFO': 'BFO:\\\\d{7}',\n",
       " 'CARO': 'CARO:\\\\d{7}',\n",
       " 'CHEBI': 'CHEBI:\\\\d{4,6}',\n",
       " 'ENVO': 'ENVO:\\\\d{7,8}',\n",
       " 'FAO': 'FAO:0000001',\n",
       " 'FOODON': 'FOODON:\\\\d{8}',\n",
       " 'GO': 'GO:\\\\d{7}',\n",
       " 'IAO': 'IAO:\\\\d{7}',\n",
       " 'NCBITaxon': 'NCBITaxon:\\\\d+',\n",
       " 'OBI': 'OBI:\\\\d{7}',\n",
       " 'PATO': 'PATO:\\\\d{7}',\n",
       " 'PCO': 'PCO:\\\\d{7}',\n",
       " 'PO': 'PO:\\\\d{7}',\n",
       " 'RO': 'RO:0002577',\n",
       " 'UBERON': 'UBERON:\\\\d{7}'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_patterns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c56612",
   "metadata": {},
   "source": [
    "## Apply some normalizastion rules to the `env_package` values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "73694ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_normalized = scoped_mapping.env_package_nomralizastion(env_package_count, 'env_package',\n",
    "                                                                   id_patterns['ENVO'])\n",
    "\n",
    "env_package_normalized = scoped_mapping.add_overrides(env_package_normalized, 'remaining_tidied', 'ep_override',\n",
    "                                                   env_package_overrides)\n",
    "\n",
    "env_package_normalized = scoped_mapping.flag_canonical(env_package_normalized, 'ep_override', 'is_canonical',\n",
    "                                                    valid_combo)\n",
    "\n",
    "env_package_normalized.to_sql('env_package_normalized', biosample_cnx, if_exists='replace', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2a7fed8",
   "metadata": {},
   "source": [
    "## What do the successful normalizations look like?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7c1fbb75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.002106\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>ep_override</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>built environment</td>\n",
       "      <td>12391</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>built</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>soil</td>\n",
       "      <td>11974</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>misc environment</td>\n",
       "      <td>11715</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>missing</td>\n",
       "      <td>8453</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>human-oral</td>\n",
       "      <td>7882</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human oral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sediment</td>\n",
       "      <td>5969</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.soil</td>\n",
       "      <td>3784</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>human-associated</td>\n",
       "      <td>3167</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.host-associated</td>\n",
       "      <td>2985</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>2499</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>microbial mat/biofilm</td>\n",
       "      <td>2129</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>microbial mat biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>plant-associated</td>\n",
       "      <td>2076</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>plant associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>1763</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-associated</td>\n",
       "      <td>1417</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.sediment</td>\n",
       "      <td>1177</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-oral</td>\n",
       "      <td>1154</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human oral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>human-vaginal</td>\n",
       "      <td>1077</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human vaginal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.plant-associated</td>\n",
       "      <td>1063</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>plant associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.microbial</td>\n",
       "      <td>732</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>microbial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.miscellaneous</td>\n",
       "      <td>525</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>wastewater|sludge</td>\n",
       "      <td>417</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wastewater sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.wastewater</td>\n",
       "      <td>385</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>wastewater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>wastewater/sludge</td>\n",
       "      <td>357</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wastewater sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>unknown</td>\n",
       "      <td>283</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Human-associated</td>\n",
       "      <td>212</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.air</td>\n",
       "      <td>205</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>air</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Human-oral</td>\n",
       "      <td>201</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human oral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>host_associated</td>\n",
       "      <td>171</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>air</td>\n",
       "      <td>152</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>air</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-skin</td>\n",
       "      <td>135</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Human-gut</td>\n",
       "      <td>111</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>wastewater sludge</td>\n",
       "      <td>90</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wastewater sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>built</td>\n",
       "      <td>61</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>built</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Human_Gut</td>\n",
       "      <td>59</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>human gut</td>\n",
       "      <td>59</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>default</td>\n",
       "      <td>51</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>microbial mat|biofilm</td>\n",
       "      <td>48</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>microbial mat biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Unspecified</td>\n",
       "      <td>42</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>miscellaneous</td>\n",
       "      <td>33</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Human-vaginal</td>\n",
       "      <td>28</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human vaginal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>human_skin</td>\n",
       "      <td>28</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Soil</td>\n",
       "      <td>19</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>wastewater</td>\n",
       "      <td>14</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wastewater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-vaginal</td>\n",
       "      <td>9</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human vaginal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>misc-environment</td>\n",
       "      <td>7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Not available</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>MIGS.plant-associated</td>\n",
       "      <td>3</td>\n",
       "      <td>MIGS</td>\n",
       "      <td></td>\n",
       "      <td>plant associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Water</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>not collected</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>host associated</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           env_package     count                lhs  \\\n",
       "0                                 None  14083847                      \n",
       "1                      host-associated     49254                      \n",
       "2                            human-gut     47921                      \n",
       "3                                water     16367                      \n",
       "4                           human-skin     13706                      \n",
       "5                    built environment     12391                      \n",
       "6                                 soil     11974                      \n",
       "7                     misc environment     11715                      \n",
       "8                              missing      8453                      \n",
       "9                           human-oral      7882                      \n",
       "10                            sediment      5969                      \n",
       "11              MIGS/MIMS/MIMARKS.soil      3784  MIGS/MIMS/MIMARKS   \n",
       "12                    human-associated      3167                      \n",
       "13   MIGS/MIMS/MIMARKS.host-associated      2985  MIGS/MIMS/MIMARKS   \n",
       "14         MIGS/MIMS/MIMARKS.human-gut      2499  MIGS/MIMS/MIMARKS   \n",
       "15               microbial mat/biofilm      2129                      \n",
       "16                    plant-associated      2076                      \n",
       "17             MIGS/MIMS/MIMARKS.water      1763  MIGS/MIMS/MIMARKS   \n",
       "18  MIGS/MIMS/MIMARKS.human-associated      1417  MIGS/MIMS/MIMARKS   \n",
       "19          MIGS/MIMS/MIMARKS.sediment      1177  MIGS/MIMS/MIMARKS   \n",
       "20        MIGS/MIMS/MIMARKS.human-oral      1154  MIGS/MIMS/MIMARKS   \n",
       "21                       human-vaginal      1077                      \n",
       "22  MIGS/MIMS/MIMARKS.plant-associated      1063  MIGS/MIMS/MIMARKS   \n",
       "23         MIGS/MIMS/MIMARKS.microbial       732  MIGS/MIMS/MIMARKS   \n",
       "24     MIGS/MIMS/MIMARKS.miscellaneous       525  MIGS/MIMS/MIMARKS   \n",
       "25                   wastewater|sludge       417                      \n",
       "26        MIGS/MIMS/MIMARKS.wastewater       385  MIGS/MIMS/MIMARKS   \n",
       "27                   wastewater/sludge       357                      \n",
       "28                             unknown       283                      \n",
       "29                    Human-associated       212                      \n",
       "30               MIGS/MIMS/MIMARKS.air       205  MIGS/MIMS/MIMARKS   \n",
       "31                          Human-oral       201                      \n",
       "32                     host_associated       171                      \n",
       "33                                 air       152                      \n",
       "34        MIGS/MIMS/MIMARKS.human-skin       135  MIGS/MIMS/MIMARKS   \n",
       "35                           Human-gut       111                      \n",
       "36                   wastewater sludge        90                      \n",
       "37                               built        61                      \n",
       "38                           Human_Gut        59                      \n",
       "39                           human gut        59                      \n",
       "40                             default        51                      \n",
       "41               microbial mat|biofilm        48                      \n",
       "42                         Unspecified        42                      \n",
       "43                       miscellaneous        33                      \n",
       "44                       Human-vaginal        28                      \n",
       "45                          human_skin        28                      \n",
       "46                                Soil        19                      \n",
       "47                          wastewater        14                      \n",
       "48     MIGS/MIMS/MIMARKS.human-vaginal         9  MIGS/MIMS/MIMARKS   \n",
       "49                    misc-environment         7                      \n",
       "50                       Not available         4                      \n",
       "51                     Host-associated         3                      \n",
       "52               MIGS.plant-associated         3               MIGS   \n",
       "53                               Water         3                      \n",
       "54                       not collected         3                      \n",
       "55                     host associated         2                      \n",
       "56                       ENVO:00000016         1                      \n",
       "57                            Sediment         1                      \n",
       "58                             Unknown         1                      \n",
       "\n",
       "          extract               ep_override  \n",
       "0                                            \n",
       "1                           host associated  \n",
       "2                                 human gut  \n",
       "3                                     water  \n",
       "4                                human skin  \n",
       "5                                     built  \n",
       "6                                      soil  \n",
       "7                             miscellaneous  \n",
       "8                  no environmental package  \n",
       "9                                human oral  \n",
       "10                                 sediment  \n",
       "11                                     soil  \n",
       "12                         human associated  \n",
       "13                          host associated  \n",
       "14                                human gut  \n",
       "15                    microbial mat biofilm  \n",
       "16                         plant associated  \n",
       "17                                    water  \n",
       "18                         human associated  \n",
       "19                                 sediment  \n",
       "20                               human oral  \n",
       "21                            human vaginal  \n",
       "22                         plant associated  \n",
       "23                                microbial  \n",
       "24                            miscellaneous  \n",
       "25                        wastewater sludge  \n",
       "26                               wastewater  \n",
       "27                        wastewater sludge  \n",
       "28                 no environmental package  \n",
       "29                         human associated  \n",
       "30                                      air  \n",
       "31                               human oral  \n",
       "32                          host associated  \n",
       "33                                      air  \n",
       "34                               human skin  \n",
       "35                                human gut  \n",
       "36                        wastewater sludge  \n",
       "37                                    built  \n",
       "38                                human gut  \n",
       "39                                human gut  \n",
       "40                 no environmental package  \n",
       "41                    microbial mat biofilm  \n",
       "42                 no environmental package  \n",
       "43                            miscellaneous  \n",
       "44                            human vaginal  \n",
       "45                               human skin  \n",
       "46                                     soil  \n",
       "47                               wastewater  \n",
       "48                            human vaginal  \n",
       "49                            miscellaneous  \n",
       "50                 no environmental package  \n",
       "51                          host associated  \n",
       "52                         plant associated  \n",
       "53                                    water  \n",
       "54                 no environmental package  \n",
       "55                          host associated  \n",
       "56  ENVO:00000016                            \n",
       "57                                 sediment  \n",
       "58                 no environmental package  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    ep_override\n",
    "from\n",
    "    env_package_normalized\n",
    "where\n",
    "    is_canonical = 1\n",
    "\"\"\"\n",
    "[successful_normalizastions, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "successful_normalizastions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "740ffe3c",
   "metadata": {},
   "source": [
    "# Are there any normalization failures?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "17e08383",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.001773\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>ep_override</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>miscellaneous natural or artificial environment</td>\n",
       "      <td>611</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous natural or artificial environment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mimarks</td>\n",
       "      <td>479</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mimarks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mouse-gut</td>\n",
       "      <td>406</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mouse gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gut</td>\n",
       "      <td>172</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>biofilm</td>\n",
       "      <td>114</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>human-not providedsopharyngeal</td>\n",
       "      <td>107</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human not providedsopharyngeal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>mice gut</td>\n",
       "      <td>87</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mice gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CV</td>\n",
       "      <td>60</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>home, outdoor environment</td>\n",
       "      <td>44</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>home outdoor environment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>fermentation-associated</td>\n",
       "      <td>42</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>fermentation associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>not applicable</td>\n",
       "      <td>41</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>not applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>soil-associated</td>\n",
       "      <td>36</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>soil associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>sterile water</td>\n",
       "      <td>35</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sterile water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Mouse_Gut</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mouse gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>fresh water</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>fresh water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rock</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>rock</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>arctic</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>arctic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>freshwater sediment</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>freshwater sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>seawater</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>seawater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MIENS_16S</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miens 16s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MIMARKS_water</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mimarks water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>cave</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cave</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>plant</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>sea water</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sea water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Coastal water body (ENVO:02000049)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:02000049</td>\n",
       "      <td>coastal water body</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Deep-sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>deep sea water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MIMARKS_16S</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mimarks 16s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sea water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>env package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        env_package  count lhs        extract  \\\n",
       "0   miscellaneous natural or artificial environment    611                      \n",
       "1                                           mimarks    479                      \n",
       "2                                         mouse-gut    406                      \n",
       "3                                               gut    172                      \n",
       "4                                           biofilm    114                      \n",
       "5                    human-not providedsopharyngeal    107                      \n",
       "6                                          mice gut     87                      \n",
       "7                                                CV     60                      \n",
       "8                         home, outdoor environment     44                      \n",
       "9                           fermentation-associated     42                      \n",
       "10                                   not applicable     41                      \n",
       "11                                  soil-associated     36                      \n",
       "12                                    sterile water     35                      \n",
       "13                                        Mouse_Gut     27                      \n",
       "14                                      fresh water     27                      \n",
       "15                                             rock     12                      \n",
       "16                                           arctic     10                      \n",
       "17                              freshwater sediment      9                      \n",
       "18                                         seawater      9                      \n",
       "19                                        MIENS_16S      6                      \n",
       "20                                    MIMARKS_water      6                      \n",
       "21                                             cave      5                      \n",
       "22                                            plant      3                      \n",
       "23                                        sea water      2                      \n",
       "24               Coastal water body (ENVO:02000049)      1      ENVO:02000049   \n",
       "25                                   Deep-sea water      1                      \n",
       "26                                      MIMARKS_16S      1                      \n",
       "27                                        Sea water      1                      \n",
       "28                                      env_package      1                      \n",
       "29                                   gut microbiome      1                      \n",
       "30                  marine sediment (ENVO:00002113)      1      ENVO:00002113   \n",
       "31                                           saliva      1                      \n",
       "32                       sea water, [ENVO:00002149]      1      ENVO:00002149   \n",
       "\n",
       "                                        ep_override  \n",
       "0   miscellaneous natural or artificial environment  \n",
       "1                                           mimarks  \n",
       "2                                         mouse gut  \n",
       "3                                               gut  \n",
       "4                                           biofilm  \n",
       "5                    human not providedsopharyngeal  \n",
       "6                                          mice gut  \n",
       "7                                                cv  \n",
       "8                          home outdoor environment  \n",
       "9                           fermentation associated  \n",
       "10                                   not applicable  \n",
       "11                                  soil associated  \n",
       "12                                    sterile water  \n",
       "13                                        mouse gut  \n",
       "14                                      fresh water  \n",
       "15                                             rock  \n",
       "16                                           arctic  \n",
       "17                              freshwater sediment  \n",
       "18                                         seawater  \n",
       "19                                        miens 16s  \n",
       "20                                    mimarks water  \n",
       "21                                             cave  \n",
       "22                                            plant  \n",
       "23                                        sea water  \n",
       "24                               coastal water body  \n",
       "25                                   deep sea water  \n",
       "26                                      mimarks 16s  \n",
       "27                                        sea water  \n",
       "28                                      env package  \n",
       "29                                   gut microbiome  \n",
       "30                                  marine sediment  \n",
       "31                                           saliva  \n",
       "32                                        sea water  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    ep_override\n",
    "from\n",
    "    env_package_normalized\n",
    "where\n",
    "    is_canonical = 0\n",
    "\"\"\"\n",
    "[normalizastion_failures, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "normalizastion_failures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80370e30",
   "metadata": {},
   "source": [
    "# Utilizing taxonomy for broad subsetting\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46e3ee59",
   "metadata": {},
   "source": [
    "**This uses an SQLite database in which the transitive closure over subClassOf has already been materialized. See the README and Makefile.**\n",
    "\n",
    "Specifically, flag the biosamples whose `taxon_id` indicates they are an unclassified entity. Ignoring the others will throw out samples of multicellular organisms, like fruit flies.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bfd8416",
   "metadata": {},
   "source": [
    "## Get a listing of all taxa that are transitive subclasses of `NCBITaxon:2787823`\n",
    "\n",
    "I.e. 'unclassified entities'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c40903d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:07.726242\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:1006967</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:1041057</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:1046002</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:1046003</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:1046004</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>NCBITaxon:939928</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>NCBITaxon:941420</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>NCBITaxon:941421</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>NCBITaxon:941422</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>NCBITaxon:942017</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>994 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               subject  unclassified\n",
       "0    NCBITaxon:1006967          True\n",
       "1    NCBITaxon:1041057          True\n",
       "2    NCBITaxon:1046002          True\n",
       "3    NCBITaxon:1046003          True\n",
       "4    NCBITaxon:1046004          True\n",
       "..                 ...           ...\n",
       "989   NCBITaxon:939928          True\n",
       "990   NCBITaxon:941420          True\n",
       "991   NCBITaxon:941421          True\n",
       "992   NCBITaxon:941422          True\n",
       "993   NCBITaxon:942017          True\n",
       "\n",
       "[994 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    distinct s.subject\n",
    "from\n",
    "    entailed_edge ee\n",
    "join statements s on\n",
    "    ee.subject = s.subject\n",
    "where\n",
    "    ee.predicate = 'rdfs:subClassOf'\n",
    "    and ee.object = 'NCBITaxon:2787823'\n",
    "    and s.predicate = 'rdfs:label'\n",
    "\"\"\"\n",
    "[unclassified_taxa, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "unclassified_taxa['unclassified'] = True\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "unclassified_taxa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cad3ffdb",
   "metadata": {},
   "source": [
    "## Get taxon counuts from the Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a5d8f028",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:01.406164\n"
     ]
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    taxonomy_id biosample_taxid,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    taxonomy_id\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[biosample_tax_id_counts, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "biosample_tax_id_counts['curie'] = 'NCBITaxon:' + biosample_tax_id_counts['biosample_taxid'].astype(str)\n",
    "\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cf3a275",
   "metadata": {},
   "source": [
    "## Merge the two taxonomy dataframes\n",
    "\n",
    "I.e. flag the the Biosample records whose `taxonomy_id` field belongs to a subclass of 'unclassified entries'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b7ad29a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>curie</th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       biosample_taxid    count             curie           subject  \\\n",
       "0                 9606  6819707    NCBITaxon:9606               NaN   \n",
       "1                10090   964219   NCBITaxon:10090               NaN   \n",
       "2               408170   290862  NCBITaxon:408170  NCBITaxon:408170   \n",
       "3               410658   280666  NCBITaxon:410658  NCBITaxon:410658   \n",
       "4               646099   208741  NCBITaxon:646099  NCBITaxon:646099   \n",
       "...                ...      ...               ...               ...   \n",
       "163372          999891        1  NCBITaxon:999891               NaN   \n",
       "163373          999892        1  NCBITaxon:999892               NaN   \n",
       "163374          999898        1  NCBITaxon:999898               NaN   \n",
       "163375          999931        1  NCBITaxon:999931               NaN   \n",
       "163376          999947        1  NCBITaxon:999947               NaN   \n",
       "\n",
       "        unclassified  \n",
       "0              False  \n",
       "1              False  \n",
       "2               True  \n",
       "3               True  \n",
       "4               True  \n",
       "...              ...  \n",
       "163372         False  \n",
       "163373         False  \n",
       "163374         False  \n",
       "163375         False  \n",
       "163376         False  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biosample_tax_id_counts = biosample_tax_id_counts.merge(unclassified_taxa, left_on='curie',\n",
    "                                                        right_on='subject', how='left')\n",
    "biosample_tax_id_counts.unclassified.fillna(False, inplace=True)\n",
    "\n",
    "biosample_tax_id_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f7aff98",
   "metadata": {},
   "source": [
    "## Add labels to all taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a6acc349",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:03.939463\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>unclassified</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>False</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>False</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>True</td>\n",
       "      <td>human gut metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>True</td>\n",
       "      <td>soil metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>True</td>\n",
       "      <td>human metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Bacillus amyloliquefaciens TA208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>[Propionibacterium] humerusii P08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Peptococcaceae bacterium CEB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Barrientosiimonas humi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Angomonas ambiguus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   curie biosample_taxid    count  unclassified  \\\n",
       "0         NCBITaxon:9606            9606  6819707         False   \n",
       "1        NCBITaxon:10090           10090   964219         False   \n",
       "2       NCBITaxon:408170          408170   290862          True   \n",
       "3       NCBITaxon:410658          410658   280666          True   \n",
       "4       NCBITaxon:646099          646099   208741          True   \n",
       "...                  ...             ...      ...           ...   \n",
       "163372  NCBITaxon:999891          999891        1         False   \n",
       "163373  NCBITaxon:999892          999892        1         False   \n",
       "163374  NCBITaxon:999898          999898        1         False   \n",
       "163375  NCBITaxon:999931          999931        1         False   \n",
       "163376  NCBITaxon:999947          999947        1         False   \n",
       "\n",
       "                                    label  \n",
       "0                            Homo sapiens  \n",
       "1                            Mus musculus  \n",
       "2                    human gut metagenome  \n",
       "3                         soil metagenome  \n",
       "4                        human metagenome  \n",
       "...                                   ...  \n",
       "163372   Bacillus amyloliquefaciens TA208  \n",
       "163373  [Propionibacterium] humerusii P08  \n",
       "163374      Peptococcaceae bacterium CEB3  \n",
       "163375             Barrientosiimonas humi  \n",
       "163376                 Angomonas ambiguus  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    subject ,\n",
    "    value\n",
    "from statements\n",
    "where\n",
    "    predicate = 'rdfs:label' and subject = stanza\n",
    "\"\"\"\n",
    "[all_tax_labels, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "\n",
    "biosample_tax_id_counts = biosample_tax_id_counts.merge(all_tax_labels, left_on='curie',\n",
    "                                                        right_on='subject', how='left')\n",
    "\n",
    "biosample_tax_id_counts = biosample_tax_id_counts[['curie', 'biosample_taxid', 'count', 'unclassified', 'value']]\n",
    "biosample_tax_id_counts.columns = ['curie', 'biosample_taxid', 'count', 'unclassified', 'label']\n",
    "\n",
    "print(query_duration)\n",
    "biosample_tax_id_counts.to_sql('biobiosample_tax_id_counts', biosample_cnx, if_exists='replace', index=False)\n",
    "\n",
    "biosample_tax_id_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "554a82c0",
   "metadata": {},
   "source": [
    "**Almost all of the taxa that are common in the biosample collection are either unclassified/metagenomes or easily recognized cellular organisms. Cellular organism samples are deprioritized in this exercise**\n",
    "\n",
    "Exceptions include:\n",
    "- 32630 = synthetic construct (other entries; other sequences; artificial sequences)\n",
    "    - 'other entries' would add 16k rows on top of the 1k 'unclassified entities'\n",
    "    - metagenomes account for 331 of the 'unclassified entities'\n",
    "    - there are also a small number of uncultured/unclassified microorganisms in the biosample dataset\n",
    "- 77133 = uncultured bacterium (cellular organisms; Bacteria; environmental samples)\n",
    "    - 'cellular organisms' would add 2M rows on top of the 1k 'unclassified entities'\n",
    "    - 'cellular organisms; Bacteria; environmental samples' adds 26k\n",
    "    \n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2363d41",
   "metadata": {},
   "source": [
    "## Get a table of MIxS annotations to be mapped to ontology classes.\n",
    "\n",
    "Explicitly scope based on normalized package data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "91dcca09",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_col_to_map = 'env_broad_scale'\n",
    "scoping_col          = 'env_package_normalized.ep_override'\n",
    "scoping_value        = 'water'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80727cfd",
   "metadata": {},
   "source": [
    "**In this case, the scoping includes an inner join requirement for 'unclassified entities'**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7e777c8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>257 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count\n",
       "0                small lake biome   3264\n",
       "1    marine biome (ENVO:00000447)   1382\n",
       "2                    marine biome   1355\n",
       "3                large lake biome   1198\n",
       "4                freshwater biome   1051\n",
       "..                            ...    ...\n",
       "252                        Arctic      1\n",
       "253                 Aquatic biome      1\n",
       "254                      01000035      1\n",
       "255                      00000891      1\n",
       "256                             0      1\n",
       "\n",
       "[257 rows x 2 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = 'select ' + biosample_col_to_map + \"\"\", count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "join env_package_normalized on\n",
    "    b.env_package = env_package_normalized.env_package\n",
    "inner join biobiosample_tax_id_counts stic on\n",
    "    b.taxonomy_id = stic.biosample_taxid\n",
    "where \"\"\" + scoping_col + \" = '\" + scoping_value + \\\n",
    "    \"' group by \" + biosample_col_to_map + \"\"\"\n",
    "order by\n",
    "    count(*) desc\"\"\"\n",
    "\n",
    "[mapping_candidates, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86f43bb4",
   "metadata": {},
   "source": [
    "## The Biosample format allows for pipe-delimited environmental package lists. \n",
    "\n",
    "Separate those out into their components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "45e1859f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted  \n",
       "0                small lake biome  \n",
       "1    marine biome (ENVO:00000447)  \n",
       "2                    marine biome  \n",
       "3                large lake biome  \n",
       "4                freshwater biome  \n",
       "..                            ...  \n",
       "258                        Arctic  \n",
       "259                 Aquatic biome  \n",
       "260                      01000035  \n",
       "261                      00000891  \n",
       "262                             0  \n",
       "\n",
       "[263 rows x 4 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "multi_frames = []\n",
    "for row in mapping_candidates.itertuples(index=True, name='Pandas'):\n",
    "    split_check = row.env_broad_scale\n",
    "    if split_check is None:\n",
    "        split_check = ''\n",
    "    splitted = pd.Series(split_check.split(\"|\"))\n",
    "    splitted_count = len(splitted)\n",
    "    repeated = [split_check] * splitted_count\n",
    "    repeated = pd.Series(repeated)\n",
    "    as_frame = pd.DataFrame(dict(repeated=repeated, splitted=splitted)).reset_index()\n",
    "    multi_frames.append(as_frame)\n",
    "concat_frame = pd.concat(multi_frames)\n",
    "concat_frame = concat_frame[['repeated', 'splitted']]\n",
    "mapping_candidates = mapping_candidates.merge(concat_frame, left_on=biosample_col_to_map,\n",
    "                                              right_on='repeated', how='left')\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3af421ac",
   "metadata": {},
   "source": [
    "## Normalize a few different ways `ENVO` IDs have been entered**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ed3a89fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>envo_tidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                     envo_tidy  \n",
       "0                small lake biome              small lake biome  \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)  \n",
       "2                    marine biome                  marine biome  \n",
       "3                large lake biome              large lake biome  \n",
       "4                freshwater biome              freshwater biome  \n",
       "..                            ...                           ...  \n",
       "258                        Arctic                        Arctic  \n",
       "259                 Aquatic biome                 Aquatic biome  \n",
       "260                      01000035                      01000035  \n",
       "261                      00000891                      00000891  \n",
       "262                             0                             0  \n",
       "\n",
       "[263 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "mapping_candidates['envo_tidy'] = mapping_candidates.splitted.str.replace('envo[:_ ]', 'ENVO:', regex=True, case=False)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cc22a1a",
   "metadata": {},
   "source": [
    "# Now try to extract ontology terms that are already present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e423fcb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                     envo_tidy  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "                           string        extract  remaining_string  \\\n",
       "0                small lake biome                 small lake biome   \n",
       "1    marine biome (ENVO:00000447)  ENVO:00000447   marine biome ()   \n",
       "2                    marine biome                     marine biome   \n",
       "3                large lake biome                 large lake biome   \n",
       "4                freshwater biome                 freshwater biome   \n",
       "..                            ...            ...               ...   \n",
       "258                        Arctic                           Arctic   \n",
       "259                 Aquatic biome                    Aquatic biome   \n",
       "260                      01000035                         01000035   \n",
       "261                      00000891                         00000891   \n",
       "262                             0                                0   \n",
       "\n",
       "     remaining_tidied  \n",
       "0    small lake biome  \n",
       "1        marine biome  \n",
       "2        marine biome  \n",
       "3    large lake biome  \n",
       "4    freshwater biome  \n",
       "..                ...  \n",
       "258            arctic  \n",
       "259     aquatic biome  \n",
       "260          01000035  \n",
       "261          00000891  \n",
       "262                 0  \n",
       "\n",
       "[263 rows x 9 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_series_decomposition = scoped_mapping.decompose_series(mapping_candidates['envo_tidy'],\n",
    "                                                                 id_patterns[target_onto_prefix])\n",
    "\n",
    "mapping_candidates = pd.concat([mapping_candidates, candidate_series_decomposition], axis=1)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccd038ff",
   "metadata": {},
   "source": [
    "## Join the extracted IDs with their labels\n",
    "\n",
    "Start by connecting to the rdftab database from which the terms and label-like annotatiosn will be obtained"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "448865ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ontodb = '../semantic-sql/db/' + target_onto_prefix.lower() + '.db'\n",
    "ontocon = sqlite3.connect(ontodb)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db5d19a9",
   "metadata": {},
   "source": [
    "## extracting the labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3bcd7e47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IAO:0000111</td>\n",
       "      <td>editor preferred term~editor preferred label</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IAO:0000112</td>\n",
       "      <td>example of usage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IAO:0000114</td>\n",
       "      <td>has curation status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IAO:0000115</td>\n",
       "      <td>definition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IAO:0000116</td>\n",
       "      <td>editor note</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6774</th>\n",
       "      <td>ENVO:01001862</td>\n",
       "      <td>Solar radiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6775</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q2&gt;</td>\n",
       "      <td>Earth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q2306597&gt;</td>\n",
       "      <td>Suni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6777</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q525&gt;</td>\n",
       "      <td>Sol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6778</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q715269&gt;</td>\n",
       "      <td>World Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6779 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       subject  \\\n",
       "0                                  IAO:0000111   \n",
       "1                                  IAO:0000112   \n",
       "2                                  IAO:0000114   \n",
       "3                                  IAO:0000115   \n",
       "4                                  IAO:0000116   \n",
       "...                                        ...   \n",
       "6774                             ENVO:01001862   \n",
       "6775        <https://www.wikidata.org/wiki/Q2>   \n",
       "6776  <https://www.wikidata.org/wiki/Q2306597>   \n",
       "6777      <https://www.wikidata.org/wiki/Q525>   \n",
       "6778   <https://www.wikidata.org/wiki/Q715269>   \n",
       "\n",
       "                                             value  \n",
       "0     editor preferred term~editor preferred label  \n",
       "1                                 example of usage  \n",
       "2                              has curation status  \n",
       "3                                       definition  \n",
       "4                                      editor note  \n",
       "...                                            ...  \n",
       "6774                               Solar radiation  \n",
       "6775                                         Earth  \n",
       "6776                                          Suni  \n",
       "6777                                           Sol  \n",
       "6778                                   World Ocean  \n",
       "\n",
       "[6779 rows x 2 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    subject ,\n",
    "    value\n",
    "from\n",
    "    statements s\n",
    "where\n",
    "    predicate = 'rdfs:label'\n",
    "\"\"\"\n",
    "[onto_labels, query_duration] = scoped_mapping.timed_query(q, ontocon)\n",
    "\n",
    "onto_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "840b922f",
   "metadata": {},
   "source": [
    "## and merging "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8543ed0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                     envo_tidy  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "                           string        extract  remaining_string  \\\n",
       "0                small lake biome                 small lake biome   \n",
       "1    marine biome (ENVO:00000447)  ENVO:00000447   marine biome ()   \n",
       "2                    marine biome                     marine biome   \n",
       "3                large lake biome                 large lake biome   \n",
       "4                freshwater biome                 freshwater biome   \n",
       "..                            ...            ...               ...   \n",
       "258                        Arctic                           Arctic   \n",
       "259                 Aquatic biome                    Aquatic biome   \n",
       "260                      01000035                         01000035   \n",
       "261                      00000891                         00000891   \n",
       "262                             0                                0   \n",
       "\n",
       "     remaining_tidied        subject         value  \n",
       "0    small lake biome            NaN           NaN  \n",
       "1        marine biome  ENVO:00000447  marine biome  \n",
       "2        marine biome            NaN           NaN  \n",
       "3    large lake biome            NaN           NaN  \n",
       "4    freshwater biome            NaN           NaN  \n",
       "..                ...            ...           ...  \n",
       "258            arctic            NaN           NaN  \n",
       "259     aquatic biome            NaN           NaN  \n",
       "260          01000035            NaN           NaN  \n",
       "261          00000891            NaN           NaN  \n",
       "262                 0            NaN           NaN  \n",
       "\n",
       "[263 rows x 11 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(onto_labels, left_on='extract', right_on='subject', how='left')\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64b3baef",
   "metadata": {},
   "source": [
    "## Use cosine string distance to see if the labels match closely enough\n",
    "\n",
    "I.e. the labels claimed by the Biosample data set and the labels asserted in the ontology. if they're close enough, consider the assigned ID legit\n",
    "\n",
    "\n",
    "_How close is close enough?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "abd0a035",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "      <th>cosine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                     envo_tidy  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "                           string        extract  remaining_string  \\\n",
       "0                small lake biome                 small lake biome   \n",
       "1    marine biome (ENVO:00000447)  ENVO:00000447   marine biome ()   \n",
       "2                    marine biome                     marine biome   \n",
       "3                large lake biome                 large lake biome   \n",
       "4                freshwater biome                 freshwater biome   \n",
       "..                            ...            ...               ...   \n",
       "258                        Arctic                           Arctic   \n",
       "259                 Aquatic biome                    Aquatic biome   \n",
       "260                      01000035                         01000035   \n",
       "261                      00000891                         00000891   \n",
       "262                             0                                0   \n",
       "\n",
       "     remaining_tidied        subject         value  cosine  \n",
       "0    small lake biome            NaN                   1.0  \n",
       "1        marine biome  ENVO:00000447  marine biome     0.0  \n",
       "2        marine biome            NaN                   1.0  \n",
       "3    large lake biome            NaN                   1.0  \n",
       "4    freshwater biome            NaN                   1.0  \n",
       "..                ...            ...           ...     ...  \n",
       "258            arctic            NaN                   1.0  \n",
       "259     aquatic biome            NaN                   1.0  \n",
       "260          01000035            NaN                   1.0  \n",
       "261          00000891            NaN                   1.0  \n",
       "262                 0            NaN                   1.0  \n",
       "\n",
       "[263 rows x 12 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "my_cosine_obj = Cosine(my_string_dist_arg)\n",
    "mapping_candidates['value'] = mapping_candidates['value'].fillna('')\n",
    "mapping_candidates['cosine'] = mapping_candidates.apply(\n",
    "    lambda my_row: my_cosine_obj.distance(my_row['remaining_tidied'].lower(), my_row['value'].lower()), axis=1)\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2358a89",
   "metadata": {},
   "source": [
    "**Previously, we did a reality check on the claimed IDs and labels. If a label is claimed without any ID, that could still be a path to an ontology term.**\n",
    "\n",
    "We'll be doing some merging, so make sure column names aren't reused\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "092aced0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000873</td>\n",
       "      <td>freshwater biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00002030</td>\n",
       "      <td>aquatic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                     envo_tidy  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "                           string        extract  remaining_string  \\\n",
       "0                small lake biome                 small lake biome   \n",
       "1    marine biome (ENVO:00000447)  ENVO:00000447   marine biome ()   \n",
       "2                    marine biome                     marine biome   \n",
       "3                large lake biome                 large lake biome   \n",
       "4                freshwater biome                 freshwater biome   \n",
       "..                            ...            ...               ...   \n",
       "258                        Arctic                           Arctic   \n",
       "259                 Aquatic biome                    Aquatic biome   \n",
       "260                      01000035                         01000035   \n",
       "261                      00000891                         00000891   \n",
       "262                             0                                0   \n",
       "\n",
       "     remaining_tidied        term_id   lab_from_id  lfi_cosine  \\\n",
       "0    small lake biome            NaN                       1.0   \n",
       "1        marine biome  ENVO:00000447  marine biome         0.0   \n",
       "2        marine biome            NaN                       1.0   \n",
       "3    large lake biome            NaN                       1.0   \n",
       "4    freshwater biome            NaN                       1.0   \n",
       "..                ...            ...           ...         ...   \n",
       "258            arctic            NaN                       1.0   \n",
       "259     aquatic biome            NaN                       1.0   \n",
       "260          01000035            NaN                       1.0   \n",
       "261          00000891            NaN                       1.0   \n",
       "262                 0            NaN                       1.0   \n",
       "\n",
       "    term_id_from_lab             value  \n",
       "0                NaN               NaN  \n",
       "1      ENVO:00000447      marine biome  \n",
       "2      ENVO:00000447      marine biome  \n",
       "3                NaN               NaN  \n",
       "4      ENVO:00000873  freshwater biome  \n",
       "..               ...               ...  \n",
       "258              NaN               NaN  \n",
       "259    ENVO:00002030     aquatic biome  \n",
       "260              NaN               NaN  \n",
       "261              NaN               NaN  \n",
       "262              NaN               NaN  \n",
       "\n",
       "[263 rows x 14 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates.columns = ['env_broad_scale', 'count', 'repeated', 'splitted', 'envo_tidy', 'string', 'extract',\n",
    "                              'remaining_string', 'remaining_tidied', 'term_id', 'lab_from_id', 'lfi_cosine']\n",
    "mapping_candidates = mapping_candidates.merge(onto_labels, left_on='remaining_tidied', right_on='value', how='left')\n",
    "\n",
    "mapping_candidates.columns = ['env_broad_scale', 'count', 'repeated', 'splitted', 'envo_tidy', 'string', 'extract',\n",
    "                              'remaining_string', 'remaining_tidied', 'term_id', 'lab_from_id',\n",
    "                              'lfi_cosine', 'term_id_from_lab', 'value']\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "54ac73d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_candidates.to_sql('mapping_scratch', biosample_cnx, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41082d99",
   "metadata": {},
   "source": [
    "## Find consensus term IDs and labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4f5e3fdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mapping_candidates['consensus_id'] = mapping_candidates['term_id_from_lab']\n",
    "mapping_candidates['consensus_lab'] = mapping_candidates['value']\n",
    "\n",
    "flag = mapping_candidates['consensus_id'].isnull() &\\\n",
    "    (( ~ mapping_candidates['term_id'].isnull() &\\\n",
    "    mapping_candidates.lfi_cosine.le(my_max_string_dist)) |\\\n",
    "     (~ mapping_candidates['term_id'].isnull() &\\\n",
    "      mapping_candidates['remaining_tidied'].eq('')))\n",
    "    \n",
    "replacements = mapping_candidates.loc[flag, 'term_id']\n",
    "mapping_candidates.loc[flag, 'consensus_id'] = replacements\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, 'lab_from_id']\n",
    "mapping_candidates.loc[flag, 'consensus_lab'] = replacements\n",
    "\n",
    "flag = mapping_candidates.consensus_id.isna()\n",
    "antiflag = ~ flag\n",
    "mapping_candidates['id_or_lab_ok'] = antiflag\n",
    "\n",
    "mapping_candidates['assembled_consensus'] = mapping_candidates['consensus_lab'] + ' [' + \\\n",
    "    mapping_candidates['consensus_id'] + ']'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6af120f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "      <th>id_or_lab_ok</th>\n",
       "      <th>assembled_consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>True</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>True</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000873</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>ENVO:00000873</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>True</td>\n",
       "      <td>freshwater biome [ENVO:00000873]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00002030</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>ENVO:00002030</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>True</td>\n",
       "      <td>aquatic biome [ENVO:00002030]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                     envo_tidy  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "                           string        extract  remaining_string  \\\n",
       "0                small lake biome                 small lake biome   \n",
       "1    marine biome (ENVO:00000447)  ENVO:00000447   marine biome ()   \n",
       "2                    marine biome                     marine biome   \n",
       "3                large lake biome                 large lake biome   \n",
       "4                freshwater biome                 freshwater biome   \n",
       "..                            ...            ...               ...   \n",
       "258                        Arctic                           Arctic   \n",
       "259                 Aquatic biome                    Aquatic biome   \n",
       "260                      01000035                         01000035   \n",
       "261                      00000891                         00000891   \n",
       "262                             0                                0   \n",
       "\n",
       "     remaining_tidied        term_id   lab_from_id  lfi_cosine  \\\n",
       "0    small lake biome            NaN                       1.0   \n",
       "1        marine biome  ENVO:00000447  marine biome         0.0   \n",
       "2        marine biome            NaN                       1.0   \n",
       "3    large lake biome            NaN                       1.0   \n",
       "4    freshwater biome            NaN                       1.0   \n",
       "..                ...            ...           ...         ...   \n",
       "258            arctic            NaN                       1.0   \n",
       "259     aquatic biome            NaN                       1.0   \n",
       "260          01000035            NaN                       1.0   \n",
       "261          00000891            NaN                       1.0   \n",
       "262                 0            NaN                       1.0   \n",
       "\n",
       "    term_id_from_lab             value   consensus_id     consensus_lab  \\\n",
       "0                NaN               NaN            NaN               NaN   \n",
       "1      ENVO:00000447      marine biome  ENVO:00000447      marine biome   \n",
       "2      ENVO:00000447      marine biome  ENVO:00000447      marine biome   \n",
       "3                NaN               NaN            NaN               NaN   \n",
       "4      ENVO:00000873  freshwater biome  ENVO:00000873  freshwater biome   \n",
       "..               ...               ...            ...               ...   \n",
       "258              NaN               NaN            NaN               NaN   \n",
       "259    ENVO:00002030     aquatic biome  ENVO:00002030     aquatic biome   \n",
       "260              NaN               NaN            NaN               NaN   \n",
       "261              NaN               NaN            NaN               NaN   \n",
       "262              NaN               NaN            NaN               NaN   \n",
       "\n",
       "     id_or_lab_ok               assembled_consensus  \n",
       "0           False                               NaN  \n",
       "1            True      marine biome [ENVO:00000447]  \n",
       "2            True      marine biome [ENVO:00000447]  \n",
       "3           False                               NaN  \n",
       "4            True  freshwater biome [ENVO:00000873]  \n",
       "..            ...                               ...  \n",
       "258         False                               NaN  \n",
       "259          True     aquatic biome [ENVO:00002030]  \n",
       "260         False                               NaN  \n",
       "261         False                               NaN  \n",
       "262         False                               NaN  \n",
       "\n",
       "[263 rows x 18 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47aa8012",
   "metadata": {},
   "source": [
    "## Save these easy term mappings to SQLite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3d8015be",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates.to_sql('mapping_scratch', biosample_cnx, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13d44c8d",
   "metadata": {},
   "source": [
    "## For which Biosample annotations were no easy mappings found?\n",
    "How many Biosamples uses those annotations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c29338b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = ~ mapping_candidates.id_or_lab_ok\n",
    "needs_search = mapping_candidates.loc[flag]\n",
    "\n",
    "needs_search = needs_search[['remaining_tidied', 'count']]\n",
    "\n",
    "sum_by_needed = needs_search.groupby('remaining_tidied')['count'].sum()\n",
    "\n",
    "sum_by_needed = sum_by_needed.to_frame()\n",
    "sum_by_needed['remaining_tidied'] = sum_by_needed.index\n",
    "\n",
    "sum_by_needed = sum_by_needed.sort_values('count', ascending=False)\n",
    "sum_by_needed.reset_index(drop=True)\n",
    "\n",
    "sum_by_needed.to_sql('sum_by_needed', biosample_cnx, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "91fec97a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>small lake biome</th>\n",
       "      <td>3287</td>\n",
       "      <td>small lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>large lake biome</th>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>879</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000686</th>\n",
       "      <td>850</td>\n",
       "      <td>1000686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>surface seawater</th>\n",
       "      <td>480</td>\n",
       "      <td>surface seawater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hydrothermal vents</th>\n",
       "      <td>1</td>\n",
       "      <td>hydrothermal vents</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hanford h 101</th>\n",
       "      <td>1</td>\n",
       "      <td>hanford h 101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hanford h 100</th>\n",
       "      <td>1</td>\n",
       "      <td>hanford h 100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gulf of eilat</th>\n",
       "      <td>1</td>\n",
       "      <td>gulf of eilat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zooplankton</th>\n",
       "      <td>1</td>\n",
       "      <td>zooplankton</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>131 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    count    remaining_tidied\n",
       "remaining_tidied                             \n",
       "small lake biome     3287    small lake biome\n",
       "large lake biome     1198    large lake biome\n",
       "15                    879                  15\n",
       "1000686               850             1000686\n",
       "surface seawater      480    surface seawater\n",
       "...                   ...                 ...\n",
       "hydrothermal vents      1  hydrothermal vents\n",
       "hanford h 101           1       hanford h 101\n",
       "hanford h 100           1       hanford h 100\n",
       "gulf of eilat           1       gulf of eilat\n",
       "zooplankton             1         zooplankton\n",
       "\n",
       "[131 rows x 2 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_by_needed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53bcffc5",
   "metadata": {},
   "source": [
    "## Extract the tidied strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "02a43f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_raw_list = list(sum_by_needed['remaining_tidied'])\n",
    "ebs_raw_list.sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bd91eb2",
   "metadata": {},
   "source": [
    "## Submit those tidied strings to a search engine\n",
    "\n",
    "Specifically OLS search. This takes roughly one second per submission\n",
    "\n",
    "_Turn logging back on to show status?_\n",
    "_Print the count and pre- and post- datestamps_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "08e39629",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_search_res = scoped_mapping.search_get_annotations_wrapper(ebs_raw_list,\n",
    "                                                               bad_chars=chars_to_whiteout,\n",
    "                                                               cat_name=biosample_col_to_map,\n",
    "                                                               ontoprefix='envo,gaz',\n",
    "                                                               query_fields='',\n",
    "                                                               rr=5,\n",
    "                                                               string_dist_arg=my_string_dist_arg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae3d29dc",
   "metadata": {},
   "source": [
    "## Filter out the best of the acceptable mappings\n",
    "From a string distance perspective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b695a649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>arabian sea</td>\n",
       "      <td>arabian sea</td>\n",
       "      <td>Arabian Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00002457</td>\n",
       "      <td>Arabian Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00002457</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>atlantic ocean</td>\n",
       "      <td>atlantic ocean</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00000344</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00000344</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>subpolar coniferous forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000250</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>caribbean sea</td>\n",
       "      <td>caribbean sea</td>\n",
       "      <td>Caribbean Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00002820</td>\n",
       "      <td>Caribbean Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00002820</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>coastal water</td>\n",
       "      <td>coastal water</td>\n",
       "      <td>coastal water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002150</td>\n",
       "      <td>coastal sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_broad_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002150</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:2759</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2759</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>forest</td>\n",
       "      <td>forest</td>\n",
       "      <td>Forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00454366</td>\n",
       "      <td>Forest</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00454366</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002011</td>\n",
       "      <td>fresh water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002011</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:22224982</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_22224982</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>hydrothermal vents</td>\n",
       "      <td>hydrothermal vents</td>\n",
       "      <td>hydrothermal vent</td>\n",
       "      <td>1</td>\n",
       "      <td>0.030</td>\n",
       "      <td>ENVO:00000215</td>\n",
       "      <td>hydrothermal vent</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000215</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>marginal sea</td>\n",
       "      <td>marginal sea</td>\n",
       "      <td>marginal sea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td>sea</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000016</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>marine brine pool basin</td>\n",
       "      <td>marine brine pool basin</td>\n",
       "      <td>marine brine pool</td>\n",
       "      <td>1</td>\n",
       "      <td>0.081</td>\n",
       "      <td>ENVO:00000369</td>\n",
       "      <td>brine pool</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000369</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>mediterranean sea</td>\n",
       "      <td>mediterranean sea</td>\n",
       "      <td>Mediterranean Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00002287</td>\n",
       "      <td>Mediterranean Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00002287</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>1</td>\n",
       "      <td>0.039</td>\n",
       "      <td>ENVO:01000008</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000008</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>ocean water</td>\n",
       "      <td>ocean water</td>\n",
       "      <td>ocean water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002149</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>seawater</td>\n",
       "      <td>seawater</td>\n",
       "      <td>seawater</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002149</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1072</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>southern ocean</td>\n",
       "      <td>southern ocean</td>\n",
       "      <td>Southern Ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00000373</td>\n",
       "      <td>Southern Ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00000373</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1132</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>temperate mediterannean sea biome</td>\n",
       "      <td>temperate mediterannean sea biome</td>\n",
       "      <td>temperate mediterranean sea biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.021</td>\n",
       "      <td>ENVO:01000857</td>\n",
       "      <td>temperate mediterranean sea biome</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000857</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1190</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>water river</td>\n",
       "      <td>water river</td>\n",
       "      <td>river water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.083</td>\n",
       "      <td>ENVO:01000599</td>\n",
       "      <td>river water</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000599</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             category                                raw  \\\n",
       "0     env_broad_scale                                      \n",
       "48    env_broad_scale                        arabian sea   \n",
       "81    env_broad_scale                     atlantic ocean   \n",
       "101   env_broad_scale                      boreal forest   \n",
       "133   env_broad_scale                      caribbean sea   \n",
       "213   env_broad_scale                      coastal water   \n",
       "417   env_broad_scale                         eukaryotes   \n",
       "508   env_broad_scale                             forest   \n",
       "520   env_broad_scale                         freshwater   \n",
       "585   env_broad_scale                        hot springs   \n",
       "594   env_broad_scale                 hydrothermal vents   \n",
       "678   env_broad_scale                       marginal sea   \n",
       "724   env_broad_scale            marine brine pool basin   \n",
       "768   env_broad_scale                  mediterranean sea   \n",
       "810   env_broad_scale                     microbial mats   \n",
       "890   env_broad_scale                        ocean water   \n",
       "1032  env_broad_scale                           seawater   \n",
       "1072  env_broad_scale                     southern ocean   \n",
       "1132  env_broad_scale  temperate mediterannean sea biome   \n",
       "1190  env_broad_scale                        water river   \n",
       "\n",
       "                                  query                               name  \\\n",
       "0                                                                            \n",
       "48                          arabian sea                        Arabian Sea   \n",
       "81                       atlantic ocean                     Atlantic Ocean   \n",
       "101                       boreal forest                      boreal forest   \n",
       "133                       caribbean sea                      Caribbean Sea   \n",
       "213                       coastal water                      coastal water   \n",
       "417                          eukaryotes                         eukaryotes   \n",
       "508                              forest                             Forest   \n",
       "520                          freshwater                         freshwater   \n",
       "585                         hot springs                        Hot Springs   \n",
       "594                  hydrothermal vents                  hydrothermal vent   \n",
       "678                        marginal sea                       marginal sea   \n",
       "724             marine brine pool basin                  marine brine pool   \n",
       "768                   mediterranean sea                  Mediterranean Sea   \n",
       "810                      microbial mats                      microbial mat   \n",
       "890                         ocean water                        ocean water   \n",
       "1032                           seawater                           seawater   \n",
       "1072                     southern ocean                     Southern Ocean   \n",
       "1132  temperate mediterannean sea biome  temperate mediterranean sea biome   \n",
       "1190                        water river                        river water   \n",
       "\n",
       "      string_dist_rank  string_dist          obo_id  \\\n",
       "0                    1        0.000                   \n",
       "48                   1        0.000    GAZ:00002457   \n",
       "81                   1        0.000    GAZ:00000344   \n",
       "101                  1        0.000   ENVO:01000250   \n",
       "133                  1        0.000    GAZ:00002820   \n",
       "213                  1        0.000   ENVO:00002150   \n",
       "417                  1        0.000  NCBITaxon:2759   \n",
       "508                  1        0.000    GAZ:00454366   \n",
       "520                  1        0.000   ENVO:00002011   \n",
       "585                  1        0.000    GAZ:22224982   \n",
       "594                  1        0.030   ENVO:00000215   \n",
       "678                  1        0.000   ENVO:00000016   \n",
       "724                  1        0.081   ENVO:00000369   \n",
       "768                  1        0.000    GAZ:00002287   \n",
       "810                  1        0.039   ENVO:01000008   \n",
       "890                  1        0.000   ENVO:00002149   \n",
       "1032                 1        0.000   ENVO:00002149   \n",
       "1072                 1        0.000    GAZ:00000373   \n",
       "1132                 1        0.021   ENVO:01000857   \n",
       "1190                 1        0.083   ENVO:01000599   \n",
       "\n",
       "                                  label  search_rank ontology_prefix  \\\n",
       "0                                                  1                   \n",
       "48                          Arabian Sea            1             GAZ   \n",
       "81                       Atlantic Ocean            1             GAZ   \n",
       "101    subpolar coniferous forest biome            1            ENVO   \n",
       "133                       Caribbean Sea            1             GAZ   \n",
       "213                   coastal sea water            1            ENVO   \n",
       "417                           Eukaryota            1            ENVO   \n",
       "508                              Forest            1             GAZ   \n",
       "520                         fresh water            1            ENVO   \n",
       "585                         Hot Springs            1             GAZ   \n",
       "594                   hydrothermal vent            4            ENVO   \n",
       "678                                 sea            1            ENVO   \n",
       "724                          brine pool            1            ENVO   \n",
       "768                   Mediterranean Sea            1             GAZ   \n",
       "810                       microbial mat            1            ENVO   \n",
       "890                           sea water            1            ENVO   \n",
       "1032                          sea water            1            ENVO   \n",
       "1072                     Southern Ocean            1             GAZ   \n",
       "1132  temperate mediterranean sea biome            2            ENVO   \n",
       "1190                        river water            3            ENVO   \n",
       "\n",
       "                   scope        type  \\\n",
       "0                    NaN         NaN   \n",
       "48                 label       label   \n",
       "81                 label       label   \n",
       "101   has_narrow_synonym  annotation   \n",
       "133                label       label   \n",
       "213    has_broad_synonym  annotation   \n",
       "417              synonym     synonym   \n",
       "508                label       label   \n",
       "520              synonym     synonym   \n",
       "585                label       label   \n",
       "594                label       label   \n",
       "678   has_narrow_synonym  annotation   \n",
       "724              synonym     synonym   \n",
       "768                label       label   \n",
       "810                label       label   \n",
       "890              synonym     synonym   \n",
       "1032             synonym     synonym   \n",
       "1072               label       label   \n",
       "1132               label       label   \n",
       "1190               label       label   \n",
       "\n",
       "                                                iri ontology_name  \n",
       "0                                                                  \n",
       "48      http://purl.obolibrary.org/obo/GAZ_00002457           gaz  \n",
       "81      http://purl.obolibrary.org/obo/GAZ_00000344           gaz  \n",
       "101    http://purl.obolibrary.org/obo/ENVO_01000250          envo  \n",
       "133     http://purl.obolibrary.org/obo/GAZ_00002820           gaz  \n",
       "213    http://purl.obolibrary.org/obo/ENVO_00002150          envo  \n",
       "417   http://purl.obolibrary.org/obo/NCBITaxon_2759          envo  \n",
       "508     http://purl.obolibrary.org/obo/GAZ_00454366           gaz  \n",
       "520    http://purl.obolibrary.org/obo/ENVO_00002011          envo  \n",
       "585     http://purl.obolibrary.org/obo/GAZ_22224982           gaz  \n",
       "594    http://purl.obolibrary.org/obo/ENVO_00000215          envo  \n",
       "678    http://purl.obolibrary.org/obo/ENVO_00000016          envo  \n",
       "724    http://purl.obolibrary.org/obo/ENVO_00000369          envo  \n",
       "768     http://purl.obolibrary.org/obo/GAZ_00002287           gaz  \n",
       "810    http://purl.obolibrary.org/obo/ENVO_01000008          envo  \n",
       "890    http://purl.obolibrary.org/obo/ENVO_00002149          envo  \n",
       "1032   http://purl.obolibrary.org/obo/ENVO_00002149          envo  \n",
       "1072    http://purl.obolibrary.org/obo/GAZ_00000373           gaz  \n",
       "1132   http://purl.obolibrary.org/obo/ENVO_01000857          envo  \n",
       "1190   http://purl.obolibrary.org/obo/ENVO_01000599          envo  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_best_acceptable = scoped_mapping.get_best_acceptable(ebs_search_res, max_string_dist=my_max_string_dist)\n",
    "my_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43aa5018",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Filter out the submissions with no acceptable matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "39e98bf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Milecastle 0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000</td>\n",
       "      <td>GAZ:00456187</td>\n",
       "      <td>Milecastle 0</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00456187</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>lead(0)</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000</td>\n",
       "      <td>CHEBI:27889</td>\n",
       "      <td>lead(0)</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CHEBI_27889</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>iron(0)</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000</td>\n",
       "      <td>CHEBI:82664</td>\n",
       "      <td>iron(0)</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CHEBI_82664</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>platinum(0)</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000</td>\n",
       "      <td>CHEBI:33400</td>\n",
       "      <td>platinum(0)</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CHEBI_33400</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>uranium(0)</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000</td>\n",
       "      <td>CHEBI:49936</td>\n",
       "      <td>uranium(0)</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CHEBI_49936</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1199</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>westerlies biome</td>\n",
       "      <td>westerlies biome</td>\n",
       "      <td>EcosytemType</td>\n",
       "      <td>8</td>\n",
       "      <td>0.927</td>\n",
       "      <td>ENVO:00000428</td>\n",
       "      <td>biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000428</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1196</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>westerlies biome</td>\n",
       "      <td>westerlies biome</td>\n",
       "      <td>major habitat type</td>\n",
       "      <td>9</td>\n",
       "      <td>0.941</td>\n",
       "      <td>ENVO:00000428</td>\n",
       "      <td>biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000428</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>westerlies biome</td>\n",
       "      <td>westerlies biome</td>\n",
       "      <td>major habitat type</td>\n",
       "      <td>10</td>\n",
       "      <td>0.941</td>\n",
       "      <td>ENVO:00000428</td>\n",
       "      <td>biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000428</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>westerlies biome</td>\n",
       "      <td>westerlies biome</td>\n",
       "      <td>marine realm</td>\n",
       "      <td>11</td>\n",
       "      <td>1.000</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000447</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1206</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>zooplankton</td>\n",
       "      <td>zooplankton</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1001 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             category               raw             query                name  \\\n",
       "1     env_broad_scale                 0                 0        Milecastle 0   \n",
       "2     env_broad_scale                 0                 0             lead(0)   \n",
       "3     env_broad_scale                 0                 0             iron(0)   \n",
       "4     env_broad_scale                 0                 0         platinum(0)   \n",
       "5     env_broad_scale                 0                 0          uranium(0)   \n",
       "...               ...               ...               ...                 ...   \n",
       "1199  env_broad_scale  westerlies biome  westerlies biome        EcosytemType   \n",
       "1196  env_broad_scale  westerlies biome  westerlies biome  major habitat type   \n",
       "1197  env_broad_scale  westerlies biome  westerlies biome  major habitat type   \n",
       "1201  env_broad_scale  westerlies biome  westerlies biome        marine realm   \n",
       "1206  env_broad_scale       zooplankton       zooplankton                       \n",
       "\n",
       "      string_dist_rank  string_dist         obo_id         label  search_rank  \\\n",
       "1                    1        1.000   GAZ:00456187  Milecastle 0            1   \n",
       "2                    2        1.000    CHEBI:27889       lead(0)            2   \n",
       "3                    3        1.000    CHEBI:82664       iron(0)            3   \n",
       "4                    4        1.000    CHEBI:33400   platinum(0)            4   \n",
       "5                    5        1.000    CHEBI:49936    uranium(0)            5   \n",
       "...                ...          ...            ...           ...          ...   \n",
       "1199                 8        0.927  ENVO:00000428         biome            1   \n",
       "1196                 9        0.941  ENVO:00000428         biome            1   \n",
       "1197                10        0.941  ENVO:00000428         biome            1   \n",
       "1201                11        1.000  ENVO:00000447  marine biome            2   \n",
       "1206                 1        1.000                                         1   \n",
       "\n",
       "     ontology_prefix                scope        type  \\\n",
       "1                GAZ                label       label   \n",
       "2               ENVO                label       label   \n",
       "3               ENVO                label       label   \n",
       "4               ENVO                label       label   \n",
       "5               ENVO                label       label   \n",
       "...              ...                  ...         ...   \n",
       "1199            ENVO  has_related_synonym  annotation   \n",
       "1196            ENVO              synonym     synonym   \n",
       "1197            ENVO      hasExactSynonym        None   \n",
       "1201            ENVO  has_related_synonym  annotation   \n",
       "1206                                  NaN         NaN   \n",
       "\n",
       "                                               iri ontology_name  \n",
       "1      http://purl.obolibrary.org/obo/GAZ_00456187           gaz  \n",
       "2       http://purl.obolibrary.org/obo/CHEBI_27889          envo  \n",
       "3       http://purl.obolibrary.org/obo/CHEBI_82664          envo  \n",
       "4       http://purl.obolibrary.org/obo/CHEBI_33400          envo  \n",
       "5       http://purl.obolibrary.org/obo/CHEBI_49936          envo  \n",
       "...                                            ...           ...  \n",
       "1199  http://purl.obolibrary.org/obo/ENVO_00000428          envo  \n",
       "1196  http://purl.obolibrary.org/obo/ENVO_00000428          envo  \n",
       "1197  http://purl.obolibrary.org/obo/ENVO_00000428          envo  \n",
       "1201  http://purl.obolibrary.org/obo/ENVO_00000447          envo  \n",
       "1206                                                              \n",
       "\n",
       "[1001 rows x 14 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_acceptable_mappings = scoped_mapping.get_no_acceptable_mappings(ebs_search_res, my_best_acceptable)\n",
    "\n",
    "no_acceptable_mappings.to_sql('no_acceptable_mappings', biosample_cnx, if_exists='replace', index=False)\n",
    "\n",
    "no_acceptable_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e07cdab1",
   "metadata": {},
   "source": [
    "## Mapping failure patterns\n",
    "\n",
    "- Some get a good hit if 'biome' is added\n",
    "\n",
    "\n",
    "_How to manually review and then add back in?__\n",
    "\n",
    "\n",
    "Add to biosample SQLite database:\n",
    "\n",
    "    - mapping_candidates -> mapping_scratch (easy ID-based and exact-tidied-label-based)\n",
    "    - ebs_search_results (no acceptable + all acceptable)?\n",
    "    - my_best_acceptable\n",
    "    - no_acceptable_mappings\n",
    "    \n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70be6244",
   "metadata": {},
   "source": [
    "## Try searching the failures against all ontologies in OLS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ac5c4f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "still_unmapped = list(set(list(no_acceptable_mappings['raw'])))\n",
    "still_unmapped.sort()\n",
    "\n",
    "salvage_search_res = scoped_mapping.search_get_annotations_wrapper(still_unmapped,\n",
    "                                                               bad_chars='._-',\n",
    "                                                               cat_name='salvage',\n",
    "                                                               ontoprefix='',\n",
    "                                                               query_fields='',\n",
    "                                                               rr=5,\n",
    "                                                               string_dist_arg=2)\n",
    "\n",
    "salvage_best_acceptable = scoped_mapping.get_best_acceptable(salvage_search_res,\n",
    "                                                             max_string_dist=0.2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "453b8356",
   "metadata": {},
   "source": [
    "## We appear to be at a point of diminishing returns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d44fe45b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>salvage</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C159767</td>\n",
       "      <td>Zero Full Term Pregnancies</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C159767</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>salvage</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C113429</td>\n",
       "      <td>Fifteen</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113429</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>salvage</td>\n",
       "      <td>arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C44738</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C44738</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>salvage</td>\n",
       "      <td>brackish water river</td>\n",
       "      <td>brackish water river</td>\n",
       "      <td>brackish water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.153</td>\n",
       "      <td>ENVO:00002019</td>\n",
       "      <td>brackish water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002019</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>salvage</td>\n",
       "      <td>brine pool interface layer</td>\n",
       "      <td>brine pool interface layer</td>\n",
       "      <td>interface layer</td>\n",
       "      <td>1</td>\n",
       "      <td>0.192</td>\n",
       "      <td>ENVO:01001684</td>\n",
       "      <td>interface layer</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001684</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>salvage</td>\n",
       "      <td>coastal ocean</td>\n",
       "      <td>coastal ocean</td>\n",
       "      <td>coastal ocean water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>ENVO:00002150</td>\n",
       "      <td>coastal sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_exact_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002150</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>salvage</td>\n",
       "      <td>deep mediterranean</td>\n",
       "      <td>deep mediterranean</td>\n",
       "      <td>Mediterranean</td>\n",
       "      <td>1</td>\n",
       "      <td>0.142</td>\n",
       "      <td>LBO:0001043</td>\n",
       "      <td>Mediterranean</td>\n",
       "      <td>2</td>\n",
       "      <td>LBO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/LBO_0001043</td>\n",
       "      <td>lbo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>801</th>\n",
       "      <td>salvage</td>\n",
       "      <td>envo marine biome</td>\n",
       "      <td>envo marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.171</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENM</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000447</td>\n",
       "      <td>enm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830</th>\n",
       "      <td>salvage</td>\n",
       "      <td>estuarine</td>\n",
       "      <td>estuarine</td>\n",
       "      <td>estuarine mud</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>ENVO:00002160</td>\n",
       "      <td>estuarine mud</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002160</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914</th>\n",
       "      <td>salvage</td>\n",
       "      <td>for marine biome</td>\n",
       "      <td>for marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.144</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000447</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>salvage</td>\n",
       "      <td>freshwater reservoir</td>\n",
       "      <td>freshwater reservoir</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>1</td>\n",
       "      <td>0.200</td>\n",
       "      <td>ENVO:00002011</td>\n",
       "      <td>fresh water</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002011</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>960</th>\n",
       "      <td>salvage</td>\n",
       "      <td>gulf</td>\n",
       "      <td>gulf</td>\n",
       "      <td>gulf</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>CHEBI:152556</td>\n",
       "      <td>D-gulofuranose</td>\n",
       "      <td>1</td>\n",
       "      <td>CHEBI</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CHEBI_152556</td>\n",
       "      <td>chebi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1130</th>\n",
       "      <td>salvage</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00000891</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENM</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000891</td>\n",
       "      <td>enm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1331</th>\n",
       "      <td>salvage</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:408172</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_408172</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1360</th>\n",
       "      <td>salvage</td>\n",
       "      <td>marine pelagic</td>\n",
       "      <td>marine pelagic</td>\n",
       "      <td>marine pelagic zone</td>\n",
       "      <td>1</td>\n",
       "      <td>0.132</td>\n",
       "      <td>ENVO:00000208</td>\n",
       "      <td>marine pelagic zone</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000208</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1445</th>\n",
       "      <td>salvage</td>\n",
       "      <td>marine water</td>\n",
       "      <td>marine water</td>\n",
       "      <td>marine waterbody</td>\n",
       "      <td>1</td>\n",
       "      <td>0.144</td>\n",
       "      <td>ENVO:00001999</td>\n",
       "      <td>marine water body</td>\n",
       "      <td>5</td>\n",
       "      <td>RBO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00001999</td>\n",
       "      <td>rbo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1453</th>\n",
       "      <td>salvage</td>\n",
       "      <td>meditterean sea biome</td>\n",
       "      <td>meditterean sea biome</td>\n",
       "      <td>mediterranean sea biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.113</td>\n",
       "      <td>ENVO:01000047</td>\n",
       "      <td>mediterranean sea biome</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000047</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1534</th>\n",
       "      <td>salvage</td>\n",
       "      <td>microorganism</td>\n",
       "      <td>microorganism</td>\n",
       "      <td>Microorganism</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C14329</td>\n",
       "      <td>Microorganism</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14329</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664</th>\n",
       "      <td>salvage</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>CO_357:2100001</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>CO_357</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://www.cropontology.org/rdf/CO_357:2100001</td>\n",
       "      <td>co_357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1694</th>\n",
       "      <td>salvage</td>\n",
       "      <td>ne atlantic</td>\n",
       "      <td>ne atlantic</td>\n",
       "      <td>ne atlantic ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>0.180</td>\n",
       "      <td>GAZ:00013759</td>\n",
       "      <td>Northeast Atlantic Ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00013759</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1721</th>\n",
       "      <td>salvage</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Not applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C48660</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C48660</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1827</th>\n",
       "      <td>salvage</td>\n",
       "      <td>oligotrophic</td>\n",
       "      <td>oligotrophic</td>\n",
       "      <td>oligotrophic lake</td>\n",
       "      <td>1</td>\n",
       "      <td>0.171</td>\n",
       "      <td>ENVO:01000774</td>\n",
       "      <td>oligotrophic lake</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000774</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1860</th>\n",
       "      <td>salvage</td>\n",
       "      <td>pacific coast</td>\n",
       "      <td>pacific coast</td>\n",
       "      <td>Pacific Coast States</td>\n",
       "      <td>1</td>\n",
       "      <td>0.181</td>\n",
       "      <td>NCIT:C43455</td>\n",
       "      <td>Pacific Coast States</td>\n",
       "      <td>4</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C43455</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1864</th>\n",
       "      <td>salvage</td>\n",
       "      <td>phragmites australis rhizosphere</td>\n",
       "      <td>phragmites australis rhizosphere</td>\n",
       "      <td>Phragmites australis</td>\n",
       "      <td>1</td>\n",
       "      <td>0.175</td>\n",
       "      <td>NCBITaxon:29695</td>\n",
       "      <td>Phragmites australis</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_29695</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1944</th>\n",
       "      <td>salvage</td>\n",
       "      <td>protist</td>\n",
       "      <td>protist</td>\n",
       "      <td>protist</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C77914</td>\n",
       "      <td>Protista</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C77914</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2006</th>\n",
       "      <td>salvage</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00000892</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENM</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000892</td>\n",
       "      <td>enm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2133</th>\n",
       "      <td>salvage</td>\n",
       "      <td>supraglacial lake water</td>\n",
       "      <td>supraglacial lake water</td>\n",
       "      <td>supraglacial lake</td>\n",
       "      <td>1</td>\n",
       "      <td>0.134</td>\n",
       "      <td>ENVO:03000035</td>\n",
       "      <td>supraglacial lake</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000035</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2255</th>\n",
       "      <td>salvage</td>\n",
       "      <td>tundra</td>\n",
       "      <td>tundra</td>\n",
       "      <td>tundra</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00000112</td>\n",
       "      <td>tundra</td>\n",
       "      <td>1</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000112</td>\n",
       "      <td>micro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2387</th>\n",
       "      <td>salvage</td>\n",
       "      <td>zooplankton</td>\n",
       "      <td>zooplankton</td>\n",
       "      <td>Zooplankton</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>OMIT:0015869</td>\n",
       "      <td>Zooplankton</td>\n",
       "      <td>1</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0015869</td>\n",
       "      <td>omit</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     category                               raw  \\\n",
       "17    salvage                                 0   \n",
       "78    salvage                                15   \n",
       "111   salvage                            arctic   \n",
       "169   salvage              brackish water river   \n",
       "194   salvage        brine pool interface layer   \n",
       "307   salvage                     coastal ocean   \n",
       "527   salvage                deep mediterranean   \n",
       "801   salvage                 envo marine biome   \n",
       "830   salvage                         estuarine   \n",
       "914   salvage                  for marine biome   \n",
       "955   salvage              freshwater reservoir   \n",
       "960   salvage                              gulf   \n",
       "1130  salvage                  large lake biome   \n",
       "1331  salvage                 marine metagenome   \n",
       "1360  salvage                    marine pelagic   \n",
       "1445  salvage                      marine water   \n",
       "1453  salvage             meditterean sea biome   \n",
       "1534  salvage                     microorganism   \n",
       "1664  salvage                                na   \n",
       "1694  salvage                       ne atlantic   \n",
       "1721  salvage                    not applicable   \n",
       "1827  salvage                      oligotrophic   \n",
       "1860  salvage                     pacific coast   \n",
       "1864  salvage  phragmites australis rhizosphere   \n",
       "1944  salvage                           protist   \n",
       "2006  salvage                  small lake biome   \n",
       "2133  salvage           supraglacial lake water   \n",
       "2255  salvage                            tundra   \n",
       "2387  salvage                       zooplankton   \n",
       "\n",
       "                                 query                     name  \\\n",
       "17                                   0                        0   \n",
       "78                                  15                       15   \n",
       "111                             arctic                   Arctic   \n",
       "169               brackish water river           brackish water   \n",
       "194         brine pool interface layer          interface layer   \n",
       "307                      coastal ocean      coastal ocean water   \n",
       "527                 deep mediterranean            Mediterranean   \n",
       "801                  envo marine biome             marine biome   \n",
       "830                          estuarine            estuarine mud   \n",
       "914                   for marine biome             marine biome   \n",
       "955               freshwater reservoir               freshwater   \n",
       "960                               gulf                     gulf   \n",
       "1130                  large lake biome         large lake biome   \n",
       "1331                 marine metagenome        marine metagenome   \n",
       "1360                    marine pelagic      marine pelagic zone   \n",
       "1445                      marine water         marine waterbody   \n",
       "1453             meditterean sea biome  mediterranean sea biome   \n",
       "1534                     microorganism            Microorganism   \n",
       "1664                                na                       NA   \n",
       "1694                       ne atlantic        ne atlantic ocean   \n",
       "1721                    not applicable           Not applicable   \n",
       "1827                      oligotrophic        oligotrophic lake   \n",
       "1860                     pacific coast     Pacific Coast States   \n",
       "1864  phragmites australis rhizosphere     Phragmites australis   \n",
       "1944                           protist                  protist   \n",
       "2006                  small lake biome         small lake biome   \n",
       "2133           supraglacial lake water        supraglacial lake   \n",
       "2255                            tundra                   tundra   \n",
       "2387                       zooplankton              Zooplankton   \n",
       "\n",
       "      string_dist_rank  string_dist            obo_id  \\\n",
       "17                   1        0.000      NCIT:C159767   \n",
       "78                   1        0.000      NCIT:C113429   \n",
       "111                  1        0.000       NCIT:C44738   \n",
       "169                  1        0.153     ENVO:00002019   \n",
       "194                  1        0.192     ENVO:01001684   \n",
       "307                  1        0.184     ENVO:00002150   \n",
       "527                  1        0.142       LBO:0001043   \n",
       "801                  1        0.171     ENVO:00000447   \n",
       "830                  1        0.184     ENVO:00002160   \n",
       "914                  1        0.144     ENVO:00000447   \n",
       "955                  1        0.200     ENVO:00002011   \n",
       "960                  1        0.000      CHEBI:152556   \n",
       "1130                 1        0.000     ENVO:00000891   \n",
       "1331                 1        0.000  NCBITaxon:408172   \n",
       "1360                 1        0.132     ENVO:00000208   \n",
       "1445                 1        0.144     ENVO:00001999   \n",
       "1453                 1        0.113     ENVO:01000047   \n",
       "1534                 1        0.000       NCIT:C14329   \n",
       "1664                 1        0.000    CO_357:2100001   \n",
       "1694                 1        0.180      GAZ:00013759   \n",
       "1721                 1        0.000       NCIT:C48660   \n",
       "1827                 1        0.171     ENVO:01000774   \n",
       "1860                 1        0.181       NCIT:C43455   \n",
       "1864                 1        0.175   NCBITaxon:29695   \n",
       "1944                 1        0.000       NCIT:C77914   \n",
       "2006                 1        0.000     ENVO:00000892   \n",
       "2133                 1        0.134     ENVO:03000035   \n",
       "2255                 1        0.000     ENVO:00000112   \n",
       "2387                 1        0.000      OMIT:0015869   \n",
       "\n",
       "                           label  search_rank ontology_prefix  \\\n",
       "17    Zero Full Term Pregnancies            3            NCIT   \n",
       "78                       Fifteen            5            NCIT   \n",
       "111                       Arctic            1            NCIT   \n",
       "169               brackish water            1            ENVO   \n",
       "194              interface layer            3            ENVO   \n",
       "307            coastal sea water            1            ENVO   \n",
       "527                Mediterranean            2             LBO   \n",
       "801                 marine biome            1             ENM   \n",
       "830                estuarine mud            1            ENVO   \n",
       "914                 marine biome            4            ENVO   \n",
       "955                  fresh water            5            ENVO   \n",
       "960               D-gulofuranose            1           CHEBI   \n",
       "1130            large lake biome            1             ENM   \n",
       "1331           marine metagenome            1       NCBITAXON   \n",
       "1360         marine pelagic zone            3            ENVO   \n",
       "1445           marine water body            5             RBO   \n",
       "1453     mediterranean sea biome            5            ENVO   \n",
       "1534               Microorganism            1            NCIT   \n",
       "1664                          NA            1          CO_357   \n",
       "1694    Northeast Atlantic Ocean            1             GAZ   \n",
       "1721              Not Applicable            1            NCIT   \n",
       "1827           oligotrophic lake            3            ENVO   \n",
       "1860        Pacific Coast States            4            NCIT   \n",
       "1864        Phragmites australis            1       NCBITAXON   \n",
       "1944                    Protista            1            NCIT   \n",
       "2006            small lake biome            1             ENM   \n",
       "2133           supraglacial lake            1            ENVO   \n",
       "2255                      tundra            1           MICRO   \n",
       "2387                 Zooplankton            1            OMIT   \n",
       "\n",
       "                  scope        type  \\\n",
       "17      hasExactSynonym        None   \n",
       "78              synonym     synonym   \n",
       "111               label       label   \n",
       "169               label       label   \n",
       "194               label       label   \n",
       "307   has_exact_synonym  annotation   \n",
       "527               label       label   \n",
       "801               label       label   \n",
       "830               label       label   \n",
       "914               label       label   \n",
       "955             synonym     synonym   \n",
       "960             synonym     synonym   \n",
       "1130              label       label   \n",
       "1331              label       label   \n",
       "1360              label       label   \n",
       "1445            synonym     synonym   \n",
       "1453              label       label   \n",
       "1534              label       label   \n",
       "1664              label       label   \n",
       "1694            synonym     synonym   \n",
       "1721    hasExactSynonym        None   \n",
       "1827              label       label   \n",
       "1860              label       label   \n",
       "1864              label       label   \n",
       "1944            synonym     synonym   \n",
       "2006              label       label   \n",
       "2133              label       label   \n",
       "2255              label       label   \n",
       "2387              label       label   \n",
       "\n",
       "                                                  iri ontology_name  \n",
       "17        http://purl.obolibrary.org/obo/NCIT_C159767          ncit  \n",
       "78        http://purl.obolibrary.org/obo/NCIT_C113429          ncit  \n",
       "111        http://purl.obolibrary.org/obo/NCIT_C44738          ncit  \n",
       "169      http://purl.obolibrary.org/obo/ENVO_00002019          envo  \n",
       "194      http://purl.obolibrary.org/obo/ENVO_01001684          envo  \n",
       "307      http://purl.obolibrary.org/obo/ENVO_00002150          envo  \n",
       "527        http://purl.obolibrary.org/obo/LBO_0001043           lbo  \n",
       "801      http://purl.obolibrary.org/obo/ENVO_00000447           enm  \n",
       "830      http://purl.obolibrary.org/obo/ENVO_00002160          envo  \n",
       "914      http://purl.obolibrary.org/obo/ENVO_00000447          envo  \n",
       "955      http://purl.obolibrary.org/obo/ENVO_00002011          envo  \n",
       "960       http://purl.obolibrary.org/obo/CHEBI_152556         chebi  \n",
       "1130     http://purl.obolibrary.org/obo/ENVO_00000891           enm  \n",
       "1331  http://purl.obolibrary.org/obo/NCBITaxon_408172     ncbitaxon  \n",
       "1360     http://purl.obolibrary.org/obo/ENVO_00000208          envo  \n",
       "1445     http://purl.obolibrary.org/obo/ENVO_00001999           rbo  \n",
       "1453     http://purl.obolibrary.org/obo/ENVO_01000047          envo  \n",
       "1534       http://purl.obolibrary.org/obo/NCIT_C14329          ncit  \n",
       "1664   http://www.cropontology.org/rdf/CO_357:2100001        co_357  \n",
       "1694      http://purl.obolibrary.org/obo/GAZ_00013759           gaz  \n",
       "1721       http://purl.obolibrary.org/obo/NCIT_C48660          ncit  \n",
       "1827     http://purl.obolibrary.org/obo/ENVO_01000774          envo  \n",
       "1860       http://purl.obolibrary.org/obo/NCIT_C43455          ncit  \n",
       "1864   http://purl.obolibrary.org/obo/NCBITaxon_29695     ncbitaxon  \n",
       "1944       http://purl.obolibrary.org/obo/NCIT_C77914          ncit  \n",
       "2006     http://purl.obolibrary.org/obo/ENVO_00000892           enm  \n",
       "2133     http://purl.obolibrary.org/obo/ENVO_03000035          envo  \n",
       "2255     http://purl.obolibrary.org/obo/ENVO_00000112         micro  \n",
       "2387      http://purl.obolibrary.org/obo/OMIT_0015869          omit  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "salvage_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ea200e4",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1473206f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ols_category</th>\n",
       "      <th>ols_raw</th>\n",
       "      <th>ols_query</th>\n",
       "      <th>ols_name</th>\n",
       "      <th>ols_string_dist_rank</th>\n",
       "      <th>ols_string_dist</th>\n",
       "      <th>ols_obo_id</th>\n",
       "      <th>ols_label</th>\n",
       "      <th>ols_search_rank</th>\n",
       "      <th>ols_ontology_prefix</th>\n",
       "      <th>ols_scope</th>\n",
       "      <th>ols_type</th>\n",
       "      <th>ols_iri</th>\n",
       "      <th>ols_ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>arabian sea</td>\n",
       "      <td>arabian sea</td>\n",
       "      <td>Arabian Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00002457</td>\n",
       "      <td>Arabian Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00002457</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>atlantic ocean</td>\n",
       "      <td>atlantic ocean</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00000344</td>\n",
       "      <td>Atlantic Ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00000344</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>subpolar coniferous forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000250</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>caribbean sea</td>\n",
       "      <td>caribbean sea</td>\n",
       "      <td>Caribbean Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00002820</td>\n",
       "      <td>Caribbean Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00002820</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>coastal water</td>\n",
       "      <td>coastal water</td>\n",
       "      <td>coastal water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002150</td>\n",
       "      <td>coastal sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_broad_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002150</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:2759</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2759</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>forest</td>\n",
       "      <td>forest</td>\n",
       "      <td>Forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00454366</td>\n",
       "      <td>Forest</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00454366</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002011</td>\n",
       "      <td>fresh water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002011</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:22224982</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_22224982</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>hydrothermal vents</td>\n",
       "      <td>hydrothermal vents</td>\n",
       "      <td>hydrothermal vent</td>\n",
       "      <td>1</td>\n",
       "      <td>0.030</td>\n",
       "      <td>ENVO:00000215</td>\n",
       "      <td>hydrothermal vent</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000215</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>marginal sea</td>\n",
       "      <td>marginal sea</td>\n",
       "      <td>marginal sea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td>sea</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000016</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>724</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>marine brine pool basin</td>\n",
       "      <td>marine brine pool basin</td>\n",
       "      <td>marine brine pool</td>\n",
       "      <td>1</td>\n",
       "      <td>0.081</td>\n",
       "      <td>ENVO:00000369</td>\n",
       "      <td>brine pool</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000369</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>mediterranean sea</td>\n",
       "      <td>mediterranean sea</td>\n",
       "      <td>Mediterranean Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00002287</td>\n",
       "      <td>Mediterranean Sea</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00002287</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>1</td>\n",
       "      <td>0.039</td>\n",
       "      <td>ENVO:01000008</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000008</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>890</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>ocean water</td>\n",
       "      <td>ocean water</td>\n",
       "      <td>ocean water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002149</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>seawater</td>\n",
       "      <td>seawater</td>\n",
       "      <td>seawater</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002149</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1072</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>southern ocean</td>\n",
       "      <td>southern ocean</td>\n",
       "      <td>Southern Ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00000373</td>\n",
       "      <td>Southern Ocean</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00000373</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1132</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>temperate mediterannean sea biome</td>\n",
       "      <td>temperate mediterannean sea biome</td>\n",
       "      <td>temperate mediterranean sea biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.021</td>\n",
       "      <td>ENVO:01000857</td>\n",
       "      <td>temperate mediterranean sea biome</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000857</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1190</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>water river</td>\n",
       "      <td>water river</td>\n",
       "      <td>river water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.083</td>\n",
       "      <td>ENVO:01000599</td>\n",
       "      <td>river water</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000599</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ols_category                            ols_raw  \\\n",
       "0     env_broad_scale                                      \n",
       "48    env_broad_scale                        arabian sea   \n",
       "81    env_broad_scale                     atlantic ocean   \n",
       "101   env_broad_scale                      boreal forest   \n",
       "133   env_broad_scale                      caribbean sea   \n",
       "213   env_broad_scale                      coastal water   \n",
       "417   env_broad_scale                         eukaryotes   \n",
       "508   env_broad_scale                             forest   \n",
       "520   env_broad_scale                         freshwater   \n",
       "585   env_broad_scale                        hot springs   \n",
       "594   env_broad_scale                 hydrothermal vents   \n",
       "678   env_broad_scale                       marginal sea   \n",
       "724   env_broad_scale            marine brine pool basin   \n",
       "768   env_broad_scale                  mediterranean sea   \n",
       "810   env_broad_scale                     microbial mats   \n",
       "890   env_broad_scale                        ocean water   \n",
       "1032  env_broad_scale                           seawater   \n",
       "1072  env_broad_scale                     southern ocean   \n",
       "1132  env_broad_scale  temperate mediterannean sea biome   \n",
       "1190  env_broad_scale                        water river   \n",
       "\n",
       "                              ols_query                           ols_name  \\\n",
       "0                                                                            \n",
       "48                          arabian sea                        Arabian Sea   \n",
       "81                       atlantic ocean                     Atlantic Ocean   \n",
       "101                       boreal forest                      boreal forest   \n",
       "133                       caribbean sea                      Caribbean Sea   \n",
       "213                       coastal water                      coastal water   \n",
       "417                          eukaryotes                         eukaryotes   \n",
       "508                              forest                             Forest   \n",
       "520                          freshwater                         freshwater   \n",
       "585                         hot springs                        Hot Springs   \n",
       "594                  hydrothermal vents                  hydrothermal vent   \n",
       "678                        marginal sea                       marginal sea   \n",
       "724             marine brine pool basin                  marine brine pool   \n",
       "768                   mediterranean sea                  Mediterranean Sea   \n",
       "810                      microbial mats                      microbial mat   \n",
       "890                         ocean water                        ocean water   \n",
       "1032                           seawater                           seawater   \n",
       "1072                     southern ocean                     Southern Ocean   \n",
       "1132  temperate mediterannean sea biome  temperate mediterranean sea biome   \n",
       "1190                        water river                        river water   \n",
       "\n",
       "      ols_string_dist_rank  ols_string_dist      ols_obo_id  \\\n",
       "0                        1            0.000                   \n",
       "48                       1            0.000    GAZ:00002457   \n",
       "81                       1            0.000    GAZ:00000344   \n",
       "101                      1            0.000   ENVO:01000250   \n",
       "133                      1            0.000    GAZ:00002820   \n",
       "213                      1            0.000   ENVO:00002150   \n",
       "417                      1            0.000  NCBITaxon:2759   \n",
       "508                      1            0.000    GAZ:00454366   \n",
       "520                      1            0.000   ENVO:00002011   \n",
       "585                      1            0.000    GAZ:22224982   \n",
       "594                      1            0.030   ENVO:00000215   \n",
       "678                      1            0.000   ENVO:00000016   \n",
       "724                      1            0.081   ENVO:00000369   \n",
       "768                      1            0.000    GAZ:00002287   \n",
       "810                      1            0.039   ENVO:01000008   \n",
       "890                      1            0.000   ENVO:00002149   \n",
       "1032                     1            0.000   ENVO:00002149   \n",
       "1072                     1            0.000    GAZ:00000373   \n",
       "1132                     1            0.021   ENVO:01000857   \n",
       "1190                     1            0.083   ENVO:01000599   \n",
       "\n",
       "                              ols_label  ols_search_rank ols_ontology_prefix  \\\n",
       "0                                                      1                       \n",
       "48                          Arabian Sea                1                 GAZ   \n",
       "81                       Atlantic Ocean                1                 GAZ   \n",
       "101    subpolar coniferous forest biome                1                ENVO   \n",
       "133                       Caribbean Sea                1                 GAZ   \n",
       "213                   coastal sea water                1                ENVO   \n",
       "417                           Eukaryota                1                ENVO   \n",
       "508                              Forest                1                 GAZ   \n",
       "520                         fresh water                1                ENVO   \n",
       "585                         Hot Springs                1                 GAZ   \n",
       "594                   hydrothermal vent                4                ENVO   \n",
       "678                                 sea                1                ENVO   \n",
       "724                          brine pool                1                ENVO   \n",
       "768                   Mediterranean Sea                1                 GAZ   \n",
       "810                       microbial mat                1                ENVO   \n",
       "890                           sea water                1                ENVO   \n",
       "1032                          sea water                1                ENVO   \n",
       "1072                     Southern Ocean                1                 GAZ   \n",
       "1132  temperate mediterranean sea biome                2                ENVO   \n",
       "1190                        river water                3                ENVO   \n",
       "\n",
       "               ols_scope    ols_type  \\\n",
       "0                    NaN         NaN   \n",
       "48                 label       label   \n",
       "81                 label       label   \n",
       "101   has_narrow_synonym  annotation   \n",
       "133                label       label   \n",
       "213    has_broad_synonym  annotation   \n",
       "417              synonym     synonym   \n",
       "508                label       label   \n",
       "520              synonym     synonym   \n",
       "585                label       label   \n",
       "594                label       label   \n",
       "678   has_narrow_synonym  annotation   \n",
       "724              synonym     synonym   \n",
       "768                label       label   \n",
       "810                label       label   \n",
       "890              synonym     synonym   \n",
       "1032             synonym     synonym   \n",
       "1072               label       label   \n",
       "1132               label       label   \n",
       "1190               label       label   \n",
       "\n",
       "                                            ols_iri ols_ontology_name  \n",
       "0                                                                      \n",
       "48      http://purl.obolibrary.org/obo/GAZ_00002457               gaz  \n",
       "81      http://purl.obolibrary.org/obo/GAZ_00000344               gaz  \n",
       "101    http://purl.obolibrary.org/obo/ENVO_01000250              envo  \n",
       "133     http://purl.obolibrary.org/obo/GAZ_00002820               gaz  \n",
       "213    http://purl.obolibrary.org/obo/ENVO_00002150              envo  \n",
       "417   http://purl.obolibrary.org/obo/NCBITaxon_2759              envo  \n",
       "508     http://purl.obolibrary.org/obo/GAZ_00454366               gaz  \n",
       "520    http://purl.obolibrary.org/obo/ENVO_00002011              envo  \n",
       "585     http://purl.obolibrary.org/obo/GAZ_22224982               gaz  \n",
       "594    http://purl.obolibrary.org/obo/ENVO_00000215              envo  \n",
       "678    http://purl.obolibrary.org/obo/ENVO_00000016              envo  \n",
       "724    http://purl.obolibrary.org/obo/ENVO_00000369              envo  \n",
       "768     http://purl.obolibrary.org/obo/GAZ_00002287               gaz  \n",
       "810    http://purl.obolibrary.org/obo/ENVO_01000008              envo  \n",
       "890    http://purl.obolibrary.org/obo/ENVO_00002149              envo  \n",
       "1032   http://purl.obolibrary.org/obo/ENVO_00002149              envo  \n",
       "1072    http://purl.obolibrary.org/obo/GAZ_00000373               gaz  \n",
       "1132   http://purl.obolibrary.org/obo/ENVO_01000857              envo  \n",
       "1190   http://purl.obolibrary.org/obo/ENVO_01000599              envo  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_best_acceptable.columns = 'ols_' + my_best_acceptable.columns\n",
    "my_best_acceptable.to_sql('best_acceptable', biosample_cnx, if_exists='replace', index=False)\n",
    "my_best_acceptable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ac6219e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "big_merge = mapping_candidates.merge(my_best_acceptable, how='outer', left_on='remaining_tidied', right_on='ols_raw')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "cb3982a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "big_merge.to_sql('big_merge', biosample_cnx, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d2c242b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = ~ big_merge['id_or_lab_ok'] & ~ big_merge['ols_obo_id'].eq('') & ~ big_merge['ols_obo_id'].isna()\n",
    "\n",
    "replacement = big_merge.loc[flag, 'ols_obo_id']\n",
    "\n",
    "big_merge.loc[flag, 'consensus_id'] = replacement\n",
    "\n",
    "replacement = big_merge.loc[flag, 'ols_label']\n",
    "\n",
    "big_merge.loc[flag, 'consensus_lab'] = replacement\n",
    "\n",
    "replacement = big_merge.loc[flag, 'consensus_lab'] + ' [' + big_merge.loc[flag, 'consensus_id'] + ']'\n",
    "\n",
    "big_merge.loc[flag, 'assembled_consensus'] = replacement\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b333c111",
   "metadata": {},
   "outputs": [],
   "source": [
    "big_merge.to_sql('big_merge', biosample_cnx, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e772701",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
