{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d34414a6",
   "metadata": {},
   "source": [
    "- don't forget to consult \"model\" too\n",
    "- uncultured vs metagenome tax id?\n",
    "- what about \"gold\" or \"path\"? WON\"T FIND IN NCBI BIOSAMPLE\n",
    "- are there samples with a host_taxid and a (sample) taxonomy_id from \"unclassified\"?\n",
    "- \"x biome\" vs \"x\"\n",
    "- be careful with GAZ matches\n",
    "- take note of non-local salvage hits\n",
    "- what to do about hits against obsolete classes... does OLS provide a mechanism to avoid?\n",
    "- is it worth trimming and concatenating \"x1\", \"x2\", \"x3\", where x is a mappable term? numbers may be low.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fac2962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy\n",
    "import scoped_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3300da64",
   "metadata": {},
   "source": [
    "## User-provided data\n",
    "See repo README for notes on setting up SQLite databases of OBO ontologies with semantic-sql, relation-graph and rdftab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5684af0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://www.ncbi.nlm.nih.gov/biosample/docs/packages/?format=xml\n",
    "# see also https://www.ncbi.nlm.nih.gov/biosample/docs/packages/\n",
    "biosample_packages_file = \"../target/biosample_packages.xml\"\n",
    "\n",
    "# from ftp://ftp.ncbi.nlm.nih.gov//biosample/biosample_set.xml.gz\n",
    "# via harmonized_table.db.gz\n",
    "# in https://drive.google.com/drive/u/0/folders/1eL0v0stoduahjDpoDJIk3z2pJBAU4b2Y\n",
    "biosample_sqlite_file   = \"../target/harmonized_table.db\"\n",
    "\n",
    "# see readme\n",
    "ncbitaxon_sqlite_file   = \"../semantic-sql/db/ncbitaxon.db\"\n",
    "envo_sqlite_file        = \"../semantic-sql/db/envo.db\"\n",
    "\n",
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)\n",
    "ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)\n",
    "envo_cnx      = sqlite3.connect(envo_sqlite_file)\n",
    "\n",
    "target_onto_prefix = \"ENVO\"\n",
    "chars_to_whiteout  = \"._-\"\n",
    "my_query_fields    = \"\"  # OLS weighted default\n",
    "my_row_req         = 3\n",
    "my_string_dist_arg = 2\n",
    "my_max_string_dist = 0.1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca683f5f",
   "metadata": {},
   "source": [
    "## it nice to see everything accounted for\n",
    "but we can prioritize NMDC for now:\n",
    "\n",
    "Soil 15,777\n",
    "Sediment 7,147\n",
    "Plant-associated 3,142\n",
    "\n",
    "Could some of these \"no environmental package\" mappings be losing important granularity?\n",
    "\n",
    "map `None` and '' to \"no environmental package\"?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "44b5188b",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_overrides = {\n",
    "    \"built environment\": \"built\",\n",
    "    \"misc environment\": \"miscellaneous\",\n",
    "    \"missing\": \"no environmental package\",\n",
    "    \"unknown\": \"no environmental package\",\n",
    "    \"default\": \"no environmental package\",\n",
    "    \"unspecified\": \"no environmental package\",\n",
    "    \"not available\": \"no environmental package\",\n",
    "    \"not collected\": \"no environmental package\",\n",
    "    \"miscellaneous natural or artificial environment\": \"miscellaneous\",\n",
    "    \"not applicable\": \"no environmental package\",\n",
    "    \"soil-associated\": \"soil\",\n",
    "    \"soil associated\": \"soil\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2d7f8bf",
   "metadata": {},
   "source": [
    "## What Biosample field should be mapped to ontology classes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2b79c519",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_col_to_map = \"env_broad_scale\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87e3b6f1",
   "metadata": {},
   "source": [
    "## How should the mapping effort be scoped?\n",
    "Additionally taxonomic filters may be applied below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ec160085",
   "metadata": {},
   "outputs": [],
   "source": [
    "# scoping_col = \"env_package_normalized.rt_override\"\n",
    "scoping_col = \"env_package_normalized.EnvPackage\"\n",
    "scoping_value = \"soil\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a568c1",
   "metadata": {},
   "source": [
    "## Settings for manual review of OLS search-based results\n",
    "Results based on merging Biosample annotations to ontology classes by label or embedded ID are not exported for review at this time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "20cdb474",
   "metadata": {},
   "outputs": [],
   "source": [
    "ols_review_file = \"../local/ols_review.tsv\"\n",
    "ols_review_seperator = \"\\t\"\n",
    "strategy_col = \"strategy\"\n",
    "include_col = \"include\"\n",
    "first_pass_include_val = True\n",
    "first_pass_strategy_val = \"env_braod_scale vs envo and gaz @ 0.1\"\n",
    "salvage_include_val = False\n",
    "salvage_strategy_val = \"env_braod_scale vs all of ols @ 0.2\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7edfbe0",
   "metadata": {},
   "source": [
    "## Settings for SSSOM output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9d71fa5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sssom_subject_prefix = \"biosample_ebs\"\n",
    "sssom_file = \"biosample_ebs_water_packages_unclassified_taxa_sssom.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c54707",
   "metadata": {},
   "source": [
    "## Print a sample of the data we're working with\n",
    "Specifically, INSDC/NCBI Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "caff574e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_package</th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMN00000002</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445970</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BIOSAMPLE:SAMN00000003</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445972</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIOSAMPLE:SAMN00000004</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>449673</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BIOSAMPLE:SAMN00000005</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>6526</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BIOSAMPLE:SAMN00000006</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>9483</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BIOSAMPLE:SAMN00000007</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445974</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BIOSAMPLE:SAMN00000008</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>411461</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BIOSAMPLE:SAMN00000009</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>13616</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BIOSAMPLE:SAMN00000010</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>451639</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BIOSAMPLE:SAMN00000011</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>451638</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id env_package      package  \\\n",
       "0  BIOSAMPLE:SAMN00000002     missing  MIGS.ba.5.0   \n",
       "1  BIOSAMPLE:SAMN00000003     missing  MIGS.ba.5.0   \n",
       "2  BIOSAMPLE:SAMN00000004     missing  MIGS.ba.5.0   \n",
       "3  BIOSAMPLE:SAMN00000005        None  Generic.1.0   \n",
       "4  BIOSAMPLE:SAMN00000006        None  Generic.1.0   \n",
       "5  BIOSAMPLE:SAMN00000007     missing  MIGS.ba.5.0   \n",
       "6  BIOSAMPLE:SAMN00000008     missing  MIGS.ba.5.0   \n",
       "7  BIOSAMPLE:SAMN00000009        None  Generic.1.0   \n",
       "8  BIOSAMPLE:SAMN00000010        None  Generic.1.0   \n",
       "9  BIOSAMPLE:SAMN00000011        None  Generic.1.0   \n",
       "\n",
       "                                   package_name host_taxid taxonomy_id  \\\n",
       "0  MIGS: cultured bacteria/archaea; version 5.0       9606      445970   \n",
       "1  MIGS: cultured bacteria/archaea; version 5.0       9606      445972   \n",
       "2  MIGS: cultured bacteria/archaea; version 5.0       9606      449673   \n",
       "3                                       Generic       None        6526   \n",
       "4                                       Generic       None        9483   \n",
       "5  MIGS: cultured bacteria/archaea; version 5.0       9606      445974   \n",
       "6  MIGS: cultured bacteria/archaea; version 5.0       9606      411461   \n",
       "7                                       Generic       None       13616   \n",
       "8                                       Generic       None      451639   \n",
       "9                                       Generic       None      451638   \n",
       "\n",
       "                     env_broad_scale  \\\n",
       "0  terrestrial biome [ENVO:00000446]   \n",
       "1  terrestrial biome [ENVO:00000446]   \n",
       "2  terrestrial biome [ENVO:00000446]   \n",
       "3                               None   \n",
       "4                               None   \n",
       "5  terrestrial biome [ENVO:00000446]   \n",
       "6  terrestrial biome [ENVO:00000446]   \n",
       "7                               None   \n",
       "8                               None   \n",
       "9                               None   \n",
       "\n",
       "                            env_local_scale  \\\n",
       "0  human-associated habitat [ENVO:00009003]   \n",
       "1  human-associated habitat [ENVO:00009003]   \n",
       "2  human-associated habitat [ENVO:00009003]   \n",
       "3                                      None   \n",
       "4                                      None   \n",
       "5  human-associated habitat [ENVO:00009003]   \n",
       "6  human-associated habitat [ENVO:00009003]   \n",
       "7                                      None   \n",
       "8                                      None   \n",
       "9                                      None   \n",
       "\n",
       "                           env_medium  \n",
       "0  biological product [ENVO:02000043]  \n",
       "1  biological product [ENVO:02000043]  \n",
       "2  biological product [ENVO:02000043]  \n",
       "3                                None  \n",
       "4                                None  \n",
       "5  biological product [ENVO:02000043]  \n",
       "6  biological product [ENVO:02000043]  \n",
       "7                                None  \n",
       "8                                None  \n",
       "9                                None  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    id,\n",
    "    env_package,\n",
    "    package,\n",
    "    package_name,\n",
    "    host_taxid,\n",
    "    taxonomy_id,\n",
    "    env_broad_scale,\n",
    "    env_local_scale,\n",
    "    env_medium\n",
    "    from biosample b\n",
    "limit 10\n",
    "\"\"\"\n",
    "biosample_first_ten = pd.read_sql(q, biosample_cnx)\n",
    "biosample_first_ten"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef577997",
   "metadata": {},
   "source": [
    "## Get the canonical checklist and package terms from NCBI\n",
    "\n",
    "Unfortunately it doesn't do a very good job of differentiating checklists (MIMAG, MIMARKS, etc.) from packages (soil, water, etc.)\n",
    "\n",
    "_What about .ba, .euk, etc?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c6cc9596",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>DisplayName</th>\n",
       "      <th>ShortName</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>EnvPackageDisplay</th>\n",
       "      <th>NotAppropriateFor</th>\n",
       "      <th>Description</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2.cl.1.0</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated; versi...</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for SARS-CoV-2 samples that are relevant t...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>Pathogen: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Clinical or host-associated pathogen</td>\n",
       "      <td>SAMN02928182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>Pathogen: environmental/food/other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Environmental, food or other pathogen</td>\n",
       "      <td>SAMN02730065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Microbe.1.0</td>\n",
       "      <td>Microbe; version 1.0</td>\n",
       "      <td>Microbe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Use for bacteria or other unicellular microbes...</td>\n",
       "      <td>SAMN02911891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>MIUVIG.plant-associated.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, plant-assoc...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>MIUVIG.sediment.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, sediment; v...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>MIUVIG.soil.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, soil; versi...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>MIUVIG.wastewater.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, wastewater;...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>MIUVIG.water.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, water; vers...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Name  \\\n",
       "0                    Generic.1.0   \n",
       "1              SARS-CoV-2.cl.1.0   \n",
       "2                Pathogen.cl.1.0   \n",
       "3               Pathogen.env.1.0   \n",
       "4                    Microbe.1.0   \n",
       "..                           ...   \n",
       "149  MIUVIG.plant-associated.5.0   \n",
       "150          MIUVIG.sediment.5.0   \n",
       "151              MIUVIG.soil.5.0   \n",
       "152        MIUVIG.wastewater.5.0   \n",
       "153             MIUVIG.water.5.0   \n",
       "\n",
       "                                           DisplayName  \\\n",
       "0                                              Generic   \n",
       "1    SARS-CoV-2: clinical or host-associated; versi...   \n",
       "2    Pathogen: clinical or host-associated; version...   \n",
       "3      Pathogen: environmental/food/other; version 1.0   \n",
       "4                                 Microbe; version 1.0   \n",
       "..                                                 ...   \n",
       "149  MIUVIG: uncultivated virus genome, plant-assoc...   \n",
       "150  MIUVIG: uncultivated virus genome, sediment; v...   \n",
       "151  MIUVIG: uncultivated virus genome, soil; versi...   \n",
       "152  MIUVIG: uncultivated virus genome, wastewater;...   \n",
       "153  MIUVIG: uncultivated virus genome, water; vers...   \n",
       "\n",
       "                                   ShortName        EnvPackage  \\\n",
       "0                                                                \n",
       "1    SARS-CoV-2: clinical or host-associated                     \n",
       "2      Pathogen: clinical or host-associated                     \n",
       "3         Pathogen: environmental/food/other                     \n",
       "4                                    Microbe                     \n",
       "..                                       ...               ...   \n",
       "149         MIUVIG Uncultivated Virus Genome  plant-associated   \n",
       "150         MIUVIG Uncultivated Virus Genome          sediment   \n",
       "151         MIUVIG Uncultivated Virus Genome              soil   \n",
       "152         MIUVIG Uncultivated Virus Genome        wastewater   \n",
       "153         MIUVIG Uncultivated Virus Genome             water   \n",
       "\n",
       "    EnvPackageDisplay                 NotAppropriateFor  \\\n",
       "0                                                         \n",
       "1                      wgs_single;wgs_batch;wgs_diploid   \n",
       "2                                                         \n",
       "3                                                         \n",
       "4                                                         \n",
       "..                ...                               ...   \n",
       "149  plant-associated  wgs_single;wgs_batch;wgs_diploid   \n",
       "150          sediment  wgs_single;wgs_batch;wgs_diploid   \n",
       "151              soil  wgs_single;wgs_batch;wgs_diploid   \n",
       "152        wastewater  wgs_single;wgs_batch;wgs_diploid   \n",
       "153             water  wgs_single;wgs_batch;wgs_diploid   \n",
       "\n",
       "                                           Description       Example  \n",
       "0                                              Generic                \n",
       "1    Use for SARS-CoV-2 samples that are relevant t...                \n",
       "2                 Clinical or host-associated pathogen  SAMN02928182  \n",
       "3                Environmental, food or other pathogen  SAMN02730065  \n",
       "4    Use for bacteria or other unicellular microbes...  SAMN02911891  \n",
       "..                                                 ...           ...  \n",
       "149  Use for uncultivated virus genome identified i...                \n",
       "150  Use for uncultivated virus genome identified i...                \n",
       "151  Use for uncultivated virus genome identified i...                \n",
       "152  Use for uncultivated virus genome identified i...                \n",
       "153  Use for uncultivated virus genome identified i...                \n",
       "\n",
       "[154 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "package_dictionary = scoped_mapping.get_package_dictionary(biosample_packages_file)\n",
    "package_dictionary.to_sql(\n",
    "    \"package_dictionary\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "package_dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa900ad8",
   "metadata": {},
   "source": [
    "## Do the Biosample checklist/package fields match any of the canonical values?\n",
    "\n",
    "Start by counting the Biosample rows/records. See XXX notes on extracting this *harmonized* database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0cfa7d0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.302293\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_row_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14300584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   biosample_row_count\n",
       "0             14300584"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select count(*) as biosample_row_count\n",
    "from biosample b\n",
    "\"\"\"\n",
    "[biosample_row_count, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=False\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "biosample_row_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc896a9c",
   "metadata": {},
   "source": [
    "## How many of those rows can be inner-joined with the canonical checklists/packages?\n",
    "Specifically, joining `biosample.package_name` with `package_dictionary.DisplayName`\n",
    "\n",
    "_Note that indices are built as part of the makefile_\n",
    "\n",
    "\n",
    "- create index biosample_package_name_idx on biosample(package_name);\n",
    "- create index package_dictionary_DisplayName_idx on package_dictionary(DisplayName);\n",
    "- create index biosample_package_idx on biosample(package);\n",
    "- create index biosample_p_pn_idx on biosample(package, package_name);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "03a8b90c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-06-09 09:52:57.256159\n",
      "2021-06-09 09:52:58.825005\n",
      "0:00:01.568846\n",
      "0:00:01.568846\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>canonical_package_name_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14300584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   canonical_package_name_count\n",
       "0                      14300584"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    count(*) as canonical_package_name_count\n",
    "from\n",
    "    biosample b\n",
    "inner join package_dictionary pd on\n",
    "    b.package_name = pd.DisplayName\n",
    "\"\"\"\n",
    "[canonical_package_name_count, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "canonical_package_name_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5951b7f1",
   "metadata": {},
   "source": [
    "## Combinations of `package` and `package_name` values in the Biosample dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "83485b99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-06-09 09:52:58.831158\n",
      "2021-06-09 09:53:01.800620\n",
      "0:00:02.969462\n",
      "0:00:02.969462\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Beta-lactamase.1.0</td>\n",
       "      <td>Beta-lactamase; version 1.0</td>\n",
       "      <td>556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>10186430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Human.1.0</td>\n",
       "      <td>Human; version 1.0</td>\n",
       "      <td>368893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Invertebrate.1.0</td>\n",
       "      <td>Invertebrate; version 1.0</td>\n",
       "      <td>131085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>28686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>Model.organism.animal.1.0</td>\n",
       "      <td>Model organism or animal; version 1.0</td>\n",
       "      <td>382980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>511040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>246497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>Plant.1.0</td>\n",
       "      <td>Plant; version 1.0</td>\n",
       "      <td>385939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Virus.1.0</td>\n",
       "      <td>Virus; version 1.0</td>\n",
       "      <td>15560</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       package  \\\n",
       "0           Beta-lactamase.1.0   \n",
       "1                  Generic.1.0   \n",
       "2                    Human.1.0   \n",
       "3             Invertebrate.1.0   \n",
       "4                  MIGS.ba.5.0   \n",
       "..                         ...   \n",
       "124  Model.organism.animal.1.0   \n",
       "125            Pathogen.cl.1.0   \n",
       "126           Pathogen.env.1.0   \n",
       "127                  Plant.1.0   \n",
       "128                  Virus.1.0   \n",
       "\n",
       "                                          package_name     count  \n",
       "0                          Beta-lactamase; version 1.0       556  \n",
       "1                                              Generic  10186430  \n",
       "2                                   Human; version 1.0    368893  \n",
       "3                            Invertebrate; version 1.0    131085  \n",
       "4         MIGS: cultured bacteria/archaea; version 5.0     28686  \n",
       "..                                                 ...       ...  \n",
       "124              Model organism or animal; version 1.0    382980  \n",
       "125  Pathogen: clinical or host-associated; version...    511040  \n",
       "126    Pathogen: environmental/food/other; version 1.0    246497  \n",
       "127                                 Plant; version 1.0    385939  \n",
       "128                                 Virus; version 1.0     15560  \n",
       "\n",
       "[129 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    package,\n",
    "    package_name,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    package ,\n",
    "    package_name\n",
    "order by\n",
    "    package ,\n",
    "    package_name\n",
    "\"\"\"\n",
    "[package_name_combos, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "package_name_combos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e15812f",
   "metadata": {},
   "source": [
    "## What about the Biosample `env_package` values?\n",
    "Are they also a small, highly regular set, like the `package` and `package_name` combinations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d042df82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:01.133042\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count\n",
       "0                              None  14083847\n",
       "1                   host-associated     49254\n",
       "2                         human-gut     47921\n",
       "3                             water     16367\n",
       "4                        human-skin     13706\n",
       "..                              ...       ...\n",
       "87                      env_package         1\n",
       "88                   gut microbiome         1\n",
       "89  marine sediment (ENVO:00002113)         1\n",
       "90                           saliva         1\n",
       "91       sea water, [ENVO:00002149]         1\n",
       "\n",
       "[92 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    env_package\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[env_package_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "env_package_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ab5ec5d",
   "metadata": {},
   "source": [
    "## `env_package` is a mixture of `ENVO` term ids and strings\n",
    "Those strings may or may not be the term's labels. There are many redundancies due to small spelling and punctuation variations.\n",
    "\n",
    "Start by tidying `env_package` etc. values recognized by INSDC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "82ba2408",
   "metadata": {},
   "outputs": [],
   "source": [
    "package_dictionary = scoped_mapping.make_tidy_col(\n",
    "    package_dictionary, \"EnvPackage\", \"eptidy\"\n",
    ")\n",
    "package_dictionary = scoped_mapping.make_tidy_col(\n",
    "    package_dictionary, \"EnvPackageDisplay\", \"epdtidy\"\n",
    ")\n",
    "\n",
    "# update in sqlite\n",
    "package_dictionary.to_sql(\n",
    "    \"package_dictionary\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70be9eb2",
   "metadata": {},
   "source": [
    "## Determine ID patterns for common ontologies, like `ENVO`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5d157f3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.043757\n"
     ]
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    distinct stanza\n",
    "    from statements s\n",
    "where\n",
    "    predicate = 'rdf:type'\n",
    "    and \"object\" = 'owl:Class'\n",
    "    and stanza = subject\"\"\"\n",
    "# include non-envo IDs that come from envo?\n",
    "[ids_from_envo, query_duration] = scoped_mapping.timed_query(q, envo_cnx)\n",
    "print(query_duration)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dcf3bb61",
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_from_envo = scoped_mapping.add_prefix_col(ids_from_envo, \"stanza\", \"prefix\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cb925483",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BFO': 'BFO:\\\\d{7}',\n",
       " 'CARO': 'CARO:\\\\d{7}',\n",
       " 'CHEBI': 'CHEBI:\\\\d{4,6}',\n",
       " 'ENVO': 'ENVO:\\\\d{7,8}',\n",
       " 'FAO': 'FAO:0000001',\n",
       " 'FOODON': 'FOODON:\\\\d{8}',\n",
       " 'GO': 'GO:\\\\d{7}',\n",
       " 'IAO': 'IAO:\\\\d{7}',\n",
       " 'NCBITaxon': 'NCBITaxon:\\\\d+',\n",
       " 'OBI': 'OBI:\\\\d{7}',\n",
       " 'PATO': 'PATO:\\\\d{7}',\n",
       " 'PCO': 'PCO:\\\\d{7}',\n",
       " 'PO': 'PO:\\\\d{7}',\n",
       " 'RO': 'RO:0002577',\n",
       " 'UBERON': 'UBERON:\\\\d{7}'}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_patterns = scoped_mapping.get_multi_term_patterns(ids_from_envo, \"stanza\", \"prefix\")\n",
    "\n",
    "id_patterns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c56612",
   "metadata": {},
   "source": [
    "## Apply some normalization rules to the `env_package` values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cf400a7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>rhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td>env package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>marine sediment ()</td>\n",
       "      <td>marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water,</td>\n",
       "      <td>sea water</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count lhs  \\\n",
       "0                              None  14083847       \n",
       "1                   host-associated     49254       \n",
       "2                         human-gut     47921       \n",
       "3                             water     16367       \n",
       "4                        human-skin     13706       \n",
       "..                              ...       ...  ..   \n",
       "87                      env_package         1       \n",
       "88                   gut microbiome         1       \n",
       "89  marine sediment (ENVO:00002113)         1       \n",
       "90                           saliva         1       \n",
       "91       sea water, [ENVO:00002149]         1       \n",
       "\n",
       "                                rhs        extract    remaining_string  \\\n",
       "0                              None                                      \n",
       "1                   host-associated                    host-associated   \n",
       "2                         human-gut                          human-gut   \n",
       "3                             water                              water   \n",
       "4                        human-skin                         human-skin   \n",
       "..                              ...            ...                 ...   \n",
       "87                      env_package                        env_package   \n",
       "88                   gut microbiome                     gut microbiome   \n",
       "89  marine sediment (ENVO:00002113)  ENVO:00002113  marine sediment ()   \n",
       "90                           saliva                             saliva   \n",
       "91       sea water, [ENVO:00002149]  ENVO:00002149         sea water,    \n",
       "\n",
       "   remaining_tidied  \n",
       "0                    \n",
       "1   host associated  \n",
       "2         human gut  \n",
       "3             water  \n",
       "4        human skin  \n",
       "..              ...  \n",
       "87      env package  \n",
       "88   gut microbiome  \n",
       "89  marine sediment  \n",
       "90           saliva  \n",
       "91        sea water  \n",
       "\n",
       "[92 rows x 7 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_package_normalized = scoped_mapping.env_package_nomralizastion(\n",
    "    env_package_count, \"env_package\", id_patterns[\"ENVO\"]\n",
    ")\n",
    "\n",
    "# getting rid of redundant? 'string' column\n",
    "env_package_normalized = env_package_normalized[\n",
    "    [\n",
    "        \"env_package\",\n",
    "        \"count\",\n",
    "        \"lhs\",\n",
    "        \"rhs\",\n",
    "        \"extract\",\n",
    "        \"remaining_string\",\n",
    "        \"remaining_tidied\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "env_package_normalized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f5dd839d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eptidy</th>\n",
       "      <th>epdtidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>microbial</td>\n",
       "      <td>microbial mat biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>miscellaneous</td>\n",
       "      <td>miscellaneous or artificial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater sludge</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           eptidy                      epdtidy\n",
       "21      microbial        microbial mat biofilm\n",
       "22  miscellaneous  miscellaneous or artificial\n",
       "26     wastewater            wastewater sludge"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pick \"EnvPackage\"/\"eptidy\" from package dictionary as canonical,\n",
    "# not \"EnvPackageDisplay\"/\"epdtidy\"\n",
    "# but stillwant to support making XXX values from YYY canonical according to \"EnvPackageDisplay\"/\"epdtidy\"\n",
    "# so make a mapping/override table\n",
    "\n",
    "epd_to_ep = package_dictionary[[\"eptidy\", \"epdtidy\"]]\n",
    "# drop duplicates\n",
    "epd_to_ep = epd_to_ep.drop_duplicates()\n",
    "\n",
    "# drop blank eptidy rows\n",
    "ep_blank_flag = epd_to_ep[\"eptidy\"].eq(\"\")\n",
    "epd_to_ep = epd_to_ep.loc[~ep_blank_flag]\n",
    "\n",
    "# drop rows where eptidy and epdtidy are the same\n",
    "identical_flag = epd_to_ep[\"eptidy\"] == epd_to_ep[\"epdtidy\"]\n",
    "epd_to_ep = epd_to_ep.loc[~identical_flag]\n",
    "\n",
    "epd_to_ep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "716b9f06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'microbial mat biofilm': 'microbial',\n",
       " 'miscellaneous or artificial': 'miscellaneous',\n",
       " 'wastewater sludge': 'wastewater'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# and add to manualy asserted overrides above\n",
    "overrides_supplement = dict(zip(epd_to_ep[\"epdtidy\"], epd_to_ep[\"eptidy\"]))\n",
    "\n",
    "overrides_supplement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b32b62b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'built environment': 'built',\n",
       " 'misc environment': 'miscellaneous',\n",
       " 'missing': 'no environmental package',\n",
       " 'unknown': 'no environmental package',\n",
       " 'default': 'no environmental package',\n",
       " 'unspecified': 'no environmental package',\n",
       " 'not available': 'no environmental package',\n",
       " 'not collected': 'no environmental package',\n",
       " 'miscellaneous natural or artificial environment': 'miscellaneous',\n",
       " 'not applicable': 'no environmental package',\n",
       " 'soil-associated': 'soil',\n",
       " 'soil associated': 'soil',\n",
       " 'microbial mat biofilm': 'microbial',\n",
       " 'miscellaneous or artificial': 'miscellaneous',\n",
       " 'wastewater sludge': 'wastewater'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# not getting soil-associated\n",
    "env_package_overrides.update(overrides_supplement)\n",
    "\n",
    "env_package_overrides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "24307595",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_normalized = scoped_mapping.add_overrides(\n",
    "    env_package_normalized, \"remaining_tidied\", \"rt_override\", env_package_overrides\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fe0899ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>eptidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>No environmental package</td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>air</td>\n",
       "      <td>air</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>built</td>\n",
       "      <td>built</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>human-associated</td>\n",
       "      <td>human associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>human-oral</td>\n",
       "      <td>human oral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>human-vaginal</td>\n",
       "      <td>human vaginal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>microbial</td>\n",
       "      <td>microbial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>miscellaneous</td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>plant-associated</td>\n",
       "      <td>plant associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  EnvPackage                    eptidy\n",
       "0                                                     \n",
       "12  No environmental package  no environmental package\n",
       "13                       air                       air\n",
       "14                     built                     built\n",
       "15           host-associated           host associated\n",
       "16          human-associated          human associated\n",
       "17                 human-gut                 human gut\n",
       "18                human-oral                human oral\n",
       "19                human-skin                human skin\n",
       "20             human-vaginal             human vaginal\n",
       "21                 microbial                 microbial\n",
       "22             miscellaneous             miscellaneous\n",
       "23          plant-associated          plant associated\n",
       "24                  sediment                  sediment\n",
       "25                      soil                      soil\n",
       "26                wastewater                wastewater\n",
       "27                     water                     water"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "denorm_frame = package_dictionary[[\"EnvPackage\", \"eptidy\"]]\n",
    "denorm_frame = denorm_frame.drop_duplicates()\n",
    "denorm_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "73694ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_normalized = env_package_normalized.merge(\n",
    "    denorm_frame, how=\"left\", left_on=\"rt_override\", right_on=\"eptidy\"\n",
    ")\n",
    "\n",
    "env_package_normalized = env_package_normalized[\n",
    "    [\n",
    "        \"env_package\",\n",
    "        \"count\",\n",
    "        \"lhs\",\n",
    "        \"rhs\",\n",
    "        \"extract\",\n",
    "        \"remaining_string\",\n",
    "        \"remaining_tidied\",\n",
    "        \"rt_override\",\n",
    "        \"EnvPackage\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "non_canonical_flag = env_package_normalized[\"EnvPackage\"].isna()\n",
    "env_package_normalized[\"is_canonical\"] = True\n",
    "env_package_normalized.loc[non_canonical_flag, \"is_canonical\"] = False\n",
    "\n",
    "# env_package = env_package annotation from NCBI Biosample file XXX\n",
    "# count = number of biosamples using that env_package annotation\n",
    "# lhs = checklist info\n",
    "# rhs = potential package info\n",
    "# extract = potential OBO ID from rhs column (currently harcoded and only looking for ENVO IDs)\n",
    "# remaining_string = rhs/string, with potential OBO IDs removed\n",
    "# remaining_tidied = remaining_string with case, whitespace and punctuation normailzastion\n",
    "# rt_override = some remaining_tidied values can be replaced according to env_package_overrides\n",
    "# EnvPackage = corresponding de-normalized value from package_dictionary\n",
    "# is_canonical = false when EnvPackage is NaN\n",
    "\n",
    "env_package_normalized.to_sql(\n",
    "    \"env_package_normalized\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "\n",
    "env_package_normalized.to_csv(\n",
    "    \"biosample_env_package_normalizastion.tsv\", sep=ols_review_seperator, index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f1b57e51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>rhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>rt_override</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>is_canonical</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "      <td>host associated</td>\n",
       "      <td>host-associated</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "      <td>human gut</td>\n",
       "      <td>human-gut</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "      <td>human skin</td>\n",
       "      <td>human-skin</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td>env package</td>\n",
       "      <td>env package</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>marine sediment ()</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water,</td>\n",
       "      <td>sea water</td>\n",
       "      <td>sea water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count lhs  \\\n",
       "0                              None  14083847       \n",
       "1                   host-associated     49254       \n",
       "2                         human-gut     47921       \n",
       "3                             water     16367       \n",
       "4                        human-skin     13706       \n",
       "..                              ...       ...  ..   \n",
       "87                      env_package         1       \n",
       "88                   gut microbiome         1       \n",
       "89  marine sediment (ENVO:00002113)         1       \n",
       "90                           saliva         1       \n",
       "91       sea water, [ENVO:00002149]         1       \n",
       "\n",
       "                                rhs        extract    remaining_string  \\\n",
       "0                              None                                      \n",
       "1                   host-associated                    host-associated   \n",
       "2                         human-gut                          human-gut   \n",
       "3                             water                              water   \n",
       "4                        human-skin                         human-skin   \n",
       "..                              ...            ...                 ...   \n",
       "87                      env_package                        env_package   \n",
       "88                   gut microbiome                     gut microbiome   \n",
       "89  marine sediment (ENVO:00002113)  ENVO:00002113  marine sediment ()   \n",
       "90                           saliva                             saliva   \n",
       "91       sea water, [ENVO:00002149]  ENVO:00002149         sea water,    \n",
       "\n",
       "   remaining_tidied      rt_override       EnvPackage  is_canonical  \n",
       "0                                                              True  \n",
       "1   host associated  host associated  host-associated          True  \n",
       "2         human gut        human gut        human-gut          True  \n",
       "3             water            water            water          True  \n",
       "4        human skin       human skin       human-skin          True  \n",
       "..              ...              ...              ...           ...  \n",
       "87      env package      env package              NaN         False  \n",
       "88   gut microbiome   gut microbiome              NaN         False  \n",
       "89  marine sediment  marine sediment              NaN         False  \n",
       "90           saliva           saliva              NaN         False  \n",
       "91        sea water        sea water              NaN         False  \n",
       "\n",
       "[92 rows x 10 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_package_normalized"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2a7fed8",
   "metadata": {},
   "source": [
    "## What do the successful normalizations look like?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7c1fbb75",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.002170\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>EnvPackage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>not collected</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>host associated</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No environmental package</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        env_package     count lhs        extract                EnvPackage\n",
       "0              None  14083847                                             \n",
       "1   host-associated     49254                              host-associated\n",
       "2         human-gut     47921                                    human-gut\n",
       "3             water     16367                                        water\n",
       "4        human-skin     13706                                   human-skin\n",
       "..              ...       ...  ..            ...                       ...\n",
       "57    not collected         3                     No environmental package\n",
       "58  host associated         2                              host-associated\n",
       "59    ENVO:00000016         1      ENVO:00000016                          \n",
       "60         Sediment         1                                     sediment\n",
       "61          Unknown         1                     No environmental package\n",
       "\n",
       "[62 rows x 5 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    EnvPackage\n",
    "from\n",
    "    env_package_normalized\n",
    "where\n",
    "    is_canonical = 1\n",
    "\"\"\"\n",
    "[successful_normalizastions, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "successful_normalizastions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "740ffe3c",
   "metadata": {},
   "source": [
    "# Are there any normalization failures?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "17e08383",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.002603\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>EnvPackage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mimarks</td>\n",
       "      <td>479</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse-gut</td>\n",
       "      <td>406</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gut</td>\n",
       "      <td>172</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>biofilm</td>\n",
       "      <td>114</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-not providedsopharyngeal</td>\n",
       "      <td>107</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>mice gut</td>\n",
       "      <td>87</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CV</td>\n",
       "      <td>60</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>home, outdoor environment</td>\n",
       "      <td>44</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fermentation-associated</td>\n",
       "      <td>42</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sterile water</td>\n",
       "      <td>35</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Mouse_Gut</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>fresh water</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rock</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>arctic</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>freshwater sediment</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>seawater</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MIENS_16S</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MIMARKS_water</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>cave</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>plant</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>sea water</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Coastal water body (ENVO:02000049)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:02000049</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Deep-sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MIMARKS_16S</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           env_package  count lhs        extract EnvPackage\n",
       "0                              mimarks    479                          None\n",
       "1                            mouse-gut    406                          None\n",
       "2                                  gut    172                          None\n",
       "3                              biofilm    114                          None\n",
       "4       human-not providedsopharyngeal    107                          None\n",
       "5                             mice gut     87                          None\n",
       "6                                   CV     60                          None\n",
       "7            home, outdoor environment     44                          None\n",
       "8              fermentation-associated     42                          None\n",
       "9                        sterile water     35                          None\n",
       "10                           Mouse_Gut     27                          None\n",
       "11                         fresh water     27                          None\n",
       "12                                rock     12                          None\n",
       "13                              arctic     10                          None\n",
       "14                 freshwater sediment      9                          None\n",
       "15                            seawater      9                          None\n",
       "16                           MIENS_16S      6                          None\n",
       "17                       MIMARKS_water      6                          None\n",
       "18                                cave      5                          None\n",
       "19                               plant      3                          None\n",
       "20                           sea water      2                          None\n",
       "21  Coastal water body (ENVO:02000049)      1      ENVO:02000049       None\n",
       "22                      Deep-sea water      1                          None\n",
       "23                         MIMARKS_16S      1                          None\n",
       "24                           Sea water      1                          None\n",
       "25                         env_package      1                          None\n",
       "26                      gut microbiome      1                          None\n",
       "27     marine sediment (ENVO:00002113)      1      ENVO:00002113       None\n",
       "28                              saliva      1                          None\n",
       "29          sea water, [ENVO:00002149]      1      ENVO:00002149       None"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    EnvPackage\n",
    "from\n",
    "    env_package_normalized\n",
    "where\n",
    "    is_canonical = 0\n",
    "\"\"\"\n",
    "[normalizastion_failures, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "normalizastion_failures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80370e30",
   "metadata": {},
   "source": [
    "# Utilizing taxonomy for broad subsetting\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46e3ee59",
   "metadata": {},
   "source": [
    "**This uses an SQLite database in which the transitive closure over subClassOf has already been materialized. See the README and Makefile.**\n",
    "\n",
    "Specifically, flag the Biosamples whose `taxon_id` indicates they are an unclassified entity. Ignoring the others will throw out samples of multicellular organisms, like fruit flies.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bfd8416",
   "metadata": {},
   "source": [
    "## Get a listing of all taxa that are transitive subclasses of `NCBITaxon:2787823`\n",
    "\n",
    "I.e. 'unclassified entities'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c40903d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:13.252446\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:1006967</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:1041057</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:1046002</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:1046003</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:1046004</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>NCBITaxon:939928</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>NCBITaxon:941420</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>NCBITaxon:941421</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>NCBITaxon:941422</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>NCBITaxon:942017</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>994 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               subject  unclassified\n",
       "0    NCBITaxon:1006967          True\n",
       "1    NCBITaxon:1041057          True\n",
       "2    NCBITaxon:1046002          True\n",
       "3    NCBITaxon:1046003          True\n",
       "4    NCBITaxon:1046004          True\n",
       "..                 ...           ...\n",
       "989   NCBITaxon:939928          True\n",
       "990   NCBITaxon:941420          True\n",
       "991   NCBITaxon:941421          True\n",
       "992   NCBITaxon:941422          True\n",
       "993   NCBITaxon:942017          True\n",
       "\n",
       "[994 rows x 2 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    distinct s.subject\n",
    "from\n",
    "    entailed_edge ee\n",
    "join statements s on\n",
    "    ee.subject = s.subject\n",
    "where\n",
    "    ee.predicate = 'rdfs:subClassOf'\n",
    "    and ee.object = 'NCBITaxon:2787823'\n",
    "    and s.predicate = 'rdfs:label'\n",
    "\"\"\"\n",
    "[unclassified_taxa, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "unclassified_taxa[\"unclassified\"] = True\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "unclassified_taxa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cad3ffdb",
   "metadata": {},
   "source": [
    "## Get taxon counts from the Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a5d8f028",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:01.530258\n"
     ]
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    taxonomy_id biosample_taxid,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    taxonomy_id\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[biosample_tax_id_counts, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "biosample_tax_id_counts[\"curie\"] = \"NCBITaxon:\" + biosample_tax_id_counts[\n",
    "    \"biosample_taxid\"\n",
    "].astype(str)\n",
    "\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cf3a275",
   "metadata": {},
   "source": [
    "## Merge the two taxonomy dataframes\n",
    "\n",
    "I.e. flag the the Biosample records whose `taxonomy_id` field belongs to a subclass of 'unclassified entries'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b7ad29a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>curie</th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       biosample_taxid    count             curie           subject  \\\n",
       "0                 9606  6819707    NCBITaxon:9606               NaN   \n",
       "1                10090   964219   NCBITaxon:10090               NaN   \n",
       "2               408170   290862  NCBITaxon:408170  NCBITaxon:408170   \n",
       "3               410658   280666  NCBITaxon:410658  NCBITaxon:410658   \n",
       "4               646099   208741  NCBITaxon:646099  NCBITaxon:646099   \n",
       "...                ...      ...               ...               ...   \n",
       "163372          999891        1  NCBITaxon:999891               NaN   \n",
       "163373          999892        1  NCBITaxon:999892               NaN   \n",
       "163374          999898        1  NCBITaxon:999898               NaN   \n",
       "163375          999931        1  NCBITaxon:999931               NaN   \n",
       "163376          999947        1  NCBITaxon:999947               NaN   \n",
       "\n",
       "        unclassified  \n",
       "0              False  \n",
       "1              False  \n",
       "2               True  \n",
       "3               True  \n",
       "4               True  \n",
       "...              ...  \n",
       "163372         False  \n",
       "163373         False  \n",
       "163374         False  \n",
       "163375         False  \n",
       "163376         False  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biosample_tax_id_counts = biosample_tax_id_counts.merge(\n",
    "    unclassified_taxa, left_on=\"curie\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "biosample_tax_id_counts.unclassified.fillna(False, inplace=True)\n",
    "\n",
    "biosample_tax_id_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f7aff98",
   "metadata": {},
   "source": [
    "## Add labels to all taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a6acc349",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:10.064933\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>unclassified</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>False</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>False</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>True</td>\n",
       "      <td>human gut metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>True</td>\n",
       "      <td>soil metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>True</td>\n",
       "      <td>human metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Bacillus amyloliquefaciens TA208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>[Propionibacterium] humerusii P08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Peptococcaceae bacterium CEB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Barrientosiimonas humi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Angomonas ambiguus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   curie biosample_taxid    count  unclassified  \\\n",
       "0         NCBITaxon:9606            9606  6819707         False   \n",
       "1        NCBITaxon:10090           10090   964219         False   \n",
       "2       NCBITaxon:408170          408170   290862          True   \n",
       "3       NCBITaxon:410658          410658   280666          True   \n",
       "4       NCBITaxon:646099          646099   208741          True   \n",
       "...                  ...             ...      ...           ...   \n",
       "163372  NCBITaxon:999891          999891        1         False   \n",
       "163373  NCBITaxon:999892          999892        1         False   \n",
       "163374  NCBITaxon:999898          999898        1         False   \n",
       "163375  NCBITaxon:999931          999931        1         False   \n",
       "163376  NCBITaxon:999947          999947        1         False   \n",
       "\n",
       "                                    label  \n",
       "0                            Homo sapiens  \n",
       "1                            Mus musculus  \n",
       "2                    human gut metagenome  \n",
       "3                         soil metagenome  \n",
       "4                        human metagenome  \n",
       "...                                   ...  \n",
       "163372   Bacillus amyloliquefaciens TA208  \n",
       "163373  [Propionibacterium] humerusii P08  \n",
       "163374      Peptococcaceae bacterium CEB3  \n",
       "163375             Barrientosiimonas humi  \n",
       "163376                 Angomonas ambiguus  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    subject ,\n",
    "    value\n",
    "from statements\n",
    "where\n",
    "    predicate = 'rdfs:label' and subject = stanza\n",
    "\"\"\"\n",
    "[all_tax_labels, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "\n",
    "biosample_tax_id_counts = biosample_tax_id_counts.merge(\n",
    "    all_tax_labels, left_on=\"curie\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "\n",
    "biosample_tax_id_counts = biosample_tax_id_counts[\n",
    "    [\"curie\", \"biosample_taxid\", \"count\", \"unclassified\", \"value\"]\n",
    "]\n",
    "biosample_tax_id_counts.columns = [\n",
    "    \"curie\",\n",
    "    \"biosample_taxid\",\n",
    "    \"count\",\n",
    "    \"unclassified\",\n",
    "    \"label\",\n",
    "]\n",
    "\n",
    "print(query_duration)\n",
    "biosample_tax_id_counts.to_sql(\n",
    "    \"biobiosample_tax_id_counts\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "\n",
    "biosample_tax_id_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "554a82c0",
   "metadata": {},
   "source": [
    "**Almost all of the taxa that are common in the biosample collection are either unclassified/metagenomes or easily recognized cellular organisms. Cellular organism samples are de-prioritized in this exercise**\n",
    "\n",
    "Exceptions include:\n",
    "- 32630 = synthetic construct (other entries; other sequences; artificial sequences)\n",
    "    - 'other entries' would add 16k rows on top of the 1k 'unclassified entities'\n",
    "    - metagenomes account for 331 of the 'unclassified entities'\n",
    "    - there are also a small number of uncultured/unclassified microorganisms in the biosample dataset\n",
    "- 77133 = uncultured bacterium (cellular organisms; Bacteria; environmental samples)\n",
    "    - 'cellular organisms' would add 2M rows on top of the 1k 'unclassified entities'\n",
    "    - 'cellular organisms; Bacteria; environmental samples' adds 26k\n",
    "    \n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2363d41",
   "metadata": {},
   "source": [
    "## Get a table of MIxS annotations to be mapped to ontology classes.\n",
    "\n",
    "Explicitly scope based on normalized package data. These values were set at the top of this notebook.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "91dcca09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env_broad_scale\n",
      "env_package_normalized.EnvPackage\n",
      "soil\n"
     ]
    }
   ],
   "source": [
    "print(biosample_col_to_map)\n",
    "print(scoping_col)\n",
    "print(scoping_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80727cfd",
   "metadata": {},
   "source": [
    "**In this case, the scoping includes an inner join requirement for 'unclassified entities'**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7e777c8a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>4856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>terestrial</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_broad_scale  count\n",
       "0            ENVO:cropland biome   4856\n",
       "1                 cropland biome   1530\n",
       "2                    urban biome    973\n",
       "3                   tundra biome    516\n",
       "4              terrestrial biome    483\n",
       "..                           ...    ...\n",
       "239                   terestrial      1\n",
       "240            thermokarst ponds      1\n",
       "241                 urban boiome      1\n",
       "242  wastewater treatment system      1\n",
       "243         western Amazon basin      1\n",
       "\n",
       "[244 rows x 2 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = (\n",
    "    \"select \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\", count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "join env_package_normalized on\n",
    "    b.env_package = env_package_normalized.env_package\n",
    "inner join biobiosample_tax_id_counts stic on\n",
    "    b.taxonomy_id = stic.biosample_taxid\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"' group by \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\"\n",
    "order by\n",
    "    count(*) desc\"\"\"\n",
    ")\n",
    "\n",
    "[mapping_candidates, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86f43bb4",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## The Biosample format allows for pipe-delimited environmental package lists. \n",
    "\n",
    "Separate those out into their components.\n",
    "\n",
    "----\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "45e1859f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       repeated                     splitted  part_count  seq\n",
       "0           ENVO:cropland biome          ENVO:cropland biome           1    1\n",
       "0                cropland biome               cropland biome           1    1\n",
       "0                   urban biome                  urban biome           1    1\n",
       "0                  tundra biome                 tundra biome           1    1\n",
       "0             terrestrial biome            terrestrial biome           1    1\n",
       "..                          ...                          ...         ...  ...\n",
       "0                    terestrial                   terestrial           1    1\n",
       "0             thermokarst ponds            thermokarst ponds           1    1\n",
       "0                  urban boiome                 urban boiome           1    1\n",
       "0   wastewater treatment system  wastewater treatment system           1    1\n",
       "0          western Amazon basin         western Amazon basin           1    1\n",
       "\n",
       "[244 rows x 4 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "multi_frames = []\n",
    "for row in mapping_candidates.itertuples(index=True, name=\"Pandas\"):\n",
    "    split_check = row.env_broad_scale\n",
    "    if split_check is None:\n",
    "        split_check = \"\"\n",
    "    splitted = pd.Series(split_check.split(\"|\"))\n",
    "    splitted_count = len(splitted)\n",
    "    repeated = [split_check] * splitted_count\n",
    "    repeated = pd.Series(repeated)\n",
    "    as_frame = pd.DataFrame(dict(repeated=repeated, splitted=splitted)).reset_index()\n",
    "    seq_list = list(range(1, splitted_count + 1))\n",
    "    as_frame[\"part_count\"] = splitted_count\n",
    "    as_frame[\"seq\"] = seq_list\n",
    "    multi_frames.append(as_frame)\n",
    "concat_frame = pd.concat(multi_frames)\n",
    "concat_frame = concat_frame[[\"repeated\", \"splitted\", \"part_count\", \"seq\"]]\n",
    "concat_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c67a272e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>4856</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1530</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>973</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>516</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>483</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>terestrial</td>\n",
       "      <td>1</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_broad_scale  count                     repeated  \\\n",
       "0            ENVO:cropland biome   4856          ENVO:cropland biome   \n",
       "1                 cropland biome   1530               cropland biome   \n",
       "2                    urban biome    973                  urban biome   \n",
       "3                   tundra biome    516                 tundra biome   \n",
       "4              terrestrial biome    483            terrestrial biome   \n",
       "..                           ...    ...                          ...   \n",
       "239                   terestrial      1                   terestrial   \n",
       "240            thermokarst ponds      1            thermokarst ponds   \n",
       "241                 urban boiome      1                 urban boiome   \n",
       "242  wastewater treatment system      1  wastewater treatment system   \n",
       "243         western Amazon basin      1         western Amazon basin   \n",
       "\n",
       "                        splitted  part_count  seq  \n",
       "0            ENVO:cropland biome         1.0  1.0  \n",
       "1                 cropland biome         1.0  1.0  \n",
       "2                    urban biome         1.0  1.0  \n",
       "3                   tundra biome         1.0  1.0  \n",
       "4              terrestrial biome         1.0  1.0  \n",
       "..                           ...         ...  ...  \n",
       "239                   terestrial         1.0  1.0  \n",
       "240            thermokarst ponds         1.0  1.0  \n",
       "241                 urban boiome         1.0  1.0  \n",
       "242  wastewater treatment system         1.0  1.0  \n",
       "243         western Amazon basin         1.0  1.0  \n",
       "\n",
       "[244 rows x 6 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    concat_frame, left_on=biosample_col_to_map, right_on=\"repeated\", how=\"left\"\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3af421ac",
   "metadata": {},
   "source": [
    "## Normalize a few different ways `ENVO` IDs have been entered\n",
    "In the Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ed3a89fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>envo_tidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>4856</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1530</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cropland biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>973</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>516</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>tundra biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>483</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terrestrial biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>terestrial</td>\n",
       "      <td>1</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terestrial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban boiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>western Amazon basin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_broad_scale  count                     repeated  \\\n",
       "0            ENVO:cropland biome   4856          ENVO:cropland biome   \n",
       "1                 cropland biome   1530               cropland biome   \n",
       "2                    urban biome    973                  urban biome   \n",
       "3                   tundra biome    516                 tundra biome   \n",
       "4              terrestrial biome    483            terrestrial biome   \n",
       "..                           ...    ...                          ...   \n",
       "239                   terestrial      1                   terestrial   \n",
       "240            thermokarst ponds      1            thermokarst ponds   \n",
       "241                 urban boiome      1                 urban boiome   \n",
       "242  wastewater treatment system      1  wastewater treatment system   \n",
       "243         western Amazon basin      1         western Amazon basin   \n",
       "\n",
       "                        splitted  part_count  seq                    envo_tidy  \n",
       "0            ENVO:cropland biome         1.0  1.0          ENVO:cropland biome  \n",
       "1                 cropland biome         1.0  1.0               cropland biome  \n",
       "2                    urban biome         1.0  1.0                  urban biome  \n",
       "3                   tundra biome         1.0  1.0                 tundra biome  \n",
       "4              terrestrial biome         1.0  1.0            terrestrial biome  \n",
       "..                           ...         ...  ...                          ...  \n",
       "239                   terestrial         1.0  1.0                   terestrial  \n",
       "240            thermokarst ponds         1.0  1.0            thermokarst ponds  \n",
       "241                 urban boiome         1.0  1.0                 urban boiome  \n",
       "242  wastewater treatment system         1.0  1.0  wastewater treatment system  \n",
       "243         western Amazon basin         1.0  1.0         western Amazon basin  \n",
       "\n",
       "[244 rows x 7 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# HARCODED/SINGLE PREFIX\n",
    "# this helps by standardizing what could be an ID\n",
    "# but if the RHS is text, then envo should just be removed\n",
    "mapping_candidates[\"envo_tidy\"] = mapping_candidates.splitted.str.replace(\n",
    "    \"envo[:_ ]\", \"ENVO:\", regex=True, case=False\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cc22a1a",
   "metadata": {},
   "source": [
    "# Now try to extract ontology terms that are already present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e423fcb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>4856</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1530</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>973</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td></td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>516</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td></td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>483</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td></td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>terestrial</td>\n",
       "      <td>1</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td></td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td></td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td></td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td></td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td></td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_broad_scale  count                     repeated  \\\n",
       "0            ENVO:cropland biome   4856          ENVO:cropland biome   \n",
       "1                 cropland biome   1530               cropland biome   \n",
       "2                    urban biome    973                  urban biome   \n",
       "3                   tundra biome    516                 tundra biome   \n",
       "4              terrestrial biome    483            terrestrial biome   \n",
       "..                           ...    ...                          ...   \n",
       "239                   terestrial      1                   terestrial   \n",
       "240            thermokarst ponds      1            thermokarst ponds   \n",
       "241                 urban boiome      1                 urban boiome   \n",
       "242  wastewater treatment system      1  wastewater treatment system   \n",
       "243         western Amazon basin      1         western Amazon basin   \n",
       "\n",
       "                        splitted  part_count  seq  \\\n",
       "0            ENVO:cropland biome         1.0  1.0   \n",
       "1                 cropland biome         1.0  1.0   \n",
       "2                    urban biome         1.0  1.0   \n",
       "3                   tundra biome         1.0  1.0   \n",
       "4              terrestrial biome         1.0  1.0   \n",
       "..                           ...         ...  ...   \n",
       "239                   terestrial         1.0  1.0   \n",
       "240            thermokarst ponds         1.0  1.0   \n",
       "241                 urban boiome         1.0  1.0   \n",
       "242  wastewater treatment system         1.0  1.0   \n",
       "243         western Amazon basin         1.0  1.0   \n",
       "\n",
       "                       envo_tidy                       string extract  \\\n",
       "0            ENVO:cropland biome          ENVO:cropland biome           \n",
       "1                 cropland biome               cropland biome           \n",
       "2                    urban biome                  urban biome           \n",
       "3                   tundra biome                 tundra biome           \n",
       "4              terrestrial biome            terrestrial biome           \n",
       "..                           ...                          ...     ...   \n",
       "239                   terestrial                   terestrial           \n",
       "240            thermokarst ponds            thermokarst ponds           \n",
       "241                 urban boiome                 urban boiome           \n",
       "242  wastewater treatment system  wastewater treatment system           \n",
       "243         western Amazon basin         western Amazon basin           \n",
       "\n",
       "                remaining_string             remaining_tidied  \n",
       "0            ENVO:cropland biome               cropland biome  \n",
       "1                 cropland biome               cropland biome  \n",
       "2                    urban biome                  urban biome  \n",
       "3                   tundra biome                 tundra biome  \n",
       "4              terrestrial biome            terrestrial biome  \n",
       "..                           ...                          ...  \n",
       "239                   terestrial                   terestrial  \n",
       "240            thermokarst ponds            thermokarst ponds  \n",
       "241                 urban boiome                 urban boiome  \n",
       "242  wastewater treatment system  wastewater treatment system  \n",
       "243         western Amazon basin         western amazon basin  \n",
       "\n",
       "[244 rows x 11 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_series_decomposition = scoped_mapping.decompose_series(\n",
    "    mapping_candidates[\"envo_tidy\"], id_patterns[target_onto_prefix]\n",
    ")\n",
    "\n",
    "mapping_candidates = pd.concat(\n",
    "    [mapping_candidates, candidate_series_decomposition], axis=1\n",
    ")\n",
    "\n",
    "# SEE \"HARCODED/SINGLE PREFIX\" ABOVE\n",
    "mapping_candidates[\"remaining_tidied\"] = mapping_candidates.remaining_tidied.str.replace(\n",
    "    \"envo[:_ ]\", \"\", regex=True, case=False\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccd038ff",
   "metadata": {},
   "source": [
    "## Join the extracted IDs with their labels\n",
    "\n",
    "Start by connecting to the rdftab database from which the terms and label-like annotations will be obtained"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "448865ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "ontodb = \"../semantic-sql/db/\" + target_onto_prefix.lower() + \".db\"\n",
    "ontocon = sqlite3.connect(ontodb)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db5d19a9",
   "metadata": {},
   "source": [
    "## extracting the labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3bcd7e47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IAO:0000111</td>\n",
       "      <td>editor preferred term~editor preferred label</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IAO:0000112</td>\n",
       "      <td>example of usage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IAO:0000114</td>\n",
       "      <td>has curation status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IAO:0000115</td>\n",
       "      <td>definition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IAO:0000116</td>\n",
       "      <td>editor note</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6774</th>\n",
       "      <td>ENVO:01001862</td>\n",
       "      <td>Solar radiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6775</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q2&gt;</td>\n",
       "      <td>Earth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q2306597&gt;</td>\n",
       "      <td>Suni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6777</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q525&gt;</td>\n",
       "      <td>Sol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6778</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q715269&gt;</td>\n",
       "      <td>World Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6779 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       subject  \\\n",
       "0                                  IAO:0000111   \n",
       "1                                  IAO:0000112   \n",
       "2                                  IAO:0000114   \n",
       "3                                  IAO:0000115   \n",
       "4                                  IAO:0000116   \n",
       "...                                        ...   \n",
       "6774                             ENVO:01001862   \n",
       "6775        <https://www.wikidata.org/wiki/Q2>   \n",
       "6776  <https://www.wikidata.org/wiki/Q2306597>   \n",
       "6777      <https://www.wikidata.org/wiki/Q525>   \n",
       "6778   <https://www.wikidata.org/wiki/Q715269>   \n",
       "\n",
       "                                             value  \n",
       "0     editor preferred term~editor preferred label  \n",
       "1                                 example of usage  \n",
       "2                              has curation status  \n",
       "3                                       definition  \n",
       "4                                      editor note  \n",
       "...                                            ...  \n",
       "6774                               Solar radiation  \n",
       "6775                                         Earth  \n",
       "6776                                          Suni  \n",
       "6777                                           Sol  \n",
       "6778                                   World Ocean  \n",
       "\n",
       "[6779 rows x 2 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    subject ,\n",
    "    value\n",
    "from\n",
    "    statements s\n",
    "where\n",
    "    predicate = 'rdfs:label'\n",
    "\"\"\"\n",
    "[onto_labels, query_duration] = scoped_mapping.timed_query(q, ontocon)\n",
    "\n",
    "onto_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "840b922f",
   "metadata": {},
   "source": [
    "## and merging "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8543ed0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>4856</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1530</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>973</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td></td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>516</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td></td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>483</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td></td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>terestrial</td>\n",
       "      <td>1</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td></td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td></td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td></td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td></td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td></td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_broad_scale  count                     repeated  \\\n",
       "0            ENVO:cropland biome   4856          ENVO:cropland biome   \n",
       "1                 cropland biome   1530               cropland biome   \n",
       "2                    urban biome    973                  urban biome   \n",
       "3                   tundra biome    516                 tundra biome   \n",
       "4              terrestrial biome    483            terrestrial biome   \n",
       "..                           ...    ...                          ...   \n",
       "239                   terestrial      1                   terestrial   \n",
       "240            thermokarst ponds      1            thermokarst ponds   \n",
       "241                 urban boiome      1                 urban boiome   \n",
       "242  wastewater treatment system      1  wastewater treatment system   \n",
       "243         western Amazon basin      1         western Amazon basin   \n",
       "\n",
       "                        splitted  part_count  seq  \\\n",
       "0            ENVO:cropland biome         1.0  1.0   \n",
       "1                 cropland biome         1.0  1.0   \n",
       "2                    urban biome         1.0  1.0   \n",
       "3                   tundra biome         1.0  1.0   \n",
       "4              terrestrial biome         1.0  1.0   \n",
       "..                           ...         ...  ...   \n",
       "239                   terestrial         1.0  1.0   \n",
       "240            thermokarst ponds         1.0  1.0   \n",
       "241                 urban boiome         1.0  1.0   \n",
       "242  wastewater treatment system         1.0  1.0   \n",
       "243         western Amazon basin         1.0  1.0   \n",
       "\n",
       "                       envo_tidy                       string extract  \\\n",
       "0            ENVO:cropland biome          ENVO:cropland biome           \n",
       "1                 cropland biome               cropland biome           \n",
       "2                    urban biome                  urban biome           \n",
       "3                   tundra biome                 tundra biome           \n",
       "4              terrestrial biome            terrestrial biome           \n",
       "..                           ...                          ...     ...   \n",
       "239                   terestrial                   terestrial           \n",
       "240            thermokarst ponds            thermokarst ponds           \n",
       "241                 urban boiome                 urban boiome           \n",
       "242  wastewater treatment system  wastewater treatment system           \n",
       "243         western Amazon basin         western Amazon basin           \n",
       "\n",
       "                remaining_string             remaining_tidied subject value  \n",
       "0            ENVO:cropland biome               cropland biome     NaN   NaN  \n",
       "1                 cropland biome               cropland biome     NaN   NaN  \n",
       "2                    urban biome                  urban biome     NaN   NaN  \n",
       "3                   tundra biome                 tundra biome     NaN   NaN  \n",
       "4              terrestrial biome            terrestrial biome     NaN   NaN  \n",
       "..                           ...                          ...     ...   ...  \n",
       "239                   terestrial                   terestrial     NaN   NaN  \n",
       "240            thermokarst ponds            thermokarst ponds     NaN   NaN  \n",
       "241                 urban boiome                 urban boiome     NaN   NaN  \n",
       "242  wastewater treatment system  wastewater treatment system     NaN   NaN  \n",
       "243         western Amazon basin         western amazon basin     NaN   NaN  \n",
       "\n",
       "[244 rows x 13 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"extract\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64b3baef",
   "metadata": {},
   "source": [
    "## Use cosine string distance to see if the labels match closely enough\n",
    "\n",
    "I.e. the labels claimed by the Biosample data set and the labels asserted in the ontology. if they're close enough, consider the assigned ID legit\n",
    "\n",
    "\n",
    "_How close is close enough?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "abd0a035",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "      <th>cosine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>4856</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1530</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>973</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td></td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>516</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td></td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>483</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td></td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>terestrial</td>\n",
       "      <td>1</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td></td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td></td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td></td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td></td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td></td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_broad_scale  count                     repeated  \\\n",
       "0            ENVO:cropland biome   4856          ENVO:cropland biome   \n",
       "1                 cropland biome   1530               cropland biome   \n",
       "2                    urban biome    973                  urban biome   \n",
       "3                   tundra biome    516                 tundra biome   \n",
       "4              terrestrial biome    483            terrestrial biome   \n",
       "..                           ...    ...                          ...   \n",
       "239                   terestrial      1                   terestrial   \n",
       "240            thermokarst ponds      1            thermokarst ponds   \n",
       "241                 urban boiome      1                 urban boiome   \n",
       "242  wastewater treatment system      1  wastewater treatment system   \n",
       "243         western Amazon basin      1         western Amazon basin   \n",
       "\n",
       "                        splitted  part_count  seq  \\\n",
       "0            ENVO:cropland biome         1.0  1.0   \n",
       "1                 cropland biome         1.0  1.0   \n",
       "2                    urban biome         1.0  1.0   \n",
       "3                   tundra biome         1.0  1.0   \n",
       "4              terrestrial biome         1.0  1.0   \n",
       "..                           ...         ...  ...   \n",
       "239                   terestrial         1.0  1.0   \n",
       "240            thermokarst ponds         1.0  1.0   \n",
       "241                 urban boiome         1.0  1.0   \n",
       "242  wastewater treatment system         1.0  1.0   \n",
       "243         western Amazon basin         1.0  1.0   \n",
       "\n",
       "                       envo_tidy                       string extract  \\\n",
       "0            ENVO:cropland biome          ENVO:cropland biome           \n",
       "1                 cropland biome               cropland biome           \n",
       "2                    urban biome                  urban biome           \n",
       "3                   tundra biome                 tundra biome           \n",
       "4              terrestrial biome            terrestrial biome           \n",
       "..                           ...                          ...     ...   \n",
       "239                   terestrial                   terestrial           \n",
       "240            thermokarst ponds            thermokarst ponds           \n",
       "241                 urban boiome                 urban boiome           \n",
       "242  wastewater treatment system  wastewater treatment system           \n",
       "243         western Amazon basin         western Amazon basin           \n",
       "\n",
       "                remaining_string             remaining_tidied subject value  \\\n",
       "0            ENVO:cropland biome               cropland biome     NaN         \n",
       "1                 cropland biome               cropland biome     NaN         \n",
       "2                    urban biome                  urban biome     NaN         \n",
       "3                   tundra biome                 tundra biome     NaN         \n",
       "4              terrestrial biome            terrestrial biome     NaN         \n",
       "..                           ...                          ...     ...   ...   \n",
       "239                   terestrial                   terestrial     NaN         \n",
       "240            thermokarst ponds            thermokarst ponds     NaN         \n",
       "241                 urban boiome                 urban boiome     NaN         \n",
       "242  wastewater treatment system  wastewater treatment system     NaN         \n",
       "243         western Amazon basin         western amazon basin     NaN         \n",
       "\n",
       "     cosine  \n",
       "0       1.0  \n",
       "1       1.0  \n",
       "2       1.0  \n",
       "3       1.0  \n",
       "4       1.0  \n",
       "..      ...  \n",
       "239     1.0  \n",
       "240     1.0  \n",
       "241     1.0  \n",
       "242     1.0  \n",
       "243     1.0  \n",
       "\n",
       "[244 rows x 14 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_cosine_obj = Cosine(my_string_dist_arg)\n",
    "mapping_candidates[\"value\"] = mapping_candidates[\"value\"].fillna(\"\")\n",
    "mapping_candidates[\"remaining_tidied\"] = mapping_candidates[\"remaining_tidied\"].fillna(\"\")\n",
    "mapping_candidates[\"cosine\"] = mapping_candidates.apply(\n",
    "    lambda my_row: my_cosine_obj.distance(\n",
    "        my_row[\"remaining_tidied\"].lower(), my_row[\"value\"].lower()\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "7fd54a7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2358a89",
   "metadata": {},
   "source": [
    "**Previously, we did a reality check on the claimed IDs and labels. If a label is claimed without any ID, that could still be a path to an ontology term.**\n",
    "\n",
    "We'll be doing some merging, so make sure column names are meaningful and unique\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "092aced0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>4856</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000245</td>\n",
       "      <td>cropland biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1530</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000245</td>\n",
       "      <td>cropland biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>973</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td></td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000249</td>\n",
       "      <td>urban biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>516</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td></td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000180</td>\n",
       "      <td>tundra biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>483</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td></td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>terestrial</td>\n",
       "      <td>1</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td></td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td></td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td></td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td></td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td></td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_broad_scale  count                     repeated  \\\n",
       "0            ENVO:cropland biome   4856          ENVO:cropland biome   \n",
       "1                 cropland biome   1530               cropland biome   \n",
       "2                    urban biome    973                  urban biome   \n",
       "3                   tundra biome    516                 tundra biome   \n",
       "4              terrestrial biome    483            terrestrial biome   \n",
       "..                           ...    ...                          ...   \n",
       "239                   terestrial      1                   terestrial   \n",
       "240            thermokarst ponds      1            thermokarst ponds   \n",
       "241                 urban boiome      1                 urban boiome   \n",
       "242  wastewater treatment system      1  wastewater treatment system   \n",
       "243         western Amazon basin      1         western Amazon basin   \n",
       "\n",
       "                        splitted  part_count  seq  \\\n",
       "0            ENVO:cropland biome         1.0  1.0   \n",
       "1                 cropland biome         1.0  1.0   \n",
       "2                    urban biome         1.0  1.0   \n",
       "3                   tundra biome         1.0  1.0   \n",
       "4              terrestrial biome         1.0  1.0   \n",
       "..                           ...         ...  ...   \n",
       "239                   terestrial         1.0  1.0   \n",
       "240            thermokarst ponds         1.0  1.0   \n",
       "241                 urban boiome         1.0  1.0   \n",
       "242  wastewater treatment system         1.0  1.0   \n",
       "243         western Amazon basin         1.0  1.0   \n",
       "\n",
       "                       envo_tidy                       string extract  \\\n",
       "0            ENVO:cropland biome          ENVO:cropland biome           \n",
       "1                 cropland biome               cropland biome           \n",
       "2                    urban biome                  urban biome           \n",
       "3                   tundra biome                 tundra biome           \n",
       "4              terrestrial biome            terrestrial biome           \n",
       "..                           ...                          ...     ...   \n",
       "239                   terestrial                   terestrial           \n",
       "240            thermokarst ponds            thermokarst ponds           \n",
       "241                 urban boiome                 urban boiome           \n",
       "242  wastewater treatment system  wastewater treatment system           \n",
       "243         western Amazon basin         western Amazon basin           \n",
       "\n",
       "                remaining_string             remaining_tidied term_id  \\\n",
       "0            ENVO:cropland biome               cropland biome     NaN   \n",
       "1                 cropland biome               cropland biome     NaN   \n",
       "2                    urban biome                  urban biome     NaN   \n",
       "3                   tundra biome                 tundra biome     NaN   \n",
       "4              terrestrial biome            terrestrial biome     NaN   \n",
       "..                           ...                          ...     ...   \n",
       "239                   terestrial                   terestrial     NaN   \n",
       "240            thermokarst ponds            thermokarst ponds     NaN   \n",
       "241                 urban boiome                 urban boiome     NaN   \n",
       "242  wastewater treatment system  wastewater treatment system     NaN   \n",
       "243         western Amazon basin         western amazon basin     NaN   \n",
       "\n",
       "    lab_from_id  lfi_cosine term_id_from_lab              value  \n",
       "0                       1.0    ENVO:01000245     cropland biome  \n",
       "1                       1.0    ENVO:01000245     cropland biome  \n",
       "2                       1.0    ENVO:01000249        urban biome  \n",
       "3                       1.0    ENVO:01000180       tundra biome  \n",
       "4                       1.0    ENVO:00000446  terrestrial biome  \n",
       "..          ...         ...              ...                ...  \n",
       "239                     1.0              NaN                NaN  \n",
       "240                     1.0              NaN                NaN  \n",
       "241                     1.0              NaN                NaN  \n",
       "242                     1.0              NaN                NaN  \n",
       "243                     1.0              NaN                NaN  \n",
       "\n",
       "[244 rows x 16 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates.columns = [\n",
    "    \"env_broad_scale\",\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"envo_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "]\n",
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"remaining_tidied\", right_on=\"value\", how=\"left\"\n",
    ")\n",
    "\n",
    "mapping_candidates.columns = [\n",
    "    \"env_broad_scale\",\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"envo_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "    \"term_id_from_lab\",\n",
    "    \"value\",\n",
    "]\n",
    "\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41082d99",
   "metadata": {},
   "source": [
    "## Find consensus term IDs and labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4f5e3fdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates[\"consensus_id\"] = mapping_candidates[\"term_id_from_lab\"]\n",
    "mapping_candidates[\"consensus_lab\"] = mapping_candidates[\"value\"]\n",
    "\n",
    "flag = mapping_candidates[\"consensus_id\"].isnull() & (\n",
    "    (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates.lfi_cosine.le(my_max_string_dist)\n",
    "    )\n",
    "    | (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates[\"remaining_tidied\"].eq(\"\")\n",
    "    )\n",
    ")\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"term_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_id\"] = replacements\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"lab_from_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_lab\"] = replacements\n",
    "\n",
    "flag = mapping_candidates.consensus_id.isna()\n",
    "antiflag = ~flag\n",
    "mapping_candidates[\"id_or_lab_ok\"] = antiflag\n",
    "\n",
    "mapping_candidates[\"assembled_consensus\"] = (\n",
    "    mapping_candidates[\"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + mapping_candidates[\"consensus_id\"]\n",
    "    + \"]\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6af120f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>envo_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "      <th>id_or_lab_ok</th>\n",
       "      <th>assembled_consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>4856</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000245</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>ENVO:01000245</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>True</td>\n",
       "      <td>cropland biome [ENVO:01000245]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1530</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td></td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000245</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>ENVO:01000245</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>True</td>\n",
       "      <td>cropland biome [ENVO:01000245]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>973</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td></td>\n",
       "      <td>urban biome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000249</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>ENVO:01000249</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>True</td>\n",
       "      <td>urban biome [ENVO:01000249]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>516</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td></td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000180</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>ENVO:01000180</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>True</td>\n",
       "      <td>tundra biome [ENVO:01000180]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>483</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td></td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>True</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>terestrial</td>\n",
       "      <td>1</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td></td>\n",
       "      <td>terestrial</td>\n",
       "      <td>terestrial</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td></td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>thermokarst ponds</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td></td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>urban boiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td></td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td></td>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_broad_scale  count                     repeated  \\\n",
       "0            ENVO:cropland biome   4856          ENVO:cropland biome   \n",
       "1                 cropland biome   1530               cropland biome   \n",
       "2                    urban biome    973                  urban biome   \n",
       "3                   tundra biome    516                 tundra biome   \n",
       "4              terrestrial biome    483            terrestrial biome   \n",
       "..                           ...    ...                          ...   \n",
       "239                   terestrial      1                   terestrial   \n",
       "240            thermokarst ponds      1            thermokarst ponds   \n",
       "241                 urban boiome      1                 urban boiome   \n",
       "242  wastewater treatment system      1  wastewater treatment system   \n",
       "243         western Amazon basin      1         western Amazon basin   \n",
       "\n",
       "                        splitted  part_count  seq  \\\n",
       "0            ENVO:cropland biome         1.0  1.0   \n",
       "1                 cropland biome         1.0  1.0   \n",
       "2                    urban biome         1.0  1.0   \n",
       "3                   tundra biome         1.0  1.0   \n",
       "4              terrestrial biome         1.0  1.0   \n",
       "..                           ...         ...  ...   \n",
       "239                   terestrial         1.0  1.0   \n",
       "240            thermokarst ponds         1.0  1.0   \n",
       "241                 urban boiome         1.0  1.0   \n",
       "242  wastewater treatment system         1.0  1.0   \n",
       "243         western Amazon basin         1.0  1.0   \n",
       "\n",
       "                       envo_tidy                       string extract  \\\n",
       "0            ENVO:cropland biome          ENVO:cropland biome           \n",
       "1                 cropland biome               cropland biome           \n",
       "2                    urban biome                  urban biome           \n",
       "3                   tundra biome                 tundra biome           \n",
       "4              terrestrial biome            terrestrial biome           \n",
       "..                           ...                          ...     ...   \n",
       "239                   terestrial                   terestrial           \n",
       "240            thermokarst ponds            thermokarst ponds           \n",
       "241                 urban boiome                 urban boiome           \n",
       "242  wastewater treatment system  wastewater treatment system           \n",
       "243         western Amazon basin         western Amazon basin           \n",
       "\n",
       "                remaining_string             remaining_tidied term_id  \\\n",
       "0            ENVO:cropland biome               cropland biome     NaN   \n",
       "1                 cropland biome               cropland biome     NaN   \n",
       "2                    urban biome                  urban biome     NaN   \n",
       "3                   tundra biome                 tundra biome     NaN   \n",
       "4              terrestrial biome            terrestrial biome     NaN   \n",
       "..                           ...                          ...     ...   \n",
       "239                   terestrial                   terestrial     NaN   \n",
       "240            thermokarst ponds            thermokarst ponds     NaN   \n",
       "241                 urban boiome                 urban boiome     NaN   \n",
       "242  wastewater treatment system  wastewater treatment system     NaN   \n",
       "243         western Amazon basin         western amazon basin     NaN   \n",
       "\n",
       "    lab_from_id  lfi_cosine term_id_from_lab              value  \\\n",
       "0                       1.0    ENVO:01000245     cropland biome   \n",
       "1                       1.0    ENVO:01000245     cropland biome   \n",
       "2                       1.0    ENVO:01000249        urban biome   \n",
       "3                       1.0    ENVO:01000180       tundra biome   \n",
       "4                       1.0    ENVO:00000446  terrestrial biome   \n",
       "..          ...         ...              ...                ...   \n",
       "239                     1.0              NaN                NaN   \n",
       "240                     1.0              NaN                NaN   \n",
       "241                     1.0              NaN                NaN   \n",
       "242                     1.0              NaN                NaN   \n",
       "243                     1.0              NaN                NaN   \n",
       "\n",
       "      consensus_id      consensus_lab  id_or_lab_ok  \\\n",
       "0    ENVO:01000245     cropland biome          True   \n",
       "1    ENVO:01000245     cropland biome          True   \n",
       "2    ENVO:01000249        urban biome          True   \n",
       "3    ENVO:01000180       tundra biome          True   \n",
       "4    ENVO:00000446  terrestrial biome          True   \n",
       "..             ...                ...           ...   \n",
       "239            NaN                NaN         False   \n",
       "240            NaN                NaN         False   \n",
       "241            NaN                NaN         False   \n",
       "242            NaN                NaN         False   \n",
       "243            NaN                NaN         False   \n",
       "\n",
       "                   assembled_consensus  \n",
       "0       cropland biome [ENVO:01000245]  \n",
       "1       cropland biome [ENVO:01000245]  \n",
       "2          urban biome [ENVO:01000249]  \n",
       "3         tundra biome [ENVO:01000180]  \n",
       "4    terrestrial biome [ENVO:00000446]  \n",
       "..                                 ...  \n",
       "239                                NaN  \n",
       "240                                NaN  \n",
       "241                                NaN  \n",
       "242                                NaN  \n",
       "243                                NaN  \n",
       "\n",
       "[244 rows x 20 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "bcad296d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_candidates.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47aa8012",
   "metadata": {},
   "source": [
    "## Save these easy term mappings to SQLite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3d8015be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_candidates.to_sql('mapping_scratch', biosample_cnx, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13d44c8d",
   "metadata": {},
   "source": [
    "## For which Biosample annotations were no easy mappings found?\n",
    "How many Biosamples uses those annotations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c29338b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = ~mapping_candidates.id_or_lab_ok\n",
    "needs_search = mapping_candidates.loc[flag]\n",
    "\n",
    "needs_search = needs_search[[\"remaining_tidied\", \"count\"]]\n",
    "\n",
    "sum_by_needed = needs_search.groupby(\"remaining_tidied\")[\"count\"].sum()\n",
    "\n",
    "sum_by_needed = sum_by_needed.to_frame()\n",
    "sum_by_needed[\"remaining_tidied\"] = sum_by_needed.index\n",
    "\n",
    "sum_by_needed = sum_by_needed.sort_values(\"count\", ascending=False)\n",
    "sum_by_needed.reset_index(drop=True)\n",
    "\n",
    "sum_by_needed.to_sql(\"sum_by_needed\", biosample_cnx, if_exists=\"replace\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "91fec97a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>forest</th>\n",
       "      <td>456</td>\n",
       "      <td>forest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tundra</th>\n",
       "      <td>347</td>\n",
       "      <td>tundra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hot springs</th>\n",
       "      <td>172</td>\n",
       "      <td>hot springs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tropical and subtropical moist broadleaf forest biome</th>\n",
       "      <td>170</td>\n",
       "      <td>tropical and subtropical moist broadleaf fores...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>arctic</th>\n",
       "      <td>164</td>\n",
       "      <td>arctic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coastal saline 9</th>\n",
       "      <td>1</td>\n",
       "      <td>coastal saline 9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coastal saline 8</th>\n",
       "      <td>1</td>\n",
       "      <td>coastal saline 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coastal saline 7</th>\n",
       "      <td>1</td>\n",
       "      <td>coastal saline 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coastal saline 6</th>\n",
       "      <td>1</td>\n",
       "      <td>coastal saline 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>western amazon basin</th>\n",
       "      <td>1</td>\n",
       "      <td>western amazon basin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    count  \\\n",
       "remaining_tidied                                            \n",
       "forest                                                456   \n",
       "tundra                                                347   \n",
       "hot springs                                           172   \n",
       "tropical and subtropical moist broadleaf forest...    170   \n",
       "arctic                                                164   \n",
       "...                                                   ...   \n",
       "coastal saline 9                                        1   \n",
       "coastal saline 8                                        1   \n",
       "coastal saline 7                                        1   \n",
       "coastal saline 6                                        1   \n",
       "western amazon basin                                    1   \n",
       "\n",
       "                                                                                     remaining_tidied  \n",
       "remaining_tidied                                                                                       \n",
       "forest                                                                                         forest  \n",
       "tundra                                                                                         tundra  \n",
       "hot springs                                                                               hot springs  \n",
       "tropical and subtropical moist broadleaf forest...  tropical and subtropical moist broadleaf fores...  \n",
       "arctic                                                                                         arctic  \n",
       "...                                                                                               ...  \n",
       "coastal saline 9                                                                     coastal saline 9  \n",
       "coastal saline 8                                                                     coastal saline 8  \n",
       "coastal saline 7                                                                     coastal saline 7  \n",
       "coastal saline 6                                                                     coastal saline 6  \n",
       "western amazon basin                                                             western amazon basin  \n",
       "\n",
       "[163 rows x 2 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_by_needed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53bcffc5",
   "metadata": {},
   "source": [
    "## Extract the tidied strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "02a43f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_raw_list = list(sum_by_needed[\"remaining_tidied\"])\n",
    "ebs_raw_list.sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bd91eb2",
   "metadata": {},
   "source": [
    "## Submit those tidied strings to a search engine\n",
    "\n",
    "Specifically OLS search. This takes roughly one second per unique post-tidied submission\n",
    "\n",
    "_Turn logging back on to show status?_\n",
    "_Print the count and pre- and post- datestamps_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "08e39629",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    ebs_raw_list,\n",
    "    bad_chars=chars_to_whiteout,\n",
    "    cat_name=biosample_col_to_map,\n",
    "    ontoprefix=\"envo,gaz\",\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=my_string_dist_arg,\n",
    ")\n",
    "ebs_search_res.to_sql(\n",
    "    \"ols_search_results\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae3d29dc",
   "metadata": {},
   "source": [
    "## Filter out the best of the acceptable mappings\n",
    "From a string distance perspective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "b695a649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>basin</td>\n",
       "      <td>basin</td>\n",
       "      <td>Basin</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:22223655</td>\n",
       "      <td>Basin</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_22223655</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>subpolar coniferous forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000250</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01000433</td>\n",
       "      <td>needleleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000433</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>forest</td>\n",
       "      <td>forest</td>\n",
       "      <td>Forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00454366</td>\n",
       "      <td>Forest</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00454366</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>grassland</td>\n",
       "      <td>grassland</td>\n",
       "      <td>Grassland</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01001206</td>\n",
       "      <td>grassland ecosystem</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001206</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>790</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:22224982</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_22224982</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>826</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>leaf litter</td>\n",
       "      <td>leaf litter</td>\n",
       "      <td>leaf litter</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01000628</td>\n",
       "      <td>plant litter</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000628</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1017</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>ria de aveiro</td>\n",
       "      <td>ria de aveiro</td>\n",
       "      <td>Ria de Aveiro</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00146341</td>\n",
       "      <td>Ria de Aveiro</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00146341</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1031</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>rice paddy</td>\n",
       "      <td>rice paddy</td>\n",
       "      <td>rice paddy</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000296</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1038</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>rice paddy soils</td>\n",
       "      <td>rice paddy soils</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.034</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005740</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1157</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>taiga</td>\n",
       "      <td>taiga</td>\n",
       "      <td>taiga</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>subpolar coniferous forest biome</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000250</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1233</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>temperate broadleaf and mixed forest</td>\n",
       "      <td>temperate broadleaf and mixed forest</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.052</td>\n",
       "      <td>ENVO:01000389</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000389</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>temperate broadleaf and mixed forests</td>\n",
       "      <td>temperate broadleaf and mixed forests</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.064</td>\n",
       "      <td>ENVO:01000389</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000389</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1327</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>tropical and subtropical coniferous forest biome</td>\n",
       "      <td>tropical and subtropical coniferous forest biome</td>\n",
       "      <td>subtropical coniferous forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.084</td>\n",
       "      <td>ENVO:01000209</td>\n",
       "      <td>subtropical coniferous forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000209</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             category                                               raw  \\\n",
       "197   env_broad_scale                                          bacteria   \n",
       "240   env_broad_scale                                             basin   \n",
       "296   env_broad_scale                                     boreal forest   \n",
       "557   env_broad_scale                                 coniferous forest   \n",
       "688   env_broad_scale                                            forest   \n",
       "773   env_broad_scale                                         grassland   \n",
       "790   env_broad_scale                                       hot springs   \n",
       "826   env_broad_scale                                       leaf litter   \n",
       "1017  env_broad_scale                                     ria de aveiro   \n",
       "1031  env_broad_scale                                        rice paddy   \n",
       "1038  env_broad_scale                                  rice paddy soils   \n",
       "1157  env_broad_scale                                             taiga   \n",
       "1233  env_broad_scale              temperate broadleaf and mixed forest   \n",
       "1247  env_broad_scale             temperate broadleaf and mixed forests   \n",
       "1327  env_broad_scale  tropical and subtropical coniferous forest biome   \n",
       "\n",
       "                                                 query  \\\n",
       "197                                           bacteria   \n",
       "240                                              basin   \n",
       "296                                      boreal forest   \n",
       "557                                  coniferous forest   \n",
       "688                                             forest   \n",
       "773                                          grassland   \n",
       "790                                        hot springs   \n",
       "826                                        leaf litter   \n",
       "1017                                     ria de aveiro   \n",
       "1031                                        rice paddy   \n",
       "1038                                  rice paddy soils   \n",
       "1157                                             taiga   \n",
       "1233              temperate broadleaf and mixed forest   \n",
       "1247             temperate broadleaf and mixed forests   \n",
       "1327  tropical and subtropical coniferous forest biome   \n",
       "\n",
       "                                     name  string_dist_rank  string_dist  \\\n",
       "197                              Bacteria                 1        0.000   \n",
       "240                                 Basin                 1        0.000   \n",
       "296                         boreal forest                 1        0.000   \n",
       "557                     coniferous forest                 1        0.000   \n",
       "688                                Forest                 1        0.000   \n",
       "773                             Grassland                 1        0.000   \n",
       "790                           Hot Springs                 1        0.000   \n",
       "826                           leaf litter                 1        0.000   \n",
       "1017                        Ria de Aveiro                 1        0.000   \n",
       "1031                           rice paddy                 1        0.000   \n",
       "1038                      rice paddy soil                 1        0.034   \n",
       "1157                                taiga                 1        0.000   \n",
       "1233     temperate mixed broadleaf forest                 1        0.052   \n",
       "1247     temperate mixed broadleaf forest                 1        0.064   \n",
       "1327  subtropical coniferous forest biome                 1        0.084   \n",
       "\n",
       "             obo_id                                label  search_rank  \\\n",
       "197     NCBITaxon:2                             Bacteria            1   \n",
       "240    GAZ:22223655                                Basin            1   \n",
       "296   ENVO:01000250     subpolar coniferous forest biome            1   \n",
       "557   ENVO:01000433                    needleleaf forest            1   \n",
       "688    GAZ:00454366                               Forest            1   \n",
       "773   ENVO:01001206                  grassland ecosystem            1   \n",
       "790    GAZ:22224982                          Hot Springs            1   \n",
       "826   ENVO:01000628                         plant litter            1   \n",
       "1017   GAZ:00146341                        Ria de Aveiro            1   \n",
       "1031  ENVO:00000296                           rice field            3   \n",
       "1038  ENVO:00005740                     paddy field soil            2   \n",
       "1157  ENVO:01000250     subpolar coniferous forest biome            2   \n",
       "1233  ENVO:01000389     temperate mixed broadleaf forest            1   \n",
       "1247  ENVO:01000389     temperate mixed broadleaf forest            1   \n",
       "1327  ENVO:01000209  subtropical coniferous forest biome            1   \n",
       "\n",
       "     ontology_prefix                scope        type  \\\n",
       "197             ENVO                label       label   \n",
       "240              GAZ                label       label   \n",
       "296             ENVO   has_narrow_synonym  annotation   \n",
       "557             ENVO              synonym     synonym   \n",
       "688              GAZ                label       label   \n",
       "773             ENVO  has_related_synonym  annotation   \n",
       "790              GAZ                label       label   \n",
       "826             ENVO              synonym     synonym   \n",
       "1017             GAZ                label       label   \n",
       "1031            ENVO  has_related_synonym  annotation   \n",
       "1038            ENVO  has_related_synonym  annotation   \n",
       "1157            ENVO   has_narrow_synonym  annotation   \n",
       "1233            ENVO                label       label   \n",
       "1247            ENVO                label       label   \n",
       "1327            ENVO                label       label   \n",
       "\n",
       "                                               iri ontology_name  \n",
       "197     http://purl.obolibrary.org/obo/NCBITaxon_2          envo  \n",
       "240    http://purl.obolibrary.org/obo/GAZ_22223655           gaz  \n",
       "296   http://purl.obolibrary.org/obo/ENVO_01000250          envo  \n",
       "557   http://purl.obolibrary.org/obo/ENVO_01000433          envo  \n",
       "688    http://purl.obolibrary.org/obo/GAZ_00454366           gaz  \n",
       "773   http://purl.obolibrary.org/obo/ENVO_01001206          envo  \n",
       "790    http://purl.obolibrary.org/obo/GAZ_22224982           gaz  \n",
       "826   http://purl.obolibrary.org/obo/ENVO_01000628          envo  \n",
       "1017   http://purl.obolibrary.org/obo/GAZ_00146341           gaz  \n",
       "1031  http://purl.obolibrary.org/obo/ENVO_00000296          envo  \n",
       "1038  http://purl.obolibrary.org/obo/ENVO_00005740          envo  \n",
       "1157  http://purl.obolibrary.org/obo/ENVO_01000250          envo  \n",
       "1233  http://purl.obolibrary.org/obo/ENVO_01000389          envo  \n",
       "1247  http://purl.obolibrary.org/obo/ENVO_01000389          envo  \n",
       "1327  http://purl.obolibrary.org/obo/ENVO_01000209          envo  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    ebs_search_res, max_string_dist=my_max_string_dist\n",
    ")\n",
    "my_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43aa5018",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Filter out the submissions with no acceptable matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "39e98bf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>a semi arid grassland soil fungi</td>\n",
       "      <td>a semi arid grassland soil fungi</td>\n",
       "      <td>grassland soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.297</td>\n",
       "      <td>ENVO:00005750</td>\n",
       "      <td>grassland soil</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005750</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>a semi arid grassland soil fungi</td>\n",
       "      <td>a semi arid grassland soil fungi</td>\n",
       "      <td>flooded grassland biome</td>\n",
       "      <td>2</td>\n",
       "      <td>0.551</td>\n",
       "      <td>ENVO:01000195</td>\n",
       "      <td>flooded grassland biome</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000195</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>a semi arid grassland soil fungi</td>\n",
       "      <td>a semi arid grassland soil fungi</td>\n",
       "      <td>semi-arid climate</td>\n",
       "      <td>3</td>\n",
       "      <td>0.662</td>\n",
       "      <td>ENVO:01000378</td>\n",
       "      <td>arid subtropical</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000378</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>a semi arid grassland soil fungi</td>\n",
       "      <td>a semi arid grassland soil fungi</td>\n",
       "      <td>semi-arid climate</td>\n",
       "      <td>4</td>\n",
       "      <td>0.662</td>\n",
       "      <td>ENVO:01000378</td>\n",
       "      <td>arid subtropical</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>hasRelatedSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000378</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1375</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>River Amazon</td>\n",
       "      <td>7</td>\n",
       "      <td>0.539</td>\n",
       "      <td>GAZ:00047130</td>\n",
       "      <td>River Amazon</td>\n",
       "      <td>4</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00047130</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1371</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>Amazonia</td>\n",
       "      <td>8</td>\n",
       "      <td>0.588</td>\n",
       "      <td>GAZ:00006844</td>\n",
       "      <td>Amazonia</td>\n",
       "      <td>2</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00006844</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1379</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>Guiana Basin</td>\n",
       "      <td>9</td>\n",
       "      <td>0.671</td>\n",
       "      <td>GAZ:00143796</td>\n",
       "      <td>Guiana Basin</td>\n",
       "      <td>5</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00143796</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>rio amazonas {language: portuguese}</td>\n",
       "      <td>10</td>\n",
       "      <td>0.709</td>\n",
       "      <td>GAZ:00047130</td>\n",
       "      <td>River Amazon</td>\n",
       "      <td>4</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00047130</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1380</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>makaroff deep {alternative name}</td>\n",
       "      <td>11</td>\n",
       "      <td>0.810</td>\n",
       "      <td>GAZ:00143796</td>\n",
       "      <td>Guiana Basin</td>\n",
       "      <td>5</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00143796</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1126 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             category                               raw  \\\n",
       "0     env_broad_scale                                     \n",
       "8     env_broad_scale  a semi arid grassland soil fungi   \n",
       "10    env_broad_scale  a semi arid grassland soil fungi   \n",
       "7     env_broad_scale  a semi arid grassland soil fungi   \n",
       "6     env_broad_scale  a semi arid grassland soil fungi   \n",
       "...               ...                               ...   \n",
       "1375  env_broad_scale              western amazon basin   \n",
       "1371  env_broad_scale              western amazon basin   \n",
       "1379  env_broad_scale              western amazon basin   \n",
       "1376  env_broad_scale              western amazon basin   \n",
       "1380  env_broad_scale              western amazon basin   \n",
       "\n",
       "                                 query                                 name  \\\n",
       "0                                                                             \n",
       "8     a semi arid grassland soil fungi                       grassland soil   \n",
       "10    a semi arid grassland soil fungi              flooded grassland biome   \n",
       "7     a semi arid grassland soil fungi                    semi-arid climate   \n",
       "6     a semi arid grassland soil fungi                    semi-arid climate   \n",
       "...                                ...                                  ...   \n",
       "1375              western amazon basin                         River Amazon   \n",
       "1371              western amazon basin                             Amazonia   \n",
       "1379              western amazon basin                         Guiana Basin   \n",
       "1376              western amazon basin  rio amazonas {language: portuguese}   \n",
       "1380              western amazon basin     makaroff deep {alternative name}   \n",
       "\n",
       "      string_dist_rank  string_dist         obo_id                    label  \\\n",
       "0                    1        0.000                                           \n",
       "8                    1        0.297  ENVO:00005750           grassland soil   \n",
       "10                   2        0.551  ENVO:01000195  flooded grassland biome   \n",
       "7                    3        0.662  ENVO:01000378         arid subtropical   \n",
       "6                    4        0.662  ENVO:01000378         arid subtropical   \n",
       "...                ...          ...            ...                      ...   \n",
       "1375                 7        0.539   GAZ:00047130             River Amazon   \n",
       "1371                 8        0.588   GAZ:00006844                 Amazonia   \n",
       "1379                 9        0.671   GAZ:00143796             Guiana Basin   \n",
       "1376                10        0.709   GAZ:00047130             River Amazon   \n",
       "1380                11        0.810   GAZ:00143796             Guiana Basin   \n",
       "\n",
       "      search_rank ontology_prefix                scope        type  \\\n",
       "0               1                                  NaN         NaN   \n",
       "8               2            ENVO                label       label   \n",
       "10              4            ENVO                label       label   \n",
       "7               1            ENVO  has_related_synonym  annotation   \n",
       "6               1            ENVO    hasRelatedSynonym        None   \n",
       "...           ...             ...                  ...         ...   \n",
       "1375            4             GAZ                label       label   \n",
       "1371            2             GAZ                label       label   \n",
       "1379            5             GAZ                label       label   \n",
       "1376            4             GAZ              synonym     synonym   \n",
       "1380            5             GAZ              synonym     synonym   \n",
       "\n",
       "                                               iri ontology_name  \n",
       "0                                                                 \n",
       "8     http://purl.obolibrary.org/obo/ENVO_00005750          envo  \n",
       "10    http://purl.obolibrary.org/obo/ENVO_01000195          envo  \n",
       "7     http://purl.obolibrary.org/obo/ENVO_01000378          envo  \n",
       "6     http://purl.obolibrary.org/obo/ENVO_01000378          envo  \n",
       "...                                            ...           ...  \n",
       "1375   http://purl.obolibrary.org/obo/GAZ_00047130           gaz  \n",
       "1371   http://purl.obolibrary.org/obo/GAZ_00006844           gaz  \n",
       "1379   http://purl.obolibrary.org/obo/GAZ_00143796           gaz  \n",
       "1376   http://purl.obolibrary.org/obo/GAZ_00047130           gaz  \n",
       "1380   http://purl.obolibrary.org/obo/GAZ_00143796           gaz  \n",
       "\n",
       "[1126 rows x 14 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_acceptable_mappings = scoped_mapping.get_no_acceptable_mappings(\n",
    "    ebs_search_res, my_best_acceptable\n",
    ")\n",
    "\n",
    "no_acceptable_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70be6244",
   "metadata": {},
   "source": [
    "## Try searching the failures against all ontologies in OLS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ac5c4f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "still_unmapped = list(set(list(no_acceptable_mappings[\"raw\"])))\n",
    "still_unmapped.sort()\n",
    "\n",
    "salvage_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    still_unmapped,\n",
    "    bad_chars=\"._-\",\n",
    "    cat_name=\"salvage\",\n",
    "    ontoprefix=\"\",\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=2,\n",
    ")\n",
    "\n",
    "salvage_search_res.to_sql(\n",
    "    \"salvage_search_res\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "453b8356",
   "metadata": {},
   "source": [
    "## We appear to be at a point of diminishing returns\n",
    "At the very least, will require some review"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "d44fe45b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>salvage</td>\n",
       "      <td>arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C44738</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C44738</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>salvage</td>\n",
       "      <td>artificial soil</td>\n",
       "      <td>artificial soil</td>\n",
       "      <td>Artificial</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>NCIT:C61464</td>\n",
       "      <td>Artificial</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C61464</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>salvage</td>\n",
       "      <td>bamboo</td>\n",
       "      <td>bamboo</td>\n",
       "      <td>Bamboo Dam</td>\n",
       "      <td>1</td>\n",
       "      <td>0.191</td>\n",
       "      <td>GAZ:00148513</td>\n",
       "      <td>Bamboo Dam</td>\n",
       "      <td>2</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00148513</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>salvage</td>\n",
       "      <td>beech forest</td>\n",
       "      <td>beech forest</td>\n",
       "      <td>beech forest soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.171</td>\n",
       "      <td>ENVO:00005770</td>\n",
       "      <td>beech forest soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005770</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983</th>\n",
       "      <td>salvage</td>\n",
       "      <td>cold desert soil</td>\n",
       "      <td>cold desert soil</td>\n",
       "      <td>cold desert</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>ENVO:01000382</td>\n",
       "      <td>cold desert</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000382</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1026</th>\n",
       "      <td>salvage</td>\n",
       "      <td>cotton</td>\n",
       "      <td>cotton</td>\n",
       "      <td>cotton</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C77117</td>\n",
       "      <td>Cotton Rat</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C77117</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1072</th>\n",
       "      <td>salvage</td>\n",
       "      <td>crop</td>\n",
       "      <td>crop</td>\n",
       "      <td>Crop</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C111162</td>\n",
       "      <td>Crop</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C111162</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>salvage</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:03000039</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000039</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1361</th>\n",
       "      <td>salvage</td>\n",
       "      <td>deciduous broadleaf forest</td>\n",
       "      <td>deciduous broadleaf forest</td>\n",
       "      <td>tropical deciduous broadleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.133</td>\n",
       "      <td>ENVO:01000387</td>\n",
       "      <td>tropical deciduous broadleaf forest</td>\n",
       "      <td>4</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000387</td>\n",
       "      <td>micro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390</th>\n",
       "      <td>salvage</td>\n",
       "      <td>deyeuxia arundinacea</td>\n",
       "      <td>deyeuxia arundinacea</td>\n",
       "      <td>Deyeuxia arundinacea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:1767996</td>\n",
       "      <td>Deyeuxia arundinacea</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1767996</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1397</th>\n",
       "      <td>salvage</td>\n",
       "      <td>ectomycorrhizae</td>\n",
       "      <td>ectomycorrhizae</td>\n",
       "      <td>ectomycorrhiza</td>\n",
       "      <td>1</td>\n",
       "      <td>0.036</td>\n",
       "      <td>BTO:0002237</td>\n",
       "      <td>ectomycorrhiza</td>\n",
       "      <td>1</td>\n",
       "      <td>BTO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/BTO_0002237</td>\n",
       "      <td>bto</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>salvage</td>\n",
       "      <td>experimental</td>\n",
       "      <td>experimental</td>\n",
       "      <td>experimental</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C28041</td>\n",
       "      <td>Investigational</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C28041</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1576</th>\n",
       "      <td>salvage</td>\n",
       "      <td>fusarium</td>\n",
       "      <td>fusarium</td>\n",
       "      <td>Fusarium</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C77185</td>\n",
       "      <td>Fusarium</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C77185</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1647</th>\n",
       "      <td>salvage</td>\n",
       "      <td>healthy</td>\n",
       "      <td>healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C115935</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C115935</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1704</th>\n",
       "      <td>salvage</td>\n",
       "      <td>konjac</td>\n",
       "      <td>konjac</td>\n",
       "      <td>konjac</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>FOODON:03414400</td>\n",
       "      <td>konjac</td>\n",
       "      <td>1</td>\n",
       "      <td>FOODON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/FOODON_03414400</td>\n",
       "      <td>foodon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1744</th>\n",
       "      <td>salvage</td>\n",
       "      <td>lab experiment</td>\n",
       "      <td>lab experiment</td>\n",
       "      <td>Experiment</td>\n",
       "      <td>1</td>\n",
       "      <td>0.168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Experiment</td>\n",
       "      <td>1</td>\n",
       "      <td>REPR</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>https://w3id.org/reproduceme#Experiment</td>\n",
       "      <td>reproduceme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1820</th>\n",
       "      <td>salvage</td>\n",
       "      <td>microorganism</td>\n",
       "      <td>microorganism</td>\n",
       "      <td>Microorganism</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C14329</td>\n",
       "      <td>Microorganism</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14329</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979</th>\n",
       "      <td>salvage</td>\n",
       "      <td>montane shrubland</td>\n",
       "      <td>montane shrubland</td>\n",
       "      <td>montane shrubland biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.134</td>\n",
       "      <td>ENVO:01000216</td>\n",
       "      <td>montane shrubland biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000216</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2009</th>\n",
       "      <td>salvage</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C48660</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C48660</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2064</th>\n",
       "      <td>salvage</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005740</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156</th>\n",
       "      <td>salvage</td>\n",
       "      <td>plant species clypeola jonthlaspi</td>\n",
       "      <td>plant species clypeola jonthlaspi</td>\n",
       "      <td>Clypeola jonthlaspi</td>\n",
       "      <td>1</td>\n",
       "      <td>0.157</td>\n",
       "      <td>NCBITaxon:457792</td>\n",
       "      <td>Clypeola jonthlaspi</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_457792</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2164</th>\n",
       "      <td>salvage</td>\n",
       "      <td>polar desert</td>\n",
       "      <td>polar desert</td>\n",
       "      <td>polar desert biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.196</td>\n",
       "      <td>ENVO:01000186</td>\n",
       "      <td>polar desert biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000186</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2175</th>\n",
       "      <td>salvage</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C76196</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C76196</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2321</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rainforest</td>\n",
       "      <td>rainforest</td>\n",
       "      <td>Rainforest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>OMIT:0028792</td>\n",
       "      <td>Rainforest</td>\n",
       "      <td>1</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0028792</td>\n",
       "      <td>omit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2447</th>\n",
       "      <td>salvage</td>\n",
       "      <td>spruce forest</td>\n",
       "      <td>spruce forest</td>\n",
       "      <td>spruce forest soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.160</td>\n",
       "      <td>ENVO:00005784</td>\n",
       "      <td>spruce forest soil</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005784</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500</th>\n",
       "      <td>salvage</td>\n",
       "      <td>temperate broadleaf and mixed forest biome</td>\n",
       "      <td>temperate broadleaf and mixed forest biome</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.111</td>\n",
       "      <td>ENVO:01000389</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000389</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513</th>\n",
       "      <td>salvage</td>\n",
       "      <td>temperate coniferous forests</td>\n",
       "      <td>temperate coniferous forests</td>\n",
       "      <td>temperate coniferous forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.102</td>\n",
       "      <td>ENVO:01000211</td>\n",
       "      <td>temperate coniferous forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000211</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2524</th>\n",
       "      <td>salvage</td>\n",
       "      <td>temperate deciduous forest</td>\n",
       "      <td>temperate deciduous forest</td>\n",
       "      <td>temperate deciduous broadleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.146</td>\n",
       "      <td>ENVO:01000385</td>\n",
       "      <td>temperate deciduous broadleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000385</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2568</th>\n",
       "      <td>salvage</td>\n",
       "      <td>terrestrial</td>\n",
       "      <td>terrestrial</td>\n",
       "      <td>terrestrial realm</td>\n",
       "      <td>1</td>\n",
       "      <td>0.151</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000446</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2602</th>\n",
       "      <td>salvage</td>\n",
       "      <td>tropical and subtropical moist broadleaf fores...</td>\n",
       "      <td>tropical and subtropical moist broadleaf fores...</td>\n",
       "      <td>subtropical broadleaf forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.107</td>\n",
       "      <td>ENVO:01000201</td>\n",
       "      <td>subtropical broadleaf forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENM</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000201</td>\n",
       "      <td>enm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2643</th>\n",
       "      <td>salvage</td>\n",
       "      <td>tundra</td>\n",
       "      <td>tundra</td>\n",
       "      <td>tundra</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00000112</td>\n",
       "      <td>tundra</td>\n",
       "      <td>1</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000112</td>\n",
       "      <td>micro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2787</th>\n",
       "      <td>salvage</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment plant</td>\n",
       "      <td>1</td>\n",
       "      <td>0.153</td>\n",
       "      <td>ENVO:00002043</td>\n",
       "      <td>wastewater treatment plant</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002043</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2794</th>\n",
       "      <td>salvage</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>Western Basin</td>\n",
       "      <td>1</td>\n",
       "      <td>0.181</td>\n",
       "      <td>GAZ:00084622</td>\n",
       "      <td>Western Basin</td>\n",
       "      <td>3</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00084622</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     category                                                raw  \\\n",
       "106   salvage                                             arctic   \n",
       "287   salvage                                    artificial soil   \n",
       "319   salvage                                             bamboo   \n",
       "389   salvage                                       beech forest   \n",
       "983   salvage                                   cold desert soil   \n",
       "1026  salvage                                             cotton   \n",
       "1072  salvage                                               crop   \n",
       "1095  salvage                                         cryoconite   \n",
       "1361  salvage                         deciduous broadleaf forest   \n",
       "1390  salvage                               deyeuxia arundinacea   \n",
       "1397  salvage                                    ectomycorrhizae   \n",
       "1428  salvage                                       experimental   \n",
       "1576  salvage                                           fusarium   \n",
       "1647  salvage                                            healthy   \n",
       "1704  salvage                                             konjac   \n",
       "1744  salvage                                     lab experiment   \n",
       "1820  salvage                                      microorganism   \n",
       "1979  salvage                                  montane shrubland   \n",
       "2009  salvage                                     not applicable   \n",
       "2064  salvage                                         paddy soil   \n",
       "2156  salvage                  plant species clypeola jonthlaspi   \n",
       "2164  salvage                                       polar desert   \n",
       "2175  salvage                                     proteobacteria   \n",
       "2321  salvage                                         rainforest   \n",
       "2447  salvage                                      spruce forest   \n",
       "2500  salvage         temperate broadleaf and mixed forest biome   \n",
       "2513  salvage                       temperate coniferous forests   \n",
       "2524  salvage                         temperate deciduous forest   \n",
       "2568  salvage                                        terrestrial   \n",
       "2602  salvage  tropical and subtropical moist broadleaf fores...   \n",
       "2643  salvage                                             tundra   \n",
       "2787  salvage                        wastewater treatment system   \n",
       "2794  salvage                               western amazon basin   \n",
       "\n",
       "                                                  query  \\\n",
       "106                                              arctic   \n",
       "287                                     artificial soil   \n",
       "319                                              bamboo   \n",
       "389                                        beech forest   \n",
       "983                                    cold desert soil   \n",
       "1026                                             cotton   \n",
       "1072                                               crop   \n",
       "1095                                         cryoconite   \n",
       "1361                         deciduous broadleaf forest   \n",
       "1390                               deyeuxia arundinacea   \n",
       "1397                                    ectomycorrhizae   \n",
       "1428                                       experimental   \n",
       "1576                                           fusarium   \n",
       "1647                                            healthy   \n",
       "1704                                             konjac   \n",
       "1744                                     lab experiment   \n",
       "1820                                      microorganism   \n",
       "1979                                  montane shrubland   \n",
       "2009                                     not applicable   \n",
       "2064                                         paddy soil   \n",
       "2156                  plant species clypeola jonthlaspi   \n",
       "2164                                       polar desert   \n",
       "2175                                     proteobacteria   \n",
       "2321                                         rainforest   \n",
       "2447                                      spruce forest   \n",
       "2500         temperate broadleaf and mixed forest biome   \n",
       "2513                       temperate coniferous forests   \n",
       "2524                         temperate deciduous forest   \n",
       "2568                                        terrestrial   \n",
       "2602  tropical and subtropical moist broadleaf fores...   \n",
       "2643                                             tundra   \n",
       "2787                        wastewater treatment system   \n",
       "2794                               western amazon basin   \n",
       "\n",
       "                                      name  string_dist_rank  string_dist  \\\n",
       "106                                 Arctic                 1        0.000   \n",
       "287                             Artificial                 1        0.198   \n",
       "319                             Bamboo Dam                 1        0.191   \n",
       "389                      beech forest soil                 1        0.171   \n",
       "983                            cold desert                 1        0.184   \n",
       "1026                                cotton                 1        0.000   \n",
       "1072                                  Crop                 1        0.000   \n",
       "1095                       cryoconite hole                 1        0.198   \n",
       "1361   tropical deciduous broadleaf forest                 1        0.133   \n",
       "1390                  Deyeuxia arundinacea                 1        0.000   \n",
       "1397                        ectomycorrhiza                 1        0.036   \n",
       "1428                          experimental                 1        0.000   \n",
       "1576                              Fusarium                 1        0.000   \n",
       "1647                               Healthy                 1        0.000   \n",
       "1704                                konjac                 1        0.000   \n",
       "1744                            Experiment                 1        0.168   \n",
       "1820                         Microorganism                 1        0.000   \n",
       "1979               montane shrubland biome                 1        0.134   \n",
       "2009                        Not Applicable                 1        0.000   \n",
       "2064                       rice paddy soil                 1        0.198   \n",
       "2156                   Clypeola jonthlaspi                 1        0.157   \n",
       "2164                    polar desert biome                 1        0.196   \n",
       "2175                        Proteobacteria                 1        0.000   \n",
       "2321                            Rainforest                 1        0.000   \n",
       "2447                    spruce forest soil                 1        0.160   \n",
       "2500      temperate mixed broadleaf forest                 1        0.111   \n",
       "2513     temperate coniferous forest biome                 1        0.102   \n",
       "2524  temperate deciduous broadleaf forest                 1        0.146   \n",
       "2568                     terrestrial realm                 1        0.151   \n",
       "2602    subtropical broadleaf forest biome                 1        0.107   \n",
       "2643                                tundra                 1        0.000   \n",
       "2787            wastewater treatment plant                 1        0.153   \n",
       "2794                         Western Basin                 1        0.181   \n",
       "\n",
       "                 obo_id                                 label  search_rank  \\\n",
       "106         NCIT:C44738                                Arctic            1   \n",
       "287         NCIT:C61464                            Artificial            1   \n",
       "319        GAZ:00148513                            Bamboo Dam            2   \n",
       "389       ENVO:00005770                     beech forest soil            1   \n",
       "983       ENVO:01000382                           cold desert            1   \n",
       "1026        NCIT:C77117                            Cotton Rat            1   \n",
       "1072       NCIT:C111162                                  Crop            1   \n",
       "1095      ENVO:03000039                       cryoconite hole            2   \n",
       "1361      ENVO:01000387   tropical deciduous broadleaf forest            4   \n",
       "1390  NCBITaxon:1767996                  Deyeuxia arundinacea            1   \n",
       "1397        BTO:0002237                        ectomycorrhiza            1   \n",
       "1428        NCIT:C28041                       Investigational            1   \n",
       "1576        NCIT:C77185                              Fusarium            1   \n",
       "1647       NCIT:C115935                               Healthy            1   \n",
       "1704    FOODON:03414400                                konjac            1   \n",
       "1744                NaN                            Experiment            1   \n",
       "1820        NCIT:C14329                         Microorganism            1   \n",
       "1979      ENVO:01000216               montane shrubland biome            1   \n",
       "2009        NCIT:C48660                        Not Applicable            1   \n",
       "2064      ENVO:00005740                      paddy field soil            1   \n",
       "2156   NCBITaxon:457792                   Clypeola jonthlaspi            1   \n",
       "2164      ENVO:01000186                    polar desert biome            1   \n",
       "2175        NCIT:C76196                        Proteobacteria            1   \n",
       "2321       OMIT:0028792                            Rainforest            1   \n",
       "2447      ENVO:00005784                    spruce forest soil            2   \n",
       "2500      ENVO:01000389      temperate mixed broadleaf forest            1   \n",
       "2513      ENVO:01000211     temperate coniferous forest biome            1   \n",
       "2524      ENVO:01000385  temperate deciduous broadleaf forest            1   \n",
       "2568      ENVO:00000446                     terrestrial biome            1   \n",
       "2602      ENVO:01000201    subtropical broadleaf forest biome            1   \n",
       "2643      ENVO:00000112                                tundra            1   \n",
       "2787      ENVO:00002043            wastewater treatment plant            4   \n",
       "2794       GAZ:00084622                         Western Basin            3   \n",
       "\n",
       "     ontology_prefix                scope        type  \\\n",
       "106             NCIT                label       label   \n",
       "287             NCIT                label       label   \n",
       "319              GAZ                label       label   \n",
       "389             ENVO                label       label   \n",
       "983             ENVO                label       label   \n",
       "1026            NCIT              synonym     synonym   \n",
       "1072            NCIT                label       label   \n",
       "1095            ENVO                label       label   \n",
       "1361           MICRO                label       label   \n",
       "1390       NCBITAXON                label       label   \n",
       "1397             BTO                label       label   \n",
       "1428            NCIT              synonym     synonym   \n",
       "1576            NCIT                label       label   \n",
       "1647            NCIT                label       label   \n",
       "1704          FOODON                label       label   \n",
       "1744            REPR                label       label   \n",
       "1820            NCIT                label       label   \n",
       "1979            ENVO                label       label   \n",
       "2009            NCIT                label       label   \n",
       "2064            ENVO  has_related_synonym  annotation   \n",
       "2156       NCBITAXON                label       label   \n",
       "2164            ENVO                label       label   \n",
       "2175            NCIT                label       label   \n",
       "2321            OMIT                label       label   \n",
       "2447            ENVO                label       label   \n",
       "2500            ENVO                label       label   \n",
       "2513            ENVO                label       label   \n",
       "2524            ENVO                label       label   \n",
       "2568            ENVO  has_related_synonym  annotation   \n",
       "2602             ENM                label       label   \n",
       "2643           MICRO                label       label   \n",
       "2787            ENVO                label       label   \n",
       "2794             GAZ                label       label   \n",
       "\n",
       "                                                   iri ontology_name  \n",
       "106         http://purl.obolibrary.org/obo/NCIT_C44738          ncit  \n",
       "287         http://purl.obolibrary.org/obo/NCIT_C61464          ncit  \n",
       "319        http://purl.obolibrary.org/obo/GAZ_00148513           gaz  \n",
       "389       http://purl.obolibrary.org/obo/ENVO_00005770          envo  \n",
       "983       http://purl.obolibrary.org/obo/ENVO_01000382          envo  \n",
       "1026        http://purl.obolibrary.org/obo/NCIT_C77117          ncit  \n",
       "1072       http://purl.obolibrary.org/obo/NCIT_C111162          ncit  \n",
       "1095      http://purl.obolibrary.org/obo/ENVO_03000039          envo  \n",
       "1361      http://purl.obolibrary.org/obo/ENVO_01000387         micro  \n",
       "1390  http://purl.obolibrary.org/obo/NCBITaxon_1767996     ncbitaxon  \n",
       "1397        http://purl.obolibrary.org/obo/BTO_0002237           bto  \n",
       "1428        http://purl.obolibrary.org/obo/NCIT_C28041          ncit  \n",
       "1576        http://purl.obolibrary.org/obo/NCIT_C77185          ncit  \n",
       "1647       http://purl.obolibrary.org/obo/NCIT_C115935          ncit  \n",
       "1704    http://purl.obolibrary.org/obo/FOODON_03414400        foodon  \n",
       "1744           https://w3id.org/reproduceme#Experiment   reproduceme  \n",
       "1820        http://purl.obolibrary.org/obo/NCIT_C14329          ncit  \n",
       "1979      http://purl.obolibrary.org/obo/ENVO_01000216          envo  \n",
       "2009        http://purl.obolibrary.org/obo/NCIT_C48660          ncit  \n",
       "2064      http://purl.obolibrary.org/obo/ENVO_00005740          envo  \n",
       "2156   http://purl.obolibrary.org/obo/NCBITaxon_457792     ncbitaxon  \n",
       "2164      http://purl.obolibrary.org/obo/ENVO_01000186          envo  \n",
       "2175        http://purl.obolibrary.org/obo/NCIT_C76196          ncit  \n",
       "2321       http://purl.obolibrary.org/obo/OMIT_0028792          omit  \n",
       "2447      http://purl.obolibrary.org/obo/ENVO_00005784          envo  \n",
       "2500      http://purl.obolibrary.org/obo/ENVO_01000389          envo  \n",
       "2513      http://purl.obolibrary.org/obo/ENVO_01000211          envo  \n",
       "2524      http://purl.obolibrary.org/obo/ENVO_01000385          envo  \n",
       "2568      http://purl.obolibrary.org/obo/ENVO_00000446          envo  \n",
       "2602      http://purl.obolibrary.org/obo/ENVO_01000201           enm  \n",
       "2643      http://purl.obolibrary.org/obo/ENVO_00000112         micro  \n",
       "2787      http://purl.obolibrary.org/obo/ENVO_00002043          envo  \n",
       "2794       http://purl.obolibrary.org/obo/GAZ_00084622           gaz  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "salvage_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    salvage_search_res, max_string_dist=0.2\n",
    ")\n",
    "salvage_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6bd9e5e",
   "metadata": {},
   "source": [
    "## Save the mappings for review!\n",
    "\n",
    "_Note that the merge on ID or label mappings are not included here_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a2b88c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_best_acceptable[include_col] = first_pass_include_val\n",
    "salvage_best_acceptable[include_col] = salvage_include_val\n",
    "best_and_salvage = pd.concat([my_best_acceptable, salvage_best_acceptable])\n",
    "\n",
    "best_and_salvage[\"ols_curation_notes\"] = \"\"\n",
    "\n",
    "best_and_salvage.to_csv(ols_review_file, sep=ols_review_seperator, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "4035647e",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_and_salvage.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f4a3988",
   "metadata": {},
   "outputs": [],
   "source": [
    "raise SystemExit(\"Don't skip reviewing and saving!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a331a6",
   "metadata": {},
   "source": [
    "## Do some review in a spreadsheet application!\n",
    "Specifically, columns `include` and `ols_curation_notes`\n",
    "\n",
    "Then Save\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be6da72c",
   "metadata": {},
   "source": [
    "## Now read the reviewed spreadsheet back in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c94cf3b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "curated = pd.read_csv(ols_review_file, sep=ols_review_seperator)\n",
    "curated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "173ce72c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>include</th>\n",
       "      <th>ols_curation_notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ectomycorrhizae</td>\n",
       "      <td>ectomycorrhizae</td>\n",
       "      <td>ectomycorrhiza</td>\n",
       "      <td>BTO:0002237</td>\n",
       "      <td>ectomycorrhiza</td>\n",
       "      <td>BTO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>bto</td>\n",
       "      <td>False</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tropical and subtropical moist broadleaf fores...</td>\n",
       "      <td>tropical and subtropical moist broadleaf fores...</td>\n",
       "      <td>subtropical broadleaf forest biome</td>\n",
       "      <td>ENVO:01000201</td>\n",
       "      <td>subtropical broadleaf forest biome</td>\n",
       "      <td>ENM</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>enm</td>\n",
       "      <td>False</td>\n",
       "      <td>Omits tropical and moose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>beech forest</td>\n",
       "      <td>beech forest</td>\n",
       "      <td>beech forest soil</td>\n",
       "      <td>ENVO:00005770</td>\n",
       "      <td>beech forest soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "      <td>Adds soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cold desert soil</td>\n",
       "      <td>cold desert soil</td>\n",
       "      <td>cold desert</td>\n",
       "      <td>ENVO:01000382</td>\n",
       "      <td>cold desert</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "      <td>Loses soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>ENVO:03000039</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "      <td>Adds hole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>spruce forest</td>\n",
       "      <td>spruce forest</td>\n",
       "      <td>spruce forest soil</td>\n",
       "      <td>ENVO:00005784</td>\n",
       "      <td>spruce forest soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "      <td>Adds soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>temperate broadleaf and mixed forest biome</td>\n",
       "      <td>temperate broadleaf and mixed forest biome</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO:01000389</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "      <td>Pretty close</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>temperate deciduous forest</td>\n",
       "      <td>temperate deciduous forest</td>\n",
       "      <td>temperate deciduous broadleaf forest</td>\n",
       "      <td>ENVO:01000385</td>\n",
       "      <td>temperate deciduous broadleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "      <td>Adds broadleaf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>konjac</td>\n",
       "      <td>konjac</td>\n",
       "      <td>konjac</td>\n",
       "      <td>FOODON:03414400</td>\n",
       "      <td>konjac</td>\n",
       "      <td>FOODON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>foodon</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>bamboo</td>\n",
       "      <td>bamboo</td>\n",
       "      <td>Bamboo Dam</td>\n",
       "      <td>GAZ:00148513</td>\n",
       "      <td>Bamboo Dam</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>western amazon basin</td>\n",
       "      <td>Western Basin</td>\n",
       "      <td>GAZ:00084622</td>\n",
       "      <td>Western Basin</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>False</td>\n",
       "      <td>No, this is near lake erie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>deciduous broadleaf forest</td>\n",
       "      <td>deciduous broadleaf forest</td>\n",
       "      <td>tropical deciduous broadleaf forest</td>\n",
       "      <td>ENVO:01000387</td>\n",
       "      <td>tropical deciduous broadleaf forest</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>micro</td>\n",
       "      <td>False</td>\n",
       "      <td>Adds tropical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>NCIT:C44738</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>artificial soil</td>\n",
       "      <td>artificial soil</td>\n",
       "      <td>Artificial</td>\n",
       "      <td>NCIT:C61464</td>\n",
       "      <td>Artificial</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>cotton</td>\n",
       "      <td>cotton</td>\n",
       "      <td>cotton</td>\n",
       "      <td>NCIT:C77117</td>\n",
       "      <td>Cotton Rat</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>crop</td>\n",
       "      <td>crop</td>\n",
       "      <td>Crop</td>\n",
       "      <td>NCIT:C111162</td>\n",
       "      <td>Crop</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>experimental</td>\n",
       "      <td>experimental</td>\n",
       "      <td>experimental</td>\n",
       "      <td>NCIT:C28041</td>\n",
       "      <td>Investigational</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>fusarium</td>\n",
       "      <td>fusarium</td>\n",
       "      <td>Fusarium</td>\n",
       "      <td>NCIT:C77185</td>\n",
       "      <td>Fusarium</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>healthy</td>\n",
       "      <td>healthy</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NCIT:C115935</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>microorganism</td>\n",
       "      <td>microorganism</td>\n",
       "      <td>Microorganism</td>\n",
       "      <td>NCIT:C14329</td>\n",
       "      <td>Microorganism</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NCIT:C48660</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>NCIT:C76196</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rainforest</td>\n",
       "      <td>rainforest</td>\n",
       "      <td>Rainforest</td>\n",
       "      <td>OMIT:0028792</td>\n",
       "      <td>Rainforest</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>omit</td>\n",
       "      <td>False</td>\n",
       "      <td>Avoid omit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>lab experiment</td>\n",
       "      <td>lab experiment</td>\n",
       "      <td>Experiment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Experiment</td>\n",
       "      <td>REPR</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>reproduceme</td>\n",
       "      <td>False</td>\n",
       "      <td>No id, avoid repo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>montane shrubland</td>\n",
       "      <td>montane shrubland</td>\n",
       "      <td>montane shrubland biome</td>\n",
       "      <td>ENVO:01000216</td>\n",
       "      <td>montane shrubland biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>paddy soil</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>polar desert</td>\n",
       "      <td>polar desert</td>\n",
       "      <td>polar desert biome</td>\n",
       "      <td>ENVO:01000186</td>\n",
       "      <td>polar desert biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>temperate coniferous forests</td>\n",
       "      <td>temperate coniferous forests</td>\n",
       "      <td>temperate coniferous forest biome</td>\n",
       "      <td>ENVO:01000211</td>\n",
       "      <td>temperate coniferous forest biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>terrestrial</td>\n",
       "      <td>terrestrial</td>\n",
       "      <td>terrestrial realm</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment plant</td>\n",
       "      <td>ENVO:00002043</td>\n",
       "      <td>wastewater treatment plant</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>subpolar coniferous forest biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>ENVO:01000433</td>\n",
       "      <td>needleleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>grassland</td>\n",
       "      <td>grassland</td>\n",
       "      <td>Grassland</td>\n",
       "      <td>ENVO:01001206</td>\n",
       "      <td>grassland ecosystem</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>leaf litter</td>\n",
       "      <td>leaf litter</td>\n",
       "      <td>leaf litter</td>\n",
       "      <td>ENVO:01000628</td>\n",
       "      <td>plant litter</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>rice paddy</td>\n",
       "      <td>rice paddy</td>\n",
       "      <td>rice paddy</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>rice paddy soils</td>\n",
       "      <td>rice paddy soils</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>taiga</td>\n",
       "      <td>taiga</td>\n",
       "      <td>taiga</td>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>subpolar coniferous forest biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>temperate broadleaf and mixed forest</td>\n",
       "      <td>temperate broadleaf and mixed forest</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO:01000389</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>temperate broadleaf and mixed forests</td>\n",
       "      <td>temperate broadleaf and mixed forests</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO:01000389</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>tropical and subtropical coniferous forest biome</td>\n",
       "      <td>tropical and subtropical coniferous forest biome</td>\n",
       "      <td>subtropical coniferous forest biome</td>\n",
       "      <td>ENVO:01000209</td>\n",
       "      <td>subtropical coniferous forest biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>basin</td>\n",
       "      <td>basin</td>\n",
       "      <td>Basin</td>\n",
       "      <td>GAZ:22223655</td>\n",
       "      <td>Basin</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>forest</td>\n",
       "      <td>forest</td>\n",
       "      <td>Forest</td>\n",
       "      <td>GAZ:00454366</td>\n",
       "      <td>Forest</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>hot springs</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>GAZ:22224982</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ria de aveiro</td>\n",
       "      <td>ria de aveiro</td>\n",
       "      <td>Ria de Aveiro</td>\n",
       "      <td>GAZ:00146341</td>\n",
       "      <td>Ria de Aveiro</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>tundra</td>\n",
       "      <td>tundra</td>\n",
       "      <td>tundra</td>\n",
       "      <td>ENVO:00000112</td>\n",
       "      <td>tundra</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>micro</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>deyeuxia arundinacea</td>\n",
       "      <td>deyeuxia arundinacea</td>\n",
       "      <td>Deyeuxia arundinacea</td>\n",
       "      <td>NCBITaxon:1767996</td>\n",
       "      <td>Deyeuxia arundinacea</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>plant species clypeola jonthlaspi</td>\n",
       "      <td>plant species clypeola jonthlaspi</td>\n",
       "      <td>Clypeola jonthlaspi</td>\n",
       "      <td>NCBITaxon:457792</td>\n",
       "      <td>Clypeola jonthlaspi</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  raw  \\\n",
       "0                                     ectomycorrhizae   \n",
       "1   tropical and subtropical moist broadleaf fores...   \n",
       "2                                        beech forest   \n",
       "3                                    cold desert soil   \n",
       "4                                          cryoconite   \n",
       "5                                       spruce forest   \n",
       "6          temperate broadleaf and mixed forest biome   \n",
       "7                          temperate deciduous forest   \n",
       "8                                              konjac   \n",
       "9                                              bamboo   \n",
       "10                               western amazon basin   \n",
       "11                         deciduous broadleaf forest   \n",
       "12                                             arctic   \n",
       "13                                    artificial soil   \n",
       "14                                             cotton   \n",
       "15                                               crop   \n",
       "16                                       experimental   \n",
       "17                                           fusarium   \n",
       "18                                            healthy   \n",
       "19                                      microorganism   \n",
       "20                                     not applicable   \n",
       "21                                     proteobacteria   \n",
       "22                                         rainforest   \n",
       "23                                     lab experiment   \n",
       "24                                  montane shrubland   \n",
       "25                                         paddy soil   \n",
       "26                                       polar desert   \n",
       "27                       temperate coniferous forests   \n",
       "28                                        terrestrial   \n",
       "29                        wastewater treatment system   \n",
       "30                                           bacteria   \n",
       "31                                      boreal forest   \n",
       "32                                  coniferous forest   \n",
       "33                                          grassland   \n",
       "34                                        leaf litter   \n",
       "35                                         rice paddy   \n",
       "36                                   rice paddy soils   \n",
       "37                                              taiga   \n",
       "38               temperate broadleaf and mixed forest   \n",
       "39              temperate broadleaf and mixed forests   \n",
       "40   tropical and subtropical coniferous forest biome   \n",
       "41                                              basin   \n",
       "42                                             forest   \n",
       "43                                        hot springs   \n",
       "44                                      ria de aveiro   \n",
       "45                                             tundra   \n",
       "46                               deyeuxia arundinacea   \n",
       "47                  plant species clypeola jonthlaspi   \n",
       "\n",
       "                                                query  \\\n",
       "0                                     ectomycorrhizae   \n",
       "1   tropical and subtropical moist broadleaf fores...   \n",
       "2                                        beech forest   \n",
       "3                                    cold desert soil   \n",
       "4                                          cryoconite   \n",
       "5                                       spruce forest   \n",
       "6          temperate broadleaf and mixed forest biome   \n",
       "7                          temperate deciduous forest   \n",
       "8                                              konjac   \n",
       "9                                              bamboo   \n",
       "10                               western amazon basin   \n",
       "11                         deciduous broadleaf forest   \n",
       "12                                             arctic   \n",
       "13                                    artificial soil   \n",
       "14                                             cotton   \n",
       "15                                               crop   \n",
       "16                                       experimental   \n",
       "17                                           fusarium   \n",
       "18                                            healthy   \n",
       "19                                      microorganism   \n",
       "20                                     not applicable   \n",
       "21                                     proteobacteria   \n",
       "22                                         rainforest   \n",
       "23                                     lab experiment   \n",
       "24                                  montane shrubland   \n",
       "25                                         paddy soil   \n",
       "26                                       polar desert   \n",
       "27                       temperate coniferous forests   \n",
       "28                                        terrestrial   \n",
       "29                        wastewater treatment system   \n",
       "30                                           bacteria   \n",
       "31                                      boreal forest   \n",
       "32                                  coniferous forest   \n",
       "33                                          grassland   \n",
       "34                                        leaf litter   \n",
       "35                                         rice paddy   \n",
       "36                                   rice paddy soils   \n",
       "37                                              taiga   \n",
       "38               temperate broadleaf and mixed forest   \n",
       "39              temperate broadleaf and mixed forests   \n",
       "40   tropical and subtropical coniferous forest biome   \n",
       "41                                              basin   \n",
       "42                                             forest   \n",
       "43                                        hot springs   \n",
       "44                                      ria de aveiro   \n",
       "45                                             tundra   \n",
       "46                               deyeuxia arundinacea   \n",
       "47                  plant species clypeola jonthlaspi   \n",
       "\n",
       "                                    name             obo_id  \\\n",
       "0                         ectomycorrhiza        BTO:0002237   \n",
       "1     subtropical broadleaf forest biome      ENVO:01000201   \n",
       "2                      beech forest soil      ENVO:00005770   \n",
       "3                            cold desert      ENVO:01000382   \n",
       "4                        cryoconite hole      ENVO:03000039   \n",
       "5                     spruce forest soil      ENVO:00005784   \n",
       "6       temperate mixed broadleaf forest      ENVO:01000389   \n",
       "7   temperate deciduous broadleaf forest      ENVO:01000385   \n",
       "8                                 konjac    FOODON:03414400   \n",
       "9                             Bamboo Dam       GAZ:00148513   \n",
       "10                         Western Basin       GAZ:00084622   \n",
       "11   tropical deciduous broadleaf forest      ENVO:01000387   \n",
       "12                                Arctic        NCIT:C44738   \n",
       "13                            Artificial        NCIT:C61464   \n",
       "14                                cotton        NCIT:C77117   \n",
       "15                                  Crop       NCIT:C111162   \n",
       "16                          experimental        NCIT:C28041   \n",
       "17                              Fusarium        NCIT:C77185   \n",
       "18                               Healthy       NCIT:C115935   \n",
       "19                         Microorganism        NCIT:C14329   \n",
       "20                        Not Applicable        NCIT:C48660   \n",
       "21                        Proteobacteria        NCIT:C76196   \n",
       "22                            Rainforest       OMIT:0028792   \n",
       "23                            Experiment                NaN   \n",
       "24               montane shrubland biome      ENVO:01000216   \n",
       "25                       rice paddy soil      ENVO:00005740   \n",
       "26                    polar desert biome      ENVO:01000186   \n",
       "27     temperate coniferous forest biome      ENVO:01000211   \n",
       "28                     terrestrial realm      ENVO:00000446   \n",
       "29            wastewater treatment plant      ENVO:00002043   \n",
       "30                              Bacteria        NCBITaxon:2   \n",
       "31                         boreal forest      ENVO:01000250   \n",
       "32                     coniferous forest      ENVO:01000433   \n",
       "33                             Grassland      ENVO:01001206   \n",
       "34                           leaf litter      ENVO:01000628   \n",
       "35                            rice paddy      ENVO:00000296   \n",
       "36                       rice paddy soil      ENVO:00005740   \n",
       "37                                 taiga      ENVO:01000250   \n",
       "38      temperate mixed broadleaf forest      ENVO:01000389   \n",
       "39      temperate mixed broadleaf forest      ENVO:01000389   \n",
       "40   subtropical coniferous forest biome      ENVO:01000209   \n",
       "41                                 Basin       GAZ:22223655   \n",
       "42                                Forest       GAZ:00454366   \n",
       "43                           Hot Springs       GAZ:22224982   \n",
       "44                         Ria de Aveiro       GAZ:00146341   \n",
       "45                                tundra      ENVO:00000112   \n",
       "46                  Deyeuxia arundinacea  NCBITaxon:1767996   \n",
       "47                   Clypeola jonthlaspi   NCBITaxon:457792   \n",
       "\n",
       "                                   label ontology_prefix                scope  \\\n",
       "0                         ectomycorrhiza             BTO                label   \n",
       "1     subtropical broadleaf forest biome             ENM                label   \n",
       "2                      beech forest soil            ENVO                label   \n",
       "3                            cold desert            ENVO                label   \n",
       "4                        cryoconite hole            ENVO                label   \n",
       "5                     spruce forest soil            ENVO                label   \n",
       "6       temperate mixed broadleaf forest            ENVO                label   \n",
       "7   temperate deciduous broadleaf forest            ENVO                label   \n",
       "8                                 konjac          FOODON                label   \n",
       "9                             Bamboo Dam             GAZ                label   \n",
       "10                         Western Basin             GAZ                label   \n",
       "11   tropical deciduous broadleaf forest           MICRO                label   \n",
       "12                                Arctic            NCIT                label   \n",
       "13                            Artificial            NCIT                label   \n",
       "14                            Cotton Rat            NCIT              synonym   \n",
       "15                                  Crop            NCIT                label   \n",
       "16                       Investigational            NCIT              synonym   \n",
       "17                              Fusarium            NCIT                label   \n",
       "18                               Healthy            NCIT                label   \n",
       "19                         Microorganism            NCIT                label   \n",
       "20                        Not Applicable            NCIT                label   \n",
       "21                        Proteobacteria            NCIT                label   \n",
       "22                            Rainforest            OMIT                label   \n",
       "23                            Experiment            REPR                label   \n",
       "24               montane shrubland biome            ENVO                label   \n",
       "25                      paddy field soil            ENVO  has_related_synonym   \n",
       "26                    polar desert biome            ENVO                label   \n",
       "27     temperate coniferous forest biome            ENVO                label   \n",
       "28                     terrestrial biome            ENVO  has_related_synonym   \n",
       "29            wastewater treatment plant            ENVO                label   \n",
       "30                              Bacteria            ENVO                label   \n",
       "31      subpolar coniferous forest biome            ENVO   has_narrow_synonym   \n",
       "32                     needleleaf forest            ENVO              synonym   \n",
       "33                   grassland ecosystem            ENVO  has_related_synonym   \n",
       "34                          plant litter            ENVO              synonym   \n",
       "35                            rice field            ENVO  has_related_synonym   \n",
       "36                      paddy field soil            ENVO  has_related_synonym   \n",
       "37      subpolar coniferous forest biome            ENVO   has_narrow_synonym   \n",
       "38      temperate mixed broadleaf forest            ENVO                label   \n",
       "39      temperate mixed broadleaf forest            ENVO                label   \n",
       "40   subtropical coniferous forest biome            ENVO                label   \n",
       "41                                 Basin             GAZ                label   \n",
       "42                                Forest             GAZ                label   \n",
       "43                           Hot Springs             GAZ                label   \n",
       "44                         Ria de Aveiro             GAZ                label   \n",
       "45                                tundra           MICRO                label   \n",
       "46                  Deyeuxia arundinacea       NCBITAXON                label   \n",
       "47                   Clypeola jonthlaspi       NCBITAXON                label   \n",
       "\n",
       "          type ontology_name  include          ols_curation_notes  \n",
       "0        label           bto    False                Mam 20210609  \n",
       "1        label           enm    False    Omits tropical and moose  \n",
       "2        label          envo    False                   Adds soil  \n",
       "3        label          envo    False                  Loses soil  \n",
       "4        label          envo    False                   Adds hole  \n",
       "5        label          envo    False                   Adds soil  \n",
       "6        label          envo    False                Pretty close  \n",
       "7        label          envo    False              Adds broadleaf  \n",
       "8        label        foodon    False                         NaN  \n",
       "9        label           gaz    False                         NaN  \n",
       "10       label           gaz    False  No, this is near lake erie  \n",
       "11       label         micro    False               Adds tropical  \n",
       "12       label          ncit    False                  Avoid ncit  \n",
       "13       label          ncit    False                  Avoid ncit  \n",
       "14     synonym          ncit    False                  Avoid ncit  \n",
       "15       label          ncit    False                  Avoid ncit  \n",
       "16     synonym          ncit    False                  Avoid ncit  \n",
       "17       label          ncit    False                  Avoid ncit  \n",
       "18       label          ncit    False                  Avoid ncit  \n",
       "19       label          ncit    False                  Avoid ncit  \n",
       "20       label          ncit    False                  Avoid ncit  \n",
       "21       label          ncit    False                  Avoid ncit  \n",
       "22       label          omit    False                  Avoid omit  \n",
       "23       label   reproduceme    False           No id, avoid repo  \n",
       "24       label          envo     True                Mam 20210609  \n",
       "25  annotation          envo     True                Mam 20210609  \n",
       "26       label          envo     True                Mam 20210609  \n",
       "27       label          envo     True                Mam 20210609  \n",
       "28  annotation          envo     True                Mam 20210609  \n",
       "29       label          envo     True                Mam 20210609  \n",
       "30       label          envo     True                         NaN  \n",
       "31  annotation          envo     True                         NaN  \n",
       "32     synonym          envo     True                         NaN  \n",
       "33  annotation          envo     True                         NaN  \n",
       "34     synonym          envo     True                         NaN  \n",
       "35  annotation          envo     True                         NaN  \n",
       "36  annotation          envo     True                         NaN  \n",
       "37  annotation          envo     True                         NaN  \n",
       "38       label          envo     True                         NaN  \n",
       "39       label          envo     True                         NaN  \n",
       "40       label          envo     True                         NaN  \n",
       "41       label           gaz     True                         NaN  \n",
       "42       label           gaz     True                         NaN  \n",
       "43       label           gaz     True                         NaN  \n",
       "44       label           gaz     True                         NaN  \n",
       "45       label         micro     True                Mam 20210609  \n",
       "46       label     ncbitaxon     True                Mam 20210609  \n",
       "47       label     ncbitaxon     True                Mam 20210609  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# OR\n",
    "curated = pd.read_clipboard()\n",
    "curated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ea200e4",
   "metadata": {},
   "source": [
    "## Filter on `include`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1473206f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ols_raw</th>\n",
       "      <th>ols_query</th>\n",
       "      <th>ols_name</th>\n",
       "      <th>ols_obo_id</th>\n",
       "      <th>ols_label</th>\n",
       "      <th>ols_ontology_prefix</th>\n",
       "      <th>ols_scope</th>\n",
       "      <th>ols_type</th>\n",
       "      <th>ols_ontology_name</th>\n",
       "      <th>ols_include</th>\n",
       "      <th>ols_ols_curation_notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>montane shrubland</td>\n",
       "      <td>montane shrubland</td>\n",
       "      <td>montane shrubland biome</td>\n",
       "      <td>ENVO:01000216</td>\n",
       "      <td>montane shrubland biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>paddy soil</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>polar desert</td>\n",
       "      <td>polar desert</td>\n",
       "      <td>polar desert biome</td>\n",
       "      <td>ENVO:01000186</td>\n",
       "      <td>polar desert biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>temperate coniferous forests</td>\n",
       "      <td>temperate coniferous forests</td>\n",
       "      <td>temperate coniferous forest biome</td>\n",
       "      <td>ENVO:01000211</td>\n",
       "      <td>temperate coniferous forest biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>terrestrial</td>\n",
       "      <td>terrestrial</td>\n",
       "      <td>terrestrial realm</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>wastewater treatment plant</td>\n",
       "      <td>ENVO:00002043</td>\n",
       "      <td>wastewater treatment plant</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>subpolar coniferous forest biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>ENVO:01000433</td>\n",
       "      <td>needleleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>grassland</td>\n",
       "      <td>grassland</td>\n",
       "      <td>Grassland</td>\n",
       "      <td>ENVO:01001206</td>\n",
       "      <td>grassland ecosystem</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>leaf litter</td>\n",
       "      <td>leaf litter</td>\n",
       "      <td>leaf litter</td>\n",
       "      <td>ENVO:01000628</td>\n",
       "      <td>plant litter</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>rice paddy</td>\n",
       "      <td>rice paddy</td>\n",
       "      <td>rice paddy</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>rice paddy soils</td>\n",
       "      <td>rice paddy soils</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>taiga</td>\n",
       "      <td>taiga</td>\n",
       "      <td>taiga</td>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>subpolar coniferous forest biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>temperate broadleaf and mixed forest</td>\n",
       "      <td>temperate broadleaf and mixed forest</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO:01000389</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>temperate broadleaf and mixed forests</td>\n",
       "      <td>temperate broadleaf and mixed forests</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO:01000389</td>\n",
       "      <td>temperate mixed broadleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>tropical and subtropical coniferous forest biome</td>\n",
       "      <td>tropical and subtropical coniferous forest biome</td>\n",
       "      <td>subtropical coniferous forest biome</td>\n",
       "      <td>ENVO:01000209</td>\n",
       "      <td>subtropical coniferous forest biome</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>basin</td>\n",
       "      <td>basin</td>\n",
       "      <td>Basin</td>\n",
       "      <td>GAZ:22223655</td>\n",
       "      <td>Basin</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>forest</td>\n",
       "      <td>forest</td>\n",
       "      <td>Forest</td>\n",
       "      <td>GAZ:00454366</td>\n",
       "      <td>Forest</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>hot springs</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>GAZ:22224982</td>\n",
       "      <td>Hot Springs</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ria de aveiro</td>\n",
       "      <td>ria de aveiro</td>\n",
       "      <td>Ria de Aveiro</td>\n",
       "      <td>GAZ:00146341</td>\n",
       "      <td>Ria de Aveiro</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>gaz</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>tundra</td>\n",
       "      <td>tundra</td>\n",
       "      <td>tundra</td>\n",
       "      <td>ENVO:00000112</td>\n",
       "      <td>tundra</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>micro</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>deyeuxia arundinacea</td>\n",
       "      <td>deyeuxia arundinacea</td>\n",
       "      <td>Deyeuxia arundinacea</td>\n",
       "      <td>NCBITaxon:1767996</td>\n",
       "      <td>Deyeuxia arundinacea</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>plant species clypeola jonthlaspi</td>\n",
       "      <td>plant species clypeola jonthlaspi</td>\n",
       "      <td>Clypeola jonthlaspi</td>\n",
       "      <td>NCBITaxon:457792</td>\n",
       "      <td>Clypeola jonthlaspi</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>True</td>\n",
       "      <td>Mam 20210609</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             ols_raw  \\\n",
       "24                                 montane shrubland   \n",
       "25                                        paddy soil   \n",
       "26                                      polar desert   \n",
       "27                      temperate coniferous forests   \n",
       "28                                       terrestrial   \n",
       "29                       wastewater treatment system   \n",
       "30                                          bacteria   \n",
       "31                                     boreal forest   \n",
       "32                                 coniferous forest   \n",
       "33                                         grassland   \n",
       "34                                       leaf litter   \n",
       "35                                        rice paddy   \n",
       "36                                  rice paddy soils   \n",
       "37                                             taiga   \n",
       "38              temperate broadleaf and mixed forest   \n",
       "39             temperate broadleaf and mixed forests   \n",
       "40  tropical and subtropical coniferous forest biome   \n",
       "41                                             basin   \n",
       "42                                            forest   \n",
       "43                                       hot springs   \n",
       "44                                     ria de aveiro   \n",
       "45                                            tundra   \n",
       "46                              deyeuxia arundinacea   \n",
       "47                 plant species clypeola jonthlaspi   \n",
       "\n",
       "                                           ols_query  \\\n",
       "24                                 montane shrubland   \n",
       "25                                        paddy soil   \n",
       "26                                      polar desert   \n",
       "27                      temperate coniferous forests   \n",
       "28                                       terrestrial   \n",
       "29                       wastewater treatment system   \n",
       "30                                          bacteria   \n",
       "31                                     boreal forest   \n",
       "32                                 coniferous forest   \n",
       "33                                         grassland   \n",
       "34                                       leaf litter   \n",
       "35                                        rice paddy   \n",
       "36                                  rice paddy soils   \n",
       "37                                             taiga   \n",
       "38              temperate broadleaf and mixed forest   \n",
       "39             temperate broadleaf and mixed forests   \n",
       "40  tropical and subtropical coniferous forest biome   \n",
       "41                                             basin   \n",
       "42                                            forest   \n",
       "43                                       hot springs   \n",
       "44                                     ria de aveiro   \n",
       "45                                            tundra   \n",
       "46                              deyeuxia arundinacea   \n",
       "47                 plant species clypeola jonthlaspi   \n",
       "\n",
       "                               ols_name         ols_obo_id  \\\n",
       "24              montane shrubland biome      ENVO:01000216   \n",
       "25                      rice paddy soil      ENVO:00005740   \n",
       "26                   polar desert biome      ENVO:01000186   \n",
       "27    temperate coniferous forest biome      ENVO:01000211   \n",
       "28                    terrestrial realm      ENVO:00000446   \n",
       "29           wastewater treatment plant      ENVO:00002043   \n",
       "30                             Bacteria        NCBITaxon:2   \n",
       "31                        boreal forest      ENVO:01000250   \n",
       "32                    coniferous forest      ENVO:01000433   \n",
       "33                            Grassland      ENVO:01001206   \n",
       "34                          leaf litter      ENVO:01000628   \n",
       "35                           rice paddy      ENVO:00000296   \n",
       "36                      rice paddy soil      ENVO:00005740   \n",
       "37                                taiga      ENVO:01000250   \n",
       "38     temperate mixed broadleaf forest      ENVO:01000389   \n",
       "39     temperate mixed broadleaf forest      ENVO:01000389   \n",
       "40  subtropical coniferous forest biome      ENVO:01000209   \n",
       "41                                Basin       GAZ:22223655   \n",
       "42                               Forest       GAZ:00454366   \n",
       "43                          Hot Springs       GAZ:22224982   \n",
       "44                        Ria de Aveiro       GAZ:00146341   \n",
       "45                               tundra      ENVO:00000112   \n",
       "46                 Deyeuxia arundinacea  NCBITaxon:1767996   \n",
       "47                  Clypeola jonthlaspi   NCBITaxon:457792   \n",
       "\n",
       "                              ols_label ols_ontology_prefix  \\\n",
       "24              montane shrubland biome                ENVO   \n",
       "25                     paddy field soil                ENVO   \n",
       "26                   polar desert biome                ENVO   \n",
       "27    temperate coniferous forest biome                ENVO   \n",
       "28                    terrestrial biome                ENVO   \n",
       "29           wastewater treatment plant                ENVO   \n",
       "30                             Bacteria                ENVO   \n",
       "31     subpolar coniferous forest biome                ENVO   \n",
       "32                    needleleaf forest                ENVO   \n",
       "33                  grassland ecosystem                ENVO   \n",
       "34                         plant litter                ENVO   \n",
       "35                           rice field                ENVO   \n",
       "36                     paddy field soil                ENVO   \n",
       "37     subpolar coniferous forest biome                ENVO   \n",
       "38     temperate mixed broadleaf forest                ENVO   \n",
       "39     temperate mixed broadleaf forest                ENVO   \n",
       "40  subtropical coniferous forest biome                ENVO   \n",
       "41                                Basin                 GAZ   \n",
       "42                               Forest                 GAZ   \n",
       "43                          Hot Springs                 GAZ   \n",
       "44                        Ria de Aveiro                 GAZ   \n",
       "45                               tundra               MICRO   \n",
       "46                 Deyeuxia arundinacea           NCBITAXON   \n",
       "47                  Clypeola jonthlaspi           NCBITAXON   \n",
       "\n",
       "              ols_scope    ols_type ols_ontology_name  ols_include  \\\n",
       "24                label       label              envo         True   \n",
       "25  has_related_synonym  annotation              envo         True   \n",
       "26                label       label              envo         True   \n",
       "27                label       label              envo         True   \n",
       "28  has_related_synonym  annotation              envo         True   \n",
       "29                label       label              envo         True   \n",
       "30                label       label              envo         True   \n",
       "31   has_narrow_synonym  annotation              envo         True   \n",
       "32              synonym     synonym              envo         True   \n",
       "33  has_related_synonym  annotation              envo         True   \n",
       "34              synonym     synonym              envo         True   \n",
       "35  has_related_synonym  annotation              envo         True   \n",
       "36  has_related_synonym  annotation              envo         True   \n",
       "37   has_narrow_synonym  annotation              envo         True   \n",
       "38                label       label              envo         True   \n",
       "39                label       label              envo         True   \n",
       "40                label       label              envo         True   \n",
       "41                label       label               gaz         True   \n",
       "42                label       label               gaz         True   \n",
       "43                label       label               gaz         True   \n",
       "44                label       label               gaz         True   \n",
       "45                label       label             micro         True   \n",
       "46                label       label         ncbitaxon         True   \n",
       "47                label       label         ncbitaxon         True   \n",
       "\n",
       "   ols_ols_curation_notes  \n",
       "24           Mam 20210609  \n",
       "25           Mam 20210609  \n",
       "26           Mam 20210609  \n",
       "27           Mam 20210609  \n",
       "28           Mam 20210609  \n",
       "29           Mam 20210609  \n",
       "30                    NaN  \n",
       "31                    NaN  \n",
       "32                    NaN  \n",
       "33                    NaN  \n",
       "34                    NaN  \n",
       "35                    NaN  \n",
       "36                    NaN  \n",
       "37                    NaN  \n",
       "38                    NaN  \n",
       "39                    NaN  \n",
       "40                    NaN  \n",
       "41                    NaN  \n",
       "42                    NaN  \n",
       "43                    NaN  \n",
       "44                    NaN  \n",
       "45           Mam 20210609  \n",
       "46           Mam 20210609  \n",
       "47           Mam 20210609  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flag = curated[\"include\"]\n",
    "curated = curated.loc[flag]\n",
    "\n",
    "curated.columns = \"ols_\" + curated.columns\n",
    "curated.to_sql(\"curated\", biosample_cnx, if_exists=\"replace\", index=False)\n",
    "curated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "ac6219e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_search_merged = mapping_candidates.merge(\n",
    "    curated, how=\"outer\", left_on=\"remaining_tidied\", right_on=\"ols_raw\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "d2c242b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = (\n",
    "    ~merge_search_merged[\"id_or_lab_ok\"]\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].eq(\"\")\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].isna()\n",
    ")\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_obo_id\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_id\"] = replacement\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_label\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_lab\"] = replacement\n",
    "\n",
    "replacement = (\n",
    "    merge_search_merged.loc[flag, \"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + merge_search_merged.loc[flag, \"consensus_id\"]\n",
    "    + \"]\"\n",
    ")\n",
    "\n",
    "merge_search_merged.loc[flag, \"assembled_consensus\"] = replacement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "b333c111",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_search_merged.to_sql(\n",
    "    \"merge_search_merged\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "5e772701",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>assembled_consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>4856</td>\n",
       "      <td>cropland biome [ENVO:01000245]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cropland biome</td>\n",
       "      <td>1530</td>\n",
       "      <td>cropland biome [ENVO:01000245]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>urban biome</td>\n",
       "      <td>973</td>\n",
       "      <td>urban biome [ENVO:01000249]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tundra biome</td>\n",
       "      <td>516</td>\n",
       "      <td>tundra biome [ENVO:01000180]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>483</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>coastal saline 12</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>coastal saline 11</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>coastal saline 10</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>cellar mud</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>western Amazon basin</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>244 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          env_broad_scale  count                assembled_consensus\n",
       "0     ENVO:cropland biome   4856     cropland biome [ENVO:01000245]\n",
       "1          cropland biome   1530     cropland biome [ENVO:01000245]\n",
       "2             urban biome    973        urban biome [ENVO:01000249]\n",
       "3            tundra biome    516       tundra biome [ENVO:01000180]\n",
       "4       terrestrial biome    483  terrestrial biome [ENVO:00000446]\n",
       "..                    ...    ...                                ...\n",
       "186     coastal saline 12      1                                NaN\n",
       "185     coastal saline 11      1                                NaN\n",
       "184     coastal saline 10      1                                NaN\n",
       "183            cellar mud      1                                NaN\n",
       "243  western Amazon basin      1                                NaN\n",
       "\n",
       "[244 rows x 3 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary = merge_search_merged[[\"env_broad_scale\", \"count\", \"assembled_consensus\"]]\n",
    "summary = summary.sort_values(\"count\", ascending=False)\n",
    "summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5278c2b",
   "metadata": {},
   "source": [
    "## TODOs etc.\n",
    "\n",
    "Advantages and disadvantages of including non-local annotations in salvage search? Example = ???\n",
    "small/large (freshwater) lake\n",
    "\n",
    "Include those native labels in the review frame?\n",
    "\n",
    "Do more of that kind of thing (term ID/label joining) with rdftab, not OLS\n",
    "\n",
    "What to save where? new column in biosample sqlite? or move content from original column and insert the new values into that original column?\n",
    "\n",
    "SSSOM:\n",
    "- add string distance/confidence... multiple columns?\n",
    "- not all of these are OLS mapped\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "38acac10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_category</th>\n",
       "      <th>subject_label</th>\n",
       "      <th>predicate_id</th>\n",
       "      <th>object_id</th>\n",
       "      <th>object_label</th>\n",
       "      <th>match_type</th>\n",
       "      <th>creator_id</th>\n",
       "      <th>mapping_date</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>ENVO:cropland biome</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000245</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:ENVO%3Acropland%20biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000245</td>\n",
       "      <td>cropland biome</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:cropland%20biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000249</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:urban%20biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000180</td>\n",
       "      <td>tundra biome</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:tundra%20biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:terrestrial%20biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>marsh</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000035</td>\n",
       "      <td>marsh</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:marsh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>mixed forest biome</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000198</td>\n",
       "      <td>mixed forest biome</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:mixed%20forest%20biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>montane shrubland</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000216</td>\n",
       "      <td>montane shrubland biome</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:montane%20shrubland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:paddy%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>wastewater treatment system</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002043</td>\n",
       "      <td>wastewater treatment plant</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-09T10:27:37</td>\n",
       "      <td>biosample_ebs:wastewater%20treatment%20system</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    subject_category                subject_label       predicate_id  \\\n",
       "0    env_broad_scale          ENVO:cropland biome  skos:relatedMatch   \n",
       "1    env_broad_scale               cropland biome  skos:relatedMatch   \n",
       "2    env_broad_scale                  urban biome  skos:relatedMatch   \n",
       "3    env_broad_scale                 tundra biome  skos:relatedMatch   \n",
       "4    env_broad_scale            terrestrial biome  skos:relatedMatch   \n",
       "..               ...                          ...                ...   \n",
       "227  env_broad_scale                        marsh  skos:relatedMatch   \n",
       "228  env_broad_scale           mixed forest biome  skos:relatedMatch   \n",
       "230  env_broad_scale            montane shrubland  skos:relatedMatch   \n",
       "232  env_broad_scale                   paddy soil  skos:relatedMatch   \n",
       "242  env_broad_scale  wastewater treatment system  skos:relatedMatch   \n",
       "\n",
       "         object_id                object_label match_type  \\\n",
       "0    ENVO:01000245              cropland biome    Lexical   \n",
       "1    ENVO:01000245              cropland biome    Lexical   \n",
       "2    ENVO:01000249                 urban biome    Lexical   \n",
       "3    ENVO:01000180                tundra biome    Lexical   \n",
       "4    ENVO:00000446           terrestrial biome    Lexical   \n",
       "..             ...                         ...        ...   \n",
       "227  ENVO:00000035                       marsh    Lexical   \n",
       "228  ENVO:01000198          mixed forest biome    Lexical   \n",
       "230  ENVO:01000216     montane shrubland biome    Lexical   \n",
       "232  ENVO:00005740            paddy field soil    Lexical   \n",
       "242  ENVO:00002043  wastewater treatment plant    Lexical   \n",
       "\n",
       "                                     creator_id         mapping_date  \\\n",
       "0    https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "1    https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "2    https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "3    https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "4    https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "..                                          ...                  ...   \n",
       "227  https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "228  https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "230  https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "232  https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "242  https://github.com/turbomam/scoped-mapping  2021-06-09T10:27:37   \n",
       "\n",
       "                                        subject_id  \n",
       "0            biosample_ebs:ENVO%3Acropland%20biome  \n",
       "1                   biosample_ebs:cropland%20biome  \n",
       "2                      biosample_ebs:urban%20biome  \n",
       "3                     biosample_ebs:tundra%20biome  \n",
       "4                biosample_ebs:terrestrial%20biome  \n",
       "..                                             ...  \n",
       "227                            biosample_ebs:marsh  \n",
       "228           biosample_ebs:mixed%20forest%20biome  \n",
       "230              biosample_ebs:montane%20shrubland  \n",
       "232                     biosample_ebs:paddy%20soil  \n",
       "242  biosample_ebs:wastewater%20treatment%20system  \n",
       "\n",
       "[100 rows x 9 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frame_for_sssom = merge_search_merged\n",
    "\n",
    "na_flag = frame_for_sssom[\"splitted\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "na_flag = frame_for_sssom[\"consensus_id\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "raw_queries = list(frame_for_sssom[\"splitted\"])\n",
    "urlencodeds = list(map(urllib.parse.quote, raw_queries))\n",
    "urlencodeds = [sssom_subject_prefix + \":\" + item for item in urlencodeds]\n",
    "\n",
    "results_rows = len(frame_for_sssom.index)\n",
    "iso8601_stamp = datetime.now().replace(microsecond=0).isoformat()\n",
    "\n",
    "sssom_frame = {\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"subject_label\": frame_for_sssom[\"splitted\"],\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"object_id\": frame_for_sssom[\"consensus_id\"],\n",
    "    \"object_label\": frame_for_sssom[\"consensus_lab\"],\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    \"mapping_date\": [iso8601_stamp] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "    \"subject_id\": urlencodeds,\n",
    "}\n",
    "\n",
    "sssom_frame = pd.DataFrame(sssom_frame)\n",
    "\n",
    "sssom_frame.to_csv(sssom_file, sep=\"\\t\", index=False)\n",
    "\n",
    "sssom_frame\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ad82349",
   "metadata": {},
   "source": [
    "# From here down requires refactoring and generalization \n",
    "\n",
    "That may even be true for some of the blocks above this!\n",
    "\n",
    "## join together pipe-delimited env_braod_scale mappings \n",
    "what other constraints? If part_count is > 1, do we need to check for empty or null env_broad_scale, repeated or splitted?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c4ccf78a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [env_broad_scale, count, repeated, splitted, part_count, seq, consensus_id, consensus_lab]\n",
       "Index: []"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_part_flag = merge_search_merged[\"part_count\"] > 1\n",
    "\n",
    "multi_part_frame = merge_search_merged.loc[multi_part_flag]\n",
    "\n",
    "multi_part_frame = multi_part_frame[\n",
    "    [\n",
    "        \"env_broad_scale\",\n",
    "        \"count\",\n",
    "        \"repeated\",\n",
    "        \"splitted\",\n",
    "        \"part_count\",\n",
    "        \"seq\",\n",
    "        \"consensus_id\",\n",
    "        \"consensus_lab\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "multi_part_frame = multi_part_frame.sort_values(\n",
    "    [\"env_broad_scale\", \"seq\"], ascending=(True, True)\n",
    ")\n",
    "\n",
    "part_count = multi_part_frame.groupby([\"env_broad_scale\"], sort=False)[\n",
    "    \"part_count\"\n",
    "].max()\n",
    "seq_max = multi_part_frame.groupby([\"env_broad_scale\"], sort=False)[\"seq\"].max()\n",
    "\n",
    "parts_check = pd.DataFrame(dict(part_count=part_count, seq_max=seq_max)).reset_index()\n",
    "\n",
    "all_parts_flag = parts_check[\"seq_max\"] == parts_check[\"part_count\"]\n",
    "ebs_with_all_parts = parts_check.loc[all_parts_flag, \"env_broad_scale\"]\n",
    "ebs_with_all_parts\n",
    "\n",
    "all_parts_flag = multi_part_frame[\"env_broad_scale\"].isin(list(ebs_with_all_parts))\n",
    "all_parts_frame = multi_part_frame.loc[all_parts_flag]\n",
    "all_parts_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0cbd251",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_pipesep_ebs = list(set(list(all_parts_frame[\"env_broad_scale\"])))\n",
    "unique_pipesep_ebs.sort()\n",
    "repipe_dict_list = []\n",
    "for one_pipesep in unique_pipesep_ebs:\n",
    "    flag = all_parts_frame[\"env_broad_scale\"] == one_pipesep\n",
    "    temp_frame = all_parts_frame.loc[flag]\n",
    "    ids_list = list(temp_frame[\"consensus_id\"])\n",
    "    lab_list = list(temp_frame[\"consensus_lab\"])\n",
    "    s = \"|\"\n",
    "    ids_repipe = s.join(ids_list)\n",
    "    lab_repipe = s.join(lab_list)\n",
    "    repipe_dict = {\n",
    "        \"env_broad_scale\": one_pipesep,\n",
    "        \"consensus_id\": ids_repipe,\n",
    "        \"consensus_lab\": lab_repipe,\n",
    "    }\n",
    "    repipe_dict_list.append(repipe_dict)\n",
    "repipe_frame = pd.DataFrame(repipe_dict_list)\n",
    "repipe_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25424849",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "temp = sssom_frame[[\"subject_label\", \"object_id\", \"object_label\"]]\n",
    "temp.columns = repipe_frame.columns\n",
    "\n",
    "direct_and_repipe = pd.concat([temp, repipe_frame])\n",
    "direct_and_repipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ab3aaa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "direct_and_repipe.to_sql(\n",
    "    \"direct_and_repipe\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "922c09b3",
   "metadata": {},
   "source": [
    "## refactor this some more"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8de27d49",
   "metadata": {},
   "outputs": [],
   "source": [
    "q = (\n",
    "    \"\"\"\n",
    "select\n",
    "\tb.id,\n",
    "\tb.env_broad_scale ,\n",
    "\tdar.consensus_id ,\n",
    "\tdar.consensus_lab\n",
    "from\n",
    "\tdirect_and_repipe dar\n",
    "left join biosample b on\n",
    "\tb.env_broad_scale = dar.env_broad_scale\n",
    "inner join biobiosample_tax_id_counts stic on\n",
    "\tb.taxonomy_id = stic.biosample_taxid\n",
    "join env_package_normalized on\n",
    "\tb.env_package = env_package_normalized.env_package\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"'\"\n",
    ")\n",
    "\n",
    "[per_biosample_scoped_ebs_mapping_results, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a863e532",
   "metadata": {},
   "outputs": [],
   "source": [
    "per_biosample_scoped_ebs_mapping_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93d2a346",
   "metadata": {},
   "outputs": [],
   "source": [
    "per_biosample_scoped_ebs_mapping_results.to_csv(\n",
    "    \"per_biosample_scoped_ebs_mapping_results.tsv\",\n",
    "    sep=ols_review_seperator,\n",
    "    index=False,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
