{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "621bbcd9",
   "metadata": {},
   "source": [
    "be really clear about what's going into repaired tables: \n",
    "- repaired\n",
    "- already OK\n",
    "- unrepairable\n",
    "\n",
    "Only ~ 215k repaired `env_package`s ?!\n",
    "- yeah, there just aren't that many annotated in the biosample metadata\n",
    "\n",
    "\n",
    "- some of this code should probably be moved back into the package\n",
    "    - splitting and repiping \n",
    "\n",
    "- read in previous SSSOMs and don't re-map? or at least double check?\n",
    "\n",
    "- don't forget to consult \"model\" field in biosample table too\n",
    "    - for scoping, in addition to en_package and taxonomy_id?\n",
    "\n",
    "- what to do about hits against obsolete classes... does OLS provide a mechanism to avoid?\n",
    "- be careful with GAZ matches\n",
    "- avoid IDless, OMIT, NCIT, MICRO, BTO\n",
    "\n",
    "- is it worth trimming and concatenating \"x1\", \"x2\", \"x3\", where x is a mappable term? numbers may be low.\n",
    "    - example = XXX\n",
    "\n",
    "- reconcile \"x biome\" and \"x\" mappings\n",
    "- take note of non-local salvage hits\n",
    "    - like XXX\n",
    "    - small/large (freshwater) lake?\n",
    "    - include native labels in the review frame, in addition to local?\n",
    "    - should use rdftab, not additional OLS calls\n",
    "    \n",
    "\n",
    "- switch from hard-coded `env_package_overrides` to something like SSSOM?\n",
    "- mappings can come from merging strategies, first pass OLS or salvage OLS\n",
    "- reflect those methods in SSSOM\n",
    "- add string distance/confidence\n",
    "    - multiple columns for multiple methods?\n",
    "\n",
    "review these SSSOM columns\n",
    "\n",
    "```\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "```\n",
    "\n",
    "- are there samples with a host_taxid and a (sample) taxonomy_id from \"unclassified\"?\n",
    "- start at uncultured or metagenome taxonomy_id?\n",
    "\n",
    "- what scoping on or mapping of  \"gold\" or \"path\"? WON'T FIND THAT IN NCBI BIOSAMPLE\n",
    "\n",
    "\n",
    "What to save where? \n",
    "- new columns in biosample sqlite? \n",
    "- delimited file dumps\n",
    "- move content from original column and insert the new values into that original column?\n",
    "- pandas clipboard integration handy for review step\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fac2962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "\n",
    "import scoped_mapping\n",
    "\n",
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3300da64",
   "metadata": {},
   "source": [
    "# User-provided data\n",
    "See repo README for notes on setting up SQLite databases of OBO ontologies with semantic-sql, relation-graph and rdftab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5684af0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://www.ncbi.nlm.nih.gov/biosample/docs/packages/?format=xml\n",
    "# see also https://www.ncbi.nlm.nih.gov/biosample/docs/packages/\n",
    "biosample_packages_file = \"../target/biosample_packages.xml\"\n",
    "\n",
    "# from ftp://ftp.ncbi.nlm.nih.gov//biosample/biosample_set.xml.gz\n",
    "# via harmonized_table.db.gz\n",
    "# in https://drive.google.com/drive/u/0/folders/1eL0v0stoduahjDpoDJIk3z2pJBAU4b2Y\n",
    "biosample_sqlite_file   = \"../target/harmonized_table.db\"\n",
    "ncbitaxon_sqlite_file   = \"../semantic-sql/db/ncbitaxon.db\"\n",
    "\n",
    "\n",
    "# where do we require a single ontology and where can we use multiple?\n",
    "target_onto_prefix    = \"ENVO\"\n",
    "# 'gaz', \n",
    "first_pass_ontologies = ['ncbitaxon']\n",
    "\n",
    "chars_to_whiteout  = \"._-\"\n",
    "my_query_fields    = \"\"  # OLS weighted default\n",
    "my_row_req         = 3\n",
    "my_string_dist_arg = 2\n",
    "my_max_string_dist = 0.1\n",
    "salvage_max_dist   = 0.2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae0bcc99",
   "metadata": {},
   "source": [
    "## It's nice to see everything accounted for\n",
    "(In terms of manual mappings for env_package.) But we can prioritize NMDC for now:\n",
    "\n",
    "- Soil 15,777\n",
    "- Sediment 7,147\n",
    "- Plant-associated 3,142\n",
    "\n",
    "Could some of these \"no environmental package\" mappings be losing important granularity?\n",
    "\n",
    "Map `None` and '' to \"no environmental package\"?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "068fa8f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_overrides = {\n",
    "    \"built environment\": \"built\",\n",
    "    \"misc environment\": \"miscellaneous\",\n",
    "    \"missing\": \"no environmental package\",\n",
    "    \"unknown\": \"no environmental package\",\n",
    "    \"default\": \"no environmental package\",\n",
    "    \"unspecified\": \"no environmental package\",\n",
    "    \"not available\": \"no environmental package\",\n",
    "    \"not collected\": \"no environmental package\",\n",
    "    \"miscellaneous natural or artificial environment\": \"miscellaneous\",\n",
    "    \"not applicable\": \"no environmental package\",\n",
    "    \"soil-associated\": \"soil\",\n",
    "    \"soil associated\": \"soil\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2d7f8bf",
   "metadata": {},
   "source": [
    "## What Biosample field should be mapped to ontology classes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b79c519",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_col_to_map = \"env_medium\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87e3b6f1",
   "metadata": {},
   "source": [
    "## How should the mapping effort be scoped?\n",
    "Additional taxonomic filters may be applied below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ec160085",
   "metadata": {},
   "outputs": [],
   "source": [
    "scoping_col   = \"env_package_normalized.EnvPackage\"\n",
    "scoping_value = \"soil\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a568c1",
   "metadata": {},
   "source": [
    "## Settings for manual review of OLS search-based results\n",
    "Results based on merging Biosample annotations to ontology classes by label or embedded ID are not exported for review at this time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "20cdb474",
   "metadata": {},
   "outputs": [],
   "source": [
    "ols_review_file = \"../local/ols_review.tsv\"\n",
    "ols_review_seperator = \"\\t\"\n",
    "strategy_col = \"strategy\"\n",
    "include_col  = \"include\"\n",
    "\n",
    "first_pass_include_val  = True\n",
    "first_pass_strategy_val = \"env_braod_scale vs envo and gaz @ 0.1\"\n",
    "salvage_include_val     = False\n",
    "salvage_strategy_val    = \"env_braod_scale vs all of ols @ 0.2\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7edfbe0",
   "metadata": {},
   "source": [
    "## Settings for SSSOM output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d71fa5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sssom_subject_prefix = \"biosample_ebs\"\n",
    "sssom_file           = \"biosample_ebs_soil_packages_unclassified_taxa_sssom.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c8336b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "per_biosample_mapping_results_file = \"per_biosample_scoped_mapping_results.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f21778e4",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c4494ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)\n",
    "ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)\n",
    "primary_ontology_sqlite_file = \"../semantic-sql/db/\" + target_onto_prefix.lower() + \".db\"\n",
    "primary_ontology_cnx         = sqlite3.connect(primary_ontology_sqlite_file)\n",
    "first_pass_ontologies.insert(0, target_onto_prefix.lower())\n",
    "first_pass_ontologies = ','.join(first_pass_ontologies) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c54707",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Print a sample of the data we're working with\n",
    "Specifically, INSDC/NCBI Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "caff574e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_package</th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMN00000002</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445970</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BIOSAMPLE:SAMN00000003</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445972</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIOSAMPLE:SAMN00000004</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>449673</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BIOSAMPLE:SAMN00000005</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>6526</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BIOSAMPLE:SAMN00000006</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>9483</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BIOSAMPLE:SAMN00000007</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445974</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BIOSAMPLE:SAMN00000008</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>411461</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BIOSAMPLE:SAMN00000009</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>13616</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BIOSAMPLE:SAMN00000010</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>451639</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BIOSAMPLE:SAMN00000011</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>451638</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id env_package      package  \\\n",
       "0  BIOSAMPLE:SAMN00000002     missing  MIGS.ba.5.0   \n",
       "1  BIOSAMPLE:SAMN00000003     missing  MIGS.ba.5.0   \n",
       "2  BIOSAMPLE:SAMN00000004     missing  MIGS.ba.5.0   \n",
       "3  BIOSAMPLE:SAMN00000005        None  Generic.1.0   \n",
       "4  BIOSAMPLE:SAMN00000006        None  Generic.1.0   \n",
       "5  BIOSAMPLE:SAMN00000007     missing  MIGS.ba.5.0   \n",
       "6  BIOSAMPLE:SAMN00000008     missing  MIGS.ba.5.0   \n",
       "7  BIOSAMPLE:SAMN00000009        None  Generic.1.0   \n",
       "8  BIOSAMPLE:SAMN00000010        None  Generic.1.0   \n",
       "9  BIOSAMPLE:SAMN00000011        None  Generic.1.0   \n",
       "\n",
       "                                   package_name host_taxid taxonomy_id  \\\n",
       "0  MIGS: cultured bacteria/archaea; version 5.0       9606      445970   \n",
       "1  MIGS: cultured bacteria/archaea; version 5.0       9606      445972   \n",
       "2  MIGS: cultured bacteria/archaea; version 5.0       9606      449673   \n",
       "3                                       Generic       None        6526   \n",
       "4                                       Generic       None        9483   \n",
       "5  MIGS: cultured bacteria/archaea; version 5.0       9606      445974   \n",
       "6  MIGS: cultured bacteria/archaea; version 5.0       9606      411461   \n",
       "7                                       Generic       None       13616   \n",
       "8                                       Generic       None      451639   \n",
       "9                                       Generic       None      451638   \n",
       "\n",
       "                     env_broad_scale  \\\n",
       "0  terrestrial biome [ENVO:00000446]   \n",
       "1  terrestrial biome [ENVO:00000446]   \n",
       "2  terrestrial biome [ENVO:00000446]   \n",
       "3                               None   \n",
       "4                               None   \n",
       "5  terrestrial biome [ENVO:00000446]   \n",
       "6  terrestrial biome [ENVO:00000446]   \n",
       "7                               None   \n",
       "8                               None   \n",
       "9                               None   \n",
       "\n",
       "                            env_local_scale  \\\n",
       "0  human-associated habitat [ENVO:00009003]   \n",
       "1  human-associated habitat [ENVO:00009003]   \n",
       "2  human-associated habitat [ENVO:00009003]   \n",
       "3                                      None   \n",
       "4                                      None   \n",
       "5  human-associated habitat [ENVO:00009003]   \n",
       "6  human-associated habitat [ENVO:00009003]   \n",
       "7                                      None   \n",
       "8                                      None   \n",
       "9                                      None   \n",
       "\n",
       "                           env_medium  \n",
       "0  biological product [ENVO:02000043]  \n",
       "1  biological product [ENVO:02000043]  \n",
       "2  biological product [ENVO:02000043]  \n",
       "3                                None  \n",
       "4                                None  \n",
       "5  biological product [ENVO:02000043]  \n",
       "6  biological product [ENVO:02000043]  \n",
       "7                                None  \n",
       "8                                None  \n",
       "9                                None  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    id,\n",
    "    env_package,\n",
    "    package,\n",
    "    package_name,\n",
    "    host_taxid,\n",
    "    taxonomy_id,\n",
    "    env_broad_scale,\n",
    "    env_local_scale,\n",
    "    env_medium\n",
    "    from biosample b\n",
    "limit 10\n",
    "\"\"\"\n",
    "biosample_first_ten = pd.read_sql(q, biosample_cnx)\n",
    "biosample_first_ten"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef577997",
   "metadata": {},
   "source": [
    "## Get the canonical checklist and package terms from NCBI\n",
    "\n",
    "### Could (should?) also examine mixs.yaml (from where?)\n",
    "\n",
    "Unfortunately it doesn't do a very good job of differentiating checklists (MIMAG, MIMARKS, etc.) from packages (soil, water, etc.)\n",
    "\n",
    "_What about .ba, .euk, etc?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c6cc9596",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>DisplayName</th>\n",
       "      <th>ShortName</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>EnvPackageDisplay</th>\n",
       "      <th>NotAppropriateFor</th>\n",
       "      <th>Description</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2.cl.1.0</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated; versi...</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for SARS-CoV-2 samples that are relevant t...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>Pathogen: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Clinical or host-associated pathogen</td>\n",
       "      <td>SAMN02928182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>Pathogen: environmental/food/other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Environmental, food or other pathogen</td>\n",
       "      <td>SAMN02730065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Microbe.1.0</td>\n",
       "      <td>Microbe; version 1.0</td>\n",
       "      <td>Microbe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Use for bacteria or other unicellular microbes...</td>\n",
       "      <td>SAMN02911891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>MIUVIG.plant-associated.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, plant-assoc...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>MIUVIG.sediment.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, sediment; v...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>MIUVIG.soil.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, soil; versi...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>MIUVIG.wastewater.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, wastewater;...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>MIUVIG.water.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, water; vers...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Name  \\\n",
       "0                    Generic.1.0   \n",
       "1              SARS-CoV-2.cl.1.0   \n",
       "2                Pathogen.cl.1.0   \n",
       "3               Pathogen.env.1.0   \n",
       "4                    Microbe.1.0   \n",
       "..                           ...   \n",
       "149  MIUVIG.plant-associated.5.0   \n",
       "150          MIUVIG.sediment.5.0   \n",
       "151              MIUVIG.soil.5.0   \n",
       "152        MIUVIG.wastewater.5.0   \n",
       "153             MIUVIG.water.5.0   \n",
       "\n",
       "                                           DisplayName  \\\n",
       "0                                              Generic   \n",
       "1    SARS-CoV-2: clinical or host-associated; versi...   \n",
       "2    Pathogen: clinical or host-associated; version...   \n",
       "3      Pathogen: environmental/food/other; version 1.0   \n",
       "4                                 Microbe; version 1.0   \n",
       "..                                                 ...   \n",
       "149  MIUVIG: uncultivated virus genome, plant-assoc...   \n",
       "150  MIUVIG: uncultivated virus genome, sediment; v...   \n",
       "151  MIUVIG: uncultivated virus genome, soil; versi...   \n",
       "152  MIUVIG: uncultivated virus genome, wastewater;...   \n",
       "153  MIUVIG: uncultivated virus genome, water; vers...   \n",
       "\n",
       "                                   ShortName        EnvPackage  \\\n",
       "0                                                                \n",
       "1    SARS-CoV-2: clinical or host-associated                     \n",
       "2      Pathogen: clinical or host-associated                     \n",
       "3         Pathogen: environmental/food/other                     \n",
       "4                                    Microbe                     \n",
       "..                                       ...               ...   \n",
       "149         MIUVIG Uncultivated Virus Genome  plant-associated   \n",
       "150         MIUVIG Uncultivated Virus Genome          sediment   \n",
       "151         MIUVIG Uncultivated Virus Genome              soil   \n",
       "152         MIUVIG Uncultivated Virus Genome        wastewater   \n",
       "153         MIUVIG Uncultivated Virus Genome             water   \n",
       "\n",
       "    EnvPackageDisplay                 NotAppropriateFor  \\\n",
       "0                                                         \n",
       "1                      wgs_single;wgs_batch;wgs_diploid   \n",
       "2                                                         \n",
       "3                                                         \n",
       "4                                                         \n",
       "..                ...                               ...   \n",
       "149  plant-associated  wgs_single;wgs_batch;wgs_diploid   \n",
       "150          sediment  wgs_single;wgs_batch;wgs_diploid   \n",
       "151              soil  wgs_single;wgs_batch;wgs_diploid   \n",
       "152        wastewater  wgs_single;wgs_batch;wgs_diploid   \n",
       "153             water  wgs_single;wgs_batch;wgs_diploid   \n",
       "\n",
       "                                           Description       Example  \n",
       "0                                              Generic                \n",
       "1    Use for SARS-CoV-2 samples that are relevant t...                \n",
       "2                 Clinical or host-associated pathogen  SAMN02928182  \n",
       "3                Environmental, food or other pathogen  SAMN02730065  \n",
       "4    Use for bacteria or other unicellular microbes...  SAMN02911891  \n",
       "..                                                 ...           ...  \n",
       "149  Use for uncultivated virus genome identified i...                \n",
       "150  Use for uncultivated virus genome identified i...                \n",
       "151  Use for uncultivated virus genome identified i...                \n",
       "152  Use for uncultivated virus genome identified i...                \n",
       "153  Use for uncultivated virus genome identified i...                \n",
       "\n",
       "[154 rows x 8 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "package_dictionary = scoped_mapping.get_package_dictionary(biosample_packages_file)\n",
    "package_dictionary.to_sql(\n",
    "    \"package_dictionary\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "package_dictionary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa900ad8",
   "metadata": {},
   "source": [
    "## Do the Biosample checklist/package fields match any of the canonical values?\n",
    "\n",
    "Start by counting the Biosample rows/records. See XXX notes on extracting this *harmonized* database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0cfa7d0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.062162\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_row_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14300584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   biosample_row_count\n",
       "0             14300584"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select count(*) as biosample_row_count\n",
    "from biosample b\n",
    "\"\"\"\n",
    "[biosample_row_count, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=False\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "biosample_row_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc896a9c",
   "metadata": {},
   "source": [
    "## How many of those rows can be inner-joined with the canonical checklists/packages?\n",
    "Specifically, joining `biosample.package_name` with `package_dictionary.DisplayName`\n",
    "\n",
    "_Note that indices are built as part of the makefile_\n",
    "\n",
    "\n",
    "- create index biosample_package_name_idx on biosample(package_name);\n",
    "- create index package_dictionary_DisplayName_idx on package_dictionary(DisplayName);\n",
    "- create index biosample_package_idx on biosample(package);\n",
    "- create index biosample_p_pn_idx on biosample(package, package_name);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "03a8b90c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-06-11 10:59:42.867011\n",
      "2021-06-11 10:59:43.573479\n",
      "0:00:00.706468\n",
      "0:00:00.706468\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>canonical_package_name_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14300584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   canonical_package_name_count\n",
       "0                      14300584"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    count(*) as canonical_package_name_count\n",
    "from\n",
    "    biosample b\n",
    "inner join package_dictionary pd on\n",
    "    b.package_name = pd.DisplayName\n",
    "\"\"\"\n",
    "[canonical_package_name_count, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "canonical_package_name_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5951b7f1",
   "metadata": {},
   "source": [
    "## Combinations of `package` and `package_name` values in the Biosample dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "83485b99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-06-11 10:59:43.582597\n",
      "2021-06-11 10:59:44.958328\n",
      "0:00:01.375731\n",
      "0:00:01.375731\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Beta-lactamase.1.0</td>\n",
       "      <td>Beta-lactamase; version 1.0</td>\n",
       "      <td>556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>10186430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Human.1.0</td>\n",
       "      <td>Human; version 1.0</td>\n",
       "      <td>368893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Invertebrate.1.0</td>\n",
       "      <td>Invertebrate; version 1.0</td>\n",
       "      <td>131085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>28686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>Model.organism.animal.1.0</td>\n",
       "      <td>Model organism or animal; version 1.0</td>\n",
       "      <td>382980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>511040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>246497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>Plant.1.0</td>\n",
       "      <td>Plant; version 1.0</td>\n",
       "      <td>385939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Virus.1.0</td>\n",
       "      <td>Virus; version 1.0</td>\n",
       "      <td>15560</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       package  \\\n",
       "0           Beta-lactamase.1.0   \n",
       "1                  Generic.1.0   \n",
       "2                    Human.1.0   \n",
       "3             Invertebrate.1.0   \n",
       "4                  MIGS.ba.5.0   \n",
       "..                         ...   \n",
       "124  Model.organism.animal.1.0   \n",
       "125            Pathogen.cl.1.0   \n",
       "126           Pathogen.env.1.0   \n",
       "127                  Plant.1.0   \n",
       "128                  Virus.1.0   \n",
       "\n",
       "                                          package_name     count  \n",
       "0                          Beta-lactamase; version 1.0       556  \n",
       "1                                              Generic  10186430  \n",
       "2                                   Human; version 1.0    368893  \n",
       "3                            Invertebrate; version 1.0    131085  \n",
       "4         MIGS: cultured bacteria/archaea; version 5.0     28686  \n",
       "..                                                 ...       ...  \n",
       "124              Model organism or animal; version 1.0    382980  \n",
       "125  Pathogen: clinical or host-associated; version...    511040  \n",
       "126    Pathogen: environmental/food/other; version 1.0    246497  \n",
       "127                                 Plant; version 1.0    385939  \n",
       "128                                 Virus; version 1.0     15560  \n",
       "\n",
       "[129 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    package,\n",
    "    package_name,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    package ,\n",
    "    package_name\n",
    "order by\n",
    "    package ,\n",
    "    package_name\n",
    "\"\"\"\n",
    "[package_name_combos, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "package_name_combos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e15812f",
   "metadata": {},
   "source": [
    "## What about the Biosample `env_package` values?\n",
    "Are they also a small, highly regular set, like the `package` and `package_name` combinations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d042df82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.819255\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count\n",
       "0                              None  14083847\n",
       "1                   host-associated     49254\n",
       "2                         human-gut     47921\n",
       "3                             water     16367\n",
       "4                        human-skin     13706\n",
       "..                              ...       ...\n",
       "87                      env_package         1\n",
       "88                   gut microbiome         1\n",
       "89  marine sediment (ENVO:00002113)         1\n",
       "90                           saliva         1\n",
       "91       sea water, [ENVO:00002149]         1\n",
       "\n",
       "[92 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    env_package\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[env_package_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "env_package_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ab5ec5d",
   "metadata": {},
   "source": [
    "## `env_package` is a mixture of `ENVO` term ids and strings\n",
    "Those strings may or may not be the term's labels. There are many redundancies due to small spelling and punctuation variations.\n",
    "\n",
    "Start by tidying `env_package` etc. values recognized by INSDC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "82ba2408",
   "metadata": {},
   "outputs": [],
   "source": [
    "package_dictionary = scoped_mapping.make_tidy_col(\n",
    "    package_dictionary, \"EnvPackage\", \"eptidy\"\n",
    ")\n",
    "package_dictionary = scoped_mapping.make_tidy_col(\n",
    "    package_dictionary, \"EnvPackageDisplay\", \"epdtidy\"\n",
    ")\n",
    "\n",
    "# update in sqlite\n",
    "package_dictionary.to_sql(\n",
    "    \"package_dictionary\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70be9eb2",
   "metadata": {},
   "source": [
    "## Determine ID patterns for common ontologies, like `ENVO`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5d157f3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.020218\n"
     ]
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    distinct stanza\n",
    "    from statements s\n",
    "where\n",
    "    predicate = 'rdf:type'\n",
    "    and \"object\" = 'owl:Class'\n",
    "    and stanza = subject\"\"\"\n",
    "# include non-envo IDs that come from envo?\n",
    "[ids_from_selected_ontolgy, query_duration] = scoped_mapping.timed_query(\n",
    "    q, primary_ontology_cnx\n",
    ")\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dcf3bb61",
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_from_selected_ontolgy = scoped_mapping.add_prefix_col(\n",
    "    ids_from_selected_ontolgy, \"stanza\", \"prefix\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8b9e5da2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BFO': 'BFO:\\\\d{7}',\n",
       " 'CARO': 'CARO:\\\\d{7}',\n",
       " 'CHEBI': 'CHEBI:\\\\d{4,6}',\n",
       " 'ENVO': 'ENVO:\\\\d{7,8}',\n",
       " 'FAO': 'FAO:0000001',\n",
       " 'FOODON': 'FOODON:\\\\d{8}',\n",
       " 'GO': 'GO:\\\\d{7}',\n",
       " 'IAO': 'IAO:\\\\d{7}',\n",
       " 'NCBITaxon': 'NCBITaxon:\\\\d+',\n",
       " 'OBI': 'OBI:\\\\d{7}',\n",
       " 'PATO': 'PATO:\\\\d{7}',\n",
       " 'PCO': 'PCO:\\\\d{7}',\n",
       " 'PO': 'PO:\\\\d{7}',\n",
       " 'RO': 'RO:0002577',\n",
       " 'UBERON': 'UBERON:\\\\d{7}'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_patterns = scoped_mapping.get_multi_term_patterns(\n",
    "    ids_from_selected_ontolgy, \"stanza\", \"prefix\"\n",
    ")\n",
    "\n",
    "id_patterns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c56612",
   "metadata": {},
   "source": [
    "## Apply some normalization rules to the `env_package` values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8d8f0613",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>rhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td>env package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>marine sediment ()</td>\n",
       "      <td>marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water,</td>\n",
       "      <td>sea water</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count lhs  \\\n",
       "0                              None  14083847       \n",
       "1                   host-associated     49254       \n",
       "2                         human-gut     47921       \n",
       "3                             water     16367       \n",
       "4                        human-skin     13706       \n",
       "..                              ...       ...  ..   \n",
       "87                      env_package         1       \n",
       "88                   gut microbiome         1       \n",
       "89  marine sediment (ENVO:00002113)         1       \n",
       "90                           saliva         1       \n",
       "91       sea water, [ENVO:00002149]         1       \n",
       "\n",
       "                                rhs        extract    remaining_string  \\\n",
       "0                              None                                      \n",
       "1                   host-associated                    host-associated   \n",
       "2                         human-gut                          human-gut   \n",
       "3                             water                              water   \n",
       "4                        human-skin                         human-skin   \n",
       "..                              ...            ...                 ...   \n",
       "87                      env_package                        env_package   \n",
       "88                   gut microbiome                     gut microbiome   \n",
       "89  marine sediment (ENVO:00002113)  ENVO:00002113  marine sediment ()   \n",
       "90                           saliva                             saliva   \n",
       "91       sea water, [ENVO:00002149]  ENVO:00002149         sea water,    \n",
       "\n",
       "   remaining_tidied  \n",
       "0                    \n",
       "1   host associated  \n",
       "2         human gut  \n",
       "3             water  \n",
       "4        human skin  \n",
       "..              ...  \n",
       "87      env package  \n",
       "88   gut microbiome  \n",
       "89  marine sediment  \n",
       "90           saliva  \n",
       "91        sea water  \n",
       "\n",
       "[92 rows x 7 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_package_normalized = scoped_mapping.env_package_nomralizastion(\n",
    "    env_package_count, \"env_package\", id_patterns[target_onto_prefix]\n",
    ")\n",
    "\n",
    "# getting rid of redundant? 'string' column\n",
    "env_package_normalized = env_package_normalized[\n",
    "    [\n",
    "        \"env_package\",\n",
    "        \"count\",\n",
    "        \"lhs\",\n",
    "        \"rhs\",\n",
    "        \"extract\",\n",
    "        \"remaining_string\",\n",
    "        \"remaining_tidied\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "env_package_normalized"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5f346dd",
   "metadata": {},
   "source": [
    "pick \"EnvPackage\"/\"eptidy\" from package dictionary as canonical, not \"EnvPackageDisplay\"/\"epdtidy\"\n",
    "\n",
    "\n",
    "But still want to support making XXX values from YYY canonical according to \"EnvPackageDisplay\"/\"epdtidy\", so make a mapping/override table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6f91fa7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eptidy</th>\n",
       "      <th>epdtidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>microbial</td>\n",
       "      <td>microbial mat biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>miscellaneous</td>\n",
       "      <td>miscellaneous or artificial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater sludge</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           eptidy                      epdtidy\n",
       "21      microbial        microbial mat biofilm\n",
       "22  miscellaneous  miscellaneous or artificial\n",
       "26     wastewater            wastewater sludge"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epd_to_ep = package_dictionary[[\"eptidy\", \"epdtidy\"]]\n",
    "# drop duplicates\n",
    "epd_to_ep = epd_to_ep.drop_duplicates()\n",
    "\n",
    "# drop blank eptidy rows\n",
    "ep_blank_flag = epd_to_ep[\"eptidy\"].eq(\"\")\n",
    "epd_to_ep = epd_to_ep.loc[~ep_blank_flag]\n",
    "\n",
    "# drop rows where eptidy and epdtidy are the same\n",
    "identical_flag = epd_to_ep[\"eptidy\"] == epd_to_ep[\"epdtidy\"]\n",
    "epd_to_ep = epd_to_ep.loc[~identical_flag]\n",
    "\n",
    "epd_to_ep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d17acd46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'microbial mat biofilm': 'microbial',\n",
       " 'miscellaneous or artificial': 'miscellaneous',\n",
       " 'wastewater sludge': 'wastewater'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# and add to manualy asserted overrides above\n",
    "overrides_supplement = dict(zip(epd_to_ep[\"epdtidy\"], epd_to_ep[\"eptidy\"]))\n",
    "\n",
    "overrides_supplement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "77aca877",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'built environment': 'built',\n",
       " 'misc environment': 'miscellaneous',\n",
       " 'missing': 'no environmental package',\n",
       " 'unknown': 'no environmental package',\n",
       " 'default': 'no environmental package',\n",
       " 'unspecified': 'no environmental package',\n",
       " 'not available': 'no environmental package',\n",
       " 'not collected': 'no environmental package',\n",
       " 'miscellaneous natural or artificial environment': 'miscellaneous',\n",
       " 'not applicable': 'no environmental package',\n",
       " 'soil-associated': 'soil',\n",
       " 'soil associated': 'soil',\n",
       " 'microbial mat biofilm': 'microbial',\n",
       " 'miscellaneous or artificial': 'miscellaneous',\n",
       " 'wastewater sludge': 'wastewater'}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# not getting soil-associated\n",
    "env_package_overrides.update(overrides_supplement)\n",
    "\n",
    "env_package_overrides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c4c2c6f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_normalized = scoped_mapping.add_overrides(\n",
    "    env_package_normalized, \"remaining_tidied\", \"rt_override\", env_package_overrides\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3b7eb038",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>eptidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>No environmental package</td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>air</td>\n",
       "      <td>air</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>built</td>\n",
       "      <td>built</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>human-associated</td>\n",
       "      <td>human associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>human-oral</td>\n",
       "      <td>human oral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>human-vaginal</td>\n",
       "      <td>human vaginal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>microbial</td>\n",
       "      <td>microbial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>miscellaneous</td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>plant-associated</td>\n",
       "      <td>plant associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  EnvPackage                    eptidy\n",
       "0                                                     \n",
       "12  No environmental package  no environmental package\n",
       "13                       air                       air\n",
       "14                     built                     built\n",
       "15           host-associated           host associated\n",
       "16          human-associated          human associated\n",
       "17                 human-gut                 human gut\n",
       "18                human-oral                human oral\n",
       "19                human-skin                human skin\n",
       "20             human-vaginal             human vaginal\n",
       "21                 microbial                 microbial\n",
       "22             miscellaneous             miscellaneous\n",
       "23          plant-associated          plant associated\n",
       "24                  sediment                  sediment\n",
       "25                      soil                      soil\n",
       "26                wastewater                wastewater\n",
       "27                     water                     water"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "denorm_frame = package_dictionary[[\"EnvPackage\", \"eptidy\"]]\n",
    "denorm_frame = denorm_frame.drop_duplicates()\n",
    "denorm_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "73694ca0",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_normalized = env_package_normalized.merge(\n",
    "    denorm_frame, how=\"left\", left_on=\"rt_override\", right_on=\"eptidy\"\n",
    ")\n",
    "\n",
    "env_package_normalized = env_package_normalized[\n",
    "    [\n",
    "        \"env_package\",\n",
    "        \"count\",\n",
    "        \"lhs\",\n",
    "        \"rhs\",\n",
    "        \"extract\",\n",
    "        \"remaining_string\",\n",
    "        \"remaining_tidied\",\n",
    "        \"rt_override\",\n",
    "        \"EnvPackage\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "non_canonical_flag = env_package_normalized[\"EnvPackage\"].isna()\n",
    "env_package_normalized[\"is_canonical\"] = True\n",
    "env_package_normalized.loc[non_canonical_flag, \"is_canonical\"] = False\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db3fd267",
   "metadata": {},
   "source": [
    "- env_package = env_package annotation from NCBI Biosample file XXX\n",
    "- count = number of biosamples using that env_package annotation\n",
    "- lhs = checklist info\n",
    "- rhs = potential package info\n",
    "- extract = potential OBO ID from rhs column (currently harcoded and only looking for ENVO IDs)\n",
    "- remaining_string = rhs/string, with potential OBO IDs removed\n",
    "- remaining_tidied = remaining_string with case, whitespace and punctuation normailzastion\n",
    "- rt_override = some remaining_tidied values can be replaced according to env_package_overrides\n",
    "- EnvPackage = corresponding de-normalized value from package_dictionary\n",
    "- is_canonical = false when EnvPackage is NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f2d82ff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "env_package_normalized.to_sql(\n",
    "    \"env_package_normalized\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "\n",
    "env_package_normalized.to_csv(\n",
    "    \"biosample_env_package_normalizastion.tsv\", sep=ols_review_seperator, index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3766cbe7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>rhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>rt_override</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>is_canonical</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "      <td>host associated</td>\n",
       "      <td>host-associated</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "      <td>human gut</td>\n",
       "      <td>human-gut</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "      <td>human skin</td>\n",
       "      <td>human-skin</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td>env package</td>\n",
       "      <td>env package</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>marine sediment ()</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water,</td>\n",
       "      <td>sea water</td>\n",
       "      <td>sea water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count lhs  \\\n",
       "0                              None  14083847       \n",
       "1                   host-associated     49254       \n",
       "2                         human-gut     47921       \n",
       "3                             water     16367       \n",
       "4                        human-skin     13706       \n",
       "..                              ...       ...  ..   \n",
       "87                      env_package         1       \n",
       "88                   gut microbiome         1       \n",
       "89  marine sediment (ENVO:00002113)         1       \n",
       "90                           saliva         1       \n",
       "91       sea water, [ENVO:00002149]         1       \n",
       "\n",
       "                                rhs        extract    remaining_string  \\\n",
       "0                              None                                      \n",
       "1                   host-associated                    host-associated   \n",
       "2                         human-gut                          human-gut   \n",
       "3                             water                              water   \n",
       "4                        human-skin                         human-skin   \n",
       "..                              ...            ...                 ...   \n",
       "87                      env_package                        env_package   \n",
       "88                   gut microbiome                     gut microbiome   \n",
       "89  marine sediment (ENVO:00002113)  ENVO:00002113  marine sediment ()   \n",
       "90                           saliva                             saliva   \n",
       "91       sea water, [ENVO:00002149]  ENVO:00002149         sea water,    \n",
       "\n",
       "   remaining_tidied      rt_override       EnvPackage  is_canonical  \n",
       "0                                                              True  \n",
       "1   host associated  host associated  host-associated          True  \n",
       "2         human gut        human gut        human-gut          True  \n",
       "3             water            water            water          True  \n",
       "4        human skin       human skin       human-skin          True  \n",
       "..              ...              ...              ...           ...  \n",
       "87      env package      env package              NaN         False  \n",
       "88   gut microbiome   gut microbiome              NaN         False  \n",
       "89  marine sediment  marine sediment              NaN         False  \n",
       "90           saliva           saliva              NaN         False  \n",
       "91        sea water        sea water              NaN         False  \n",
       "\n",
       "[92 rows x 10 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "env_package_normalized"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2a7fed8",
   "metadata": {},
   "source": [
    "## What do the successful normalizations look like?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7c1fbb75",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.001594\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>EnvPackage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>not collected</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>host associated</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No environmental package</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        env_package     count lhs        extract                EnvPackage\n",
       "0              None  14083847                                             \n",
       "1   host-associated     49254                              host-associated\n",
       "2         human-gut     47921                                    human-gut\n",
       "3             water     16367                                        water\n",
       "4        human-skin     13706                                   human-skin\n",
       "..              ...       ...  ..            ...                       ...\n",
       "57    not collected         3                     No environmental package\n",
       "58  host associated         2                              host-associated\n",
       "59    ENVO:00000016         1      ENVO:00000016                          \n",
       "60         Sediment         1                                     sediment\n",
       "61          Unknown         1                     No environmental package\n",
       "\n",
       "[62 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    EnvPackage\n",
    "from\n",
    "    env_package_normalized\n",
    "where\n",
    "    is_canonical = 1\n",
    "\"\"\"\n",
    "[successful_normalizastions, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx\n",
    ")\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "successful_normalizastions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "740ffe3c",
   "metadata": {},
   "source": [
    "# Are there any normalization failures?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "17e08383",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.001350\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>EnvPackage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mimarks</td>\n",
       "      <td>479</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mouse-gut</td>\n",
       "      <td>406</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gut</td>\n",
       "      <td>172</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>biofilm</td>\n",
       "      <td>114</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-not providedsopharyngeal</td>\n",
       "      <td>107</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>mice gut</td>\n",
       "      <td>87</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CV</td>\n",
       "      <td>60</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>home, outdoor environment</td>\n",
       "      <td>44</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fermentation-associated</td>\n",
       "      <td>42</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sterile water</td>\n",
       "      <td>35</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Mouse_Gut</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>fresh water</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rock</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>arctic</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>freshwater sediment</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>seawater</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MIENS_16S</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MIMARKS_water</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>cave</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>plant</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>sea water</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Coastal water body (ENVO:02000049)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:02000049</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Deep-sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MIMARKS_16S</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           env_package  count lhs        extract EnvPackage\n",
       "0                              mimarks    479                          None\n",
       "1                            mouse-gut    406                          None\n",
       "2                                  gut    172                          None\n",
       "3                              biofilm    114                          None\n",
       "4       human-not providedsopharyngeal    107                          None\n",
       "5                             mice gut     87                          None\n",
       "6                                   CV     60                          None\n",
       "7            home, outdoor environment     44                          None\n",
       "8              fermentation-associated     42                          None\n",
       "9                        sterile water     35                          None\n",
       "10                           Mouse_Gut     27                          None\n",
       "11                         fresh water     27                          None\n",
       "12                                rock     12                          None\n",
       "13                              arctic     10                          None\n",
       "14                 freshwater sediment      9                          None\n",
       "15                            seawater      9                          None\n",
       "16                           MIENS_16S      6                          None\n",
       "17                       MIMARKS_water      6                          None\n",
       "18                                cave      5                          None\n",
       "19                               plant      3                          None\n",
       "20                           sea water      2                          None\n",
       "21  Coastal water body (ENVO:02000049)      1      ENVO:02000049       None\n",
       "22                      Deep-sea water      1                          None\n",
       "23                         MIMARKS_16S      1                          None\n",
       "24                           Sea water      1                          None\n",
       "25                         env_package      1                          None\n",
       "26                      gut microbiome      1                          None\n",
       "27     marine sediment (ENVO:00002113)      1      ENVO:00002113       None\n",
       "28                              saliva      1                          None\n",
       "29          sea water, [ENVO:00002149]      1      ENVO:00002149       None"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    EnvPackage\n",
    "from\n",
    "    env_package_normalized\n",
    "where\n",
    "    is_canonical = 0\n",
    "\"\"\"\n",
    "[normalizastion_failures, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "normalizastion_failures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "82953ddc",
   "metadata": {},
   "outputs": [],
   "source": [
    "cursor = biosample_cnx.cursor()\n",
    "statement = \"\"\"\n",
    "DROP TABLE if exists repaired_env_package ;\n",
    "\n",
    "CREATE TABLE repaired_env_package AS\n",
    "select\n",
    "\tb.id,\n",
    "\tepn.env_package as env_package_orig,\n",
    "\tepn.EnvPackage as env_package_rep\n",
    "from\n",
    "\tenv_package_normalized epn\n",
    "join biosample b on\n",
    "\tb.env_package = epn.env_package\n",
    "where\n",
    "\tis_canonical = 1\n",
    "\tand EnvPackage != '';\n",
    "\t\n",
    "select\n",
    "\tcount(*)\n",
    "from\n",
    "\trepaired_env_package rep\n",
    "\"\"\"\n",
    "\n",
    "cursor.executescript(statement)\n",
    "biosample_cnx.commit()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80370e30",
   "metadata": {},
   "source": [
    "# Utilizing taxonomy for broad subsetting\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46e3ee59",
   "metadata": {},
   "source": [
    "**This uses an SQLite database in which the transitive closure over subClassOf has already been materialized. See the README and Makefile.**\n",
    "\n",
    "Specifically, flag the Biosamples whose `taxon_id` indicates they are an unclassified entity. Ignoring the others will throw out samples of multicellular organisms, like fruit flies.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bfd8416",
   "metadata": {},
   "source": [
    "## Get a listing of all taxa that are transitive subclasses of `NCBITaxon:2787823`\n",
    "\n",
    "I.e. 'unclassified entities'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c40903d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:06.176164\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:1006967</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:1041057</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:1046002</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:1046003</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:1046004</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>NCBITaxon:939928</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>NCBITaxon:941420</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>NCBITaxon:941421</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>NCBITaxon:941422</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>NCBITaxon:942017</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>994 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               subject  unclassified\n",
       "0    NCBITaxon:1006967          True\n",
       "1    NCBITaxon:1041057          True\n",
       "2    NCBITaxon:1046002          True\n",
       "3    NCBITaxon:1046003          True\n",
       "4    NCBITaxon:1046004          True\n",
       "..                 ...           ...\n",
       "989   NCBITaxon:939928          True\n",
       "990   NCBITaxon:941420          True\n",
       "991   NCBITaxon:941421          True\n",
       "992   NCBITaxon:941422          True\n",
       "993   NCBITaxon:942017          True\n",
       "\n",
       "[994 rows x 2 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    distinct s.subject\n",
    "from\n",
    "    entailed_edge ee\n",
    "join statements s on\n",
    "    ee.subject = s.subject\n",
    "where\n",
    "    ee.predicate = 'rdfs:subClassOf'\n",
    "    and ee.object = 'NCBITaxon:2787823'\n",
    "    and s.predicate = 'rdfs:label'\n",
    "\"\"\"\n",
    "[unclassified_taxa, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "unclassified_taxa[\"unclassified\"] = True\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "unclassified_taxa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cad3ffdb",
   "metadata": {},
   "source": [
    "## Get taxon counts from the Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a5d8f028",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:01.172136\n"
     ]
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    taxonomy_id biosample_taxid,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    taxonomy_id\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[biosample_tax_id_counts, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "biosample_tax_id_counts[\"curie\"] = \"NCBITaxon:\" + biosample_tax_id_counts[\n",
    "    \"biosample_taxid\"\n",
    "].astype(str)\n",
    "\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cf3a275",
   "metadata": {},
   "source": [
    "## Merge the two taxonomy dataframes\n",
    "\n",
    "I.e. flag the the Biosample records whose `taxonomy_id` field belongs to a subclass of 'unclassified entries'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b7ad29a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>curie</th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       biosample_taxid    count             curie           subject  \\\n",
       "0                 9606  6819707    NCBITaxon:9606               NaN   \n",
       "1                10090   964219   NCBITaxon:10090               NaN   \n",
       "2               408170   290862  NCBITaxon:408170  NCBITaxon:408170   \n",
       "3               410658   280666  NCBITaxon:410658  NCBITaxon:410658   \n",
       "4               646099   208741  NCBITaxon:646099  NCBITaxon:646099   \n",
       "...                ...      ...               ...               ...   \n",
       "163372          999891        1  NCBITaxon:999891               NaN   \n",
       "163373          999892        1  NCBITaxon:999892               NaN   \n",
       "163374          999898        1  NCBITaxon:999898               NaN   \n",
       "163375          999931        1  NCBITaxon:999931               NaN   \n",
       "163376          999947        1  NCBITaxon:999947               NaN   \n",
       "\n",
       "        unclassified  \n",
       "0              False  \n",
       "1              False  \n",
       "2               True  \n",
       "3               True  \n",
       "4               True  \n",
       "...              ...  \n",
       "163372         False  \n",
       "163373         False  \n",
       "163374         False  \n",
       "163375         False  \n",
       "163376         False  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biosample_tax_id_counts = biosample_tax_id_counts.merge(\n",
    "    unclassified_taxa, left_on=\"curie\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "biosample_tax_id_counts.unclassified.fillna(False, inplace=True)\n",
    "\n",
    "biosample_tax_id_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f7aff98",
   "metadata": {},
   "source": [
    "## Add labels to all taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a6acc349",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:03.276666\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>unclassified</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>False</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>False</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>True</td>\n",
       "      <td>human gut metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>True</td>\n",
       "      <td>soil metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>True</td>\n",
       "      <td>human metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Bacillus amyloliquefaciens TA208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>[Propionibacterium] humerusii P08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Peptococcaceae bacterium CEB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Barrientosiimonas humi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Angomonas ambiguus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   curie biosample_taxid    count  unclassified  \\\n",
       "0         NCBITaxon:9606            9606  6819707         False   \n",
       "1        NCBITaxon:10090           10090   964219         False   \n",
       "2       NCBITaxon:408170          408170   290862          True   \n",
       "3       NCBITaxon:410658          410658   280666          True   \n",
       "4       NCBITaxon:646099          646099   208741          True   \n",
       "...                  ...             ...      ...           ...   \n",
       "163372  NCBITaxon:999891          999891        1         False   \n",
       "163373  NCBITaxon:999892          999892        1         False   \n",
       "163374  NCBITaxon:999898          999898        1         False   \n",
       "163375  NCBITaxon:999931          999931        1         False   \n",
       "163376  NCBITaxon:999947          999947        1         False   \n",
       "\n",
       "                                    label  \n",
       "0                            Homo sapiens  \n",
       "1                            Mus musculus  \n",
       "2                    human gut metagenome  \n",
       "3                         soil metagenome  \n",
       "4                        human metagenome  \n",
       "...                                   ...  \n",
       "163372   Bacillus amyloliquefaciens TA208  \n",
       "163373  [Propionibacterium] humerusii P08  \n",
       "163374      Peptococcaceae bacterium CEB3  \n",
       "163375             Barrientosiimonas humi  \n",
       "163376                 Angomonas ambiguus  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    subject ,\n",
    "    value\n",
    "from statements\n",
    "where\n",
    "    predicate = 'rdfs:label' and subject = stanza\n",
    "\"\"\"\n",
    "[all_tax_labels, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "\n",
    "biosample_tax_id_counts = biosample_tax_id_counts.merge(\n",
    "    all_tax_labels, left_on=\"curie\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "\n",
    "biosample_tax_id_counts = biosample_tax_id_counts[\n",
    "    [\"curie\", \"biosample_taxid\", \"count\", \"unclassified\", \"value\"]\n",
    "]\n",
    "biosample_tax_id_counts.columns = [\n",
    "    \"curie\",\n",
    "    \"biosample_taxid\",\n",
    "    \"count\",\n",
    "    \"unclassified\",\n",
    "    \"label\",\n",
    "]\n",
    "\n",
    "print(query_duration)\n",
    "biosample_tax_id_counts.to_sql(\n",
    "    \"biobiosample_tax_id_counts\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")\n",
    "\n",
    "biosample_tax_id_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "554a82c0",
   "metadata": {},
   "source": [
    "**Almost all of the taxa that are common in the biosample collection are either unclassified/metagenomes or easily recognized cellular organisms. Cellular organism samples are de-prioritized in this exercise**\n",
    "\n",
    "Exceptions include:\n",
    "- 32630 = synthetic construct (other entries; other sequences; artificial sequences)\n",
    "    - 'other entries' would add 16k rows on top of the 1k 'unclassified entities'\n",
    "    - metagenomes account for 331 of the 'unclassified entities'\n",
    "    - there are also a small number of uncultured/unclassified microorganisms in the biosample dataset\n",
    "- 77133 = uncultured bacterium (cellular organisms; Bacteria; environmental samples)\n",
    "    - 'cellular organisms' would add 2M rows on top of the 1k 'unclassified entities'\n",
    "    - 'cellular organisms; Bacteria; environmental samples' adds 26k\n",
    "    \n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2363d41",
   "metadata": {},
   "source": [
    "## Get a table of MIxS annotations to be mapped to ontology classes.\n",
    "\n",
    "Explicitly scope based on normalized package data. These values were set at the top of this notebook.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "91dcca09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env_medium\n",
      "env_package_normalized.EnvPackage\n",
      "soil\n"
     ]
    }
   ],
   "source": [
    "print(biosample_col_to_map)\n",
    "print(scoping_col)\n",
    "print(scoping_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80727cfd",
   "metadata": {},
   "source": [
    "**In this case, the scoping includes an inner join requirement for 'unclassified entities'**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7e777c8a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_medium  count\n",
       "0           soil  12400\n",
       "1           Soil    439\n",
       "2    rhizosphere    238\n",
       "3      peat soil    224\n",
       "4      ENVO:soil    214\n",
       "..           ...    ...\n",
       "101            4      1\n",
       "102           24      1\n",
       "103           22      1\n",
       "104           15      1\n",
       "105           13      1\n",
       "\n",
       "[106 rows x 2 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = (\n",
    "    \"select \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\", count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "join env_package_normalized on\n",
    "    b.env_package = env_package_normalized.env_package\n",
    "inner join biobiosample_tax_id_counts stic on\n",
    "    b.taxonomy_id = stic.biosample_taxid\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"' group by \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\"\n",
    "order by\n",
    "    count(*) desc\"\"\"\n",
    ")\n",
    "\n",
    "[mapping_candidates, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86f43bb4",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## The Biosample format allows for pipe-delimited environmental package lists. \n",
    "\n",
    "Separate those out into their components.\n",
    "\n",
    "----\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "45e1859f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       repeated     splitted  part_count  seq\n",
       "0          soil         soil           1    1\n",
       "0          Soil         Soil           1    1\n",
       "0   rhizosphere  rhizosphere           1    1\n",
       "0     peat soil    peat soil           1    1\n",
       "0     ENVO:soil    ENVO:soil           1    1\n",
       "..          ...          ...         ...  ...\n",
       "0             4            4           1    1\n",
       "0            24           24           1    1\n",
       "0            22           22           1    1\n",
       "0            15           15           1    1\n",
       "0            13           13           1    1\n",
       "\n",
       "[106 rows x 4 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_frames = []\n",
    "for row in mapping_candidates.itertuples(index=True, name=\"PandaZ\"):\n",
    "    split_check = row[1]\n",
    "    if split_check is None:\n",
    "        split_check = \"\"\n",
    "    splitted = pd.Series(split_check.split(\"|\"))\n",
    "    splitted_count = len(splitted)\n",
    "    repeated = [split_check] * splitted_count\n",
    "    repeated = pd.Series(repeated)\n",
    "    as_frame = pd.DataFrame(dict(repeated=repeated, splitted=splitted)).reset_index()\n",
    "    seq_list = list(range(1, splitted_count + 1))\n",
    "    as_frame[\"part_count\"] = splitted_count\n",
    "    as_frame[\"seq\"] = seq_list\n",
    "    multi_frames.append(as_frame)\n",
    "concat_frame = pd.concat(multi_frames)\n",
    "concat_frame = concat_frame[[\"repeated\", \"splitted\", \"part_count\", \"seq\"]]\n",
    "concat_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c67a272e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12400</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>224</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_medium  count     repeated     splitted  part_count  seq\n",
       "0           soil  12400         soil         soil         1.0  1.0\n",
       "1           Soil    439         Soil         Soil         1.0  1.0\n",
       "2    rhizosphere    238  rhizosphere  rhizosphere         1.0  1.0\n",
       "3      peat soil    224    peat soil    peat soil         1.0  1.0\n",
       "4      ENVO:soil    214    ENVO:soil    ENVO:soil         1.0  1.0\n",
       "..           ...    ...          ...          ...         ...  ...\n",
       "101            4      1            4            4         1.0  1.0\n",
       "102           24      1           24           24         1.0  1.0\n",
       "103           22      1           22           22         1.0  1.0\n",
       "104           15      1           15           15         1.0  1.0\n",
       "105           13      1           13           13         1.0  1.0\n",
       "\n",
       "[106 rows x 6 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    concat_frame, left_on=biosample_col_to_map, right_on=\"repeated\", how=\"left\"\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3af421ac",
   "metadata": {},
   "source": [
    "## Normalize a few different ways `ENVO` IDs have been entered\n",
    "In the Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ed3a89fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12400</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>224</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_medium  count     repeated     splitted  part_count  seq  \\\n",
       "0           soil  12400         soil         soil         1.0  1.0   \n",
       "1           Soil    439         Soil         Soil         1.0  1.0   \n",
       "2    rhizosphere    238  rhizosphere  rhizosphere         1.0  1.0   \n",
       "3      peat soil    224    peat soil    peat soil         1.0  1.0   \n",
       "4      ENVO:soil    214    ENVO:soil    ENVO:soil         1.0  1.0   \n",
       "..           ...    ...          ...          ...         ...  ...   \n",
       "101            4      1            4            4         1.0  1.0   \n",
       "102           24      1           24           24         1.0  1.0   \n",
       "103           22      1           22           22         1.0  1.0   \n",
       "104           15      1           15           15         1.0  1.0   \n",
       "105           13      1           13           13         1.0  1.0   \n",
       "\n",
       "       term_tidy  \n",
       "0           soil  \n",
       "1           Soil  \n",
       "2    rhizosphere  \n",
       "3      peat soil  \n",
       "4      ENVO:soil  \n",
       "..           ...  \n",
       "101            4  \n",
       "102           24  \n",
       "103           22  \n",
       "104           15  \n",
       "105           13  \n",
       "\n",
       "[106 rows x 7 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this helps by standardizing what could be an ID\n",
    "# but if the RHS is text, then envo should just be removed\n",
    "mapping_candidates[\"term_tidy\"] = mapping_candidates.splitted.str.replace(\n",
    "    target_onto_prefix.lower() + \"[:_ ]\",\n",
    "    target_onto_prefix + \":\",\n",
    "    regex=True,\n",
    "    case=False,\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cc22a1a",
   "metadata": {},
   "source": [
    "# Now try to extract ontology terms that are already present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e423fcb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12400</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>224</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td></td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td></td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_medium  count     repeated     splitted  part_count  seq  \\\n",
       "0           soil  12400         soil         soil         1.0  1.0   \n",
       "1           Soil    439         Soil         Soil         1.0  1.0   \n",
       "2    rhizosphere    238  rhizosphere  rhizosphere         1.0  1.0   \n",
       "3      peat soil    224    peat soil    peat soil         1.0  1.0   \n",
       "4      ENVO:soil    214    ENVO:soil    ENVO:soil         1.0  1.0   \n",
       "..           ...    ...          ...          ...         ...  ...   \n",
       "101            4      1            4            4         1.0  1.0   \n",
       "102           24      1           24           24         1.0  1.0   \n",
       "103           22      1           22           22         1.0  1.0   \n",
       "104           15      1           15           15         1.0  1.0   \n",
       "105           13      1           13           13         1.0  1.0   \n",
       "\n",
       "       term_tidy       string extract remaining_string remaining_tidied  \n",
       "0           soil         soil                     soil             soil  \n",
       "1           Soil         Soil                     Soil             soil  \n",
       "2    rhizosphere  rhizosphere              rhizosphere      rhizosphere  \n",
       "3      peat soil    peat soil                peat soil        peat soil  \n",
       "4      ENVO:soil    ENVO:soil                ENVO:soil             soil  \n",
       "..           ...          ...     ...              ...              ...  \n",
       "101            4            4                        4                4  \n",
       "102           24           24                       24               24  \n",
       "103           22           22                       22               22  \n",
       "104           15           15                       15               15  \n",
       "105           13           13                       13               13  \n",
       "\n",
       "[106 rows x 11 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_series_decomposition = scoped_mapping.decompose_series(\n",
    "    mapping_candidates[\"term_tidy\"], id_patterns[target_onto_prefix]\n",
    ")\n",
    "\n",
    "mapping_candidates = pd.concat(\n",
    "    [mapping_candidates, candidate_series_decomposition], axis=1\n",
    ")\n",
    "\n",
    "mapping_candidates[\n",
    "    \"remaining_tidied\"\n",
    "] = mapping_candidates.remaining_tidied.str.replace(\n",
    "    target_onto_prefix.lower() + \"[:_ ]\", \"\", regex=True, case=False\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccd038ff",
   "metadata": {},
   "source": [
    "## Join the extracted IDs with their labels\n",
    "\n",
    "_Start by connecting to the rdftab database from which the terms and label-like annotations will be obtained?_"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db5d19a9",
   "metadata": {},
   "source": [
    "## extracting the labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "3bcd7e47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IAO:0000111</td>\n",
       "      <td>editor preferred term~editor preferred label</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IAO:0000112</td>\n",
       "      <td>example of usage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IAO:0000114</td>\n",
       "      <td>has curation status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IAO:0000115</td>\n",
       "      <td>definition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IAO:0000116</td>\n",
       "      <td>editor note</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6774</th>\n",
       "      <td>ENVO:01001862</td>\n",
       "      <td>Solar radiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6775</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q2&gt;</td>\n",
       "      <td>Earth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q2306597&gt;</td>\n",
       "      <td>Suni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6777</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q525&gt;</td>\n",
       "      <td>Sol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6778</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q715269&gt;</td>\n",
       "      <td>World Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6779 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       subject  \\\n",
       "0                                  IAO:0000111   \n",
       "1                                  IAO:0000112   \n",
       "2                                  IAO:0000114   \n",
       "3                                  IAO:0000115   \n",
       "4                                  IAO:0000116   \n",
       "...                                        ...   \n",
       "6774                             ENVO:01001862   \n",
       "6775        <https://www.wikidata.org/wiki/Q2>   \n",
       "6776  <https://www.wikidata.org/wiki/Q2306597>   \n",
       "6777      <https://www.wikidata.org/wiki/Q525>   \n",
       "6778   <https://www.wikidata.org/wiki/Q715269>   \n",
       "\n",
       "                                             value  \n",
       "0     editor preferred term~editor preferred label  \n",
       "1                                 example of usage  \n",
       "2                              has curation status  \n",
       "3                                       definition  \n",
       "4                                      editor note  \n",
       "...                                            ...  \n",
       "6774                               Solar radiation  \n",
       "6775                                         Earth  \n",
       "6776                                          Suni  \n",
       "6777                                           Sol  \n",
       "6778                                   World Ocean  \n",
       "\n",
       "[6779 rows x 2 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    subject ,\n",
    "    value\n",
    "from\n",
    "    statements s\n",
    "where\n",
    "    predicate = 'rdfs:label'\n",
    "\"\"\"\n",
    "\n",
    "[onto_labels, query_duration] = scoped_mapping.timed_query(q, primary_ontology_cnx)\n",
    "\n",
    "onto_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "840b922f",
   "metadata": {},
   "source": [
    "## and merging "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8543ed0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12400</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>224</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td></td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td></td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_medium  count     repeated     splitted  part_count  seq  \\\n",
       "0           soil  12400         soil         soil         1.0  1.0   \n",
       "1           Soil    439         Soil         Soil         1.0  1.0   \n",
       "2    rhizosphere    238  rhizosphere  rhizosphere         1.0  1.0   \n",
       "3      peat soil    224    peat soil    peat soil         1.0  1.0   \n",
       "4      ENVO:soil    214    ENVO:soil    ENVO:soil         1.0  1.0   \n",
       "..           ...    ...          ...          ...         ...  ...   \n",
       "101            4      1            4            4         1.0  1.0   \n",
       "102           24      1           24           24         1.0  1.0   \n",
       "103           22      1           22           22         1.0  1.0   \n",
       "104           15      1           15           15         1.0  1.0   \n",
       "105           13      1           13           13         1.0  1.0   \n",
       "\n",
       "       term_tidy       string extract remaining_string remaining_tidied  \\\n",
       "0           soil         soil                     soil             soil   \n",
       "1           Soil         Soil                     Soil             soil   \n",
       "2    rhizosphere  rhizosphere              rhizosphere      rhizosphere   \n",
       "3      peat soil    peat soil                peat soil        peat soil   \n",
       "4      ENVO:soil    ENVO:soil                ENVO:soil             soil   \n",
       "..           ...          ...     ...              ...              ...   \n",
       "101            4            4                        4                4   \n",
       "102           24           24                       24               24   \n",
       "103           22           22                       22               22   \n",
       "104           15           15                       15               15   \n",
       "105           13           13                       13               13   \n",
       "\n",
       "    subject value  \n",
       "0       NaN   NaN  \n",
       "1       NaN   NaN  \n",
       "2       NaN   NaN  \n",
       "3       NaN   NaN  \n",
       "4       NaN   NaN  \n",
       "..      ...   ...  \n",
       "101     NaN   NaN  \n",
       "102     NaN   NaN  \n",
       "103     NaN   NaN  \n",
       "104     NaN   NaN  \n",
       "105     NaN   NaN  \n",
       "\n",
       "[106 rows x 13 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"extract\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64b3baef",
   "metadata": {},
   "source": [
    "## Use cosine string distance to see if the labels match closely enough\n",
    "\n",
    "I.e. the labels claimed by the Biosample data set and the labels asserted in the ontology. if they're close enough, consider the assigned ID legit\n",
    "\n",
    "\n",
    "_How close is close enough?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "abd0a035",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "      <th>cosine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12400</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>224</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td></td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td></td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_medium  count     repeated     splitted  part_count  seq  \\\n",
       "0           soil  12400         soil         soil         1.0  1.0   \n",
       "1           Soil    439         Soil         Soil         1.0  1.0   \n",
       "2    rhizosphere    238  rhizosphere  rhizosphere         1.0  1.0   \n",
       "3      peat soil    224    peat soil    peat soil         1.0  1.0   \n",
       "4      ENVO:soil    214    ENVO:soil    ENVO:soil         1.0  1.0   \n",
       "..           ...    ...          ...          ...         ...  ...   \n",
       "101            4      1            4            4         1.0  1.0   \n",
       "102           24      1           24           24         1.0  1.0   \n",
       "103           22      1           22           22         1.0  1.0   \n",
       "104           15      1           15           15         1.0  1.0   \n",
       "105           13      1           13           13         1.0  1.0   \n",
       "\n",
       "       term_tidy       string extract remaining_string remaining_tidied  \\\n",
       "0           soil         soil                     soil             soil   \n",
       "1           Soil         Soil                     Soil             soil   \n",
       "2    rhizosphere  rhizosphere              rhizosphere      rhizosphere   \n",
       "3      peat soil    peat soil                peat soil        peat soil   \n",
       "4      ENVO:soil    ENVO:soil                ENVO:soil             soil   \n",
       "..           ...          ...     ...              ...              ...   \n",
       "101            4            4                        4                4   \n",
       "102           24           24                       24               24   \n",
       "103           22           22                       22               22   \n",
       "104           15           15                       15               15   \n",
       "105           13           13                       13               13   \n",
       "\n",
       "    subject value  cosine  \n",
       "0       NaN           1.0  \n",
       "1       NaN           1.0  \n",
       "2       NaN           1.0  \n",
       "3       NaN           1.0  \n",
       "4       NaN           1.0  \n",
       "..      ...   ...     ...  \n",
       "101     NaN           1.0  \n",
       "102     NaN           1.0  \n",
       "103     NaN           1.0  \n",
       "104     NaN           1.0  \n",
       "105     NaN           1.0  \n",
       "\n",
       "[106 rows x 14 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_cosine_obj = Cosine(my_string_dist_arg)\n",
    "mapping_candidates[\"value\"] = mapping_candidates[\"value\"].fillna(\"\")\n",
    "mapping_candidates[\"remaining_tidied\"] = mapping_candidates[\"remaining_tidied\"].fillna(\n",
    "    \"\"\n",
    ")\n",
    "mapping_candidates[\"cosine\"] = mapping_candidates.apply(\n",
    "    lambda my_row: my_cosine_obj.distance(\n",
    "        my_row[\"remaining_tidied\"].lower(), my_row[\"value\"].lower()\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e968917d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2358a89",
   "metadata": {},
   "source": [
    "**Previously, we did a reality check on the claimed IDs and labels. If a label is claimed without any ID, that could still be a path to an ontology term.**\n",
    "\n",
    "We'll be doing some merging, so make sure column names are meaningful and unique\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "092aced0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12400</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>224</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td></td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td></td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_medium  count     repeated     splitted  part_count  seq  \\\n",
       "0           soil  12400         soil         soil         1.0  1.0   \n",
       "1           Soil    439         Soil         Soil         1.0  1.0   \n",
       "2    rhizosphere    238  rhizosphere  rhizosphere         1.0  1.0   \n",
       "3      peat soil    224    peat soil    peat soil         1.0  1.0   \n",
       "4      ENVO:soil    214    ENVO:soil    ENVO:soil         1.0  1.0   \n",
       "..           ...    ...          ...          ...         ...  ...   \n",
       "101            4      1            4            4         1.0  1.0   \n",
       "102           24      1           24           24         1.0  1.0   \n",
       "103           22      1           22           22         1.0  1.0   \n",
       "104           15      1           15           15         1.0  1.0   \n",
       "105           13      1           13           13         1.0  1.0   \n",
       "\n",
       "       term_tidy       string extract remaining_string remaining_tidied  \\\n",
       "0           soil         soil                     soil             soil   \n",
       "1           Soil         Soil                     Soil             soil   \n",
       "2    rhizosphere  rhizosphere              rhizosphere      rhizosphere   \n",
       "3      peat soil    peat soil                peat soil        peat soil   \n",
       "4      ENVO:soil    ENVO:soil                ENVO:soil             soil   \n",
       "..           ...          ...     ...              ...              ...   \n",
       "101            4            4                        4                4   \n",
       "102           24           24                       24               24   \n",
       "103           22           22                       22               22   \n",
       "104           15           15                       15               15   \n",
       "105           13           13                       13               13   \n",
       "\n",
       "    term_id lab_from_id  lfi_cosine term_id_from_lab        value  \n",
       "0       NaN                     1.0    ENVO:00001998         soil  \n",
       "1       NaN                     1.0    ENVO:00001998         soil  \n",
       "2       NaN                     1.0    ENVO:00005801  rhizosphere  \n",
       "3       NaN                     1.0    ENVO:00005774    peat soil  \n",
       "4       NaN                     1.0    ENVO:00001998         soil  \n",
       "..      ...         ...         ...              ...          ...  \n",
       "101     NaN                     1.0              NaN          NaN  \n",
       "102     NaN                     1.0              NaN          NaN  \n",
       "103     NaN                     1.0              NaN          NaN  \n",
       "104     NaN                     1.0              NaN          NaN  \n",
       "105     NaN                     1.0              NaN          NaN  \n",
       "\n",
       "[106 rows x 16 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates.columns = [\n",
    "    biosample_col_to_map,\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"term_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "]\n",
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"remaining_tidied\", right_on=\"value\", how=\"left\"\n",
    ")\n",
    "\n",
    "mapping_candidates.columns = [\n",
    "    biosample_col_to_map,\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"term_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "    \"term_id_from_lab\",\n",
    "    \"value\",\n",
    "]\n",
    "\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41082d99",
   "metadata": {},
   "source": [
    "## Find consensus term IDs and labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "4f5e3fdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates[\"consensus_id\"] = mapping_candidates[\"term_id_from_lab\"]\n",
    "mapping_candidates[\"consensus_lab\"] = mapping_candidates[\"value\"]\n",
    "\n",
    "flag = mapping_candidates[\"consensus_id\"].isnull() & (\n",
    "    (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates.lfi_cosine.le(my_max_string_dist)\n",
    "    )\n",
    "    | (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates[\"remaining_tidied\"].eq(\"\")\n",
    "    )\n",
    ")\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"term_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_id\"] = replacements\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"lab_from_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_lab\"] = replacements\n",
    "\n",
    "flag = mapping_candidates.consensus_id.isna()\n",
    "antiflag = ~flag\n",
    "mapping_candidates[\"id_or_lab_ok\"] = antiflag\n",
    "\n",
    "mapping_candidates[\"assembled_consensus\"] = (\n",
    "    mapping_candidates[\"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + mapping_candidates[\"consensus_id\"]\n",
    "    + \"]\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6af120f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "      <th>id_or_lab_ok</th>\n",
       "      <th>assembled_consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12400</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>True</td>\n",
       "      <td>soil [ENVO:00001998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Soil</td>\n",
       "      <td></td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>True</td>\n",
       "      <td>soil [ENVO:00001998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td></td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>True</td>\n",
       "      <td>rhizosphere [ENVO:00005801]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>224</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td></td>\n",
       "      <td>peat soil</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>True</td>\n",
       "      <td>peat soil [ENVO:00005774]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>True</td>\n",
       "      <td>soil [ENVO:00001998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>24</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td></td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td></td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td></td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_medium  count     repeated     splitted  part_count  seq  \\\n",
       "0           soil  12400         soil         soil         1.0  1.0   \n",
       "1           Soil    439         Soil         Soil         1.0  1.0   \n",
       "2    rhizosphere    238  rhizosphere  rhizosphere         1.0  1.0   \n",
       "3      peat soil    224    peat soil    peat soil         1.0  1.0   \n",
       "4      ENVO:soil    214    ENVO:soil    ENVO:soil         1.0  1.0   \n",
       "..           ...    ...          ...          ...         ...  ...   \n",
       "101            4      1            4            4         1.0  1.0   \n",
       "102           24      1           24           24         1.0  1.0   \n",
       "103           22      1           22           22         1.0  1.0   \n",
       "104           15      1           15           15         1.0  1.0   \n",
       "105           13      1           13           13         1.0  1.0   \n",
       "\n",
       "       term_tidy       string extract remaining_string remaining_tidied  \\\n",
       "0           soil         soil                     soil             soil   \n",
       "1           Soil         Soil                     Soil             soil   \n",
       "2    rhizosphere  rhizosphere              rhizosphere      rhizosphere   \n",
       "3      peat soil    peat soil                peat soil        peat soil   \n",
       "4      ENVO:soil    ENVO:soil                ENVO:soil             soil   \n",
       "..           ...          ...     ...              ...              ...   \n",
       "101            4            4                        4                4   \n",
       "102           24           24                       24               24   \n",
       "103           22           22                       22               22   \n",
       "104           15           15                       15               15   \n",
       "105           13           13                       13               13   \n",
       "\n",
       "    term_id lab_from_id  lfi_cosine term_id_from_lab        value  \\\n",
       "0       NaN                     1.0    ENVO:00001998         soil   \n",
       "1       NaN                     1.0    ENVO:00001998         soil   \n",
       "2       NaN                     1.0    ENVO:00005801  rhizosphere   \n",
       "3       NaN                     1.0    ENVO:00005774    peat soil   \n",
       "4       NaN                     1.0    ENVO:00001998         soil   \n",
       "..      ...         ...         ...              ...          ...   \n",
       "101     NaN                     1.0              NaN          NaN   \n",
       "102     NaN                     1.0              NaN          NaN   \n",
       "103     NaN                     1.0              NaN          NaN   \n",
       "104     NaN                     1.0              NaN          NaN   \n",
       "105     NaN                     1.0              NaN          NaN   \n",
       "\n",
       "      consensus_id consensus_lab  id_or_lab_ok          assembled_consensus  \n",
       "0    ENVO:00001998          soil          True         soil [ENVO:00001998]  \n",
       "1    ENVO:00001998          soil          True         soil [ENVO:00001998]  \n",
       "2    ENVO:00005801   rhizosphere          True  rhizosphere [ENVO:00005801]  \n",
       "3    ENVO:00005774     peat soil          True    peat soil [ENVO:00005774]  \n",
       "4    ENVO:00001998          soil          True         soil [ENVO:00001998]  \n",
       "..             ...           ...           ...                          ...  \n",
       "101            NaN           NaN         False                          NaN  \n",
       "102            NaN           NaN         False                          NaN  \n",
       "103            NaN           NaN         False                          NaN  \n",
       "104            NaN           NaN         False                          NaN  \n",
       "105            NaN           NaN         False                          NaN  \n",
       "\n",
       "[106 rows x 20 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "4c62116e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_candidates.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47aa8012",
   "metadata": {},
   "source": [
    "## Save these easy term mappings to SQLite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "3d8015be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_candidates.to_sql('mapping_scratch', biosample_cnx, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13d44c8d",
   "metadata": {},
   "source": [
    "## For which Biosample annotations were no easy mappings found?\n",
    "How many Biosamples uses those annotations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c29338b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = ~mapping_candidates.id_or_lab_ok\n",
    "needs_search = mapping_candidates.loc[flag]\n",
    "\n",
    "needs_search = needs_search[[\"remaining_tidied\", \"count\"]]\n",
    "\n",
    "sum_by_needed = needs_search.groupby(\"remaining_tidied\")[\"count\"].sum()\n",
    "\n",
    "sum_by_needed = sum_by_needed.to_frame()\n",
    "sum_by_needed[\"remaining_tidied\"] = sum_by_needed.index\n",
    "\n",
    "sum_by_needed = sum_by_needed.sort_values(\"count\", ascending=False)\n",
    "sum_by_needed.reset_index(drop=True)\n",
    "\n",
    "sum_by_needed.to_sql(\"sum_by_needed\", biosample_cnx, if_exists=\"replace\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "91fec97a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>organic soil</th>\n",
       "      <td>162</td>\n",
       "      <td>organic soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <td>113</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dna</th>\n",
       "      <td>99</td>\n",
       "      <td>dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mine tailings</th>\n",
       "      <td>90</td>\n",
       "      <td>mine tailings</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mineral soil</th>\n",
       "      <td>80</td>\n",
       "      <td>mineral soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mud soil</th>\n",
       "      <td>1</td>\n",
       "      <td>mud soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>paddy soil</th>\n",
       "      <td>1</td>\n",
       "      <td>paddy soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>replicate 1 soil from baodongyu of taoyuan</th>\n",
       "      <td>1</td>\n",
       "      <td>replicate 1 soil from baodongyu of taoyuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>replicate 1 soil from gushi of taoyuan</th>\n",
       "      <td>1</td>\n",
       "      <td>replicate 1 soil from gushi of taoyuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>topsoil</th>\n",
       "      <td>1</td>\n",
       "      <td>topsoil</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            count  \\\n",
       "remaining_tidied                                    \n",
       "organic soil                                  162   \n",
       "                                              113   \n",
       "dna                                            99   \n",
       "mine tailings                                  90   \n",
       "mineral soil                                   80   \n",
       "...                                           ...   \n",
       "mud soil                                        1   \n",
       "paddy soil                                      1   \n",
       "replicate 1 soil from baodongyu of taoyuan      1   \n",
       "replicate 1 soil from gushi of taoyuan          1   \n",
       "topsoil                                         1   \n",
       "\n",
       "                                                                      remaining_tidied  \n",
       "remaining_tidied                                                                        \n",
       "organic soil                                                              organic soil  \n",
       "                                                                                        \n",
       "dna                                                                                dna  \n",
       "mine tailings                                                            mine tailings  \n",
       "mineral soil                                                              mineral soil  \n",
       "...                                                                                ...  \n",
       "mud soil                                                                      mud soil  \n",
       "paddy soil                                                                  paddy soil  \n",
       "replicate 1 soil from baodongyu of taoyuan  replicate 1 soil from baodongyu of taoyuan  \n",
       "replicate 1 soil from gushi of taoyuan          replicate 1 soil from gushi of taoyuan  \n",
       "topsoil                                                                        topsoil  \n",
       "\n",
       "[65 rows x 2 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_by_needed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53bcffc5",
   "metadata": {},
   "source": [
    "## Extract the tidied strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "02a43f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_raw_list = list(sum_by_needed[\"remaining_tidied\"])\n",
    "ebs_raw_list.sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bd91eb2",
   "metadata": {},
   "source": [
    "## Submit those tidied strings to a search engine\n",
    "\n",
    "Specifically OLS search. This takes roughly one second per unique post-tidied submission\n",
    "\n",
    "_Turn logging back on to show status?_\n",
    "_Print the count and pre- and post- datestamps_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "08e39629",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    ebs_raw_list,\n",
    "    bad_chars=chars_to_whiteout,\n",
    "    cat_name=biosample_col_to_map,\n",
    "    ontoprefix=first_pass_ontologies,\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=my_string_dist_arg,\n",
    ")\n",
    "ebs_search_res.to_sql(\n",
    "    \"ols_search_results\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae3d29dc",
   "metadata": {},
   "source": [
    "## Filter out the best of the acceptable mappings\n",
    "From a string distance perspective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b695a649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>1</td>\n",
       "      <td>0.036</td>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000003</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>hasRelatedSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005774</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002229</td>\n",
       "      <td>arenosol</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002229</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       category            raw          query          name  string_dist_rank  \\\n",
       "45   env_medium       bacteria       bacteria      Bacteria                 1   \n",
       "181  env_medium  mine tailings  mine tailings  mine tailing                 1   \n",
       "301  env_medium           peat           peat          peat                 1   \n",
       "424  env_medium     sandy soil     sandy soil    sandy soil                 1   \n",
       "\n",
       "     string_dist         obo_id         label  search_rank ontology_prefix  \\\n",
       "45         0.000    NCBITaxon:2      Bacteria            1       NCBITAXON   \n",
       "181        0.036  ENVO:00000003  mine tailing            4            ENVO   \n",
       "301        0.000  ENVO:00005774     peat soil            1            ENVO   \n",
       "424        0.000  ENVO:00002229      arenosol            1            ENVO   \n",
       "\n",
       "                 scope     type                                           iri  \\\n",
       "45               label    label    http://purl.obolibrary.org/obo/NCBITaxon_2   \n",
       "181              label    label  http://purl.obolibrary.org/obo/ENVO_00000003   \n",
       "301  hasRelatedSynonym     None  http://purl.obolibrary.org/obo/ENVO_00005774   \n",
       "424            synonym  synonym  http://purl.obolibrary.org/obo/ENVO_00002229   \n",
       "\n",
       "    ontology_name  \n",
       "45      ncbitaxon  \n",
       "181          envo  \n",
       "301          envo  \n",
       "424          envo  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    ebs_search_res, max_string_dist=my_max_string_dist\n",
    ")\n",
    "my_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43aa5018",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Filter out the submissions with no acceptable matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "39e98bf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_medium</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>bacterium A9 13-13</td>\n",
       "      <td>1</td>\n",
       "      <td>0.541</td>\n",
       "      <td>NCBITaxon:1469932</td>\n",
       "      <td>bacterium A9 13-13</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1469932</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>Phage 13</td>\n",
       "      <td>2</td>\n",
       "      <td>0.622</td>\n",
       "      <td>NCBITaxon:12424</td>\n",
       "      <td>Phage 13</td>\n",
       "      <td>3</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_12424</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>Pasteurellaceae bacterium 13/13</td>\n",
       "      <td>3</td>\n",
       "      <td>0.667</td>\n",
       "      <td>NCBITaxon:1347399</td>\n",
       "      <td>Pasteurellaceae bacterium 13/13</td>\n",
       "      <td>2</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1347399</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>Bacteriophage 13</td>\n",
       "      <td>4</td>\n",
       "      <td>0.742</td>\n",
       "      <td>NCBITaxon:12424</td>\n",
       "      <td>Phage 13</td>\n",
       "      <td>3</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>equivalent name</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_12424</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>soil erosion</td>\n",
       "      <td>1</td>\n",
       "      <td>0.631</td>\n",
       "      <td>ENVO:01000706</td>\n",
       "      <td>soil erosion</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000706</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>luvisol</td>\n",
       "      <td>2</td>\n",
       "      <td>0.833</td>\n",
       "      <td>ENVO:00002248</td>\n",
       "      <td>luvisol</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002248</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>lixisol</td>\n",
       "      <td>3</td>\n",
       "      <td>0.833</td>\n",
       "      <td>ENVO:00002242</td>\n",
       "      <td>lixisol</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002242</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>stagnosol</td>\n",
       "      <td>4</td>\n",
       "      <td>0.856</td>\n",
       "      <td>ENVO:00002274</td>\n",
       "      <td>stagnosol</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002274</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>topsoil</td>\n",
       "      <td>mineral horizon</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000</td>\n",
       "      <td>ENVO:03600011</td>\n",
       "      <td>mineral horizon</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03600011</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>428 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       category      raw    query                             name  \\\n",
       "0    env_medium                                                      \n",
       "1    env_medium       13       13               bacterium A9 13-13   \n",
       "3    env_medium       13       13                         Phage 13   \n",
       "2    env_medium       13       13  Pasteurellaceae bacterium 13/13   \n",
       "4    env_medium       13       13                 Bacteriophage 13   \n",
       "..          ...      ...      ...                              ...   \n",
       "518  env_medium  topsoil  topsoil                     soil erosion   \n",
       "519  env_medium  topsoil  topsoil                          luvisol   \n",
       "520  env_medium  topsoil  topsoil                          lixisol   \n",
       "517  env_medium  topsoil  topsoil                        stagnosol   \n",
       "516  env_medium  topsoil  topsoil                  mineral horizon   \n",
       "\n",
       "     string_dist_rank  string_dist             obo_id  \\\n",
       "0                   1        0.000                      \n",
       "1                   1        0.541  NCBITaxon:1469932   \n",
       "3                   2        0.622    NCBITaxon:12424   \n",
       "2                   3        0.667  NCBITaxon:1347399   \n",
       "4                   4        0.742    NCBITaxon:12424   \n",
       "..                ...          ...                ...   \n",
       "518                 1        0.631      ENVO:01000706   \n",
       "519                 2        0.833      ENVO:00002248   \n",
       "520                 3        0.833      ENVO:00002242   \n",
       "517                 4        0.856      ENVO:00002274   \n",
       "516                 5        1.000      ENVO:03600011   \n",
       "\n",
       "                               label  search_rank ontology_prefix  \\\n",
       "0                                               1                   \n",
       "1                 bacterium A9 13-13            1       NCBITAXON   \n",
       "3                           Phage 13            3       NCBITAXON   \n",
       "2    Pasteurellaceae bacterium 13/13            2       NCBITAXON   \n",
       "4                           Phage 13            3       NCBITAXON   \n",
       "..                               ...          ...             ...   \n",
       "518                     soil erosion            3            ENVO   \n",
       "519                          luvisol            4            ENVO   \n",
       "520                          lixisol            5            ENVO   \n",
       "517                        stagnosol            2            ENVO   \n",
       "516                  mineral horizon            1            ENVO   \n",
       "\n",
       "               scope             type  \\\n",
       "0                NaN              NaN   \n",
       "1              label            label   \n",
       "3              label            label   \n",
       "2              label            label   \n",
       "4    hasExactSynonym  equivalent name   \n",
       "..               ...              ...   \n",
       "518            label            label   \n",
       "519            label            label   \n",
       "520            label            label   \n",
       "517            label            label   \n",
       "516            label            label   \n",
       "\n",
       "                                                  iri ontology_name  \n",
       "0                                                                    \n",
       "1    http://purl.obolibrary.org/obo/NCBITaxon_1469932     ncbitaxon  \n",
       "3      http://purl.obolibrary.org/obo/NCBITaxon_12424     ncbitaxon  \n",
       "2    http://purl.obolibrary.org/obo/NCBITaxon_1347399     ncbitaxon  \n",
       "4      http://purl.obolibrary.org/obo/NCBITaxon_12424     ncbitaxon  \n",
       "..                                                ...           ...  \n",
       "518      http://purl.obolibrary.org/obo/ENVO_01000706          envo  \n",
       "519      http://purl.obolibrary.org/obo/ENVO_00002248          envo  \n",
       "520      http://purl.obolibrary.org/obo/ENVO_00002242          envo  \n",
       "517      http://purl.obolibrary.org/obo/ENVO_00002274          envo  \n",
       "516      http://purl.obolibrary.org/obo/ENVO_03600011          envo  \n",
       "\n",
       "[428 rows x 14 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_acceptable_mappings = scoped_mapping.get_no_acceptable_mappings(\n",
    "    ebs_search_res, my_best_acceptable\n",
    ")\n",
    "\n",
    "no_acceptable_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70be6244",
   "metadata": {},
   "source": [
    "## Try searching the failures against all ontologies in OLS\n",
    "what would it be like to use a local OLS instance? (latency-wise)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ec662d5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['',\n",
       " '13',\n",
       " '15',\n",
       " '16s rrna gene sequence',\n",
       " '22',\n",
       " '24',\n",
       " '4',\n",
       " '6',\n",
       " 'alfisols',\n",
       " 'biomat',\n",
       " 'cryoconite',\n",
       " 'decomposed litter oe layer',\n",
       " 'dna',\n",
       " 'fermentation humic sustrate',\n",
       " 'forest soil biochar',\n",
       " 'genome',\n",
       " 'genomic bacterial dna of the ssu rrna gene v5 v7 hypervariable region',\n",
       " 'hypolith scappings from under rocks',\n",
       " 'juvenile from soil aluvial loam',\n",
       " 'juvenile from soil diluvial sand',\n",
       " 'juvenile from soil loess loam',\n",
       " 'juveniles meloidogyne hapla from soil diluvial sand',\n",
       " 'leaves',\n",
       " 'litters',\n",
       " 'mineral soil',\n",
       " 'mixed fungi',\n",
       " 'molecular ecology sample w1 3',\n",
       " 'molecular ecology sample w1 6',\n",
       " 'molecular ecology sample x1 3',\n",
       " 'molecular ecology sample x1 6',\n",
       " 'moss',\n",
       " 'mud soil',\n",
       " 'not applicable',\n",
       " 'organic soil',\n",
       " 'paddy soil',\n",
       " 'permafrost soil',\n",
       " 'picea litter',\n",
       " 'pinus litter',\n",
       " 'replicate 1 soil from baodongyu of taoyuan',\n",
       " 'replicate 1 soil from gushi of taoyuan',\n",
       " 'replicate 1 soil from jiaxing',\n",
       " 'replicate 1 soil from leizhou',\n",
       " 'replicate 1 soil from yingtan',\n",
       " 'replicate 2 soil from baodongyu of taoyuan',\n",
       " 'replicate 2 soil from gushi of taoyuan',\n",
       " 'replicate 2 soil from jiaxing',\n",
       " 'replicate 2 soil from leizhou',\n",
       " 'replicate 2 soil from yingtan',\n",
       " 'rhizosphere soil',\n",
       " 'rna',\n",
       " 'soil aluvial loam',\n",
       " 'soil and ice',\n",
       " 'soil collected during dry season',\n",
       " 'soil collected during rainy season',\n",
       " 'soil core',\n",
       " 'soil dna',\n",
       " 'soil sample',\n",
       " 'soil water',\n",
       " 'soybean soil',\n",
       " 'surface sediment water',\n",
       " 'topsoil']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "still_unmapped = list(set(list(no_acceptable_mappings[\"raw\"])))\n",
    "still_unmapped.sort()\n",
    "\n",
    "print(len(still_unmapped))\n",
    "\n",
    "still_unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "ac5c4f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "salvage_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    still_unmapped,\n",
    "    bad_chars=\"._-\",\n",
    "    cat_name=\"salvage\",\n",
    "    ontoprefix=\"\",\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=2,\n",
    ")\n",
    "\n",
    "salvage_search_res.to_sql(\n",
    "    \"salvage_search_res\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "453b8356",
   "metadata": {},
   "source": [
    "## We appear to be at a point of diminishing returns\n",
    "At the very least, will require some review"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "d44fe45b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>salvage</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C113429</td>\n",
       "      <td>Fifteen</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113429</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>salvage</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C172977</td>\n",
       "      <td>Fear Score 4</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172977</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>salvage</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C179016</td>\n",
       "      <td>COVID-19 Disease Severity 6</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179016</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>salvage</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:03000039</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000039</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>salvage</td>\n",
       "      <td>dna</td>\n",
       "      <td>dna</td>\n",
       "      <td>DNA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C449</td>\n",
       "      <td>DNA</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C449</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>salvage</td>\n",
       "      <td>genome</td>\n",
       "      <td>genome</td>\n",
       "      <td>Genome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C16629</td>\n",
       "      <td>Genome</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C16629</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>salvage</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>CO_366:0000273</td>\n",
       "      <td>leaves</td>\n",
       "      <td>1</td>\n",
       "      <td>CO_366</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://www.cropontology.org/rdf/CO_366:0000273</td>\n",
       "      <td>co_366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>salvage</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C48660</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C48660</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>salvage</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005740</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>salvage</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:00000134</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000134</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>770</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005801</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rna</td>\n",
       "      <td>rna</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>OMIT:0013221</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0013221</td>\n",
       "      <td>omit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>963</th>\n",
       "      <td>salvage</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>1</td>\n",
       "      <td>0.155</td>\n",
       "      <td>ENVO:01001192</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001192</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    category                     raw                   query  \\\n",
       "11   salvage                      15                      15   \n",
       "73   salvage                       4                       4   \n",
       "88   salvage                       6                       6   \n",
       "107  salvage              cryoconite              cryoconite   \n",
       "125  salvage                     dna                     dna   \n",
       "162  salvage                  genome                  genome   \n",
       "291  salvage                  leaves                  leaves   \n",
       "520  salvage          not applicable          not applicable   \n",
       "576  salvage              paddy soil              paddy soil   \n",
       "588  salvage         permafrost soil         permafrost soil   \n",
       "770  salvage        rhizosphere soil        rhizosphere soil   \n",
       "781  salvage                     rna                     rna   \n",
       "963  salvage  surface sediment water  surface sediment water   \n",
       "\n",
       "                 name  string_dist_rank  string_dist          obo_id  \\\n",
       "11                 15                 1        0.000    NCIT:C113429   \n",
       "73                  4                 1        0.000    NCIT:C172977   \n",
       "88                  6                 1        0.000    NCIT:C179016   \n",
       "107   cryoconite hole                 1        0.198   ENVO:03000039   \n",
       "125               DNA                 1        0.000       NCIT:C449   \n",
       "162            Genome                 1        0.000     NCIT:C16629   \n",
       "291            leaves                 1        0.000  CO_366:0000273   \n",
       "520    Not Applicable                 1        0.000     NCIT:C48660   \n",
       "576   rice paddy soil                 1        0.198   ENVO:00005740   \n",
       "588        permafrost                 1        0.198   ENVO:00000134   \n",
       "770       rhizosphere                 1        0.184   ENVO:00005801   \n",
       "781               RNA                 1        0.000    OMIT:0013221   \n",
       "963  sediment surface                 1        0.155   ENVO:01001192   \n",
       "\n",
       "                           label  search_rank ontology_prefix  \\\n",
       "11                       Fifteen            5            NCIT   \n",
       "73                  Fear Score 4            5            NCIT   \n",
       "88   COVID-19 Disease Severity 6            3            NCIT   \n",
       "107              cryoconite hole            2            ENVO   \n",
       "125                          DNA            1            NCIT   \n",
       "162                       Genome            1            NCIT   \n",
       "291                       leaves            1          CO_366   \n",
       "520               Not Applicable            1            NCIT   \n",
       "576             paddy field soil            1            ENVO   \n",
       "588                   permafrost            1            ENVO   \n",
       "770                  rhizosphere            1            ENVO   \n",
       "781                          RNA            1            OMIT   \n",
       "963             sediment surface            2            ENVO   \n",
       "\n",
       "                   scope        type  \\\n",
       "11               synonym     synonym   \n",
       "73       hasExactSynonym        None   \n",
       "88       hasExactSynonym        None   \n",
       "107                label       label   \n",
       "125                label       label   \n",
       "162                label       label   \n",
       "291                label       label   \n",
       "520                label       label   \n",
       "576  has_related_synonym  annotation   \n",
       "588                label       label   \n",
       "770                label       label   \n",
       "781                label       label   \n",
       "963                label       label   \n",
       "\n",
       "                                                iri ontology_name  \n",
       "11      http://purl.obolibrary.org/obo/NCIT_C113429          ncit  \n",
       "73      http://purl.obolibrary.org/obo/NCIT_C172977          ncit  \n",
       "88      http://purl.obolibrary.org/obo/NCIT_C179016          ncit  \n",
       "107    http://purl.obolibrary.org/obo/ENVO_03000039          envo  \n",
       "125        http://purl.obolibrary.org/obo/NCIT_C449          ncit  \n",
       "162      http://purl.obolibrary.org/obo/NCIT_C16629          ncit  \n",
       "291  http://www.cropontology.org/rdf/CO_366:0000273        co_366  \n",
       "520      http://purl.obolibrary.org/obo/NCIT_C48660          ncit  \n",
       "576    http://purl.obolibrary.org/obo/ENVO_00005740          envo  \n",
       "588    http://purl.obolibrary.org/obo/ENVO_00000134          envo  \n",
       "770    http://purl.obolibrary.org/obo/ENVO_00005801          envo  \n",
       "781     http://purl.obolibrary.org/obo/OMIT_0013221          omit  \n",
       "963    http://purl.obolibrary.org/obo/ENVO_01001192          envo  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "salvage_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    salvage_search_res, max_string_dist=salvage_max_dist\n",
    ")\n",
    "salvage_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6bd9e5e",
   "metadata": {},
   "source": [
    "## Save the mappings for review!\n",
    "\n",
    "_Note that the merge on ID or label mappings are not included here_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "a2b88c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_best_acceptable[include_col] = first_pass_include_val\n",
    "salvage_best_acceptable[include_col] = salvage_include_val\n",
    "best_and_salvage = pd.concat([my_best_acceptable, salvage_best_acceptable])\n",
    "\n",
    "best_and_salvage[\"ols_curation_notes\"] = \"\"\n",
    "\n",
    "best_and_salvage.to_csv(ols_review_file, sep=ols_review_seperator, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "131b2343",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_and_salvage.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "4f4a3988",
   "metadata": {},
   "outputs": [
    {
     "ename": "SystemExit",
     "evalue": "Don't skip reviewing and saving!",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m Don't skip reviewing and saving!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/scoped-mapping/sm_venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3449: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "raise SystemExit(\"Don't skip reviewing and saving!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a331a6",
   "metadata": {},
   "source": [
    "## Do some review in a spreadsheet application!\n",
    "Specifically, columns `include` and `ols_curation_notes`\n",
    "\n",
    "Then Save\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be6da72c",
   "metadata": {},
   "source": [
    "## Now read the reviewed spreadsheet back in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c94cf3b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "curated = pd.read_csv(ols_review_file, sep=ols_review_seperator)\n",
    "curated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49389e4e",
   "metadata": {},
   "source": [
    "## OR copy and paste"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "bd84b014",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>include</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>salvage</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>leaves</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>CO_366:0000273</td>\n",
       "      <td>leaves</td>\n",
       "      <td>1</td>\n",
       "      <td>CO_366</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://www.cropontology.org/rdf/CO_366:0000273</td>\n",
       "      <td>co_366</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>salvage</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:03000039</td>\n",
       "      <td>cryoconite hole</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000039</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>salvage</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005740</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>salvage</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost soil</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:00000134</td>\n",
       "      <td>permafrost</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000134</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere soil</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005801</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>salvage</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>surface sediment water</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>1</td>\n",
       "      <td>0.155</td>\n",
       "      <td>ENVO:01001192</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001192</td>\n",
       "      <td>envo</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>salvage</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C113429</td>\n",
       "      <td>Fifteen</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113429</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>salvage</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C172977</td>\n",
       "      <td>Fear Score 4</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172977</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>salvage</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C179016</td>\n",
       "      <td>COVID-19 Disease Severity 6</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179016</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>salvage</td>\n",
       "      <td>dna</td>\n",
       "      <td>dna</td>\n",
       "      <td>DNA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C449</td>\n",
       "      <td>DNA</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C449</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>salvage</td>\n",
       "      <td>genome</td>\n",
       "      <td>genome</td>\n",
       "      <td>Genome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C16629</td>\n",
       "      <td>Genome</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C16629</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>salvage</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C48660</td>\n",
       "      <td>Not Applicable</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C48660</td>\n",
       "      <td>ncit</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>salvage</td>\n",
       "      <td>rna</td>\n",
       "      <td>rna</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>OMIT:0013221</td>\n",
       "      <td>RNA</td>\n",
       "      <td>1</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0013221</td>\n",
       "      <td>omit</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>1</td>\n",
       "      <td>0.036</td>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000003</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>hasRelatedSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005774</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002229</td>\n",
       "      <td>arenosol</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002229</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      category                     raw                   query  \\\n",
       "0      salvage                  leaves                  leaves   \n",
       "1      salvage              cryoconite              cryoconite   \n",
       "2      salvage              paddy soil              paddy soil   \n",
       "3      salvage         permafrost soil         permafrost soil   \n",
       "4      salvage        rhizosphere soil        rhizosphere soil   \n",
       "5      salvage  surface sediment water  surface sediment water   \n",
       "6      salvage                      15                      15   \n",
       "7      salvage                       4                       4   \n",
       "8      salvage                       6                       6   \n",
       "9      salvage                     dna                     dna   \n",
       "10     salvage                  genome                  genome   \n",
       "11     salvage          not applicable          not applicable   \n",
       "12     salvage                     rna                     rna   \n",
       "13  env_medium           mine tailings           mine tailings   \n",
       "14  env_medium                    peat                    peat   \n",
       "15  env_medium              sandy soil              sandy soil   \n",
       "16  env_medium                bacteria                bacteria   \n",
       "\n",
       "                name  string_dist_rank  string_dist          obo_id  \\\n",
       "0             leaves                 1        0.000  CO_366:0000273   \n",
       "1    cryoconite hole                 1        0.198   ENVO:03000039   \n",
       "2    rice paddy soil                 1        0.198   ENVO:00005740   \n",
       "3         permafrost                 1        0.198   ENVO:00000134   \n",
       "4        rhizosphere                 1        0.184   ENVO:00005801   \n",
       "5   sediment surface                 1        0.155   ENVO:01001192   \n",
       "6                 15                 1        0.000    NCIT:C113429   \n",
       "7                  4                 1        0.000    NCIT:C172977   \n",
       "8                  6                 1        0.000    NCIT:C179016   \n",
       "9                DNA                 1        0.000       NCIT:C449   \n",
       "10            Genome                 1        0.000     NCIT:C16629   \n",
       "11    Not Applicable                 1        0.000     NCIT:C48660   \n",
       "12               RNA                 1        0.000    OMIT:0013221   \n",
       "13      mine tailing                 1        0.036   ENVO:00000003   \n",
       "14              peat                 1        0.000   ENVO:00005774   \n",
       "15        sandy soil                 1        0.000   ENVO:00002229   \n",
       "16          Bacteria                 1        0.000     NCBITaxon:2   \n",
       "\n",
       "                          label  search_rank ontology_prefix  \\\n",
       "0                        leaves            1          CO_366   \n",
       "1               cryoconite hole            2            ENVO   \n",
       "2              paddy field soil            1            ENVO   \n",
       "3                    permafrost            1            ENVO   \n",
       "4                   rhizosphere            1            ENVO   \n",
       "5              sediment surface            2            ENVO   \n",
       "6                       Fifteen            5            NCIT   \n",
       "7                  Fear Score 4            5            NCIT   \n",
       "8   COVID-19 Disease Severity 6            3            NCIT   \n",
       "9                           DNA            1            NCIT   \n",
       "10                       Genome            1            NCIT   \n",
       "11               Not Applicable            1            NCIT   \n",
       "12                          RNA            1            OMIT   \n",
       "13                 mine tailing            4            ENVO   \n",
       "14                    peat soil            1            ENVO   \n",
       "15                     arenosol            1            ENVO   \n",
       "16                     Bacteria            1       NCBITAXON   \n",
       "\n",
       "                  scope        type  \\\n",
       "0                 label       label   \n",
       "1                 label       label   \n",
       "2   has_related_synonym  annotation   \n",
       "3                 label       label   \n",
       "4                 label       label   \n",
       "5                 label       label   \n",
       "6               synonym     synonym   \n",
       "7       hasExactSynonym         NaN   \n",
       "8       hasExactSynonym         NaN   \n",
       "9                 label       label   \n",
       "10                label       label   \n",
       "11                label       label   \n",
       "12                label       label   \n",
       "13                label       label   \n",
       "14    hasRelatedSynonym         NaN   \n",
       "15              synonym     synonym   \n",
       "16                label       label   \n",
       "\n",
       "                                               iri ontology_name  include  \n",
       "0   http://www.cropontology.org/rdf/CO_366:0000273        co_366    False  \n",
       "1     http://purl.obolibrary.org/obo/ENVO_03000039          envo    False  \n",
       "2     http://purl.obolibrary.org/obo/ENVO_00005740          envo     True  \n",
       "3     http://purl.obolibrary.org/obo/ENVO_00000134          envo    False  \n",
       "4     http://purl.obolibrary.org/obo/ENVO_00005801          envo    False  \n",
       "5     http://purl.obolibrary.org/obo/ENVO_01001192          envo    False  \n",
       "6      http://purl.obolibrary.org/obo/NCIT_C113429          ncit    False  \n",
       "7      http://purl.obolibrary.org/obo/NCIT_C172977          ncit    False  \n",
       "8      http://purl.obolibrary.org/obo/NCIT_C179016          ncit    False  \n",
       "9         http://purl.obolibrary.org/obo/NCIT_C449          ncit    False  \n",
       "10      http://purl.obolibrary.org/obo/NCIT_C16629          ncit    False  \n",
       "11      http://purl.obolibrary.org/obo/NCIT_C48660          ncit    False  \n",
       "12     http://purl.obolibrary.org/obo/OMIT_0013221          omit    False  \n",
       "13    http://purl.obolibrary.org/obo/ENVO_00000003          envo     True  \n",
       "14    http://purl.obolibrary.org/obo/ENVO_00005774          envo     True  \n",
       "15    http://purl.obolibrary.org/obo/ENVO_00002229          envo     True  \n",
       "16      http://purl.obolibrary.org/obo/NCBITaxon_2     ncbitaxon     True  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "curated = pd.read_clipboard()\n",
    "curated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ea200e4",
   "metadata": {},
   "source": [
    "## Filter on `include`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1473206f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ols_category</th>\n",
       "      <th>ols_raw</th>\n",
       "      <th>ols_query</th>\n",
       "      <th>ols_name</th>\n",
       "      <th>ols_string_dist_rank</th>\n",
       "      <th>ols_string_dist</th>\n",
       "      <th>ols_obo_id</th>\n",
       "      <th>ols_label</th>\n",
       "      <th>ols_search_rank</th>\n",
       "      <th>ols_ontology_prefix</th>\n",
       "      <th>ols_scope</th>\n",
       "      <th>ols_type</th>\n",
       "      <th>ols_iri</th>\n",
       "      <th>ols_ontology_name</th>\n",
       "      <th>ols_include</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>salvage</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>rice paddy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.198</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005740</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>1</td>\n",
       "      <td>0.036</td>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000003</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>peat</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>hasRelatedSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005774</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002229</td>\n",
       "      <td>arenosol</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002229</td>\n",
       "      <td>envo</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ols_category        ols_raw      ols_query         ols_name  \\\n",
       "2       salvage     paddy soil     paddy soil  rice paddy soil   \n",
       "13   env_medium  mine tailings  mine tailings     mine tailing   \n",
       "14   env_medium           peat           peat             peat   \n",
       "15   env_medium     sandy soil     sandy soil       sandy soil   \n",
       "16   env_medium       bacteria       bacteria         Bacteria   \n",
       "\n",
       "    ols_string_dist_rank  ols_string_dist     ols_obo_id         ols_label  \\\n",
       "2                      1            0.198  ENVO:00005740  paddy field soil   \n",
       "13                     1            0.036  ENVO:00000003      mine tailing   \n",
       "14                     1            0.000  ENVO:00005774         peat soil   \n",
       "15                     1            0.000  ENVO:00002229          arenosol   \n",
       "16                     1            0.000    NCBITaxon:2          Bacteria   \n",
       "\n",
       "    ols_search_rank ols_ontology_prefix            ols_scope    ols_type  \\\n",
       "2                 1                ENVO  has_related_synonym  annotation   \n",
       "13                4                ENVO                label       label   \n",
       "14                1                ENVO    hasRelatedSynonym         NaN   \n",
       "15                1                ENVO              synonym     synonym   \n",
       "16                1           NCBITAXON                label       label   \n",
       "\n",
       "                                         ols_iri ols_ontology_name  \\\n",
       "2   http://purl.obolibrary.org/obo/ENVO_00005740              envo   \n",
       "13  http://purl.obolibrary.org/obo/ENVO_00000003              envo   \n",
       "14  http://purl.obolibrary.org/obo/ENVO_00005774              envo   \n",
       "15  http://purl.obolibrary.org/obo/ENVO_00002229              envo   \n",
       "16    http://purl.obolibrary.org/obo/NCBITaxon_2         ncbitaxon   \n",
       "\n",
       "    ols_include  \n",
       "2          True  \n",
       "13         True  \n",
       "14         True  \n",
       "15         True  \n",
       "16         True  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flag = curated[\"include\"]\n",
    "curated = curated.loc[flag]\n",
    "\n",
    "curated.columns = \"ols_\" + curated.columns\n",
    "curated.to_sql(\"curated\", biosample_cnx, if_exists=\"replace\", index=False)\n",
    "curated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "ac6219e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_search_merged = mapping_candidates.merge(\n",
    "    curated, how=\"outer\", left_on=\"remaining_tidied\", right_on=\"ols_raw\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "d2c242b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = (\n",
    "    ~merge_search_merged[\"id_or_lab_ok\"]\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].eq(\"\")\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].isna()\n",
    ")\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_obo_id\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_id\"] = replacement\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_label\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_lab\"] = replacement\n",
    "\n",
    "replacement = (\n",
    "    merge_search_merged.loc[flag, \"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + merge_search_merged.loc[flag, \"consensus_id\"]\n",
    "    + \"]\"\n",
    ")\n",
    "\n",
    "merge_search_merged.loc[flag, \"assembled_consensus\"] = replacement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "b333c111",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_search_merged.to_sql(\n",
    "    \"merge_search_merged\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "5e772701",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>assembled_consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>12400</td>\n",
       "      <td>soil [ENVO:00001998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>439</td>\n",
       "      <td>soil [ENVO:00001998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>238</td>\n",
       "      <td>rhizosphere [ENVO:00005801]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>224</td>\n",
       "      <td>peat soil [ENVO:00005774]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>214</td>\n",
       "      <td>soil [ENVO:00001998]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>replicate 1 soil from Yingtan</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>replicate 1 soil from Leizhou</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>replicate 1 soil from Jiaxing</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>replicate 1 soil from Gushi of Taoyuan</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 env_medium  count  \\\n",
       "0                                      soil  12400   \n",
       "1                                      Soil    439   \n",
       "3                               rhizosphere    238   \n",
       "5                                 peat soil    224   \n",
       "2                                 ENVO:soil    214   \n",
       "..                                      ...    ...   \n",
       "82            replicate 1 soil from Yingtan      1   \n",
       "83            replicate 1 soil from Leizhou      1   \n",
       "84            replicate 1 soil from Jiaxing      1   \n",
       "85   replicate 1 soil from Gushi of Taoyuan      1   \n",
       "105                                      13      1   \n",
       "\n",
       "             assembled_consensus  \n",
       "0           soil [ENVO:00001998]  \n",
       "1           soil [ENVO:00001998]  \n",
       "3    rhizosphere [ENVO:00005801]  \n",
       "5      peat soil [ENVO:00005774]  \n",
       "2           soil [ENVO:00001998]  \n",
       "..                           ...  \n",
       "82                           NaN  \n",
       "83                           NaN  \n",
       "84                           NaN  \n",
       "85                           NaN  \n",
       "105                          NaN  \n",
       "\n",
       "[106 rows x 3 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary = merge_search_merged[[biosample_col_to_map, \"count\", \"assembled_consensus\"]]\n",
    "summary = summary.sort_values(\"count\", ascending=False)\n",
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "38acac10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_category</th>\n",
       "      <th>subject_label</th>\n",
       "      <th>predicate_id</th>\n",
       "      <th>object_id</th>\n",
       "      <th>object_label</th>\n",
       "      <th>match_type</th>\n",
       "      <th>creator_id</th>\n",
       "      <th>mapping_date</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:ENVO%3Asoil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Rhizosphere</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:peat%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Peat soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Peat%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>sand</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000017</td>\n",
       "      <td>sand</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:sand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Leafy wood soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005783</td>\n",
       "      <td>leafy wood soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Leafy%20wood%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>surface soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:02000059</td>\n",
       "      <td>surface soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:surface%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>bulk soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005802</td>\n",
       "      <td>bulk soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:bulk%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Bulk soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005802</td>\n",
       "      <td>bulk soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Bulk%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>ENVO:00002259</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002259</td>\n",
       "      <td>agricultural soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:ENVO%3A00002259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:ENVO%3A00005740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>ENVO:00005783</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005783</td>\n",
       "      <td>leafy wood soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:ENVO%3A00005783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>ENVO:00002245</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002245</td>\n",
       "      <td>gypsisol</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:ENVO%3A00002245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>mine tailings</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>mine tailing</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:mine%20tailings</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>sediment</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>agricultural soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002259</td>\n",
       "      <td>agricultural soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:agricultural%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>histosol</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002243</td>\n",
       "      <td>histosol</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:histosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>upland soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005786</td>\n",
       "      <td>upland soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:upland%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>peat</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:peat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Spruce forest soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005784</td>\n",
       "      <td>spruce forest soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Spruce%20forest%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Mountain forest soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005769</td>\n",
       "      <td>mountain forest soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Mountain%20forest%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>humus</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000000</td>\n",
       "      <td>humus</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:humus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>sandy soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002229</td>\n",
       "      <td>arenosol</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:sandy%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>rock</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00001995</td>\n",
       "      <td>rock</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:rock</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>Grassland soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005750</td>\n",
       "      <td>grassland soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:Grassland%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>ENVO:00005740 (paddy field soil)</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:ENVO%3A00005740%20%28paddy%20fie...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:arable%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>ferralsol</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002246</td>\n",
       "      <td>ferralsol</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:ferralsol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>gravel</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000018</td>\n",
       "      <td>gravel</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:gravel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>frozen_soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01001526</td>\n",
       "      <td>frozen soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:frozen_soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>andosol [ENVO:00002232]</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002232</td>\n",
       "      <td>andosol</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:andosol%20%5BENVO%3A00002232%5D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:oil%20contaminated%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>forest soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002261</td>\n",
       "      <td>forest soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:forest%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>water</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>CHEBI:15377</td>\n",
       "      <td>water</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>paddy soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:paddy%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>loam</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002258</td>\n",
       "      <td>loam</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:loam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005780</td>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:greenhouse%20soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>anthrosol</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00002230</td>\n",
       "      <td>anthrosol</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T11:05:31</td>\n",
       "      <td>biosample_ebs:anthrosol</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   subject_category                     subject_label       predicate_id  \\\n",
       "0        env_medium                              soil  skos:relatedMatch   \n",
       "1        env_medium                              Soil  skos:relatedMatch   \n",
       "2        env_medium                         ENVO:soil  skos:relatedMatch   \n",
       "3        env_medium                       rhizosphere  skos:relatedMatch   \n",
       "4        env_medium                       Rhizosphere  skos:relatedMatch   \n",
       "5        env_medium                         peat soil  skos:relatedMatch   \n",
       "6        env_medium                         Peat soil  skos:relatedMatch   \n",
       "7        env_medium                              sand  skos:relatedMatch   \n",
       "9        env_medium                   Leafy wood soil  skos:relatedMatch   \n",
       "10       env_medium                      surface soil  skos:relatedMatch   \n",
       "11       env_medium                         bulk soil  skos:relatedMatch   \n",
       "12       env_medium                         Bulk soil  skos:relatedMatch   \n",
       "14       env_medium                     ENVO:00002259  skos:relatedMatch   \n",
       "15       env_medium                     ENVO:00005740  skos:relatedMatch   \n",
       "16       env_medium                     ENVO:00005783  skos:relatedMatch   \n",
       "17       env_medium                     ENVO:00002245  skos:relatedMatch   \n",
       "19       env_medium                     mine tailings  skos:relatedMatch   \n",
       "20       env_medium                          Sediment  skos:relatedMatch   \n",
       "21       env_medium                          sediment  skos:relatedMatch   \n",
       "22       env_medium                 agricultural soil  skos:relatedMatch   \n",
       "24       env_medium                          histosol  skos:relatedMatch   \n",
       "25       env_medium                       upland soil  skos:relatedMatch   \n",
       "28       env_medium                              peat  skos:relatedMatch   \n",
       "29       env_medium                Spruce forest soil  skos:relatedMatch   \n",
       "30       env_medium              Mountain forest soil  skos:relatedMatch   \n",
       "31       env_medium                          Bacteria  skos:relatedMatch   \n",
       "32       env_medium                          bacteria  skos:relatedMatch   \n",
       "37       env_medium                             humus  skos:relatedMatch   \n",
       "39       env_medium                        sandy soil  skos:relatedMatch   \n",
       "42       env_medium                              rock  skos:relatedMatch   \n",
       "46       env_medium                    Grassland soil  skos:relatedMatch   \n",
       "47       env_medium  ENVO:00005740 (paddy field soil)  skos:relatedMatch   \n",
       "48       env_medium                       arable soil  skos:relatedMatch   \n",
       "52       env_medium                         ferralsol  skos:relatedMatch   \n",
       "53       env_medium                            gravel  skos:relatedMatch   \n",
       "54       env_medium                       frozen_soil  skos:relatedMatch   \n",
       "56       env_medium           andosol [ENVO:00002232]  skos:relatedMatch   \n",
       "60       env_medium             oil contaminated soil  skos:relatedMatch   \n",
       "67       env_medium                       forest soil  skos:relatedMatch   \n",
       "69       env_medium                             water  skos:relatedMatch   \n",
       "87       env_medium                        paddy soil  skos:relatedMatch   \n",
       "90       env_medium                              loam  skos:relatedMatch   \n",
       "91       env_medium                   greenhouse soil  skos:relatedMatch   \n",
       "92       env_medium                         anthrosol  skos:relatedMatch   \n",
       "\n",
       "        object_id           object_label match_type  \\\n",
       "0   ENVO:00001998                   soil    Lexical   \n",
       "1   ENVO:00001998                   soil    Lexical   \n",
       "2   ENVO:00001998                   soil    Lexical   \n",
       "3   ENVO:00005801            rhizosphere    Lexical   \n",
       "4   ENVO:00005801            rhizosphere    Lexical   \n",
       "5   ENVO:00005774              peat soil    Lexical   \n",
       "6   ENVO:00005774              peat soil    Lexical   \n",
       "7   ENVO:01000017                   sand    Lexical   \n",
       "9   ENVO:00005783        leafy wood soil    Lexical   \n",
       "10  ENVO:02000059           surface soil    Lexical   \n",
       "11  ENVO:00005802              bulk soil    Lexical   \n",
       "12  ENVO:00005802              bulk soil    Lexical   \n",
       "14  ENVO:00002259      agricultural soil    Lexical   \n",
       "15  ENVO:00005740       paddy field soil    Lexical   \n",
       "16  ENVO:00005783        leafy wood soil    Lexical   \n",
       "17  ENVO:00002245               gypsisol    Lexical   \n",
       "19  ENVO:00000003           mine tailing    Lexical   \n",
       "20  ENVO:00002007               sediment    Lexical   \n",
       "21  ENVO:00002007               sediment    Lexical   \n",
       "22  ENVO:00002259      agricultural soil    Lexical   \n",
       "24  ENVO:00002243               histosol    Lexical   \n",
       "25  ENVO:00005786            upland soil    Lexical   \n",
       "28  ENVO:00005774              peat soil    Lexical   \n",
       "29  ENVO:00005784     spruce forest soil    Lexical   \n",
       "30  ENVO:00005769   mountain forest soil    Lexical   \n",
       "31    NCBITaxon:2               Bacteria    Lexical   \n",
       "32    NCBITaxon:2               Bacteria    Lexical   \n",
       "37  ENVO:01000000                  humus    Lexical   \n",
       "39  ENVO:00002229               arenosol    Lexical   \n",
       "42  ENVO:00001995                   rock    Lexical   \n",
       "46  ENVO:00005750         grassland soil    Lexical   \n",
       "47  ENVO:00005740       paddy field soil    Lexical   \n",
       "48  ENVO:00005742            arable soil    Lexical   \n",
       "52  ENVO:00002246              ferralsol    Lexical   \n",
       "53  ENVO:01000018                 gravel    Lexical   \n",
       "54  ENVO:01001526            frozen soil    Lexical   \n",
       "56  ENVO:00002232                andosol    Lexical   \n",
       "60  ENVO:00002875  oil contaminated soil    Lexical   \n",
       "67  ENVO:00002261            forest soil    Lexical   \n",
       "69    CHEBI:15377                  water    Lexical   \n",
       "87  ENVO:00005740       paddy field soil    Lexical   \n",
       "90  ENVO:00002258                   loam    Lexical   \n",
       "91  ENVO:00005780        greenhouse soil    Lexical   \n",
       "92  ENVO:00002230              anthrosol    Lexical   \n",
       "\n",
       "                                    creator_id         mapping_date  \\\n",
       "0   https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "1   https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "2   https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "3   https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "4   https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "5   https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "6   https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "7   https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "9   https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "10  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "11  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "12  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "14  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "15  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "16  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "17  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "19  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "20  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "21  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "22  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "24  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "25  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "28  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "29  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "30  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "31  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "32  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "37  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "39  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "42  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "46  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "47  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "48  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "52  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "53  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "54  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "56  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "60  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "67  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "69  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "87  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "90  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "91  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "92  https://github.com/turbomam/scoped-mapping  2021-06-11T11:05:31   \n",
       "\n",
       "                                           subject_id  \n",
       "0                                  biosample_ebs:soil  \n",
       "1                                  biosample_ebs:Soil  \n",
       "2                           biosample_ebs:ENVO%3Asoil  \n",
       "3                           biosample_ebs:rhizosphere  \n",
       "4                           biosample_ebs:Rhizosphere  \n",
       "5                           biosample_ebs:peat%20soil  \n",
       "6                           biosample_ebs:Peat%20soil  \n",
       "7                                  biosample_ebs:sand  \n",
       "9                   biosample_ebs:Leafy%20wood%20soil  \n",
       "10                       biosample_ebs:surface%20soil  \n",
       "11                          biosample_ebs:bulk%20soil  \n",
       "12                          biosample_ebs:Bulk%20soil  \n",
       "14                      biosample_ebs:ENVO%3A00002259  \n",
       "15                      biosample_ebs:ENVO%3A00005740  \n",
       "16                      biosample_ebs:ENVO%3A00005783  \n",
       "17                      biosample_ebs:ENVO%3A00002245  \n",
       "19                      biosample_ebs:mine%20tailings  \n",
       "20                             biosample_ebs:Sediment  \n",
       "21                             biosample_ebs:sediment  \n",
       "22                  biosample_ebs:agricultural%20soil  \n",
       "24                             biosample_ebs:histosol  \n",
       "25                        biosample_ebs:upland%20soil  \n",
       "28                                 biosample_ebs:peat  \n",
       "29               biosample_ebs:Spruce%20forest%20soil  \n",
       "30             biosample_ebs:Mountain%20forest%20soil  \n",
       "31                             biosample_ebs:Bacteria  \n",
       "32                             biosample_ebs:bacteria  \n",
       "37                                biosample_ebs:humus  \n",
       "39                         biosample_ebs:sandy%20soil  \n",
       "42                                 biosample_ebs:rock  \n",
       "46                     biosample_ebs:Grassland%20soil  \n",
       "47  biosample_ebs:ENVO%3A00005740%20%28paddy%20fie...  \n",
       "48                        biosample_ebs:arable%20soil  \n",
       "52                            biosample_ebs:ferralsol  \n",
       "53                               biosample_ebs:gravel  \n",
       "54                          biosample_ebs:frozen_soil  \n",
       "56      biosample_ebs:andosol%20%5BENVO%3A00002232%5D  \n",
       "60            biosample_ebs:oil%20contaminated%20soil  \n",
       "67                        biosample_ebs:forest%20soil  \n",
       "69                                biosample_ebs:water  \n",
       "87                         biosample_ebs:paddy%20soil  \n",
       "90                                 biosample_ebs:loam  \n",
       "91                    biosample_ebs:greenhouse%20soil  \n",
       "92                            biosample_ebs:anthrosol  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frame_for_sssom = merge_search_merged\n",
    "\n",
    "na_flag = frame_for_sssom[\"splitted\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "na_flag = frame_for_sssom[\"consensus_id\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "raw_queries = list(frame_for_sssom[\"splitted\"])\n",
    "urlencodeds = list(map(urllib.parse.quote, raw_queries))\n",
    "urlencodeds = [sssom_subject_prefix + \":\" + item for item in urlencodeds]\n",
    "\n",
    "results_rows = len(frame_for_sssom.index)\n",
    "iso8601_stamp = datetime.now().replace(microsecond=0).isoformat()\n",
    "\n",
    "sssom_frame = {\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"subject_label\": frame_for_sssom[\"splitted\"],\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"object_id\": frame_for_sssom[\"consensus_id\"],\n",
    "    \"object_label\": frame_for_sssom[\"consensus_lab\"],\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    \"mapping_date\": [iso8601_stamp] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "    \"subject_id\": urlencodeds,\n",
    "}\n",
    "\n",
    "sssom_frame = pd.DataFrame(sssom_frame)\n",
    "\n",
    "sssom_frame.to_csv(sssom_file, sep=ols_review_seperator, index=False)\n",
    "\n",
    "sssom_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ad82349",
   "metadata": {},
   "source": [
    "# From here down requires refactoring and generalization \n",
    "\n",
    "That may even be true for some of the blocks above this!\n",
    "\n",
    "## Join together pipe-delimited env_broad_scale mappings \n",
    "_Any other constraints? If part_count is > 1, do we need to check for empty or null env_broad_scale, repeated or splitted?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "32ad2b25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    106\n",
       "Name: part_count, dtype: int64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_part_flag = merge_search_merged[\"part_count\"] > 1\n",
    "multi_part_flag.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "d686f6ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [env_medium, count, repeated, splitted, part_count, seq, consensus_id, consensus_lab]\n",
       "Index: []"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_part_frame = merge_search_merged.loc[multi_part_flag]\n",
    "\n",
    "multi_part_frame = multi_part_frame[\n",
    "    [\n",
    "        biosample_col_to_map,\n",
    "        \"count\",\n",
    "        \"repeated\",\n",
    "        \"splitted\",\n",
    "        \"part_count\",\n",
    "        \"seq\",\n",
    "        \"consensus_id\",\n",
    "        \"consensus_lab\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "multi_part_frame = multi_part_frame.sort_values(\n",
    "    [biosample_col_to_map, \"seq\"], ascending=(True, True)\n",
    ")\n",
    "\n",
    "multi_part_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "f147c6b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: part_count, dtype: int64)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "part_count = multi_part_frame.groupby([biosample_col_to_map], sort=False)[\n",
    "    \"part_count\"\n",
    "].max()\n",
    "part_count.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "5db86116",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: seq, dtype: float64)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_max = multi_part_frame.groupby([biosample_col_to_map], sort=False)[\"seq\"].max()\n",
    "seq_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "c4ccf78a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [env_medium, count, repeated, splitted, part_count, seq, consensus_id, consensus_lab]\n",
       "Index: []"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parts_check = pd.DataFrame(dict(part_count=part_count, seq_max=seq_max)).reset_index()\n",
    "\n",
    "all_parts_flag = parts_check[\"seq_max\"] == parts_check[\"part_count\"]\n",
    "ebs_with_all_parts = parts_check.loc[all_parts_flag, biosample_col_to_map]\n",
    "ebs_with_all_parts\n",
    "\n",
    "all_parts_flag = multi_part_frame[biosample_col_to_map].isin(list(ebs_with_all_parts))\n",
    "all_parts_frame = multi_part_frame.loc[all_parts_flag]\n",
    "all_parts_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f0cbd251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_pipesep_ebs = list(set(list(all_parts_frame[biosample_col_to_map])))\n",
    "unique_pipesep_ebs.sort()\n",
    "repipe_dict_list = []\n",
    "for one_pipesep in unique_pipesep_ebs:\n",
    "    flag = all_parts_frame[biosample_col_to_map] == one_pipesep\n",
    "    temp_frame = all_parts_frame.loc[flag]\n",
    "    ids_list = list(temp_frame[\"consensus_id\"])\n",
    "    lab_list = list(temp_frame[\"consensus_lab\"])\n",
    "    s = \"|\"\n",
    "    ids_repipe = s.join(ids_list)\n",
    "    lab_repipe = s.join(lab_list)\n",
    "\n",
    "    repipe_dict = {\n",
    "        biosample_col_to_map: one_pipesep,\n",
    "        \"consensus_id\": ids_repipe,\n",
    "        \"consensus_lab\": lab_repipe,\n",
    "    }\n",
    "    repipe_dict_list.append(repipe_dict)\n",
    "repipe_frame = pd.DataFrame(repipe_dict_list)\n",
    "repipe_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "0e8d54b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Peat soil</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sand</td>\n",
       "      <td>ENVO:01000017</td>\n",
       "      <td>sand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Leafy wood soil</td>\n",
       "      <td>ENVO:00005783</td>\n",
       "      <td>leafy wood soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>surface soil</td>\n",
       "      <td>ENVO:02000059</td>\n",
       "      <td>surface soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>bulk soil</td>\n",
       "      <td>ENVO:00005802</td>\n",
       "      <td>bulk soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Bulk soil</td>\n",
       "      <td>ENVO:00005802</td>\n",
       "      <td>bulk soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENVO:00002259</td>\n",
       "      <td>ENVO:00002259</td>\n",
       "      <td>agricultural soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENVO:00005783</td>\n",
       "      <td>ENVO:00005783</td>\n",
       "      <td>leafy wood soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENVO:00002245</td>\n",
       "      <td>ENVO:00002245</td>\n",
       "      <td>gypsisol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>mine tailings</td>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>mine tailing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>sediment</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>agricultural soil</td>\n",
       "      <td>ENVO:00002259</td>\n",
       "      <td>agricultural soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>histosol</td>\n",
       "      <td>ENVO:00002243</td>\n",
       "      <td>histosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>upland soil</td>\n",
       "      <td>ENVO:00005786</td>\n",
       "      <td>upland soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>peat</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Spruce forest soil</td>\n",
       "      <td>ENVO:00005784</td>\n",
       "      <td>spruce forest soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Mountain forest soil</td>\n",
       "      <td>ENVO:00005769</td>\n",
       "      <td>mountain forest soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>humus</td>\n",
       "      <td>ENVO:01000000</td>\n",
       "      <td>humus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>sandy soil</td>\n",
       "      <td>ENVO:00002229</td>\n",
       "      <td>arenosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>rock</td>\n",
       "      <td>ENVO:00001995</td>\n",
       "      <td>rock</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Grassland soil</td>\n",
       "      <td>ENVO:00005750</td>\n",
       "      <td>grassland soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ENVO:00005740 (paddy field soil)</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>ferralsol</td>\n",
       "      <td>ENVO:00002246</td>\n",
       "      <td>ferralsol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>gravel</td>\n",
       "      <td>ENVO:01000018</td>\n",
       "      <td>gravel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>frozen_soil</td>\n",
       "      <td>ENVO:01001526</td>\n",
       "      <td>frozen soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>andosol [ENVO:00002232]</td>\n",
       "      <td>ENVO:00002232</td>\n",
       "      <td>andosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>forest soil</td>\n",
       "      <td>ENVO:00002261</td>\n",
       "      <td>forest soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>water</td>\n",
       "      <td>CHEBI:15377</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>paddy soil</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>loam</td>\n",
       "      <td>ENVO:00002258</td>\n",
       "      <td>loam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>ENVO:00005780</td>\n",
       "      <td>greenhouse soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>anthrosol</td>\n",
       "      <td>ENVO:00002230</td>\n",
       "      <td>anthrosol</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          env_medium   consensus_id          consensus_lab\n",
       "0                               soil  ENVO:00001998                   soil\n",
       "1                               Soil  ENVO:00001998                   soil\n",
       "2                          ENVO:soil  ENVO:00001998                   soil\n",
       "3                        rhizosphere  ENVO:00005801            rhizosphere\n",
       "4                        Rhizosphere  ENVO:00005801            rhizosphere\n",
       "5                          peat soil  ENVO:00005774              peat soil\n",
       "6                          Peat soil  ENVO:00005774              peat soil\n",
       "7                               sand  ENVO:01000017                   sand\n",
       "9                    Leafy wood soil  ENVO:00005783        leafy wood soil\n",
       "10                      surface soil  ENVO:02000059           surface soil\n",
       "11                         bulk soil  ENVO:00005802              bulk soil\n",
       "12                         Bulk soil  ENVO:00005802              bulk soil\n",
       "14                     ENVO:00002259  ENVO:00002259      agricultural soil\n",
       "15                     ENVO:00005740  ENVO:00005740       paddy field soil\n",
       "16                     ENVO:00005783  ENVO:00005783        leafy wood soil\n",
       "17                     ENVO:00002245  ENVO:00002245               gypsisol\n",
       "19                     mine tailings  ENVO:00000003           mine tailing\n",
       "20                          Sediment  ENVO:00002007               sediment\n",
       "21                          sediment  ENVO:00002007               sediment\n",
       "22                 agricultural soil  ENVO:00002259      agricultural soil\n",
       "24                          histosol  ENVO:00002243               histosol\n",
       "25                       upland soil  ENVO:00005786            upland soil\n",
       "28                              peat  ENVO:00005774              peat soil\n",
       "29                Spruce forest soil  ENVO:00005784     spruce forest soil\n",
       "30              Mountain forest soil  ENVO:00005769   mountain forest soil\n",
       "31                          Bacteria    NCBITaxon:2               Bacteria\n",
       "32                          bacteria    NCBITaxon:2               Bacteria\n",
       "37                             humus  ENVO:01000000                  humus\n",
       "39                        sandy soil  ENVO:00002229               arenosol\n",
       "42                              rock  ENVO:00001995                   rock\n",
       "46                    Grassland soil  ENVO:00005750         grassland soil\n",
       "47  ENVO:00005740 (paddy field soil)  ENVO:00005740       paddy field soil\n",
       "48                       arable soil  ENVO:00005742            arable soil\n",
       "52                         ferralsol  ENVO:00002246              ferralsol\n",
       "53                            gravel  ENVO:01000018                 gravel\n",
       "54                       frozen_soil  ENVO:01001526            frozen soil\n",
       "56           andosol [ENVO:00002232]  ENVO:00002232                andosol\n",
       "60             oil contaminated soil  ENVO:00002875  oil contaminated soil\n",
       "67                       forest soil  ENVO:00002261            forest soil\n",
       "69                             water    CHEBI:15377                  water\n",
       "87                        paddy soil  ENVO:00005740       paddy field soil\n",
       "90                              loam  ENVO:00002258                   loam\n",
       "91                   greenhouse soil  ENVO:00005780        greenhouse soil\n",
       "92                         anthrosol  ENVO:00002230              anthrosol"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sssom_min_cols = sssom_frame[[\"subject_label\", \"object_id\", \"object_label\"]]\n",
    "\n",
    "sssom_min_cols.columns = [biosample_col_to_map, \"consensus_id\", \"consensus_lab\"]\n",
    "\n",
    "sssom_min_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "25424849",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENVO:soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>peat soil</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Peat soil</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sand</td>\n",
       "      <td>ENVO:01000017</td>\n",
       "      <td>sand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Leafy wood soil</td>\n",
       "      <td>ENVO:00005783</td>\n",
       "      <td>leafy wood soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>surface soil</td>\n",
       "      <td>ENVO:02000059</td>\n",
       "      <td>surface soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>bulk soil</td>\n",
       "      <td>ENVO:00005802</td>\n",
       "      <td>bulk soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Bulk soil</td>\n",
       "      <td>ENVO:00005802</td>\n",
       "      <td>bulk soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENVO:00002259</td>\n",
       "      <td>ENVO:00002259</td>\n",
       "      <td>agricultural soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENVO:00005783</td>\n",
       "      <td>ENVO:00005783</td>\n",
       "      <td>leafy wood soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENVO:00002245</td>\n",
       "      <td>ENVO:00002245</td>\n",
       "      <td>gypsisol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>mine tailings</td>\n",
       "      <td>ENVO:00000003</td>\n",
       "      <td>mine tailing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>sediment</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>agricultural soil</td>\n",
       "      <td>ENVO:00002259</td>\n",
       "      <td>agricultural soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>histosol</td>\n",
       "      <td>ENVO:00002243</td>\n",
       "      <td>histosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>upland soil</td>\n",
       "      <td>ENVO:00005786</td>\n",
       "      <td>upland soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>peat</td>\n",
       "      <td>ENVO:00005774</td>\n",
       "      <td>peat soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Spruce forest soil</td>\n",
       "      <td>ENVO:00005784</td>\n",
       "      <td>spruce forest soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Mountain forest soil</td>\n",
       "      <td>ENVO:00005769</td>\n",
       "      <td>mountain forest soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>NCBITaxon:2</td>\n",
       "      <td>Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>humus</td>\n",
       "      <td>ENVO:01000000</td>\n",
       "      <td>humus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>sandy soil</td>\n",
       "      <td>ENVO:00002229</td>\n",
       "      <td>arenosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>rock</td>\n",
       "      <td>ENVO:00001995</td>\n",
       "      <td>rock</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Grassland soil</td>\n",
       "      <td>ENVO:00005750</td>\n",
       "      <td>grassland soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ENVO:00005740 (paddy field soil)</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>ferralsol</td>\n",
       "      <td>ENVO:00002246</td>\n",
       "      <td>ferralsol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>gravel</td>\n",
       "      <td>ENVO:01000018</td>\n",
       "      <td>gravel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>frozen_soil</td>\n",
       "      <td>ENVO:01001526</td>\n",
       "      <td>frozen soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>andosol [ENVO:00002232]</td>\n",
       "      <td>ENVO:00002232</td>\n",
       "      <td>andosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>forest soil</td>\n",
       "      <td>ENVO:00002261</td>\n",
       "      <td>forest soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>water</td>\n",
       "      <td>CHEBI:15377</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>paddy soil</td>\n",
       "      <td>ENVO:00005740</td>\n",
       "      <td>paddy field soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>loam</td>\n",
       "      <td>ENVO:00002258</td>\n",
       "      <td>loam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>ENVO:00005780</td>\n",
       "      <td>greenhouse soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>anthrosol</td>\n",
       "      <td>ENVO:00002230</td>\n",
       "      <td>anthrosol</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          env_medium   consensus_id          consensus_lab\n",
       "0                               soil  ENVO:00001998                   soil\n",
       "1                               Soil  ENVO:00001998                   soil\n",
       "2                          ENVO:soil  ENVO:00001998                   soil\n",
       "3                        rhizosphere  ENVO:00005801            rhizosphere\n",
       "4                        Rhizosphere  ENVO:00005801            rhizosphere\n",
       "5                          peat soil  ENVO:00005774              peat soil\n",
       "6                          Peat soil  ENVO:00005774              peat soil\n",
       "7                               sand  ENVO:01000017                   sand\n",
       "9                    Leafy wood soil  ENVO:00005783        leafy wood soil\n",
       "10                      surface soil  ENVO:02000059           surface soil\n",
       "11                         bulk soil  ENVO:00005802              bulk soil\n",
       "12                         Bulk soil  ENVO:00005802              bulk soil\n",
       "14                     ENVO:00002259  ENVO:00002259      agricultural soil\n",
       "15                     ENVO:00005740  ENVO:00005740       paddy field soil\n",
       "16                     ENVO:00005783  ENVO:00005783        leafy wood soil\n",
       "17                     ENVO:00002245  ENVO:00002245               gypsisol\n",
       "19                     mine tailings  ENVO:00000003           mine tailing\n",
       "20                          Sediment  ENVO:00002007               sediment\n",
       "21                          sediment  ENVO:00002007               sediment\n",
       "22                 agricultural soil  ENVO:00002259      agricultural soil\n",
       "24                          histosol  ENVO:00002243               histosol\n",
       "25                       upland soil  ENVO:00005786            upland soil\n",
       "28                              peat  ENVO:00005774              peat soil\n",
       "29                Spruce forest soil  ENVO:00005784     spruce forest soil\n",
       "30              Mountain forest soil  ENVO:00005769   mountain forest soil\n",
       "31                          Bacteria    NCBITaxon:2               Bacteria\n",
       "32                          bacteria    NCBITaxon:2               Bacteria\n",
       "37                             humus  ENVO:01000000                  humus\n",
       "39                        sandy soil  ENVO:00002229               arenosol\n",
       "42                              rock  ENVO:00001995                   rock\n",
       "46                    Grassland soil  ENVO:00005750         grassland soil\n",
       "47  ENVO:00005740 (paddy field soil)  ENVO:00005740       paddy field soil\n",
       "48                       arable soil  ENVO:00005742            arable soil\n",
       "52                         ferralsol  ENVO:00002246              ferralsol\n",
       "53                            gravel  ENVO:01000018                 gravel\n",
       "54                       frozen_soil  ENVO:01001526            frozen soil\n",
       "56           andosol [ENVO:00002232]  ENVO:00002232                andosol\n",
       "60             oil contaminated soil  ENVO:00002875  oil contaminated soil\n",
       "67                       forest soil  ENVO:00002261            forest soil\n",
       "69                             water    CHEBI:15377                  water\n",
       "87                        paddy soil  ENVO:00005740       paddy field soil\n",
       "90                              loam  ENVO:00002258                   loam\n",
       "91                   greenhouse soil  ENVO:00005780        greenhouse soil\n",
       "92                         anthrosol  ENVO:00002230              anthrosol"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repipe_rows = len(repipe_frame.index)\n",
    "if repipe_rows > 0:\n",
    "    direct_and_repipe = pd.concat([temp, repipe_frame])\n",
    "else:\n",
    "    direct_and_repipe = sssom_min_cols\n",
    "direct_and_repipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "5ab3aaa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "direct_and_repipe.to_sql(\n",
    "    \"direct_and_repipe\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "922c09b3",
   "metadata": {},
   "source": [
    "## refactor this even more?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "8de27d49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-06-11 11:05:31.459113\n",
      "2021-06-11 11:05:31.667968\n",
      "0:00:00.208855\n"
     ]
    }
   ],
   "source": [
    "q = (\n",
    "    \"\"\"\n",
    "select\n",
    "\tb.id, b.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\",\n",
    "\tdar.consensus_id ,\n",
    "\tdar.consensus_lab\n",
    "from\n",
    "\tdirect_and_repipe dar\n",
    "left join biosample b on\n",
    "\tb.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\" = dar.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\"\n",
    "inner join biobiosample_tax_id_counts stic on\n",
    "\tb.taxonomy_id = stic.biosample_taxid\n",
    "join env_package_normalized on\n",
    "\tb.env_package = env_package_normalized.env_package\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"'\"\n",
    ")\n",
    "\n",
    "[per_biosample_scoped_mapping_results, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "a863e532",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706552</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706550</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706543</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706553</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706542</td>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>ENVO:00005780</td>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14946</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687814</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14947</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687805</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14948</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687808</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14949</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687809</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14950</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687806</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14951 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           id             env_medium   consensus_id  \\\n",
       "0      BIOSAMPLE:SAMEA1706552  oil contaminated soil  ENVO:00002875   \n",
       "1      BIOSAMPLE:SAMEA1706550  oil contaminated soil  ENVO:00002875   \n",
       "2      BIOSAMPLE:SAMEA1706543  oil contaminated soil  ENVO:00002875   \n",
       "3      BIOSAMPLE:SAMEA1706553  oil contaminated soil  ENVO:00002875   \n",
       "4      BIOSAMPLE:SAMEA1706542        greenhouse soil  ENVO:00005780   \n",
       "...                       ...                    ...            ...   \n",
       "14946  BIOSAMPLE:SAMEA1687814            arable soil  ENVO:00005742   \n",
       "14947  BIOSAMPLE:SAMEA1687805            arable soil  ENVO:00005742   \n",
       "14948  BIOSAMPLE:SAMEA1687808            arable soil  ENVO:00005742   \n",
       "14949  BIOSAMPLE:SAMEA1687809            arable soil  ENVO:00005742   \n",
       "14950  BIOSAMPLE:SAMEA1687806            arable soil  ENVO:00005742   \n",
       "\n",
       "               consensus_lab                                   notes  \n",
       "0      oil contaminated soil  env_package_normalized.EnvPackage=soil  \n",
       "1      oil contaminated soil  env_package_normalized.EnvPackage=soil  \n",
       "2      oil contaminated soil  env_package_normalized.EnvPackage=soil  \n",
       "3      oil contaminated soil  env_package_normalized.EnvPackage=soil  \n",
       "4            greenhouse soil  env_package_normalized.EnvPackage=soil  \n",
       "...                      ...                                     ...  \n",
       "14946            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "14947            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "14948            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "14949            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "14950            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "\n",
       "[14951 rows x 5 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_biosample_scoped_mapping_results[\"notes\"] = scoping_col + \"=\" + scoping_value\n",
    "\n",
    "table_name = \"repaired_\" + biosample_col_to_map\n",
    "\n",
    "per_biosample_scoped_mapping_results.to_sql(\n",
    "    table_name, biosample_cnx, if_exists=\"append\", index=False\n",
    ")\n",
    "\n",
    "per_biosample_scoped_mapping_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "93d2a346",
   "metadata": {},
   "outputs": [],
   "source": [
    "per_biosample_scoped_mapping_results.to_csv(\n",
    "    per_biosample_mapping_results_file,\n",
    "    sep=ols_review_seperator,\n",
    "    index=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43619dd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "raise SystemExit(\"Manually do ebs, els and em, OR AUTOMTE!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33cc14b4",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "1d45b3c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-06-11 12:02:49.805963\n",
      "2021-06-11 12:02:49.849540\n",
      "0:00:00.043577\n",
      "2021-06-11 12:02:49.850565\n",
      "2021-06-11 12:02:49.888576\n",
      "0:00:00.038011\n",
      "2021-06-11 12:02:49.894644\n",
      "2021-06-11 12:02:49.915513\n",
      "0:00:00.020869\n",
      "2021-06-11 12:02:49.918733\n",
      "2021-06-11 12:02:49.956433\n",
      "0:00:00.037700\n"
     ]
    }
   ],
   "source": [
    "# hard to do with pure sql in sqlite3 (no full outer jons,\n",
    "# just nested unions https://www.sqlitetutorial.net/sqlite-full-outer-join/)\n",
    "\n",
    "# refactor\n",
    "\n",
    "\n",
    "q = \"select * from repaired_env_package where env_package_rep = 'soil'\"\n",
    "\n",
    "[repaired_env_package, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "q = \"select * from repaired_env_broad_scale\"\n",
    "\n",
    "[repaired_env_broad_scale, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "repaired_env_broad_scale.drop(\"notes\", axis=1)\n",
    "colgroup = '_ebs'\n",
    "repaired_env_broad_scale=repaired_env_broad_scale.rename(columns = {'consensus_id':'consensus_id'+colgroup})\n",
    "repaired_env_broad_scale=repaired_env_broad_scale.rename(columns = {'consensus_lab':'consensus_lab'+colgroup})\n",
    "# repaired_env_broad_scale.columns[2] = repaired_env_broad_scale.columns[2] + colgroup\n",
    "# repaired_env_broad_scale.columns[3] = repaired_env_broad_scale.columns[3] + colgroup\n",
    "\n",
    "# temp = repaired_env_broad_scale.columns\n",
    "# temp = [sub + colgroup for sub in temp]\n",
    "# temp[0] = 'id'\n",
    "# repaired_env_broad_scale.columns = temp\n",
    "\n",
    "q = \"select * from repaired_env_local_scale\"\n",
    "\n",
    "[repaired_env_local_scale, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "repaired_env_local_scale.drop(\"notes\", axis=1)\n",
    "colgroup = '_els'\n",
    "repaired_env_local_scale=repaired_env_local_scale.rename(columns = {'consensus_id':'consensus_id'+colgroup})\n",
    "repaired_env_local_scale=repaired_env_local_scale.rename(columns = {'consensus_lab':'consensus_lab'+colgroup})\n",
    "\n",
    "\n",
    "q = \"select * from repaired_env_medium\"\n",
    "\n",
    "[repaired_env_medium, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")\n",
    "\n",
    "colgroup = '_em'\n",
    "repaired_env_medium=repaired_env_medium.rename(columns = {'consensus_id':'consensus_id'+colgroup})\n",
    "repaired_env_medium=repaired_env_medium.rename(columns = {'consensus_lab':'consensus_lab'+colgroup})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "29e69885",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>consensus_id_em</th>\n",
       "      <th>consensus_lab_em</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706552</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706550</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706543</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706553</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706542</td>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>ENVO:00005780</td>\n",
       "      <td>greenhouse soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14946</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687814</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14947</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687805</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14948</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687808</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14949</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687809</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14950</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687806</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14951 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           id             env_medium consensus_id_em  \\\n",
       "0      BIOSAMPLE:SAMEA1706552  oil contaminated soil   ENVO:00002875   \n",
       "1      BIOSAMPLE:SAMEA1706550  oil contaminated soil   ENVO:00002875   \n",
       "2      BIOSAMPLE:SAMEA1706543  oil contaminated soil   ENVO:00002875   \n",
       "3      BIOSAMPLE:SAMEA1706553  oil contaminated soil   ENVO:00002875   \n",
       "4      BIOSAMPLE:SAMEA1706542        greenhouse soil   ENVO:00005780   \n",
       "...                       ...                    ...             ...   \n",
       "14946  BIOSAMPLE:SAMEA1687814            arable soil   ENVO:00005742   \n",
       "14947  BIOSAMPLE:SAMEA1687805            arable soil   ENVO:00005742   \n",
       "14948  BIOSAMPLE:SAMEA1687808            arable soil   ENVO:00005742   \n",
       "14949  BIOSAMPLE:SAMEA1687809            arable soil   ENVO:00005742   \n",
       "14950  BIOSAMPLE:SAMEA1687806            arable soil   ENVO:00005742   \n",
       "\n",
       "            consensus_lab_em                                   notes  \n",
       "0      oil contaminated soil  env_package_normalized.EnvPackage=soil  \n",
       "1      oil contaminated soil  env_package_normalized.EnvPackage=soil  \n",
       "2      oil contaminated soil  env_package_normalized.EnvPackage=soil  \n",
       "3      oil contaminated soil  env_package_normalized.EnvPackage=soil  \n",
       "4            greenhouse soil  env_package_normalized.EnvPackage=soil  \n",
       "...                      ...                                     ...  \n",
       "14946            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "14947            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "14948            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "14949            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "14950            arable soil  env_package_normalized.EnvPackage=soil  \n",
       "\n",
       "[14951 rows x 5 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repaired_env_medium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "ae494cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "\n",
    "dflist = [\n",
    "    repaired_env_package,\n",
    "    repaired_env_broad_scale,\n",
    "    repaired_env_local_scale,\n",
    "    repaired_env_medium,\n",
    "]\n",
    "nan_value = \"\"\n",
    "# dfconcat = pd.concat(dflist, join='outer', axis=1).fillna(nan_value)\n",
    "# dfconcat\n",
    "\n",
    "df_merged = reduce(\n",
    "    lambda left, right: pd.merge(left, right, on=[\"id\"], how=\"outer\"), dflist\n",
    ").fillna(nan_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "5cdf0fd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_package_orig</th>\n",
       "      <th>env_package_rep</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>consensus_id_ebs</th>\n",
       "      <th>consensus_lab_ebs</th>\n",
       "      <th>notes_x</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>consensus_id_els</th>\n",
       "      <th>consensus_lab_els</th>\n",
       "      <th>notes_y</th>\n",
       "      <th>env_medium</th>\n",
       "      <th>consensus_id_em</th>\n",
       "      <th>consensus_lab_em</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706488</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706552</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>anthropogenic habitat</td>\n",
       "      <td>ENVO:00002031</td>\n",
       "      <td>anthropogenic habitat</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706550</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>anthropogenic habitat</td>\n",
       "      <td>ENVO:00002031</td>\n",
       "      <td>anthropogenic habitat</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706543</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>anthropogenic habitat</td>\n",
       "      <td>ENVO:00002031</td>\n",
       "      <td>anthropogenic habitat</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BIOSAMPLE:SAMEA1706553</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>anthropogenic habitat</td>\n",
       "      <td>ENVO:00002031</td>\n",
       "      <td>anthropogenic habitat</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>ENVO:00002875</td>\n",
       "      <td>oil contaminated soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15808</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687814</td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15809</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687805</td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15810</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687808</td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15811</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687809</td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15812</th>\n",
       "      <td>BIOSAMPLE:SAMEA1687806</td>\n",
       "      <td>Soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>ENVO:00000446</td>\n",
       "      <td>terrestrial biome</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>ENVO:00001998</td>\n",
       "      <td>soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>ENVO:00005742</td>\n",
       "      <td>arable soil</td>\n",
       "      <td>env_package_normalized.EnvPackage=soil</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15813 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           id env_package_orig env_package_rep  \\\n",
       "0      BIOSAMPLE:SAMEA1706488             soil            soil   \n",
       "1      BIOSAMPLE:SAMEA1706552             soil            soil   \n",
       "2      BIOSAMPLE:SAMEA1706550             soil            soil   \n",
       "3      BIOSAMPLE:SAMEA1706543             soil            soil   \n",
       "4      BIOSAMPLE:SAMEA1706553             soil            soil   \n",
       "...                       ...              ...             ...   \n",
       "15808  BIOSAMPLE:SAMEA1687814             Soil            soil   \n",
       "15809  BIOSAMPLE:SAMEA1687805             Soil            soil   \n",
       "15810  BIOSAMPLE:SAMEA1687808             Soil            soil   \n",
       "15811  BIOSAMPLE:SAMEA1687809             Soil            soil   \n",
       "15812  BIOSAMPLE:SAMEA1687806             Soil            soil   \n",
       "\n",
       "         env_broad_scale consensus_id_ebs  consensus_lab_ebs  \\\n",
       "0                                                              \n",
       "1      terrestrial biome    ENVO:00000446  terrestrial biome   \n",
       "2      terrestrial biome    ENVO:00000446  terrestrial biome   \n",
       "3      terrestrial biome    ENVO:00000446  terrestrial biome   \n",
       "4      terrestrial biome    ENVO:00000446  terrestrial biome   \n",
       "...                  ...              ...                ...   \n",
       "15808  terrestrial biome    ENVO:00000446  terrestrial biome   \n",
       "15809  terrestrial biome    ENVO:00000446  terrestrial biome   \n",
       "15810  terrestrial biome    ENVO:00000446  terrestrial biome   \n",
       "15811  terrestrial biome    ENVO:00000446  terrestrial biome   \n",
       "15812  terrestrial biome    ENVO:00000446  terrestrial biome   \n",
       "\n",
       "                                      notes_x        env_local_scale  \\\n",
       "0                                                                      \n",
       "1      env_package_normalized.EnvPackage=soil  anthropogenic habitat   \n",
       "2      env_package_normalized.EnvPackage=soil  anthropogenic habitat   \n",
       "3      env_package_normalized.EnvPackage=soil  anthropogenic habitat   \n",
       "4      env_package_normalized.EnvPackage=soil  anthropogenic habitat   \n",
       "...                                       ...                    ...   \n",
       "15808  env_package_normalized.EnvPackage=soil                   soil   \n",
       "15809  env_package_normalized.EnvPackage=soil                   soil   \n",
       "15810  env_package_normalized.EnvPackage=soil                   soil   \n",
       "15811  env_package_normalized.EnvPackage=soil                   soil   \n",
       "15812  env_package_normalized.EnvPackage=soil                   soil   \n",
       "\n",
       "      consensus_id_els      consensus_lab_els  \\\n",
       "0                                               \n",
       "1        ENVO:00002031  anthropogenic habitat   \n",
       "2        ENVO:00002031  anthropogenic habitat   \n",
       "3        ENVO:00002031  anthropogenic habitat   \n",
       "4        ENVO:00002031  anthropogenic habitat   \n",
       "...                ...                    ...   \n",
       "15808    ENVO:00001998                   soil   \n",
       "15809    ENVO:00001998                   soil   \n",
       "15810    ENVO:00001998                   soil   \n",
       "15811    ENVO:00001998                   soil   \n",
       "15812    ENVO:00001998                   soil   \n",
       "\n",
       "                                      notes_y             env_medium  \\\n",
       "0                                                                      \n",
       "1      env_package_normalized.EnvPackage=soil  oil contaminated soil   \n",
       "2      env_package_normalized.EnvPackage=soil  oil contaminated soil   \n",
       "3      env_package_normalized.EnvPackage=soil  oil contaminated soil   \n",
       "4      env_package_normalized.EnvPackage=soil  oil contaminated soil   \n",
       "...                                       ...                    ...   \n",
       "15808  env_package_normalized.EnvPackage=soil            arable soil   \n",
       "15809  env_package_normalized.EnvPackage=soil            arable soil   \n",
       "15810  env_package_normalized.EnvPackage=soil            arable soil   \n",
       "15811  env_package_normalized.EnvPackage=soil            arable soil   \n",
       "15812  env_package_normalized.EnvPackage=soil            arable soil   \n",
       "\n",
       "      consensus_id_em       consensus_lab_em  \\\n",
       "0                                              \n",
       "1       ENVO:00002875  oil contaminated soil   \n",
       "2       ENVO:00002875  oil contaminated soil   \n",
       "3       ENVO:00002875  oil contaminated soil   \n",
       "4       ENVO:00002875  oil contaminated soil   \n",
       "...               ...                    ...   \n",
       "15808   ENVO:00005742            arable soil   \n",
       "15809   ENVO:00005742            arable soil   \n",
       "15810   ENVO:00005742            arable soil   \n",
       "15811   ENVO:00005742            arable soil   \n",
       "15812   ENVO:00005742            arable soil   \n",
       "\n",
       "                                        notes  \n",
       "0                                              \n",
       "1      env_package_normalized.EnvPackage=soil  \n",
       "2      env_package_normalized.EnvPackage=soil  \n",
       "3      env_package_normalized.EnvPackage=soil  \n",
       "4      env_package_normalized.EnvPackage=soil  \n",
       "...                                       ...  \n",
       "15808  env_package_normalized.EnvPackage=soil  \n",
       "15809  env_package_normalized.EnvPackage=soil  \n",
       "15810  env_package_normalized.EnvPackage=soil  \n",
       "15811  env_package_normalized.EnvPackage=soil  \n",
       "15812  env_package_normalized.EnvPackage=soil  \n",
       "\n",
       "[15813 rows x 15 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "30540015",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged.to_csv('repaired_soil_packages.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "410fce5d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e6ed73d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
