{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fac2962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy\n",
    "import scoped_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5684af0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_sqlite_file = \"../target/harmonized_table.db\"\n",
    "# TODO process these as a list?\n",
    "ncbitaxon_sqlite_file = \"../semantic-sql/db/ncbitaxon.db\"\n",
    "envo_sqlite_file = \"../semantic-sql/db/envo.db\"\n",
    "ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)\n",
    "envo_cnx = sqlite3.connect(envo_sqlite_file)\n",
    "target_onto_prefix = 'ENVO'\n",
    "chars_to_whiteout = '._-'\n",
    "my_query_fields = ''\n",
    "my_row_req = 3\n",
    "\n",
    "env_package_overrides = {\n",
    "    'built environment': 'built',\n",
    "    'misc environment': 'miscellaneous',\n",
    "    'missing': 'no environmental package',\n",
    "    'unknown': 'no environmental package',\n",
    "    'default': 'no environmental package',\n",
    "    'unspecified': 'no environmental package',\n",
    "    'not available': 'no environmental package',\n",
    "    'not collected': 'no environmental package'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9282126f",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "caff574e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_package</th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>host_taxid</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>env_medium</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMN00000002</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445970</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BIOSAMPLE:SAMN00000003</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445972</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIOSAMPLE:SAMN00000004</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>449673</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BIOSAMPLE:SAMN00000005</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>6526</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BIOSAMPLE:SAMN00000006</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>9483</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BIOSAMPLE:SAMN00000007</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>445974</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BIOSAMPLE:SAMN00000008</td>\n",
       "      <td>missing</td>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>411461</td>\n",
       "      <td>terrestrial biome [ENVO:00000446]</td>\n",
       "      <td>human-associated habitat [ENVO:00009003]</td>\n",
       "      <td>biological product [ENVO:02000043]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BIOSAMPLE:SAMN00000009</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>13616</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BIOSAMPLE:SAMN00000010</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>451639</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BIOSAMPLE:SAMN00000011</td>\n",
       "      <td>None</td>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>None</td>\n",
       "      <td>451638</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id env_package      package  \\\n",
       "0  BIOSAMPLE:SAMN00000002     missing  MIGS.ba.5.0   \n",
       "1  BIOSAMPLE:SAMN00000003     missing  MIGS.ba.5.0   \n",
       "2  BIOSAMPLE:SAMN00000004     missing  MIGS.ba.5.0   \n",
       "3  BIOSAMPLE:SAMN00000005        None  Generic.1.0   \n",
       "4  BIOSAMPLE:SAMN00000006        None  Generic.1.0   \n",
       "5  BIOSAMPLE:SAMN00000007     missing  MIGS.ba.5.0   \n",
       "6  BIOSAMPLE:SAMN00000008     missing  MIGS.ba.5.0   \n",
       "7  BIOSAMPLE:SAMN00000009        None  Generic.1.0   \n",
       "8  BIOSAMPLE:SAMN00000010        None  Generic.1.0   \n",
       "9  BIOSAMPLE:SAMN00000011        None  Generic.1.0   \n",
       "\n",
       "                                   package_name host_taxid taxonomy_id  \\\n",
       "0  MIGS: cultured bacteria/archaea; version 5.0       9606      445970   \n",
       "1  MIGS: cultured bacteria/archaea; version 5.0       9606      445972   \n",
       "2  MIGS: cultured bacteria/archaea; version 5.0       9606      449673   \n",
       "3                                       Generic       None        6526   \n",
       "4                                       Generic       None        9483   \n",
       "5  MIGS: cultured bacteria/archaea; version 5.0       9606      445974   \n",
       "6  MIGS: cultured bacteria/archaea; version 5.0       9606      411461   \n",
       "7                                       Generic       None       13616   \n",
       "8                                       Generic       None      451639   \n",
       "9                                       Generic       None      451638   \n",
       "\n",
       "                     env_broad_scale  \\\n",
       "0  terrestrial biome [ENVO:00000446]   \n",
       "1  terrestrial biome [ENVO:00000446]   \n",
       "2  terrestrial biome [ENVO:00000446]   \n",
       "3                               None   \n",
       "4                               None   \n",
       "5  terrestrial biome [ENVO:00000446]   \n",
       "6  terrestrial biome [ENVO:00000446]   \n",
       "7                               None   \n",
       "8                               None   \n",
       "9                               None   \n",
       "\n",
       "                            env_local_scale  \\\n",
       "0  human-associated habitat [ENVO:00009003]   \n",
       "1  human-associated habitat [ENVO:00009003]   \n",
       "2  human-associated habitat [ENVO:00009003]   \n",
       "3                                      None   \n",
       "4                                      None   \n",
       "5  human-associated habitat [ENVO:00009003]   \n",
       "6  human-associated habitat [ENVO:00009003]   \n",
       "7                                      None   \n",
       "8                                      None   \n",
       "9                                      None   \n",
       "\n",
       "                           env_medium  \n",
       "0  biological product [ENVO:02000043]  \n",
       "1  biological product [ENVO:02000043]  \n",
       "2  biological product [ENVO:02000043]  \n",
       "3                                None  \n",
       "4                                None  \n",
       "5  biological product [ENVO:02000043]  \n",
       "6  biological product [ENVO:02000043]  \n",
       "7                                None  \n",
       "8                                None  \n",
       "9                                None  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample of the data we're working with\n",
    "q = \"\"\"\n",
    "select\n",
    "    id,\n",
    "    env_package,\n",
    "    package,\n",
    "    package_name,\n",
    "    host_taxid,\n",
    "    taxonomy_id,\n",
    "    env_broad_scale,\n",
    "    env_local_scale,\n",
    "    env_medium\n",
    "    from biosample b\n",
    "limit 10\n",
    "\"\"\"\n",
    "biosample_first_ten = pd.read_sql(q, biosample_cnx)\n",
    "biosample_first_ten"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c6cc9596",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>DisplayName</th>\n",
       "      <th>ShortName</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>EnvPackageDisplay</th>\n",
       "      <th>NotAppropriateFor</th>\n",
       "      <th>Description</th>\n",
       "      <th>Example</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Generic</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2.cl.1.0</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated; versi...</td>\n",
       "      <td>SARS-CoV-2: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for SARS-CoV-2 samples that are relevant t...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>Pathogen: clinical or host-associated</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Clinical or host-associated pathogen</td>\n",
       "      <td>SAMN02928182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>Pathogen: environmental/food/other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Environmental, food or other pathogen</td>\n",
       "      <td>SAMN02730065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Microbe.1.0</td>\n",
       "      <td>Microbe; version 1.0</td>\n",
       "      <td>Microbe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Use for bacteria or other unicellular microbes...</td>\n",
       "      <td>SAMN02911891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>MIUVIG.plant-associated.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, plant-assoc...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>plant-associated</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>MIUVIG.sediment.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, sediment; v...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>MIUVIG.soil.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, soil; versi...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>soil</td>\n",
       "      <td>soil</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>MIUVIG.wastewater.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, wastewater;...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wastewater</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>MIUVIG.water.5.0</td>\n",
       "      <td>MIUVIG: uncultivated virus genome, water; vers...</td>\n",
       "      <td>MIUVIG Uncultivated Virus Genome</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>wgs_single;wgs_batch;wgs_diploid</td>\n",
       "      <td>Use for uncultivated virus genome identified i...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Name  \\\n",
       "0                    Generic.1.0   \n",
       "1              SARS-CoV-2.cl.1.0   \n",
       "2                Pathogen.cl.1.0   \n",
       "3               Pathogen.env.1.0   \n",
       "4                    Microbe.1.0   \n",
       "..                           ...   \n",
       "149  MIUVIG.plant-associated.5.0   \n",
       "150          MIUVIG.sediment.5.0   \n",
       "151              MIUVIG.soil.5.0   \n",
       "152        MIUVIG.wastewater.5.0   \n",
       "153             MIUVIG.water.5.0   \n",
       "\n",
       "                                           DisplayName  \\\n",
       "0                                              Generic   \n",
       "1    SARS-CoV-2: clinical or host-associated; versi...   \n",
       "2    Pathogen: clinical or host-associated; version...   \n",
       "3      Pathogen: environmental/food/other; version 1.0   \n",
       "4                                 Microbe; version 1.0   \n",
       "..                                                 ...   \n",
       "149  MIUVIG: uncultivated virus genome, plant-assoc...   \n",
       "150  MIUVIG: uncultivated virus genome, sediment; v...   \n",
       "151  MIUVIG: uncultivated virus genome, soil; versi...   \n",
       "152  MIUVIG: uncultivated virus genome, wastewater;...   \n",
       "153  MIUVIG: uncultivated virus genome, water; vers...   \n",
       "\n",
       "                                   ShortName        EnvPackage  \\\n",
       "0                                                                \n",
       "1    SARS-CoV-2: clinical or host-associated                     \n",
       "2      Pathogen: clinical or host-associated                     \n",
       "3         Pathogen: environmental/food/other                     \n",
       "4                                    Microbe                     \n",
       "..                                       ...               ...   \n",
       "149         MIUVIG Uncultivated Virus Genome  plant-associated   \n",
       "150         MIUVIG Uncultivated Virus Genome          sediment   \n",
       "151         MIUVIG Uncultivated Virus Genome              soil   \n",
       "152         MIUVIG Uncultivated Virus Genome        wastewater   \n",
       "153         MIUVIG Uncultivated Virus Genome             water   \n",
       "\n",
       "    EnvPackageDisplay                 NotAppropriateFor  \\\n",
       "0                                                         \n",
       "1                      wgs_single;wgs_batch;wgs_diploid   \n",
       "2                                                         \n",
       "3                                                         \n",
       "4                                                         \n",
       "..                ...                               ...   \n",
       "149  plant-associated  wgs_single;wgs_batch;wgs_diploid   \n",
       "150          sediment  wgs_single;wgs_batch;wgs_diploid   \n",
       "151              soil  wgs_single;wgs_batch;wgs_diploid   \n",
       "152        wastewater  wgs_single;wgs_batch;wgs_diploid   \n",
       "153             water  wgs_single;wgs_batch;wgs_diploid   \n",
       "\n",
       "                                           Description       Example  \n",
       "0                                              Generic                \n",
       "1    Use for SARS-CoV-2 samples that are relevant t...                \n",
       "2                 Clinical or host-associated pathogen  SAMN02928182  \n",
       "3                Environmental, food or other pathogen  SAMN02730065  \n",
       "4    Use for bacteria or other unicellular microbes...  SAMN02911891  \n",
       "..                                                 ...           ...  \n",
       "149  Use for uncultivated virus genome identified i...                \n",
       "150  Use for uncultivated virus genome identified i...                \n",
       "151  Use for uncultivated virus genome identified i...                \n",
       "152  Use for uncultivated virus genome identified i...                \n",
       "153  Use for uncultivated virus genome identified i...                \n",
       "\n",
       "[154 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the canonical checklist and package terms from NCBI\n",
    "# Unfortunately it doesn't do a very good job of differentiating \n",
    "# checklists (MIMAG, MIMARKS, etc.) \n",
    "# from packages (soil, water, etc.)\n",
    "# what about ba , euk, etc?\n",
    "package_dictionary = scoped_mapping.get_package_dictionary()\n",
    "package_dictionary.to_sql('package_dictionary', biosample_cnx, if_exists='replace', index=False)\n",
    "package_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0cfa7d0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.466004\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_row_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14300584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   biosample_row_count\n",
       "0             14300584"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Do the Biosample checklist/package fields match any of the cannonical values?\n",
    "# How many Biosample rows are there?\n",
    "q = \"\"\"\n",
    "select count(*) as biosample_row_count\n",
    "from biosample b\n",
    "\"\"\"\n",
    "[biosample_row_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx, print_timing=False)\n",
    "\n",
    "print(query_duration)\n",
    "biosample_row_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "03a8b90c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-05-24 21:13:36.208937\n",
      "2021-05-24 21:13:37.977819\n",
      "0:00:01.768882\n",
      "0:00:01.768882\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cannonical_package_name_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14300584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cannonical_package_name_count\n",
       "0                       14300584"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many of those rows can be inner-joined with the canonical checklists/packages?\n",
    "# Specifically, joining biosample.package_name = package_dictionary.DisplayName\n",
    "# TODO add indexing to docs and or makefile\n",
    "# create index biosample_package_name_idx on biosample(package_name);\n",
    "# create index package_dictionary_DisplayName_idx on package_dictionary(DisplayName);\n",
    "# create index biosample_package_idx on biosample(package);\n",
    "# create index biosample_p_pn_idx on biosample(package, package_name);\n",
    "q = \"\"\"\n",
    "select\n",
    "    count(*) as cannonical_package_name_count\n",
    "from\n",
    "    biosample b\n",
    "inner join package_dictionary pd on\n",
    "    b.package_name = pd.DisplayName\n",
    "\"\"\"\n",
    "[cannonical_package_name_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx, print_timing=True)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "cannonical_package_name_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "83485b99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-05-24 21:13:37.985821\n",
      "2021-05-24 21:13:41.457979\n",
      "0:00:03.472158\n",
      "0:00:03.472158\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>package_name</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Beta-lactamase.1.0</td>\n",
       "      <td>Beta-lactamase; version 1.0</td>\n",
       "      <td>556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Generic.1.0</td>\n",
       "      <td>Generic</td>\n",
       "      <td>10186430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Human.1.0</td>\n",
       "      <td>Human; version 1.0</td>\n",
       "      <td>368893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Invertebrate.1.0</td>\n",
       "      <td>Invertebrate; version 1.0</td>\n",
       "      <td>131085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIGS.ba.5.0</td>\n",
       "      <td>MIGS: cultured bacteria/archaea; version 5.0</td>\n",
       "      <td>28686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>Model.organism.animal.1.0</td>\n",
       "      <td>Model organism or animal; version 1.0</td>\n",
       "      <td>382980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Pathogen.cl.1.0</td>\n",
       "      <td>Pathogen: clinical or host-associated; version...</td>\n",
       "      <td>511040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Pathogen.env.1.0</td>\n",
       "      <td>Pathogen: environmental/food/other; version 1.0</td>\n",
       "      <td>246497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>Plant.1.0</td>\n",
       "      <td>Plant; version 1.0</td>\n",
       "      <td>385939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Virus.1.0</td>\n",
       "      <td>Virus; version 1.0</td>\n",
       "      <td>15560</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       package  \\\n",
       "0           Beta-lactamase.1.0   \n",
       "1                  Generic.1.0   \n",
       "2                    Human.1.0   \n",
       "3             Invertebrate.1.0   \n",
       "4                  MIGS.ba.5.0   \n",
       "..                         ...   \n",
       "124  Model.organism.animal.1.0   \n",
       "125            Pathogen.cl.1.0   \n",
       "126           Pathogen.env.1.0   \n",
       "127                  Plant.1.0   \n",
       "128                  Virus.1.0   \n",
       "\n",
       "                                          package_name     count  \n",
       "0                          Beta-lactamase; version 1.0       556  \n",
       "1                                              Generic  10186430  \n",
       "2                                   Human; version 1.0    368893  \n",
       "3                            Invertebrate; version 1.0    131085  \n",
       "4         MIGS: cultured bacteria/archaea; version 5.0     28686  \n",
       "..                                                 ...       ...  \n",
       "124              Model organism or animal; version 1.0    382980  \n",
       "125  Pathogen: clinical or host-associated; version...    511040  \n",
       "126    Pathogen: environmental/food/other; version 1.0    246497  \n",
       "127                                 Plant; version 1.0    385939  \n",
       "128                                 Virus; version 1.0     15560  \n",
       "\n",
       "[129 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What do the combinations of package and package_name look like in the Biosample dataset?\n",
    "q = \"\"\"\n",
    "select\n",
    "    package,\n",
    "    package_name,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    package ,\n",
    "    package_name\n",
    "order by\n",
    "    package ,\n",
    "    package_name\n",
    "\"\"\"\n",
    "[package_name_combos, query_duration] = scoped_mapping.timed_query(q, biosample_cnx, print_timing=True)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "package_name_combos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d042df82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:01.153799\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count\n",
       "0                              None  14083847\n",
       "1                   host-associated     49254\n",
       "2                         human-gut     47921\n",
       "3                             water     16367\n",
       "4                        human-skin     13706\n",
       "..                              ...       ...\n",
       "87                      env_package         1\n",
       "88                   gut microbiome         1\n",
       "89  marine sediment (ENVO:00002113)         1\n",
       "90                           saliva         1\n",
       "91       sea water, [ENVO:00002149]         1\n",
       "\n",
       "[92 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What about the Biosample env_package values?\n",
    "# Are they also a small, highly regular set?\n",
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    env_package\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[env_package_count, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "env_package_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "82ba2408",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " 'air',\n",
       " 'built',\n",
       " 'host associated',\n",
       " 'human associated',\n",
       " 'human gut',\n",
       " 'human oral',\n",
       " 'human skin',\n",
       " 'human vaginal',\n",
       " 'microbial',\n",
       " 'microbial mat biofilm',\n",
       " 'miscellaneous',\n",
       " 'miscellaneous or artificial',\n",
       " 'no environmental package',\n",
       " 'plant associated',\n",
       " 'sediment',\n",
       " 'soil',\n",
       " 'wastewater',\n",
       " 'wastewater sludge',\n",
       " 'water']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# env_package is going to need some cleanup\n",
    "# First, get a set of all canonical env_package values\n",
    "package_dictionary = scoped_mapping.make_tidy_col(package_dictionary, 'EnvPackage', 'eptidy')\n",
    "package_dictionary =scoped_mapping.make_tidy_col(package_dictionary, 'EnvPackageDisplay', 'epdtidy')\n",
    "# update in sqlite\n",
    "package_dictionary.to_sql('package_dictionary', biosample_cnx, if_exists='replace', index=False)\n",
    "valid_combo = []\n",
    "valid_combo = scoped_mapping.add_unique_to_list(valid_combo, package_dictionary['eptidy'])\n",
    "valid_combo = scoped_mapping.add_unique_to_list(valid_combo, package_dictionary['epdtidy'])\n",
    "\n",
    "valid_combo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5d157f3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.030432\n"
     ]
    }
   ],
   "source": [
    "# determine ID patterns\n",
    "q = \"\"\"\n",
    "select\n",
    "    distinct stanza\n",
    "    from statements s\n",
    "where\n",
    "    predicate = 'rdf:type'\n",
    "    and \"object\" = 'owl:Class'\n",
    "    and stanza = subject\"\"\"\n",
    "# include non-envo IDs that come from envo?\n",
    "[ids_from_envo, query_duration] = scoped_mapping.timed_query(q, envo_cnx)\n",
    "print(query_duration)\n",
    "ids_from_envo = scoped_mapping.add_prefix_col(ids_from_envo, 'stanza', 'prefix')\n",
    "\n",
    "id_patterns = scoped_mapping.get_multi_term_patterns(ids_from_envo, 'stanza', 'prefix')\n",
    "\n",
    "env_package_normalized = scoped_mapping.env_package_nomralizastion(env_package_count, 'env_package',\n",
    "                                                                target_onto_prefix, id_patterns['ENVO'])\n",
    "\n",
    "env_package_normalized = scoped_mapping.add_overrides(env_package_normalized, 'remaining_tidied', 'ep_override',\n",
    "                                                   env_package_overrides)\n",
    "\n",
    "env_package_normalized = scoped_mapping.flag_canonical(env_package_normalized, 'ep_override', 'is_canonical',\n",
    "                                                    valid_combo)\n",
    "\n",
    "env_package_normalized.to_sql('env_package_normalized', biosample_cnx, if_exists='replace', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7c1fbb75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.002367\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>ep_override</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>built environment</td>\n",
       "      <td>12391</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>built</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>soil</td>\n",
       "      <td>11974</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>misc environment</td>\n",
       "      <td>11715</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>missing</td>\n",
       "      <td>8453</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>human-oral</td>\n",
       "      <td>7882</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human oral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sediment</td>\n",
       "      <td>5969</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.soil</td>\n",
       "      <td>3784</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>human-associated</td>\n",
       "      <td>3167</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.host-associated</td>\n",
       "      <td>2985</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>2499</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>microbial mat/biofilm</td>\n",
       "      <td>2129</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>microbial mat biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>plant-associated</td>\n",
       "      <td>2076</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>plant associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>1763</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-associated</td>\n",
       "      <td>1417</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.sediment</td>\n",
       "      <td>1177</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-oral</td>\n",
       "      <td>1154</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human oral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>human-vaginal</td>\n",
       "      <td>1077</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human vaginal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.plant-associated</td>\n",
       "      <td>1063</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>plant associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.microbial</td>\n",
       "      <td>732</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>microbial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.miscellaneous</td>\n",
       "      <td>525</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>wastewater|sludge</td>\n",
       "      <td>417</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wastewater sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.wastewater</td>\n",
       "      <td>385</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>wastewater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>wastewater/sludge</td>\n",
       "      <td>357</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wastewater sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>unknown</td>\n",
       "      <td>283</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Human-associated</td>\n",
       "      <td>212</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.air</td>\n",
       "      <td>205</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>air</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Human-oral</td>\n",
       "      <td>201</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human oral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>host_associated</td>\n",
       "      <td>171</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>air</td>\n",
       "      <td>152</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>air</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-skin</td>\n",
       "      <td>135</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Human-gut</td>\n",
       "      <td>111</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>wastewater sludge</td>\n",
       "      <td>90</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wastewater sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>built</td>\n",
       "      <td>61</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>built</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Human_Gut</td>\n",
       "      <td>59</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>human gut</td>\n",
       "      <td>59</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>default</td>\n",
       "      <td>51</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>microbial mat|biofilm</td>\n",
       "      <td>48</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>microbial mat biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Unspecified</td>\n",
       "      <td>42</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>miscellaneous</td>\n",
       "      <td>33</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Human-vaginal</td>\n",
       "      <td>28</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human vaginal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>human_skin</td>\n",
       "      <td>28</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human skin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Soil</td>\n",
       "      <td>19</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>wastewater</td>\n",
       "      <td>14</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>wastewater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-vaginal</td>\n",
       "      <td>9</td>\n",
       "      <td>MIGS/MIMS/MIMARKS</td>\n",
       "      <td></td>\n",
       "      <td>human vaginal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>misc-environment</td>\n",
       "      <td>7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Not available</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Host-associated</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>MIGS.plant-associated</td>\n",
       "      <td>3</td>\n",
       "      <td>MIGS</td>\n",
       "      <td></td>\n",
       "      <td>plant associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Water</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>not collected</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>host associated</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>host associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Unknown</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no environmental package</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           env_package     count                lhs  \\\n",
       "0                                 None  14083847                      \n",
       "1                      host-associated     49254                      \n",
       "2                            human-gut     47921                      \n",
       "3                                water     16367                      \n",
       "4                           human-skin     13706                      \n",
       "5                    built environment     12391                      \n",
       "6                                 soil     11974                      \n",
       "7                     misc environment     11715                      \n",
       "8                              missing      8453                      \n",
       "9                           human-oral      7882                      \n",
       "10                            sediment      5969                      \n",
       "11              MIGS/MIMS/MIMARKS.soil      3784  MIGS/MIMS/MIMARKS   \n",
       "12                    human-associated      3167                      \n",
       "13   MIGS/MIMS/MIMARKS.host-associated      2985  MIGS/MIMS/MIMARKS   \n",
       "14         MIGS/MIMS/MIMARKS.human-gut      2499  MIGS/MIMS/MIMARKS   \n",
       "15               microbial mat/biofilm      2129                      \n",
       "16                    plant-associated      2076                      \n",
       "17             MIGS/MIMS/MIMARKS.water      1763  MIGS/MIMS/MIMARKS   \n",
       "18  MIGS/MIMS/MIMARKS.human-associated      1417  MIGS/MIMS/MIMARKS   \n",
       "19          MIGS/MIMS/MIMARKS.sediment      1177  MIGS/MIMS/MIMARKS   \n",
       "20        MIGS/MIMS/MIMARKS.human-oral      1154  MIGS/MIMS/MIMARKS   \n",
       "21                       human-vaginal      1077                      \n",
       "22  MIGS/MIMS/MIMARKS.plant-associated      1063  MIGS/MIMS/MIMARKS   \n",
       "23         MIGS/MIMS/MIMARKS.microbial       732  MIGS/MIMS/MIMARKS   \n",
       "24     MIGS/MIMS/MIMARKS.miscellaneous       525  MIGS/MIMS/MIMARKS   \n",
       "25                   wastewater|sludge       417                      \n",
       "26        MIGS/MIMS/MIMARKS.wastewater       385  MIGS/MIMS/MIMARKS   \n",
       "27                   wastewater/sludge       357                      \n",
       "28                             unknown       283                      \n",
       "29                    Human-associated       212                      \n",
       "30               MIGS/MIMS/MIMARKS.air       205  MIGS/MIMS/MIMARKS   \n",
       "31                          Human-oral       201                      \n",
       "32                     host_associated       171                      \n",
       "33                                 air       152                      \n",
       "34        MIGS/MIMS/MIMARKS.human-skin       135  MIGS/MIMS/MIMARKS   \n",
       "35                           Human-gut       111                      \n",
       "36                   wastewater sludge        90                      \n",
       "37                               built        61                      \n",
       "38                           Human_Gut        59                      \n",
       "39                           human gut        59                      \n",
       "40                             default        51                      \n",
       "41               microbial mat|biofilm        48                      \n",
       "42                         Unspecified        42                      \n",
       "43                       miscellaneous        33                      \n",
       "44                       Human-vaginal        28                      \n",
       "45                          human_skin        28                      \n",
       "46                                Soil        19                      \n",
       "47                          wastewater        14                      \n",
       "48     MIGS/MIMS/MIMARKS.human-vaginal         9  MIGS/MIMS/MIMARKS   \n",
       "49                    misc-environment         7                      \n",
       "50                       Not available         4                      \n",
       "51                     Host-associated         3                      \n",
       "52               MIGS.plant-associated         3               MIGS   \n",
       "53                               Water         3                      \n",
       "54                       not collected         3                      \n",
       "55                     host associated         2                      \n",
       "56                       ENVO:00000016         1                      \n",
       "57                            Sediment         1                      \n",
       "58                             Unknown         1                      \n",
       "\n",
       "          extract               ep_override  \n",
       "0                                            \n",
       "1                           host associated  \n",
       "2                                 human gut  \n",
       "3                                     water  \n",
       "4                                human skin  \n",
       "5                                     built  \n",
       "6                                      soil  \n",
       "7                             miscellaneous  \n",
       "8                  no environmental package  \n",
       "9                                human oral  \n",
       "10                                 sediment  \n",
       "11                                     soil  \n",
       "12                         human associated  \n",
       "13                          host associated  \n",
       "14                                human gut  \n",
       "15                    microbial mat biofilm  \n",
       "16                         plant associated  \n",
       "17                                    water  \n",
       "18                         human associated  \n",
       "19                                 sediment  \n",
       "20                               human oral  \n",
       "21                            human vaginal  \n",
       "22                         plant associated  \n",
       "23                                microbial  \n",
       "24                            miscellaneous  \n",
       "25                        wastewater sludge  \n",
       "26                               wastewater  \n",
       "27                        wastewater sludge  \n",
       "28                 no environmental package  \n",
       "29                         human associated  \n",
       "30                                      air  \n",
       "31                               human oral  \n",
       "32                          host associated  \n",
       "33                                      air  \n",
       "34                               human skin  \n",
       "35                                human gut  \n",
       "36                        wastewater sludge  \n",
       "37                                    built  \n",
       "38                                human gut  \n",
       "39                                human gut  \n",
       "40                 no environmental package  \n",
       "41                    microbial mat biofilm  \n",
       "42                 no environmental package  \n",
       "43                            miscellaneous  \n",
       "44                            human vaginal  \n",
       "45                               human skin  \n",
       "46                                     soil  \n",
       "47                               wastewater  \n",
       "48                            human vaginal  \n",
       "49                            miscellaneous  \n",
       "50                 no environmental package  \n",
       "51                          host associated  \n",
       "52                         plant associated  \n",
       "53                                    water  \n",
       "54                 no environmental package  \n",
       "55                          host associated  \n",
       "56  ENVO:00000016                            \n",
       "57                                 sediment  \n",
       "58                 no environmental package  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What do the successful normalizations look like?\n",
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    ep_override\n",
    "from\n",
    "    env_package_normalized\n",
    "where\n",
    "    is_canonical = 1\n",
    "\"\"\"\n",
    "[successful_normalizastions, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "successful_normalizastions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "17e08383",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:00.002144\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>ep_override</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>miscellaneous natural or artificial environment</td>\n",
       "      <td>611</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miscellaneous natural or artificial environment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mimarks</td>\n",
       "      <td>479</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mimarks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mouse-gut</td>\n",
       "      <td>406</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mouse gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gut</td>\n",
       "      <td>172</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>biofilm</td>\n",
       "      <td>114</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>biofilm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>human-not providedsopharyngeal</td>\n",
       "      <td>107</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>human not providedsopharyngeal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>mice gut</td>\n",
       "      <td>87</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mice gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CV</td>\n",
       "      <td>60</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>home, outdoor environment</td>\n",
       "      <td>44</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>home outdoor environment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>fermentation-associated</td>\n",
       "      <td>42</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>fermentation associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>not applicable</td>\n",
       "      <td>41</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>not applicable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>soil-associated</td>\n",
       "      <td>36</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>soil associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>sterile water</td>\n",
       "      <td>35</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sterile water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Mouse_Gut</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mouse gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>fresh water</td>\n",
       "      <td>27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>fresh water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rock</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>rock</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>arctic</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>arctic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>freshwater sediment</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>freshwater sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>seawater</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>seawater</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MIENS_16S</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miens 16s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MIMARKS_water</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mimarks water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>cave</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cave</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>plant</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>plant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>sea water</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sea water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Coastal water body (ENVO:02000049)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:02000049</td>\n",
       "      <td>coastal water body</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Deep-sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>deep sea water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MIMARKS_16S</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mimarks 16s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Sea water</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sea water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>env package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        env_package  count lhs        extract  \\\n",
       "0   miscellaneous natural or artificial environment    611                      \n",
       "1                                           mimarks    479                      \n",
       "2                                         mouse-gut    406                      \n",
       "3                                               gut    172                      \n",
       "4                                           biofilm    114                      \n",
       "5                    human-not providedsopharyngeal    107                      \n",
       "6                                          mice gut     87                      \n",
       "7                                                CV     60                      \n",
       "8                         home, outdoor environment     44                      \n",
       "9                           fermentation-associated     42                      \n",
       "10                                   not applicable     41                      \n",
       "11                                  soil-associated     36                      \n",
       "12                                    sterile water     35                      \n",
       "13                                        Mouse_Gut     27                      \n",
       "14                                      fresh water     27                      \n",
       "15                                             rock     12                      \n",
       "16                                           arctic     10                      \n",
       "17                              freshwater sediment      9                      \n",
       "18                                         seawater      9                      \n",
       "19                                        MIENS_16S      6                      \n",
       "20                                    MIMARKS_water      6                      \n",
       "21                                             cave      5                      \n",
       "22                                            plant      3                      \n",
       "23                                        sea water      2                      \n",
       "24               Coastal water body (ENVO:02000049)      1      ENVO:02000049   \n",
       "25                                   Deep-sea water      1                      \n",
       "26                                      MIMARKS_16S      1                      \n",
       "27                                        Sea water      1                      \n",
       "28                                      env_package      1                      \n",
       "29                                   gut microbiome      1                      \n",
       "30                  marine sediment (ENVO:00002113)      1      ENVO:00002113   \n",
       "31                                           saliva      1                      \n",
       "32                       sea water, [ENVO:00002149]      1      ENVO:00002149   \n",
       "\n",
       "                                        ep_override  \n",
       "0   miscellaneous natural or artificial environment  \n",
       "1                                           mimarks  \n",
       "2                                         mouse gut  \n",
       "3                                               gut  \n",
       "4                                           biofilm  \n",
       "5                    human not providedsopharyngeal  \n",
       "6                                          mice gut  \n",
       "7                                                cv  \n",
       "8                          home outdoor environment  \n",
       "9                           fermentation associated  \n",
       "10                                   not applicable  \n",
       "11                                  soil associated  \n",
       "12                                    sterile water  \n",
       "13                                        mouse gut  \n",
       "14                                      fresh water  \n",
       "15                                             rock  \n",
       "16                                           arctic  \n",
       "17                              freshwater sediment  \n",
       "18                                         seawater  \n",
       "19                                        miens 16s  \n",
       "20                                    mimarks water  \n",
       "21                                             cave  \n",
       "22                                            plant  \n",
       "23                                        sea water  \n",
       "24                               coastal water body  \n",
       "25                                   deep sea water  \n",
       "26                                      mimarks 16s  \n",
       "27                                        sea water  \n",
       "28                                      env package  \n",
       "29                                   gut microbiome  \n",
       "30                                  marine sediment  \n",
       "31                                           saliva  \n",
       "32                                        sea water  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Are there any normalization failures?\n",
    "q = \"\"\"\n",
    "select\n",
    "    env_package,\n",
    "    count,\n",
    "    lhs,\n",
    "    extract,\n",
    "    ep_override\n",
    "from\n",
    "    env_package_normalized\n",
    "where\n",
    "    is_canonical = 0\n",
    "\"\"\"\n",
    "[normalizastion_failures, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "normalizastion_failures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c40903d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:10.028503\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:1006967</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:1041057</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:1046002</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:1046003</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:1046004</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>NCBITaxon:939928</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>NCBITaxon:941420</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>NCBITaxon:941421</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>NCBITaxon:941422</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>NCBITaxon:942017</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>994 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               subject  unclassified\n",
       "0    NCBITaxon:1006967          True\n",
       "1    NCBITaxon:1041057          True\n",
       "2    NCBITaxon:1046002          True\n",
       "3    NCBITaxon:1046003          True\n",
       "4    NCBITaxon:1046004          True\n",
       "..                 ...           ...\n",
       "989   NCBITaxon:939928          True\n",
       "990   NCBITaxon:941420          True\n",
       "991   NCBITaxon:941421          True\n",
       "992   NCBITaxon:941422          True\n",
       "993   NCBITaxon:942017          True\n",
       "\n",
       "[994 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# utilizing ncbtitaxon for broad subsetting\n",
    "# specifically, flag the biosamples whose taxon_id indicates they are an unclassified entity\n",
    "# ignoring the others will throw out samples OF multicellular organisms, like fruit flies\n",
    "# Add previous notes about what kinds of samples are missed by this bifurcation\n",
    "# like bacteria.unclassified_bacteria\n",
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    distinct s.subject\n",
    "from\n",
    "    entailed_edge ee\n",
    "join statements s on\n",
    "    ee.subject = s.subject\n",
    "where\n",
    "    ee.predicate = 'rdfs:subClassOf'\n",
    "    and ee.object = 'NCBITaxon:2787823'\n",
    "    and s.predicate = 'rdfs:label'\n",
    "\"\"\"\n",
    "[unclassified_taxa, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "unclassified_taxa['unclassified'] = True\n",
    "\n",
    "print(query_duration)\n",
    "\n",
    "unclassified_taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a5d8f028",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:01.352609\n"
     ]
    }
   ],
   "source": [
    "\n",
    "q = \"\"\"\n",
    "select\n",
    "    taxonomy_id biosample_taxid,\n",
    "    count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "group by\n",
    "    taxonomy_id\n",
    "order by\n",
    "    count(*) desc\n",
    "\"\"\"\n",
    "[biosample_tax_id_counts, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "biosample_tax_id_counts['curie'] = 'NCBITaxon:' + biosample_tax_id_counts['biosample_taxid'].astype(str)\n",
    "\n",
    "print(query_duration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b7ad29a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>curie</th>\n",
       "      <th>subject</th>\n",
       "      <th>unclassified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       biosample_taxid    count             curie           subject  \\\n",
       "0                 9606  6819707    NCBITaxon:9606               NaN   \n",
       "1                10090   964219   NCBITaxon:10090               NaN   \n",
       "2               408170   290862  NCBITaxon:408170  NCBITaxon:408170   \n",
       "3               410658   280666  NCBITaxon:410658  NCBITaxon:410658   \n",
       "4               646099   208741  NCBITaxon:646099  NCBITaxon:646099   \n",
       "...                ...      ...               ...               ...   \n",
       "163372          999891        1  NCBITaxon:999891               NaN   \n",
       "163373          999892        1  NCBITaxon:999892               NaN   \n",
       "163374          999898        1  NCBITaxon:999898               NaN   \n",
       "163375          999931        1  NCBITaxon:999931               NaN   \n",
       "163376          999947        1  NCBITaxon:999947               NaN   \n",
       "\n",
       "        unclassified  \n",
       "0              False  \n",
       "1              False  \n",
       "2               True  \n",
       "3               True  \n",
       "4               True  \n",
       "...              ...  \n",
       "163372         False  \n",
       "163373         False  \n",
       "163374         False  \n",
       "163375         False  \n",
       "163376         False  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the two taxon id datasets\n",
    "# I.e. flag the the Biosample records whose taxonomy_id field belongs to a subclass of 'unclassified entries'.\n",
    "biosample_tax_id_counts = biosample_tax_id_counts.merge(unclassified_taxa, left_on='curie',\n",
    "                                                        right_on='subject', how='left')\n",
    "biosample_tax_id_counts.unclassified.fillna(False, inplace=True)\n",
    "\n",
    "biosample_tax_id_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a6acc349",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0:00:05.466133\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>curie</th>\n",
       "      <th>biosample_taxid</th>\n",
       "      <th>count</th>\n",
       "      <th>unclassified</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>6819707</td>\n",
       "      <td>False</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>10090</td>\n",
       "      <td>964219</td>\n",
       "      <td>False</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCBITaxon:408170</td>\n",
       "      <td>408170</td>\n",
       "      <td>290862</td>\n",
       "      <td>True</td>\n",
       "      <td>human gut metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCBITaxon:410658</td>\n",
       "      <td>410658</td>\n",
       "      <td>280666</td>\n",
       "      <td>True</td>\n",
       "      <td>soil metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCBITaxon:646099</td>\n",
       "      <td>646099</td>\n",
       "      <td>208741</td>\n",
       "      <td>True</td>\n",
       "      <td>human metagenome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163372</th>\n",
       "      <td>NCBITaxon:999891</td>\n",
       "      <td>999891</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Bacillus amyloliquefaciens TA208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163373</th>\n",
       "      <td>NCBITaxon:999892</td>\n",
       "      <td>999892</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>[Propionibacterium] humerusii P08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163374</th>\n",
       "      <td>NCBITaxon:999898</td>\n",
       "      <td>999898</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Peptococcaceae bacterium CEB3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163375</th>\n",
       "      <td>NCBITaxon:999931</td>\n",
       "      <td>999931</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Barrientosiimonas humi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163376</th>\n",
       "      <td>NCBITaxon:999947</td>\n",
       "      <td>999947</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>Angomonas ambiguus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163377 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   curie biosample_taxid    count  unclassified  \\\n",
       "0         NCBITaxon:9606            9606  6819707         False   \n",
       "1        NCBITaxon:10090           10090   964219         False   \n",
       "2       NCBITaxon:408170          408170   290862          True   \n",
       "3       NCBITaxon:410658          410658   280666          True   \n",
       "4       NCBITaxon:646099          646099   208741          True   \n",
       "...                  ...             ...      ...           ...   \n",
       "163372  NCBITaxon:999891          999891        1         False   \n",
       "163373  NCBITaxon:999892          999892        1         False   \n",
       "163374  NCBITaxon:999898          999898        1         False   \n",
       "163375  NCBITaxon:999931          999931        1         False   \n",
       "163376  NCBITaxon:999947          999947        1         False   \n",
       "\n",
       "                                    label  \n",
       "0                            Homo sapiens  \n",
       "1                            Mus musculus  \n",
       "2                    human gut metagenome  \n",
       "3                         soil metagenome  \n",
       "4                        human metagenome  \n",
       "...                                   ...  \n",
       "163372   Bacillus amyloliquefaciens TA208  \n",
       "163373  [Propionibacterium] humerusii P08  \n",
       "163374      Peptococcaceae bacterium CEB3  \n",
       "163375             Barrientosiimonas humi  \n",
       "163376                 Angomonas ambiguus  \n",
       "\n",
       "[163377 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# should really add labels to all of them\n",
    "q = \"\"\"\n",
    "select\n",
    "    subject ,\n",
    "    value\n",
    "from statements\n",
    "where\n",
    "    predicate = 'rdfs:label' and subject = stanza\n",
    "\"\"\"\n",
    "[all_tax_labels, query_duration] = scoped_mapping.timed_query(q, ncbitaxon_cnx)\n",
    "\n",
    "biosample_tax_id_counts = biosample_tax_id_counts.merge(all_tax_labels, left_on='curie',\n",
    "                                                        right_on='subject', how='left')\n",
    "\n",
    "biosample_tax_id_counts = biosample_tax_id_counts[['curie', 'biosample_taxid', 'count', 'unclassified', 'value']]\n",
    "biosample_tax_id_counts.columns = ['curie', 'biosample_taxid', 'count', 'unclassified', 'label']\n",
    "\n",
    "print(query_duration)\n",
    "biosample_tax_id_counts.to_sql('biobiosample_tax_id_counts', biosample_cnx, if_exists='replace', index=False)\n",
    "\n",
    "biosample_tax_id_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "554a82c0",
   "metadata": {},
   "source": [
    "Almost all of the taxa that are common in the biosample collection are either unclassified/metagenomes or easily recognized cellular organisms\n",
    "\n",
    "exceptions include:\n",
    "- 32630 = synthetic construct (other entries; other sequences; artificial sequences)\n",
    "    - 'other entries' would add 16k rows on top of the 1k 'unclassified entities'\n",
    "    - metagenomes account for 331 of the 'unclassified entities'\n",
    "    - there are also a small number of uncultured/unclassified microorganisms in the biosample dataset\n",
    "- 77133 = uncultured bacterium (cellular organisms; Bacteria; environmental samples)\n",
    "    - 'cellular organisms' would add 2M rows on top of the 1k 'unclassified entities'\n",
    "    - 'cellular organisms; Bacteria; environmental samples' adds 26k\n",
    "    \n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7e777c8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>257 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count\n",
       "0                small lake biome   3264\n",
       "1    marine biome (ENVO:00000447)   1382\n",
       "2                    marine biome   1355\n",
       "3                large lake biome   1198\n",
       "4                freshwater biome   1051\n",
       "..                            ...    ...\n",
       "252                        Arctic      1\n",
       "253                 Aquatic biome      1\n",
       "254                      01000035      1\n",
       "255                      00000891      1\n",
       "256                             0      1\n",
       "\n",
       "[257 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get a table of scoped mixs annotations to be mapped to ontology classes.\n",
    "biosample_col_to_map = 'env_broad_scale'\n",
    "scoping_col = 'env_package_normalized.ep_override'\n",
    "scoping_value = 'water'\n",
    "# In this case, the scoping includes an inner join requirement for 'unclassified entities'\n",
    "\n",
    "q = 'select ' + biosample_col_to_map + \"\"\", count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "join env_package_normalized on\n",
    "    b.env_package = env_package_normalized.env_package\n",
    "inner join biobiosample_tax_id_counts stic on\n",
    "    b.taxonomy_id = stic.biosample_taxid\n",
    "where \"\"\" + scoping_col + \" = '\" + scoping_value + \\\n",
    "    \"' group by \" + biosample_col_to_map + \"\"\"\n",
    "order by\n",
    "    count(*) desc\"\"\"\n",
    "[mapping_candidates, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "45e1859f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted  \n",
       "0                small lake biome  \n",
       "1    marine biome (ENVO:00000447)  \n",
       "2                    marine biome  \n",
       "3                large lake biome  \n",
       "4                freshwater biome  \n",
       "..                            ...  \n",
       "258                        Arctic  \n",
       "259                 Aquatic biome  \n",
       "260                      01000035  \n",
       "261                      00000891  \n",
       "262                             0  \n",
       "\n",
       "[263 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The Biosample format allows for pipe-delimited environmental package lists\n",
    "# Separate those out into their components\n",
    "multi_frames = []\n",
    "for row in mapping_candidates.itertuples(index=True, name='Pandas'):\n",
    "    split_check = row.env_broad_scale\n",
    "    if split_check is None:\n",
    "        split_check = ''\n",
    "    splitted = pd.Series(split_check.split(\"|\"))\n",
    "    splitted_count = len(splitted)\n",
    "    repeated = [split_check] * splitted_count\n",
    "    repeated = pd.Series(repeated)\n",
    "    as_frame = pd.DataFrame(dict(repeated=repeated, splitted=splitted)).reset_index()\n",
    "    multi_frames.append(as_frame)\n",
    "concat_frame = pd.concat(multi_frames)\n",
    "concat_frame = concat_frame[['repeated', 'splitted']]\n",
    "mapping_candidates = mapping_candidates.merge(concat_frame, left_on=biosample_col_to_map,\n",
    "                                              right_on='repeated', how='left')\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e423fcb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                        string  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "           extract  remaining_string  remaining_tidied  \n",
       "0                   small lake biome  small lake biome  \n",
       "1    ENVO:00000447   marine biome ()      marine biome  \n",
       "2                       marine biome      marine biome  \n",
       "3                   large lake biome  large lake biome  \n",
       "4                   freshwater biome  freshwater biome  \n",
       "..             ...               ...               ...  \n",
       "258                           Arctic            arctic  \n",
       "259                    Aquatic biome     aquatic biome  \n",
       "260                         01000035          01000035  \n",
       "261                         00000891          00000891  \n",
       "262                                0                 0  \n",
       "\n",
       "[263 rows x 8 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# do the spliting and extraction here\n",
    "\n",
    "# Now try to extract ontology terms that are already present\n",
    "candidate_series_decomposition = scoped_mapping.decompose_series(mapping_candidates['splitted'], id_patterns[target_onto_prefix])\n",
    "mapping_candidates = pd.concat([mapping_candidates, candidate_series_decomposition], axis=1)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "448865ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# And join the extracted IDs with their labels\n",
    "# start by conencting to the rdftab database \n",
    "# from which the terms and label-like annotatiosn will be obtained\n",
    "ontodb = '../semantic-sql/db/' + target_onto_prefix.lower() + '.db'\n",
    "ontocon = sqlite3.connect(ontodb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3bcd7e47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IAO:0000111</td>\n",
       "      <td>editor preferred term~editor preferred label</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IAO:0000112</td>\n",
       "      <td>example of usage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IAO:0000114</td>\n",
       "      <td>has curation status</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IAO:0000115</td>\n",
       "      <td>definition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IAO:0000116</td>\n",
       "      <td>editor note</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6774</th>\n",
       "      <td>ENVO:01001862</td>\n",
       "      <td>Solar radiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6775</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q2&gt;</td>\n",
       "      <td>Earth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6776</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q2306597&gt;</td>\n",
       "      <td>Suni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6777</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q525&gt;</td>\n",
       "      <td>Sol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6778</th>\n",
       "      <td>&lt;https://www.wikidata.org/wiki/Q715269&gt;</td>\n",
       "      <td>World Ocean</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6779 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       subject  \\\n",
       "0                                  IAO:0000111   \n",
       "1                                  IAO:0000112   \n",
       "2                                  IAO:0000114   \n",
       "3                                  IAO:0000115   \n",
       "4                                  IAO:0000116   \n",
       "...                                        ...   \n",
       "6774                             ENVO:01001862   \n",
       "6775        <https://www.wikidata.org/wiki/Q2>   \n",
       "6776  <https://www.wikidata.org/wiki/Q2306597>   \n",
       "6777      <https://www.wikidata.org/wiki/Q525>   \n",
       "6778   <https://www.wikidata.org/wiki/Q715269>   \n",
       "\n",
       "                                             value  \n",
       "0     editor preferred term~editor preferred label  \n",
       "1                                 example of usage  \n",
       "2                              has curation status  \n",
       "3                                       definition  \n",
       "4                                      editor note  \n",
       "...                                            ...  \n",
       "6774                               Solar radiation  \n",
       "6775                                         Earth  \n",
       "6776                                          Suni  \n",
       "6777                                           Sol  \n",
       "6778                                   World Ocean  \n",
       "\n",
       "[6779 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "select\n",
    "    subject ,\n",
    "    value\n",
    "from\n",
    "    statements s\n",
    "where\n",
    "    predicate = 'rdfs:label'\n",
    "\"\"\"\n",
    "[onto_labels, query_duration] = scoped_mapping.timed_query(q, ontocon)\n",
    "\n",
    "onto_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8543ed0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                        string  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "           extract  remaining_string  remaining_tidied        subject  \\\n",
       "0                   small lake biome  small lake biome            NaN   \n",
       "1    ENVO:00000447   marine biome ()      marine biome  ENVO:00000447   \n",
       "2                       marine biome      marine biome            NaN   \n",
       "3                   large lake biome  large lake biome            NaN   \n",
       "4                   freshwater biome  freshwater biome            NaN   \n",
       "..             ...               ...               ...            ...   \n",
       "258                           Arctic            arctic            NaN   \n",
       "259                    Aquatic biome     aquatic biome            NaN   \n",
       "260                         01000035          01000035            NaN   \n",
       "261                         00000891          00000891            NaN   \n",
       "262                                0                 0            NaN   \n",
       "\n",
       "            value  \n",
       "0             NaN  \n",
       "1    marine biome  \n",
       "2             NaN  \n",
       "3             NaN  \n",
       "4             NaN  \n",
       "..            ...  \n",
       "258           NaN  \n",
       "259           NaN  \n",
       "260           NaN  \n",
       "261           NaN  \n",
       "262           NaN  \n",
       "\n",
       "[263 rows x 10 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(onto_labels, left_on='extract', right_on='subject', how='left')\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "abd0a035",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "      <th>cosine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                        string  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "           extract  remaining_string  remaining_tidied        subject  \\\n",
       "0                   small lake biome  small lake biome            NaN   \n",
       "1    ENVO:00000447   marine biome ()      marine biome  ENVO:00000447   \n",
       "2                       marine biome      marine biome            NaN   \n",
       "3                   large lake biome  large lake biome            NaN   \n",
       "4                   freshwater biome  freshwater biome            NaN   \n",
       "..             ...               ...               ...            ...   \n",
       "258                           Arctic            arctic            NaN   \n",
       "259                    Aquatic biome     aquatic biome            NaN   \n",
       "260                         01000035          01000035            NaN   \n",
       "261                         00000891          00000891            NaN   \n",
       "262                                0                 0            NaN   \n",
       "\n",
       "            value  cosine  \n",
       "0                     1.0  \n",
       "1    marine biome     0.0  \n",
       "2                     1.0  \n",
       "3                     1.0  \n",
       "4                     1.0  \n",
       "..            ...     ...  \n",
       "258                   1.0  \n",
       "259                   1.0  \n",
       "260                   1.0  \n",
       "261                   1.0  \n",
       "262                   1.0  \n",
       "\n",
       "[263 rows x 11 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use cosine string distance to see if the labels match\n",
    "# I.e. the labels claimed by the Biosample data set and the labels asserted in the ontology\n",
    "# if they're close enough, consider the assigned ID legit\n",
    "# how close is close enough?\n",
    "my_cosine_obj = Cosine(1)\n",
    "mapping_candidates['value'] = mapping_candidates['value'].fillna('')\n",
    "mapping_candidates['cosine'] = mapping_candidates.apply(\n",
    "    lambda my_row: my_cosine_obj.distance(my_row['remaining_tidied'].lower(), my_row['value'].lower()), axis=1)\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "092aced0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000873</td>\n",
       "      <td>freshwater biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00002030</td>\n",
       "      <td>aquatic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                        string  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "           extract  remaining_string  remaining_tidied        term_id  \\\n",
       "0                   small lake biome  small lake biome            NaN   \n",
       "1    ENVO:00000447   marine biome ()      marine biome  ENVO:00000447   \n",
       "2                       marine biome      marine biome            NaN   \n",
       "3                   large lake biome  large lake biome            NaN   \n",
       "4                   freshwater biome  freshwater biome            NaN   \n",
       "..             ...               ...               ...            ...   \n",
       "258                           Arctic            arctic            NaN   \n",
       "259                    Aquatic biome     aquatic biome            NaN   \n",
       "260                         01000035          01000035            NaN   \n",
       "261                         00000891          00000891            NaN   \n",
       "262                                0                 0            NaN   \n",
       "\n",
       "      lab_from_id  lfi_cosine term_id_from_lab             value  \n",
       "0                         1.0              NaN               NaN  \n",
       "1    marine biome         0.0    ENVO:00000447      marine biome  \n",
       "2                         1.0    ENVO:00000447      marine biome  \n",
       "3                         1.0              NaN               NaN  \n",
       "4                         1.0    ENVO:00000873  freshwater biome  \n",
       "..            ...         ...              ...               ...  \n",
       "258                       1.0              NaN               NaN  \n",
       "259                       1.0    ENVO:00002030     aquatic biome  \n",
       "260                       1.0              NaN               NaN  \n",
       "261                       1.0              NaN               NaN  \n",
       "262                       1.0              NaN               NaN  \n",
       "\n",
       "[263 rows x 13 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get ready to join in the other direction\n",
    "# I.e. trying to find ontology term IDs based on perfect label matches. Be careful not to reuse column names.\n",
    "mapping_candidates.columns = ['env_broad_scale', 'count', 'repeated', 'splitted', 'string', 'extract',\n",
    "                              'remaining_string', 'remaining_tidied', 'term_id', 'lab_from_id', 'lfi_cosine']\n",
    "mapping_candidates = mapping_candidates.merge(onto_labels, left_on='remaining_tidied', right_on='value', how='left')\n",
    "mapping_candidates.columns = ['env_broad_scale', 'count', 'repeated', 'splitted', 'string', 'extract',\n",
    "                              'remaining_string', 'remaining_tidied', 'term_id', 'lab_from_id',\n",
    "                              'lfi_cosine', 'term_id_from_lab', 'value']\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4f5e3fdd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-27-1875bdd846cc>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mapping_candidates['consensus_id'][mapping_candidates['consensus_id'].isnull()] = \\\n",
      "<ipython-input-27-1875bdd846cc>:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mapping_candidates['consensus_lab'][mapping_candidates['consensus_lab'] == ''] = \\\n"
     ]
    }
   ],
   "source": [
    "# Record a consensus\n",
    "# If either merging on codes or labels was successful.\n",
    "# cosines for first pass check on assigned IDs still haven't been filtered?\n",
    "mapping_candidates['consensus_id'] = mapping_candidates['term_id']\n",
    "mapping_candidates['consensus_id'][mapping_candidates['consensus_id'].isnull()] = \\\n",
    "    mapping_candidates['term_id_from_lab'][mapping_candidates['consensus_id'].isnull()]\n",
    "mapping_candidates['consensus_lab'] = mapping_candidates['lab_from_id']\n",
    "mapping_candidates['consensus_lab'][mapping_candidates['consensus_lab'] == ''] = \\\n",
    "    mapping_candidates['value'][mapping_candidates['consensus_lab'] == '']\n",
    "# mapping_candidates.to_sql('mapping_scratch', biosample_cnx, if_exists='replace', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1c7644d",
   "metadata": {},
   "source": [
    "```\n",
    "<ipython-input-49-3e62557cf6d9>:5: SettingWithCopyWarning: \n",
    "A value is trying to be set on a copy of a slice from a DataFrame\n",
    "\n",
    "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
    "  mapping_candidates['consensus_id'][mapping_candidates['consensus_id'].isnull()] = \\\n",
    "<ipython-input-49-3e62557cf6d9>:8: SettingWithCopyWarning: \n",
    "A value is trying to be set on a copy of a slice from a DataFrame\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6af120f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_broad_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>3264</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td></td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>1382</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>marine biome (ENVO:00000447)</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome ()</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>marine biome</td>\n",
       "      <td>1355</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td></td>\n",
       "      <td>marine biome</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "      <td>ENVO:00000447</td>\n",
       "      <td>marine biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>large lake biome</td>\n",
       "      <td>1198</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td></td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>large lake biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>1051</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td></td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000873</td>\n",
       "      <td>freshwater biome</td>\n",
       "      <td>ENVO:00000873</td>\n",
       "      <td>freshwater biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Arctic</td>\n",
       "      <td>1</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td>Arctic</td>\n",
       "      <td></td>\n",
       "      <td>Arctic</td>\n",
       "      <td>arctic</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>1</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td></td>\n",
       "      <td>Aquatic biome</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00002030</td>\n",
       "      <td>aquatic biome</td>\n",
       "      <td>ENVO:00002030</td>\n",
       "      <td>aquatic biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>01000035</td>\n",
       "      <td>1</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td></td>\n",
       "      <td>01000035</td>\n",
       "      <td>01000035</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>00000891</td>\n",
       "      <td>1</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td></td>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>263 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  env_broad_scale  count                      repeated  \\\n",
       "0                small lake biome   3264              small lake biome   \n",
       "1    marine biome (ENVO:00000447)   1382  marine biome (ENVO:00000447)   \n",
       "2                    marine biome   1355                  marine biome   \n",
       "3                large lake biome   1198              large lake biome   \n",
       "4                freshwater biome   1051              freshwater biome   \n",
       "..                            ...    ...                           ...   \n",
       "258                        Arctic      1                        Arctic   \n",
       "259                 Aquatic biome      1                 Aquatic biome   \n",
       "260                      01000035      1                      01000035   \n",
       "261                      00000891      1                      00000891   \n",
       "262                             0      1                             0   \n",
       "\n",
       "                         splitted                        string  \\\n",
       "0                small lake biome              small lake biome   \n",
       "1    marine biome (ENVO:00000447)  marine biome (ENVO:00000447)   \n",
       "2                    marine biome                  marine biome   \n",
       "3                large lake biome              large lake biome   \n",
       "4                freshwater biome              freshwater biome   \n",
       "..                            ...                           ...   \n",
       "258                        Arctic                        Arctic   \n",
       "259                 Aquatic biome                 Aquatic biome   \n",
       "260                      01000035                      01000035   \n",
       "261                      00000891                      00000891   \n",
       "262                             0                             0   \n",
       "\n",
       "           extract  remaining_string  remaining_tidied        term_id  \\\n",
       "0                   small lake biome  small lake biome            NaN   \n",
       "1    ENVO:00000447   marine biome ()      marine biome  ENVO:00000447   \n",
       "2                       marine biome      marine biome            NaN   \n",
       "3                   large lake biome  large lake biome            NaN   \n",
       "4                   freshwater biome  freshwater biome            NaN   \n",
       "..             ...               ...               ...            ...   \n",
       "258                           Arctic            arctic            NaN   \n",
       "259                    Aquatic biome     aquatic biome            NaN   \n",
       "260                         01000035          01000035            NaN   \n",
       "261                         00000891          00000891            NaN   \n",
       "262                                0                 0            NaN   \n",
       "\n",
       "      lab_from_id  lfi_cosine term_id_from_lab             value  \\\n",
       "0                         1.0              NaN               NaN   \n",
       "1    marine biome         0.0    ENVO:00000447      marine biome   \n",
       "2                         1.0    ENVO:00000447      marine biome   \n",
       "3                         1.0              NaN               NaN   \n",
       "4                         1.0    ENVO:00000873  freshwater biome   \n",
       "..            ...         ...              ...               ...   \n",
       "258                       1.0              NaN               NaN   \n",
       "259                       1.0    ENVO:00002030     aquatic biome   \n",
       "260                       1.0              NaN               NaN   \n",
       "261                       1.0              NaN               NaN   \n",
       "262                       1.0              NaN               NaN   \n",
       "\n",
       "      consensus_id     consensus_lab  \n",
       "0              NaN               NaN  \n",
       "1    ENVO:00000447      marine biome  \n",
       "2    ENVO:00000447      marine biome  \n",
       "3              NaN               NaN  \n",
       "4    ENVO:00000873  freshwater biome  \n",
       "..             ...               ...  \n",
       "258            NaN               NaN  \n",
       "259  ENVO:00002030     aquatic biome  \n",
       "260            NaN               NaN  \n",
       "261            NaN               NaN  \n",
       "262            NaN               NaN  \n",
       "\n",
       "[263 rows x 15 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c29338b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                 3\n",
       "polar biome envo 01000339        3\n",
       "small lake biome                 2\n",
       "atlantic ocean                   2\n",
       "clams fluids                     2\n",
       "                                ..\n",
       "00000891                         1\n",
       "temperate shelf and sea biome    1\n",
       "cyanobacterial culture           1\n",
       "deep mediterranean               1\n",
       "protist                          1\n",
       "Name: remaining_tidied, Length: 131, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For which Biosample annotations were not mappings by merging found?\n",
    "# It looks like remaining_tidied is retaining too much punctuation\n",
    "# and loosing useful digits (relative to remaining_string)?\n",
    "# Should try harder to parse not-quite-right embedded IDs like ...\n",
    "needs_search = mapping_candidates.remaining_tidied[mapping_candidates.consensus_id.isna()]\n",
    "needs_search_counts = needs_search.value_counts()\n",
    "\n",
    "needs_search_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "02a43f7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " 'polar biome envo 01000339',\n",
       " 'small lake biome',\n",
       " 'atlantic ocean',\n",
       " 'clams fluids',\n",
       " 'envo 01000023',\n",
       " 'marine',\n",
       " 'seawater',\n",
       " 'freshwater',\n",
       " 'marine pelagic',\n",
       " 'oceanic',\n",
       " 'shrmip culture water',\n",
       " 'water river',\n",
       " 'phragmites australis rhizosphere',\n",
       " 'lake pond',\n",
       " 'temperate glacier close to sea level',\n",
       " 'temperate mediterannean sea biome',\n",
       " 'deep near bottom layer of freshwater lake',\n",
       " 'deep water coral reef',\n",
       " 'marine biome coastal',\n",
       " 'hanford h 100',\n",
       " 'surface marine water',\n",
       " 'saltwater',\n",
       " 'coastal marine estuary',\n",
       " 'plankton community',\n",
       " 'envo 01000043',\n",
       " 'env 00000447',\n",
       " 'brackish water river',\n",
       " 'ocean this term applies to plant growth forms',\n",
       " 'coastal sea area',\n",
       " 'snow transect pooled',\n",
       " 'sponge symbiont',\n",
       " 'tundra',\n",
       " 'arctic ocean associated',\n",
       " 'arabian sea',\n",
       " 'marine biome 00000447',\n",
       " 'envo mediterranean forests woodlands and shrub biome',\n",
       " 'supraglacial lake water',\n",
       " 'microorganism',\n",
       " 'mediterranean sea',\n",
       " 'large lake biome',\n",
       " 'envo montane grasslands and shrubland biome',\n",
       " 'euxinic lake',\n",
       " 'deep sea marine',\n",
       " 'ocean water',\n",
       " 'na',\n",
       " 'atolls of the maldives',\n",
       " 'gulf of eilat',\n",
       " 'enriched culture in triptone casein soja medium',\n",
       " 'oxygen deficient zone',\n",
       " 'surface seawater',\n",
       " 'eukaryotes',\n",
       " '0',\n",
       " 'coastal',\n",
       " '15',\n",
       " 'terrestrial hot spring',\n",
       " 'boreal forest',\n",
       " 'meromictic lakes are unusual hydrological formations where bodies of water follow a seasonal cycle of stratifi cation but do not mix completely',\n",
       " 'shrimp culture water',\n",
       " 'marginal sea',\n",
       " 'ne atlantic',\n",
       " 'arctic',\n",
       " 'marine near barrier island',\n",
       " 'eutrophic pond',\n",
       " 'aquatic',\n",
       " 'zooplankton',\n",
       " 'deep sea',\n",
       " 'bacterioplankton',\n",
       " 'antarctic coastal associated',\n",
       " 'plastic marine debris',\n",
       " 'freshwater great lakes',\n",
       " 'subtropical coastal water',\n",
       " 'outer slope of the reef',\n",
       " '01000035',\n",
       " 'large river',\n",
       " 'mixed layer surface ocean',\n",
       " 'hydrothermal vents',\n",
       " 'test field amp kinds of river',\n",
       " 'brine pool interface layer',\n",
       " 'water from coastal marshes',\n",
       " 'marine habitat',\n",
       " 'drinking water distribution system',\n",
       " 'diatom culture',\n",
       " 'this is a filtered seawater sample',\n",
       " 'pacific coast',\n",
       " 'produced water',\n",
       " 'envo marine biome',\n",
       " 'southern ocean',\n",
       " 'gulf',\n",
       " 'costal lagoon periodically connected to the atlantic ocean',\n",
       " 'envo 01000047',\n",
       " 'clean river',\n",
       " 'hanford h 101',\n",
       " 'process water',\n",
       " 'deep ocean',\n",
       " 'managed aquifer recharge system',\n",
       " 'marine water',\n",
       " 'estuarine plume',\n",
       " 'oil deposit',\n",
       " 'microbial mats',\n",
       " 'marine brine pool basin',\n",
       " 'marine metagenome',\n",
       " 'forest',\n",
       " 'coastal ocean',\n",
       " 'sandstone aquifer',\n",
       " 'estuarine',\n",
       " 'westerlies biome',\n",
       " 'hot springs',\n",
       " 'costal lagoon periodicay coneccted to the atlantic ocean',\n",
       " 'soil and water environments',\n",
       " '1000686',\n",
       " 'deep mediterranean thetis halocline',\n",
       " 'coastal water',\n",
       " 'antarctic peninsula associated',\n",
       " 'not applicable',\n",
       " 'freshwater reservoir',\n",
       " 'ocean temperate oligotrophic',\n",
       " 'freshwater biofilm',\n",
       " 'sea plankton',\n",
       " 'caribbean sea',\n",
       " 'coral lagoon',\n",
       " 'in situ sea water',\n",
       " 'aquatic bacteria',\n",
       " 'brackish amp fresh water mixed',\n",
       " 'oligotrophic',\n",
       " 'ocean beach',\n",
       " '00000891',\n",
       " 'temperate shelf and sea biome',\n",
       " 'cyanobacterial culture',\n",
       " 'deep mediterranean',\n",
       " 'protist']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use a search engine\n",
    "# For the mixs annotations that didn't already have cannonical IDs or labels\n",
    "ebs_raw_list = list(needs_search_counts.index)\n",
    "ebs_raw_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d9c944a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>raw</th>\n",
       "      <th>woed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>polar biome envo 01000339</th>\n",
       "      <td>polar biome envo 01000339</td>\n",
       "      <td>polar biome envo 01000339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>small lake biome</th>\n",
       "      <td>small lake biome</td>\n",
       "      <td>small lake biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atlantic ocean</th>\n",
       "      <td>atlantic ocean</td>\n",
       "      <td>atlantic ocean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clams fluids</th>\n",
       "      <td>clams fluids</td>\n",
       "      <td>clams fluids</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>00000891</th>\n",
       "      <td>00000891</td>\n",
       "      <td>00000891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>temperate shelf and sea biome</th>\n",
       "      <td>temperate shelf and sea biome</td>\n",
       "      <td>temperate shelf and sea biome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cyanobacterial culture</th>\n",
       "      <td>cyanobacterial culture</td>\n",
       "      <td>cyanobacterial culture</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>deep mediterranean</th>\n",
       "      <td>deep mediterranean</td>\n",
       "      <td>deep mediterranean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protist</th>\n",
       "      <td>protist</td>\n",
       "      <td>protist</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>131 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         raw  \\\n",
       "                                                               \n",
       "polar biome envo 01000339          polar biome envo 01000339   \n",
       "small lake biome                            small lake biome   \n",
       "atlantic ocean                                atlantic ocean   \n",
       "clams fluids                                    clams fluids   \n",
       "...                                                      ...   \n",
       "00000891                                            00000891   \n",
       "temperate shelf and sea biome  temperate shelf and sea biome   \n",
       "cyanobacterial culture                cyanobacterial culture   \n",
       "deep mediterranean                        deep mediterranean   \n",
       "protist                                              protist   \n",
       "\n",
       "                                                        woed  \n",
       "                                                              \n",
       "polar biome envo 01000339          polar biome envo 01000339  \n",
       "small lake biome                            small lake biome  \n",
       "atlantic ocean                                atlantic ocean  \n",
       "clams fluids                                    clams fluids  \n",
       "...                                                      ...  \n",
       "00000891                                            00000891  \n",
       "temperate shelf and sea biome  temperate shelf and sea biome  \n",
       "cyanobacterial culture                cyanobacterial culture  \n",
       "deep mediterranean                        deep mediterranean  \n",
       "protist                                              protist  \n",
       "\n",
       "[131 rows x 2 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get whiteout frame and relateds\n",
    "ebs_wo_frame = scoped_mapping.get_whiteout_frame(ebs_raw_list, replaced_chars=chars_to_whiteout)\n",
    "\n",
    "ebs_wo_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e0c0f419",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['',\n",
       " '0',\n",
       " '00000891',\n",
       " '01000035',\n",
       " '1000686',\n",
       " '15',\n",
       " 'antarctic coastal associated',\n",
       " 'antarctic peninsula associated',\n",
       " 'aquatic',\n",
       " 'aquatic bacteria',\n",
       " 'arabian sea',\n",
       " 'arctic',\n",
       " 'arctic ocean associated',\n",
       " 'atlantic ocean',\n",
       " 'atolls of the maldives',\n",
       " 'bacterioplankton',\n",
       " 'boreal forest',\n",
       " 'brackish amp fresh water mixed',\n",
       " 'brackish water river',\n",
       " 'brine pool interface layer',\n",
       " 'caribbean sea',\n",
       " 'clams fluids',\n",
       " 'clean river',\n",
       " 'coastal',\n",
       " 'coastal marine estuary',\n",
       " 'coastal ocean',\n",
       " 'coastal sea area',\n",
       " 'coastal water',\n",
       " 'coral lagoon',\n",
       " 'costal lagoon periodically connected to the atlantic ocean',\n",
       " 'costal lagoon periodicay coneccted to the atlantic ocean',\n",
       " 'cyanobacterial culture',\n",
       " 'deep mediterranean',\n",
       " 'deep mediterranean thetis halocline',\n",
       " 'deep near bottom layer of freshwater lake',\n",
       " 'deep ocean',\n",
       " 'deep sea',\n",
       " 'deep sea marine',\n",
       " 'deep water coral reef',\n",
       " 'diatom culture',\n",
       " 'drinking water distribution system',\n",
       " 'enriched culture in triptone casein soja medium',\n",
       " 'env 00000447',\n",
       " 'envo 01000023',\n",
       " 'envo 01000043',\n",
       " 'envo 01000047',\n",
       " 'envo marine biome',\n",
       " 'envo mediterranean forests woodlands and shrub biome',\n",
       " 'envo montane grasslands and shrubland biome',\n",
       " 'estuarine',\n",
       " 'estuarine plume',\n",
       " 'eukaryotes',\n",
       " 'eutrophic pond',\n",
       " 'euxinic lake',\n",
       " 'forest',\n",
       " 'freshwater',\n",
       " 'freshwater biofilm',\n",
       " 'freshwater great lakes',\n",
       " 'freshwater reservoir',\n",
       " 'gulf',\n",
       " 'gulf of eilat',\n",
       " 'hanford h 100',\n",
       " 'hanford h 101',\n",
       " 'hot springs',\n",
       " 'hydrothermal vents',\n",
       " 'in situ sea water',\n",
       " 'lake pond',\n",
       " 'large lake biome',\n",
       " 'large river',\n",
       " 'managed aquifer recharge system',\n",
       " 'marginal sea',\n",
       " 'marine',\n",
       " 'marine biome 00000447',\n",
       " 'marine biome coastal',\n",
       " 'marine brine pool basin',\n",
       " 'marine habitat',\n",
       " 'marine metagenome',\n",
       " 'marine near barrier island',\n",
       " 'marine pelagic',\n",
       " 'marine water',\n",
       " 'mediterranean sea',\n",
       " 'meromictic lakes are unusual hydrological formations where bodies of water follow a seasonal cycle of stratifi cation but do not mix completely',\n",
       " 'microbial mats',\n",
       " 'microorganism',\n",
       " 'mixed layer surface ocean',\n",
       " 'na',\n",
       " 'ne atlantic',\n",
       " 'not applicable',\n",
       " 'ocean beach',\n",
       " 'ocean temperate oligotrophic',\n",
       " 'ocean this term applies to plant growth forms',\n",
       " 'ocean water',\n",
       " 'oceanic',\n",
       " 'oil deposit',\n",
       " 'oligotrophic',\n",
       " 'outer slope of the reef',\n",
       " 'oxygen deficient zone',\n",
       " 'pacific coast',\n",
       " 'phragmites australis rhizosphere',\n",
       " 'plankton community',\n",
       " 'plastic marine debris',\n",
       " 'polar biome envo 01000339',\n",
       " 'process water',\n",
       " 'produced water',\n",
       " 'protist',\n",
       " 'saltwater',\n",
       " 'sandstone aquifer',\n",
       " 'sea plankton',\n",
       " 'seawater',\n",
       " 'shrimp culture water',\n",
       " 'shrmip culture water',\n",
       " 'small lake biome',\n",
       " 'snow transect pooled',\n",
       " 'soil and water environments',\n",
       " 'southern ocean',\n",
       " 'sponge symbiont',\n",
       " 'subtropical coastal water',\n",
       " 'supraglacial lake water',\n",
       " 'surface marine water',\n",
       " 'surface seawater',\n",
       " 'temperate glacier close to sea level',\n",
       " 'temperate mediterannean sea biome',\n",
       " 'temperate shelf and sea biome',\n",
       " 'terrestrial hot spring',\n",
       " 'test field amp kinds of river',\n",
       " 'this is a filtered seawater sample',\n",
       " 'tundra',\n",
       " 'water from coastal marshes',\n",
       " 'water river',\n",
       " 'westerlies biome',\n",
       " 'zooplankton']"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ebs_wo_list = scoped_mapping.get_wo_list(ebs_wo_frame)\n",
    "ebs_wo_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "08e39629",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>cosine_rank</th>\n",
       "      <th>cosine_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>boreal forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01000250</td>\n",
       "      <td>subpolar coniferous forest biome</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000250</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>coastal marine estuary</td>\n",
       "      <td>coastal marine estuary</td>\n",
       "      <td>estuarine coastal surface layer</td>\n",
       "      <td>1</td>\n",
       "      <td>0.041</td>\n",
       "      <td>ENVO:01001302</td>\n",
       "      <td>estuarine coastal surface layer</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001302</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>coastal sea area</td>\n",
       "      <td>coastal sea area</td>\n",
       "      <td>coastal area</td>\n",
       "      <td>1</td>\n",
       "      <td>0.024</td>\n",
       "      <td>ENVO:00000303</td>\n",
       "      <td>sea coast</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000303</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>coastal water</td>\n",
       "      <td>coastal water</td>\n",
       "      <td>coastal water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002150</td>\n",
       "      <td>coastal sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_broad_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002150</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>deep water coral reef</td>\n",
       "      <td>deep water coral reef</td>\n",
       "      <td>marine coral reef deep fore reef</td>\n",
       "      <td>1</td>\n",
       "      <td>0.047</td>\n",
       "      <td>ENVO:01000149</td>\n",
       "      <td>marine coral reef deep fore reef</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000149</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:2759</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2759</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>744</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>freshwater</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002011</td>\n",
       "      <td>fresh water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002011</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>770</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>freshwater great lakes</td>\n",
       "      <td>freshwater great lakes</td>\n",
       "      <td>freshwater lake</td>\n",
       "      <td>1</td>\n",
       "      <td>0.029</td>\n",
       "      <td>ENVO:00000021</td>\n",
       "      <td>freshwater lake</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000021</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>hot springs</td>\n",
       "      <td>hot spring</td>\n",
       "      <td>1</td>\n",
       "      <td>0.035</td>\n",
       "      <td>ENVO:00000051</td>\n",
       "      <td>hot spring</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000051</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>hydrothermal vents</td>\n",
       "      <td>hydrothermal vents</td>\n",
       "      <td>hydrothermal vent</td>\n",
       "      <td>1</td>\n",
       "      <td>0.019</td>\n",
       "      <td>ENVO:00000215</td>\n",
       "      <td>hydrothermal vent</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000215</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>marginal sea</td>\n",
       "      <td>marginal sea</td>\n",
       "      <td>marginal sea</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00000016</td>\n",
       "      <td>sea</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_narrow_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000016</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1001</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>marine brine pool basin</td>\n",
       "      <td>marine brine pool basin</td>\n",
       "      <td>marine brine pool</td>\n",
       "      <td>1</td>\n",
       "      <td>0.038</td>\n",
       "      <td>ENVO:00000369</td>\n",
       "      <td>brine pool</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00000369</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1078</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>mediterranean sea</td>\n",
       "      <td>mediterranean sea</td>\n",
       "      <td>mediterranean</td>\n",
       "      <td>1</td>\n",
       "      <td>0.039</td>\n",
       "      <td>ENVO:01000207</td>\n",
       "      <td>mediterranean</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000207</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1094</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>1</td>\n",
       "      <td>0.025</td>\n",
       "      <td>ENVO:01000008</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000008</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>ocean water</td>\n",
       "      <td>ocean water</td>\n",
       "      <td>ocean water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002149</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1377</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>seawater</td>\n",
       "      <td>seawater</td>\n",
       "      <td>seawater</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002149</td>\n",
       "      <td>sea water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002149</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1475</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>supraglacial lake water</td>\n",
       "      <td>supraglacial lake water</td>\n",
       "      <td>supraglacial lake</td>\n",
       "      <td>1</td>\n",
       "      <td>0.043</td>\n",
       "      <td>ENVO:03000035</td>\n",
       "      <td>supraglacial lake</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000035</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1519</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>temperate mediterannean sea biome</td>\n",
       "      <td>temperate mediterannean sea biome</td>\n",
       "      <td>temperate mediterranean sea biome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.008</td>\n",
       "      <td>ENVO:01000857</td>\n",
       "      <td>temperate mediterranean sea biome</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000857</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1593</th>\n",
       "      <td>env_broad_scale</td>\n",
       "      <td>water river</td>\n",
       "      <td>water river</td>\n",
       "      <td>river water</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01000599</td>\n",
       "      <td>river water</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000599</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             category                                raw  \\\n",
       "0     env_broad_scale                                      \n",
       "234   env_broad_scale                      boreal forest   \n",
       "337   env_broad_scale             coastal marine estuary   \n",
       "369   env_broad_scale                   coastal sea area   \n",
       "377   env_broad_scale                      coastal water   \n",
       "598   env_broad_scale              deep water coral reef   \n",
       "663   env_broad_scale                         eukaryotes   \n",
       "744   env_broad_scale                         freshwater   \n",
       "770   env_broad_scale             freshwater great lakes   \n",
       "827   env_broad_scale                        hot springs   \n",
       "844   env_broad_scale                 hydrothermal vents   \n",
       "957   env_broad_scale                       marginal sea   \n",
       "1001  env_broad_scale            marine brine pool basin   \n",
       "1078  env_broad_scale                  mediterranean sea   \n",
       "1094  env_broad_scale                     microbial mats   \n",
       "1205  env_broad_scale                        ocean water   \n",
       "1377  env_broad_scale                           seawater   \n",
       "1475  env_broad_scale            supraglacial lake water   \n",
       "1519  env_broad_scale  temperate mediterannean sea biome   \n",
       "1593  env_broad_scale                        water river   \n",
       "\n",
       "                                  query                               name  \\\n",
       "0                                                                            \n",
       "234                       boreal forest                      boreal forest   \n",
       "337              coastal marine estuary    estuarine coastal surface layer   \n",
       "369                    coastal sea area                       coastal area   \n",
       "377                       coastal water                      coastal water   \n",
       "598               deep water coral reef   marine coral reef deep fore reef   \n",
       "663                          eukaryotes                         eukaryotes   \n",
       "744                          freshwater                         freshwater   \n",
       "770              freshwater great lakes                    freshwater lake   \n",
       "827                         hot springs                         hot spring   \n",
       "844                  hydrothermal vents                  hydrothermal vent   \n",
       "957                        marginal sea                       marginal sea   \n",
       "1001            marine brine pool basin                  marine brine pool   \n",
       "1078                  mediterranean sea                      mediterranean   \n",
       "1094                     microbial mats                      microbial mat   \n",
       "1205                        ocean water                        ocean water   \n",
       "1377                           seawater                           seawater   \n",
       "1475            supraglacial lake water                  supraglacial lake   \n",
       "1519  temperate mediterannean sea biome  temperate mediterranean sea biome   \n",
       "1593                        water river                        river water   \n",
       "\n",
       "      cosine_rank  cosine_dist          obo_id  \\\n",
       "0               1        0.000                   \n",
       "234             1        0.000   ENVO:01000250   \n",
       "337             1        0.041   ENVO:01001302   \n",
       "369             1        0.024   ENVO:00000303   \n",
       "377             1        0.000   ENVO:00002150   \n",
       "598             1        0.047   ENVO:01000149   \n",
       "663             1        0.000  NCBITaxon:2759   \n",
       "744             1        0.000   ENVO:00002011   \n",
       "770             1        0.029   ENVO:00000021   \n",
       "827             1        0.035   ENVO:00000051   \n",
       "844             1        0.019   ENVO:00000215   \n",
       "957             1        0.000   ENVO:00000016   \n",
       "1001            1        0.038   ENVO:00000369   \n",
       "1078            1        0.039   ENVO:01000207   \n",
       "1094            1        0.025   ENVO:01000008   \n",
       "1205            1        0.000   ENVO:00002149   \n",
       "1377            1        0.000   ENVO:00002149   \n",
       "1475            1        0.043   ENVO:03000035   \n",
       "1519            1        0.008   ENVO:01000857   \n",
       "1593            1        0.000   ENVO:01000599   \n",
       "\n",
       "                                  label  search_rank ontology_prefix  \\\n",
       "0                                                  1                   \n",
       "234    subpolar coniferous forest biome            1            ENVO   \n",
       "337     estuarine coastal surface layer            2            ENVO   \n",
       "369                           sea coast            4            ENVO   \n",
       "377                   coastal sea water            1            ENVO   \n",
       "598    marine coral reef deep fore reef            2            ENVO   \n",
       "663                           Eukaryota            1            ENVO   \n",
       "744                         fresh water            1            ENVO   \n",
       "770                     freshwater lake            5            ENVO   \n",
       "827                          hot spring            2            ENVO   \n",
       "844                   hydrothermal vent            3            ENVO   \n",
       "957                                 sea            1            ENVO   \n",
       "1001                         brine pool            1            ENVO   \n",
       "1078                      mediterranean            5            ENVO   \n",
       "1094                      microbial mat            1            ENVO   \n",
       "1205                          sea water            1            ENVO   \n",
       "1377                          sea water            1            ENVO   \n",
       "1475                  supraglacial lake            1            ENVO   \n",
       "1519  temperate mediterranean sea biome            2            ENVO   \n",
       "1593                        river water            1            ENVO   \n",
       "\n",
       "                    scope        type  \\\n",
       "0                     NaN         NaN   \n",
       "234    has_narrow_synonym  annotation   \n",
       "337                 label       label   \n",
       "369   has_related_synonym  annotation   \n",
       "377     has_broad_synonym  annotation   \n",
       "598                 label       label   \n",
       "663               synonym     synonym   \n",
       "744               synonym     synonym   \n",
       "770                 label       label   \n",
       "827       hasExactSynonym        None   \n",
       "844                 label       label   \n",
       "957    has_narrow_synonym  annotation   \n",
       "1001              synonym     synonym   \n",
       "1078                label       label   \n",
       "1094                label       label   \n",
       "1205              synonym     synonym   \n",
       "1377              synonym     synonym   \n",
       "1475                label       label   \n",
       "1519                label       label   \n",
       "1593                label       label   \n",
       "\n",
       "                                                iri ontology_name  \n",
       "0                                                                  \n",
       "234    http://purl.obolibrary.org/obo/ENVO_01000250          envo  \n",
       "337    http://purl.obolibrary.org/obo/ENVO_01001302          envo  \n",
       "369    http://purl.obolibrary.org/obo/ENVO_00000303          envo  \n",
       "377    http://purl.obolibrary.org/obo/ENVO_00002150          envo  \n",
       "598    http://purl.obolibrary.org/obo/ENVO_01000149          envo  \n",
       "663   http://purl.obolibrary.org/obo/NCBITaxon_2759          envo  \n",
       "744    http://purl.obolibrary.org/obo/ENVO_00002011          envo  \n",
       "770    http://purl.obolibrary.org/obo/ENVO_00000021          envo  \n",
       "827    http://purl.obolibrary.org/obo/ENVO_00000051          envo  \n",
       "844    http://purl.obolibrary.org/obo/ENVO_00000215          envo  \n",
       "957    http://purl.obolibrary.org/obo/ENVO_00000016          envo  \n",
       "1001   http://purl.obolibrary.org/obo/ENVO_00000369          envo  \n",
       "1078   http://purl.obolibrary.org/obo/ENVO_01000207          envo  \n",
       "1094   http://purl.obolibrary.org/obo/ENVO_01000008          envo  \n",
       "1205   http://purl.obolibrary.org/obo/ENVO_00002149          envo  \n",
       "1377   http://purl.obolibrary.org/obo/ENVO_00002149          envo  \n",
       "1475   http://purl.obolibrary.org/obo/ENVO_03000035          envo  \n",
       "1519   http://purl.obolibrary.org/obo/ENVO_01000857          envo  \n",
       "1593   http://purl.obolibrary.org/obo/ENVO_01000599          envo  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# slow... turn logging back on to show status?\n",
    "ebs_search_res = scoped_mapping.search_get_annotations_wrapper(ebs_wo_list, bad_chars=chars_to_whiteout, cat_name=biosample_col_to_map,\n",
    "                                                ontoprefix=target_onto_prefix.lower(), query_fields='', rr=5)\n",
    "my_best_acceptable = scoped_mapping.get_best_acceptable(ebs_search_res)\n",
    "\n",
    "my_best_acceptable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39e98bf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "no_acceptable_mappings = scoped_mapping.get_no_acceptable_mappings(ebs_search_res, my_best_acceptable)\n",
    "\n",
    "no_acceptable_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e07cdab1",
   "metadata": {},
   "source": [
    "- Some broad scales look like place names\n",
    "- Some get a good hit if 'biome' is added\n",
    "- how to manually review and then add back in?\n",
    "- add to biosample SQLite database:\n",
    "    - no_acceptable_mappings\n",
    "    - my_best_acceptable\n",
    "    - ebs_search_results (no acceptable + all acceptable)?\n",
    "    - mapping_candidates -> mapping_scratch (ID-based and exact-tidied-label-based)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac6219e6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
