{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "621bbcd9",
   "metadata": {},
   "source": [
    "be really clear about what's going into repaired tables: \n",
    "- repaired\n",
    "- already OK\n",
    "- unrepairable\n",
    "\n",
    "Only ~ 215k repaired `env_package`s ?!\n",
    "- yeah, there just aren't that many annotated in the biosample metadata\n",
    "\n",
    "\n",
    "- some of this code should probably be moved back into the package\n",
    "    - splitting and repiping \n",
    "\n",
    "- read in previous SSSOMs and don't re-map? or at least double check?\n",
    "\n",
    "- don't forget to consult \"model\" field in biosample table too\n",
    "    - for scoping, in addition to en_package and taxonomy_id?\n",
    "\n",
    "- what to do about hits against obsolete classes... does OLS provide a mechanism to avoid?\n",
    "- be careful with GAZ matches\n",
    "- avoid IDless, OMIT, NCIT, MICRO, BTO\n",
    "\n",
    "- is it worth trimming and concatenating \"x1\", \"x2\", \"x3\", where x is a mappable term? numbers may be low.\n",
    "    - example = XXX\n",
    "\n",
    "- reconcile \"x biome\" and \"x\" mappings\n",
    "- take note of non-local salvage hits\n",
    "    - like XXX\n",
    "    - small/large (freshwater) lake?\n",
    "    - include native labels in the review frame, in addition to local?\n",
    "    - should use rdftab, not additional OLS calls\n",
    "    \n",
    "\n",
    "- switch from hard-coded `env_package_overrides` to something like SSSOM?\n",
    "- mappings can come from merging strategies, first pass OLS or salvage OLS\n",
    "- reflect those methods in SSSOM\n",
    "- add string distance/confidence\n",
    "    - multiple columns for multiple methods?\n",
    "\n",
    "review these SSSOM columns\n",
    "\n",
    "```\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "```\n",
    "\n",
    "- are there samples with a host_taxid and a (sample) taxonomy_id from \"unclassified\"?\n",
    "- start at uncultured or metagenome taxonomy_id?\n",
    "\n",
    "- what scoping on or mapping of  \"gold\" or \"path\"? WON'T FIND THAT IN NCBI BIOSAMPLE\n",
    "\n",
    "\n",
    "What to save where? \n",
    "- new columns in biosample sqlite? \n",
    "- delimited file dumps\n",
    "- move content from original column and insert the new values into that original column?\n",
    "- pandas clipboard integration handy for review step\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fac2962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "\n",
    "import scoped_mapping\n",
    "\n",
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3300da64",
   "metadata": {},
   "source": [
    "# Settings required from user:\n",
    "\n",
    "See repo README for notes on setting up SQLite databases of OBO ontologies with semantic-sql, relation-graph and rdftab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87e3b6f1",
   "metadata": {},
   "source": [
    "### How should the mapping effort be scoped?\n",
    "Additional taxonomic filters may be applied below\n",
    "- soil\n",
    "- sediment\n",
    "- plant-associated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ec160085",
   "metadata": {},
   "outputs": [],
   "source": [
    "scoping_col = \"env_package_normalization.EnvPackage\"\n",
    "scoping_value = \"sediment\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2d7f8bf",
   "metadata": {},
   "source": [
    "### What Biosample field should be mapped to ontology classes?\n",
    "- env_broad_scale\n",
    "- env_local_scale\n",
    "- env_medium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b79c519",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_col_to_map = \"env_medium\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c138145d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# where do we require a single ontology and where can we use multiple?\n",
    "primary_onto_prefix = \"ENVO\"\n",
    "# 'gaz',\n",
    "# need to rank results\n",
    "# additional_preferred_ontologies = [\"ncbitaxon\", \"obi\", \"pato\", \"efo\"]\n",
    "additional_preferred_ontologies = [\"ncbitaxon\", \"po\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5684af0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://www.ncbi.nlm.nih.gov/biosample/docs/packages/?format=xml\n",
    "# see also https://www.ncbi.nlm.nih.gov/biosample/docs/packages/\n",
    "biosample_packages_file = \"../target/biosample_packages.xml\"\n",
    "\n",
    "# from ftp://ftp.ncbi.nlm.nih.gov//biosample/biosample_set.xml.gz\n",
    "# via harmonized_table.db.gz\n",
    "# in https://drive.google.com/drive/u/0/folders/1eL0v0stoduahjDpoDJIk3z2pJBAU4b2Y\n",
    "biosample_sqlite_file = \"../target/harmonized_table.db\"\n",
    "ncbitaxon_sqlite_file = \"../semantic-sql/db/ncbitaxon.db\"\n",
    "\n",
    "chars_to_whiteout = \"._-\"\n",
    "my_query_fields = \"\"  # OLS weighted default\n",
    "my_row_req = 3\n",
    "my_string_dist_arg = 2\n",
    "my_max_string_dist = 0.1\n",
    "salvage_max_dist = 0.2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a568c1",
   "metadata": {},
   "source": [
    "### Settings for manual review of OLS search-based results\n",
    "Results based on merging Biosample annotations to ontology classes by label or embedded ID are not exported for review at this time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "20cdb474",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ols_review_file = \"../local/ols_review.tsv\"\n",
    "ols_review_seperator = \"\\t\"\n",
    "\n",
    "strategy_col = \"strategy\"\n",
    "include_col = \"include\"\n",
    "\n",
    "first_pass_include_val = True\n",
    "# first_pass_strategy_val = \"env_braod_scale vs envo and gaz @ 0.1\"\n",
    "salvage_include_val = False\n",
    "# salvage_strategy_val    = \"env_braod_scale vs all of ols @ 0.2\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7edfbe0",
   "metadata": {},
   "source": [
    "### Settings for SSSOM output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d71fa5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sssom_subject_prefix = biosample_col_to_map + \"_\" + scoping_col + \"_\" + scoping_value\n",
    "sssom_file = \"../\" + sssom_subject_prefix + \"_SSSOM.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c8336b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# per_biosample_mapping_results_file = \"per_biosample_scoped_mapping_results.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f21778e4",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c4494ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)\n",
    "ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)\n",
    "primary_ontology_sqlite_file = (\n",
    "    \"../semantic-sql/db/\" + primary_onto_prefix.lower() + \".db\"\n",
    ")\n",
    "primary_ontology_cnx = sqlite3.connect(primary_ontology_sqlite_file)\n",
    "additional_preferred_ontologies.insert(0, primary_onto_prefix.lower())\n",
    "additional_preferred_ontologies_str = \",\".join(additional_preferred_ontologies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "48241248",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ontology</th>\n",
       "      <th>id_pattern</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFO</td>\n",
       "      <td>BFO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CARO</td>\n",
       "      <td>CARO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEBI</td>\n",
       "      <td>CHEBI:\\d{4,6}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO</td>\n",
       "      <td>ENVO:\\d{7,8}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAO</td>\n",
       "      <td>FAO:0000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FOODON</td>\n",
       "      <td>FOODON:\\d{8}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO</td>\n",
       "      <td>GO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IAO</td>\n",
       "      <td>IAO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NCBITaxon</td>\n",
       "      <td>NCBITaxon:\\d+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OBI</td>\n",
       "      <td>OBI:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PATO</td>\n",
       "      <td>PATO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PCO</td>\n",
       "      <td>PCO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PO</td>\n",
       "      <td>PO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RO</td>\n",
       "      <td>RO:0002577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>UBERON</td>\n",
       "      <td>UBERON:\\d{7}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ontology     id_pattern\n",
       "0         BFO      BFO:\\d{7}\n",
       "1        CARO     CARO:\\d{7}\n",
       "2       CHEBI  CHEBI:\\d{4,6}\n",
       "3        ENVO   ENVO:\\d{7,8}\n",
       "4         FAO    FAO:0000001\n",
       "5      FOODON   FOODON:\\d{8}\n",
       "6          GO       GO:\\d{7}\n",
       "7         IAO      IAO:\\d{7}\n",
       "8   NCBITaxon  NCBITaxon:\\d+\n",
       "9         OBI      OBI:\\d{7}\n",
       "10       PATO     PATO:\\d{7}\n",
       "11        PCO      PCO:\\d{7}\n",
       "12         PO       PO:\\d{7}\n",
       "13         RO     RO:0002577\n",
       "14     UBERON   UBERON:\\d{7}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"select * from id_patterns\"\n",
    "[id_patterns_frame, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "id_patterns_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6cea64bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_patterns = dict(zip(id_patterns_frame.ontology, id_patterns_frame.id_pattern))\n",
    "# id_patterns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db3fd267",
   "metadata": {},
   "source": [
    "- env_package = env_package annotation from NCBI Biosample file XXX\n",
    "- count = number of biosamples using that env_package annotation\n",
    "- lhs = checklist info\n",
    "- rhs = potential package info\n",
    "- extract = potential OBO ID from rhs column (currently harcoded and only looking for ENVO IDs)\n",
    "- remaining_string = rhs/string, with potential OBO IDs removed\n",
    "- remaining_tidied = remaining_string with case, whitespace and punctuation normailzastion\n",
    "- rt_override = some remaining_tidied values can be replaced according to env_package_overrides\n",
    "- EnvPackage = corresponding de-normalized value from package_dictionary\n",
    "- is_canonical = false when EnvPackage is NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3766cbe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "q = \"select * from env_package_normalization\"\n",
    "[env_package_normalization, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx\n",
    ")\n",
    "\n",
    "# env_package_normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2363d41",
   "metadata": {},
   "source": [
    "## Get a table of MIxS annotations to be mapped to ontology classes.\n",
    "\n",
    "Explicitly scope based on normalized package data. These values were set at the top of this notebook.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "91dcca09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env_medium\n",
      "env_package_normalization.EnvPackage\n",
      "sediment\n"
     ]
    }
   ],
   "source": [
    "print(biosample_col_to_map)\n",
    "print(scoping_col)\n",
    "print(scoping_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80727cfd",
   "metadata": {},
   "source": [
    "**In this case, the scoping includes an inner join requirement for 'unclassified entities'**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7e777c8a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mud</td>\n",
       "      <td>3917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine sediment</td>\n",
       "      <td>854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sediment</td>\n",
       "      <td>702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Sequences</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              env_medium  count\n",
       "0                                    mud   3917\n",
       "1                        marine sediment    854\n",
       "2                               sediment    702\n",
       "3                               Sediment    450\n",
       "4   deep marine sediment [ENVO:00002113]    185\n",
       "..                                   ...    ...\n",
       "71                             Sequences      1\n",
       "72                       Marine sediment      1\n",
       "73                         Lake Sediment      1\n",
       "74                  Environmental_sample      1\n",
       "75                             Crude Oil      1\n",
       "\n",
       "[76 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = (\n",
    "    \"select \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\", count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "join env_package_normalization on\n",
    "    b.env_package = env_package_normalization.env_package\n",
    "inner join biosample_sample_taxon_summary stic on\n",
    "    b.taxonomy_id = stic.biosample_taxid\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"' group by \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\"\n",
    "order by\n",
    "    count(*) desc\"\"\"\n",
    ")\n",
    "\n",
    "[mapping_candidates, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86f43bb4",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## The Biosample format allows for pipe-delimited environmental package lists. \n",
    "\n",
    "Separate those out into their components.\n",
    "\n",
    "----\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "45e1859f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                repeated  \\\n",
       "0                                    mud   \n",
       "0                        marine sediment   \n",
       "0                               sediment   \n",
       "0                               Sediment   \n",
       "0   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "0                              Sequences   \n",
       "0                        Marine sediment   \n",
       "0                          Lake Sediment   \n",
       "0                   Environmental_sample   \n",
       "0                              Crude Oil   \n",
       "\n",
       "                                splitted  part_count  seq  \n",
       "0                                    mud           1    1  \n",
       "0                        marine sediment           1    1  \n",
       "0                               sediment           1    1  \n",
       "0                               Sediment           1    1  \n",
       "0   deep marine sediment [ENVO:00002113]           1    1  \n",
       "..                                   ...         ...  ...  \n",
       "0                              Sequences           1    1  \n",
       "0                        Marine sediment           1    1  \n",
       "0                          Lake Sediment           1    1  \n",
       "0                   Environmental_sample           1    1  \n",
       "0                              Crude Oil           1    1  \n",
       "\n",
       "[76 rows x 4 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_frames = []\n",
    "for row in mapping_candidates.itertuples(index=True, name=\"PandaZ\"):\n",
    "    split_check = row[1]\n",
    "    if split_check is None:\n",
    "        split_check = \"\"\n",
    "    splitted = pd.Series(split_check.split(\"|\"))\n",
    "    splitted_count = len(splitted)\n",
    "    repeated = [split_check] * splitted_count\n",
    "    repeated = pd.Series(repeated)\n",
    "    as_frame = pd.DataFrame(dict(repeated=repeated, splitted=splitted)).reset_index()\n",
    "    seq_list = list(range(1, splitted_count + 1))\n",
    "    as_frame[\"part_count\"] = splitted_count\n",
    "    as_frame[\"seq\"] = seq_list\n",
    "    multi_frames.append(as_frame)\n",
    "concat_frame = pd.concat(multi_frames)\n",
    "concat_frame = concat_frame[[\"repeated\", \"splitted\", \"part_count\", \"seq\"]]\n",
    "concat_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c67a272e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    concat_frame, left_on=biosample_col_to_map, right_on=\"repeated\", how=\"left\"\n",
    ")\n",
    "\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3af421ac",
   "metadata": {},
   "source": [
    "## Normalize a few different ways `OBO` IDs have been entered\n",
    "In the Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ed3a89fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # this helps by standardizing what could be an ID\n",
    "# but if the RHS is text, then envo should just be removed\n",
    "mapping_candidates[\"term_tidy\"] = mapping_candidates.splitted.str.replace(\n",
    "    primary_onto_prefix.lower() + \"[:_ ]\",\n",
    "    primary_onto_prefix + \":\",\n",
    "    regex=True,\n",
    "    case=False,\n",
    ")\n",
    "\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cc22a1a",
   "metadata": {},
   "source": [
    "# Now try to extract ontology terms that are already present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e423fcb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mud</td>\n",
       "      <td>3917</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td></td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine sediment</td>\n",
       "      <td>854</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sediment</td>\n",
       "      <td>702</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>450</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Sediment</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>185</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>deep marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Sequences</td>\n",
       "      <td>1</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td></td>\n",
       "      <td>Sequences</td>\n",
       "      <td>sequences</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>lake sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td></td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>environmental sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td></td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>crude oil</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              env_medium  count  \\\n",
       "0                                    mud   3917   \n",
       "1                        marine sediment    854   \n",
       "2                               sediment    702   \n",
       "3                               Sediment    450   \n",
       "4   deep marine sediment [ENVO:00002113]    185   \n",
       "..                                   ...    ...   \n",
       "71                             Sequences      1   \n",
       "72                       Marine sediment      1   \n",
       "73                         Lake Sediment      1   \n",
       "74                  Environmental_sample      1   \n",
       "75                             Crude Oil      1   \n",
       "\n",
       "                                repeated  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "71                             Sequences   \n",
       "72                       Marine sediment   \n",
       "73                         Lake Sediment   \n",
       "74                  Environmental_sample   \n",
       "75                             Crude Oil   \n",
       "\n",
       "                                splitted  part_count  seq  \\\n",
       "0                                    mud         1.0  1.0   \n",
       "1                        marine sediment         1.0  1.0   \n",
       "2                               sediment         1.0  1.0   \n",
       "3                               Sediment         1.0  1.0   \n",
       "4   deep marine sediment [ENVO:00002113]         1.0  1.0   \n",
       "..                                   ...         ...  ...   \n",
       "71                             Sequences         1.0  1.0   \n",
       "72                       Marine sediment         1.0  1.0   \n",
       "73                         Lake Sediment         1.0  1.0   \n",
       "74                  Environmental_sample         1.0  1.0   \n",
       "75                             Crude Oil         1.0  1.0   \n",
       "\n",
       "                               term_tidy  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "71                             Sequences   \n",
       "72                       Marine sediment   \n",
       "73                         Lake Sediment   \n",
       "74                  Environmental_sample   \n",
       "75                             Crude Oil   \n",
       "\n",
       "                                  string        extract  \\\n",
       "0                                    mud                  \n",
       "1                        marine sediment                  \n",
       "2                               sediment                  \n",
       "3                               Sediment                  \n",
       "4   deep marine sediment [ENVO:00002113]  ENVO:00002113   \n",
       "..                                   ...            ...   \n",
       "71                             Sequences                  \n",
       "72                       Marine sediment                  \n",
       "73                         Lake Sediment                  \n",
       "74                  Environmental_sample                  \n",
       "75                             Crude Oil                  \n",
       "\n",
       "         remaining_string      remaining_tidied  \n",
       "0                     mud                   mud  \n",
       "1         marine sediment       marine sediment  \n",
       "2                sediment              sediment  \n",
       "3                Sediment              sediment  \n",
       "4   deep marine sediment   deep marine sediment  \n",
       "..                    ...                   ...  \n",
       "71              Sequences             sequences  \n",
       "72        Marine sediment       marine sediment  \n",
       "73          Lake Sediment         lake sediment  \n",
       "74   Environmental_sample  environmental sample  \n",
       "75              Crude Oil             crude oil  \n",
       "\n",
       "[76 rows x 11 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_series_decomposition = scoped_mapping.decompose_series(\n",
    "    mapping_candidates[\"term_tidy\"], id_patterns[primary_onto_prefix]\n",
    ")\n",
    "\n",
    "# candidate_series_decomposition\n",
    "\n",
    "mapping_candidates = pd.concat(\n",
    "    [mapping_candidates, candidate_series_decomposition], axis=1\n",
    ")\n",
    "\n",
    "mapping_candidates[\n",
    "    \"remaining_tidied\"\n",
    "] = mapping_candidates.remaining_tidied.str.replace(\n",
    "    primary_onto_prefix.lower() + \"[:_ ]\", \"\", regex=True, case=False\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccd038ff",
   "metadata": {},
   "source": [
    "## Join the extracted IDs with their labels\n",
    "\n",
    "_Start by connecting to the rdftab database from which the terms and label-like annotations will be obtained?_"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db5d19a9",
   "metadata": {},
   "source": [
    "## extracting the labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3bcd7e47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFO:0000001</td>\n",
       "      <td>entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BFO:0000002</td>\n",
       "      <td>continuant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BFO:0000003</td>\n",
       "      <td>occurrent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BFO:0000004</td>\n",
       "      <td>independent continuant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BFO:0000006</td>\n",
       "      <td>spatial region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8350</th>\n",
       "      <td>PO:0030132</td>\n",
       "      <td>cincinnus inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8351</th>\n",
       "      <td>PO:0030133</td>\n",
       "      <td>compound cincinnus inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8352</th>\n",
       "      <td>PO:0030134</td>\n",
       "      <td>bostryx inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8353</th>\n",
       "      <td>PO:0030135</td>\n",
       "      <td>drepanium inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8354</th>\n",
       "      <td>PO:0030136</td>\n",
       "      <td>compound drepanium inflorescence</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8355 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          subject                             value\n",
       "0     BFO:0000001                            entity\n",
       "1     BFO:0000002                        continuant\n",
       "2     BFO:0000003                         occurrent\n",
       "3     BFO:0000004            independent continuant\n",
       "4     BFO:0000006                    spatial region\n",
       "...           ...                               ...\n",
       "8350   PO:0030132           cincinnus inflorescence\n",
       "8351   PO:0030133  compound cincinnus inflorescence\n",
       "8352   PO:0030134             bostryx inflorescence\n",
       "8353   PO:0030135           drepanium inflorescence\n",
       "8354   PO:0030136  compound drepanium inflorescence\n",
       "\n",
       "[8355 rows x 2 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"select stanza as subject, value from ids_labs_selected_ontolgies\"\n",
    "\n",
    "[onto_labels, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "onto_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "840b922f",
   "metadata": {},
   "source": [
    "## and merging "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d82daf6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_candidates.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8543ed0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mud</td>\n",
       "      <td>3917</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td></td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine sediment</td>\n",
       "      <td>854</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sediment</td>\n",
       "      <td>702</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>450</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>185</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Sequences</td>\n",
       "      <td>1</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td></td>\n",
       "      <td>Sequences</td>\n",
       "      <td>sequences</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>lake sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td></td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>environmental sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td></td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>crude oil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              env_medium  count  \\\n",
       "0                                    mud   3917   \n",
       "1                        marine sediment    854   \n",
       "2                               sediment    702   \n",
       "3                               Sediment    450   \n",
       "4   deep marine sediment [ENVO:00002113]    185   \n",
       "..                                   ...    ...   \n",
       "71                             Sequences      1   \n",
       "72                       Marine sediment      1   \n",
       "73                         Lake Sediment      1   \n",
       "74                  Environmental_sample      1   \n",
       "75                             Crude Oil      1   \n",
       "\n",
       "                                repeated  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "71                             Sequences   \n",
       "72                       Marine sediment   \n",
       "73                         Lake Sediment   \n",
       "74                  Environmental_sample   \n",
       "75                             Crude Oil   \n",
       "\n",
       "                                splitted  part_count  seq  \\\n",
       "0                                    mud         1.0  1.0   \n",
       "1                        marine sediment         1.0  1.0   \n",
       "2                               sediment         1.0  1.0   \n",
       "3                               Sediment         1.0  1.0   \n",
       "4   deep marine sediment [ENVO:00002113]         1.0  1.0   \n",
       "..                                   ...         ...  ...   \n",
       "71                             Sequences         1.0  1.0   \n",
       "72                       Marine sediment         1.0  1.0   \n",
       "73                         Lake Sediment         1.0  1.0   \n",
       "74                  Environmental_sample         1.0  1.0   \n",
       "75                             Crude Oil         1.0  1.0   \n",
       "\n",
       "                               term_tidy  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "71                             Sequences   \n",
       "72                       Marine sediment   \n",
       "73                         Lake Sediment   \n",
       "74                  Environmental_sample   \n",
       "75                             Crude Oil   \n",
       "\n",
       "                                  string        extract  \\\n",
       "0                                    mud                  \n",
       "1                        marine sediment                  \n",
       "2                               sediment                  \n",
       "3                               Sediment                  \n",
       "4   deep marine sediment [ENVO:00002113]  ENVO:00002113   \n",
       "..                                   ...            ...   \n",
       "71                             Sequences                  \n",
       "72                       Marine sediment                  \n",
       "73                         Lake Sediment                  \n",
       "74                  Environmental_sample                  \n",
       "75                             Crude Oil                  \n",
       "\n",
       "         remaining_string      remaining_tidied        subject  \\\n",
       "0                     mud                   mud            NaN   \n",
       "1         marine sediment       marine sediment            NaN   \n",
       "2                sediment              sediment            NaN   \n",
       "3                Sediment              sediment            NaN   \n",
       "4   deep marine sediment   deep marine sediment  ENVO:00002113   \n",
       "..                    ...                   ...            ...   \n",
       "71              Sequences             sequences            NaN   \n",
       "72        Marine sediment       marine sediment            NaN   \n",
       "73          Lake Sediment         lake sediment            NaN   \n",
       "74   Environmental_sample  environmental sample            NaN   \n",
       "75              Crude Oil             crude oil            NaN   \n",
       "\n",
       "                   value  \n",
       "0                    NaN  \n",
       "1                    NaN  \n",
       "2                    NaN  \n",
       "3                    NaN  \n",
       "4   deep marine sediment  \n",
       "..                   ...  \n",
       "71                   NaN  \n",
       "72                   NaN  \n",
       "73                   NaN  \n",
       "74                   NaN  \n",
       "75                   NaN  \n",
       "\n",
       "[76 rows x 13 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"extract\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64b3baef",
   "metadata": {},
   "source": [
    "## Use cosine string distance to see if the labels match closely enough\n",
    "\n",
    "I.e. the labels claimed by the Biosample data set and the labels asserted in the ontology. if they're close enough, consider the assigned ID legit\n",
    "\n",
    "\n",
    "_How close is close enough?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "abd0a035",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "      <th>cosine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mud</td>\n",
       "      <td>3917</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td></td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine sediment</td>\n",
       "      <td>854</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sediment</td>\n",
       "      <td>702</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>450</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>185</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Sequences</td>\n",
       "      <td>1</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td></td>\n",
       "      <td>Sequences</td>\n",
       "      <td>sequences</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>lake sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td></td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>environmental sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td></td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>crude oil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              env_medium  count  \\\n",
       "0                                    mud   3917   \n",
       "1                        marine sediment    854   \n",
       "2                               sediment    702   \n",
       "3                               Sediment    450   \n",
       "4   deep marine sediment [ENVO:00002113]    185   \n",
       "..                                   ...    ...   \n",
       "71                             Sequences      1   \n",
       "72                       Marine sediment      1   \n",
       "73                         Lake Sediment      1   \n",
       "74                  Environmental_sample      1   \n",
       "75                             Crude Oil      1   \n",
       "\n",
       "                                repeated  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "71                             Sequences   \n",
       "72                       Marine sediment   \n",
       "73                         Lake Sediment   \n",
       "74                  Environmental_sample   \n",
       "75                             Crude Oil   \n",
       "\n",
       "                                splitted  part_count  seq  \\\n",
       "0                                    mud         1.0  1.0   \n",
       "1                        marine sediment         1.0  1.0   \n",
       "2                               sediment         1.0  1.0   \n",
       "3                               Sediment         1.0  1.0   \n",
       "4   deep marine sediment [ENVO:00002113]         1.0  1.0   \n",
       "..                                   ...         ...  ...   \n",
       "71                             Sequences         1.0  1.0   \n",
       "72                       Marine sediment         1.0  1.0   \n",
       "73                         Lake Sediment         1.0  1.0   \n",
       "74                  Environmental_sample         1.0  1.0   \n",
       "75                             Crude Oil         1.0  1.0   \n",
       "\n",
       "                               term_tidy  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "71                             Sequences   \n",
       "72                       Marine sediment   \n",
       "73                         Lake Sediment   \n",
       "74                  Environmental_sample   \n",
       "75                             Crude Oil   \n",
       "\n",
       "                                  string        extract  \\\n",
       "0                                    mud                  \n",
       "1                        marine sediment                  \n",
       "2                               sediment                  \n",
       "3                               Sediment                  \n",
       "4   deep marine sediment [ENVO:00002113]  ENVO:00002113   \n",
       "..                                   ...            ...   \n",
       "71                             Sequences                  \n",
       "72                       Marine sediment                  \n",
       "73                         Lake Sediment                  \n",
       "74                  Environmental_sample                  \n",
       "75                             Crude Oil                  \n",
       "\n",
       "         remaining_string      remaining_tidied        subject  \\\n",
       "0                     mud                   mud            NaN   \n",
       "1         marine sediment       marine sediment            NaN   \n",
       "2                sediment              sediment            NaN   \n",
       "3                Sediment              sediment            NaN   \n",
       "4   deep marine sediment   deep marine sediment  ENVO:00002113   \n",
       "..                    ...                   ...            ...   \n",
       "71              Sequences             sequences            NaN   \n",
       "72        Marine sediment       marine sediment            NaN   \n",
       "73          Lake Sediment         lake sediment            NaN   \n",
       "74   Environmental_sample  environmental sample            NaN   \n",
       "75              Crude Oil             crude oil            NaN   \n",
       "\n",
       "                   value  cosine  \n",
       "0                            1.0  \n",
       "1                            1.0  \n",
       "2                            1.0  \n",
       "3                            1.0  \n",
       "4   deep marine sediment     0.0  \n",
       "..                   ...     ...  \n",
       "71                           1.0  \n",
       "72                           1.0  \n",
       "73                           1.0  \n",
       "74                           1.0  \n",
       "75                           1.0  \n",
       "\n",
       "[76 rows x 14 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_cosine_obj = Cosine(my_string_dist_arg)\n",
    "mapping_candidates[\"value\"] = mapping_candidates[\"value\"].fillna(\"\")\n",
    "mapping_candidates[\"remaining_tidied\"] = mapping_candidates[\"remaining_tidied\"].fillna(\n",
    "    \"\"\n",
    ")\n",
    "mapping_candidates[\"cosine\"] = mapping_candidates.apply(\n",
    "    lambda my_row: my_cosine_obj.distance(\n",
    "        my_row[\"remaining_tidied\"].lower(), my_row[\"value\"].lower()\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2358a89",
   "metadata": {},
   "source": [
    "**Previously, we did a reality check on the claimed IDs and labels. If a label is claimed without any ID, that could still be a path to an ontology term.**\n",
    "\n",
    "We'll be doing some merging, so make sure column names are meaningful and unique\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "092aced0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mud</td>\n",
       "      <td>3917</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td></td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000001</td>\n",
       "      <td>mud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine sediment</td>\n",
       "      <td>854</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sediment</td>\n",
       "      <td>702</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>450</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>185</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Sequences</td>\n",
       "      <td>1</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td></td>\n",
       "      <td>Sequences</td>\n",
       "      <td>sequences</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>lake sediment</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000546</td>\n",
       "      <td>lake sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td></td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>environmental sample</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td></td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>crude oil</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              env_medium  count  \\\n",
       "0                                    mud   3917   \n",
       "1                        marine sediment    854   \n",
       "2                               sediment    702   \n",
       "3                               Sediment    450   \n",
       "4   deep marine sediment [ENVO:00002113]    185   \n",
       "..                                   ...    ...   \n",
       "71                             Sequences      1   \n",
       "72                       Marine sediment      1   \n",
       "73                         Lake Sediment      1   \n",
       "74                  Environmental_sample      1   \n",
       "75                             Crude Oil      1   \n",
       "\n",
       "                                repeated  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "71                             Sequences   \n",
       "72                       Marine sediment   \n",
       "73                         Lake Sediment   \n",
       "74                  Environmental_sample   \n",
       "75                             Crude Oil   \n",
       "\n",
       "                                splitted  part_count  seq  \\\n",
       "0                                    mud         1.0  1.0   \n",
       "1                        marine sediment         1.0  1.0   \n",
       "2                               sediment         1.0  1.0   \n",
       "3                               Sediment         1.0  1.0   \n",
       "4   deep marine sediment [ENVO:00002113]         1.0  1.0   \n",
       "..                                   ...         ...  ...   \n",
       "71                             Sequences         1.0  1.0   \n",
       "72                       Marine sediment         1.0  1.0   \n",
       "73                         Lake Sediment         1.0  1.0   \n",
       "74                  Environmental_sample         1.0  1.0   \n",
       "75                             Crude Oil         1.0  1.0   \n",
       "\n",
       "                               term_tidy  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "71                             Sequences   \n",
       "72                       Marine sediment   \n",
       "73                         Lake Sediment   \n",
       "74                  Environmental_sample   \n",
       "75                             Crude Oil   \n",
       "\n",
       "                                  string        extract  \\\n",
       "0                                    mud                  \n",
       "1                        marine sediment                  \n",
       "2                               sediment                  \n",
       "3                               Sediment                  \n",
       "4   deep marine sediment [ENVO:00002113]  ENVO:00002113   \n",
       "..                                   ...            ...   \n",
       "71                             Sequences                  \n",
       "72                       Marine sediment                  \n",
       "73                         Lake Sediment                  \n",
       "74                  Environmental_sample                  \n",
       "75                             Crude Oil                  \n",
       "\n",
       "         remaining_string      remaining_tidied        term_id  \\\n",
       "0                     mud                   mud            NaN   \n",
       "1         marine sediment       marine sediment            NaN   \n",
       "2                sediment              sediment            NaN   \n",
       "3                Sediment              sediment            NaN   \n",
       "4   deep marine sediment   deep marine sediment  ENVO:00002113   \n",
       "..                    ...                   ...            ...   \n",
       "71              Sequences             sequences            NaN   \n",
       "72        Marine sediment       marine sediment            NaN   \n",
       "73          Lake Sediment         lake sediment            NaN   \n",
       "74   Environmental_sample  environmental sample            NaN   \n",
       "75              Crude Oil             crude oil            NaN   \n",
       "\n",
       "             lab_from_id  lfi_cosine term_id_from_lab                 value  \n",
       "0                                1.0    ENVO:01000001                   mud  \n",
       "1                                1.0    ENVO:03000033       marine sediment  \n",
       "2                                1.0    ENVO:00002007              sediment  \n",
       "3                                1.0    ENVO:00002007              sediment  \n",
       "4   deep marine sediment         0.0    ENVO:00002113  deep marine sediment  \n",
       "..                   ...         ...              ...                   ...  \n",
       "71                               1.0              NaN                   NaN  \n",
       "72                               1.0    ENVO:03000033       marine sediment  \n",
       "73                               1.0    ENVO:00000546         lake sediment  \n",
       "74                               1.0              NaN                   NaN  \n",
       "75                               1.0              NaN                   NaN  \n",
       "\n",
       "[76 rows x 16 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates.columns = [\n",
    "    biosample_col_to_map,\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"term_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "]\n",
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"remaining_tidied\", right_on=\"value\", how=\"left\"\n",
    ")\n",
    "\n",
    "mapping_candidates.columns = [\n",
    "    biosample_col_to_map,\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"term_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "    \"term_id_from_lab\",\n",
    "    \"value\",\n",
    "]\n",
    "\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e968917d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates = mapping_candidates.drop_duplicates()\n",
    "\n",
    "mc_tifl_split = mapping_candidates[\"term_id_from_lab\"].str.split(\":\")\n",
    "\n",
    "import math\n",
    "\n",
    "prefix_when_possible = []\n",
    "for one_split in mc_tifl_split:\n",
    "    if type(one_split) == list:\n",
    "        os_first = one_split[0]\n",
    "        prefix_when_possible.append(os_first)\n",
    "    else:\n",
    "        prefix_when_possible.append(\"\")\n",
    "\n",
    "prefix_when_possible = [each_string.lower() for each_string in prefix_when_possible]\n",
    "\n",
    "# prefix_when_possible"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c10ced03",
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"onto_prefix\" in mapping_candidates.columns:\n",
    "    mapping_candidates = mapping_candidates.drop(\"onto_prefix\", 1)\n",
    "\n",
    "mapping_candidates.loc[:, \"onto_prefix\"] = pd.Series(prefix_when_possible).values\n",
    "\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9720bade",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>onto_prefix</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>envo</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>po</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  onto_prefix  rank\n",
       "0        envo     0\n",
       "1   ncbitaxon     1\n",
       "2          po     2"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranks = list(range(len(additional_preferred_ontologies)))\n",
    "\n",
    "d = {\"onto_prefix\": additional_preferred_ontologies, \"rank\": ranks}\n",
    "onto_rank_frame = pd.DataFrame(d)\n",
    "\n",
    "onto_rank_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "536a6f9c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_rank_frame, on=\"onto_prefix\", how=\"left\"\n",
    ")\n",
    "\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "46ce3f1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_maps = mapping_candidates.groupby(biosample_col_to_map)[\"rank\"].min().reset_index()\n",
    "# best_maps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2667e541",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    best_maps, on=[biosample_col_to_map, \"rank\"]\n",
    ")\n",
    "# mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41082d99",
   "metadata": {},
   "source": [
    "## Find consensus term IDs and labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4f5e3fdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates[\"consensus_id\"] = mapping_candidates[\"term_id_from_lab\"]\n",
    "mapping_candidates[\"consensus_lab\"] = mapping_candidates[\"value\"]\n",
    "\n",
    "flag = mapping_candidates[\"consensus_id\"].isnull() & (\n",
    "    (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates.lfi_cosine.le(my_max_string_dist)\n",
    "    )\n",
    "    | (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates[\"remaining_tidied\"].eq(\"\")\n",
    "    )\n",
    ")\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"term_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_id\"] = replacements\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"lab_from_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_lab\"] = replacements\n",
    "\n",
    "flag = mapping_candidates.consensus_id.isna()\n",
    "antiflag = ~flag\n",
    "mapping_candidates[\"id_or_lab_ok\"] = antiflag\n",
    "\n",
    "mapping_candidates[\"assembled_consensus\"] = (\n",
    "    mapping_candidates[\"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + mapping_candidates[\"consensus_id\"]\n",
    "    + \"]\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6af120f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_medium</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>...</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>onto_prefix</th>\n",
       "      <th>rank</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "      <th>id_or_lab_ok</th>\n",
       "      <th>assembled_consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mud</td>\n",
       "      <td>3917</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>mud</td>\n",
       "      <td>mud</td>\n",
       "      <td></td>\n",
       "      <td>mud</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:01000001</td>\n",
       "      <td>mud</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:01000001</td>\n",
       "      <td>mud</td>\n",
       "      <td>True</td>\n",
       "      <td>mud [ENVO:01000001]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>marine sediment</td>\n",
       "      <td>854</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>True</td>\n",
       "      <td>marine sediment [ENVO:03000033]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sediment</td>\n",
       "      <td>702</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td></td>\n",
       "      <td>sediment</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "      <td>True</td>\n",
       "      <td>sediment [ENVO:00002007]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sediment</td>\n",
       "      <td>450</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Sediment</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "      <td>True</td>\n",
       "      <td>sediment [ENVO:00002007]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>185</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>...</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00002113</td>\n",
       "      <td>deep marine sediment</td>\n",
       "      <td>True</td>\n",
       "      <td>deep marine sediment [ENVO:00002113]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>Sequences</td>\n",
       "      <td>1</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Sequences</td>\n",
       "      <td>Sequences</td>\n",
       "      <td></td>\n",
       "      <td>Sequences</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td></td>\n",
       "      <td>Marine sediment</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>True</td>\n",
       "      <td>marine sediment [ENVO:03000033]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td></td>\n",
       "      <td>Lake Sediment</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000546</td>\n",
       "      <td>lake sediment</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00000546</td>\n",
       "      <td>lake sediment</td>\n",
       "      <td>True</td>\n",
       "      <td>lake sediment [ENVO:00000546]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td></td>\n",
       "      <td>Environmental_sample</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td></td>\n",
       "      <td>Crude Oil</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              env_medium  count  \\\n",
       "0                                    mud   3917   \n",
       "1                        marine sediment    854   \n",
       "2                               sediment    702   \n",
       "3                               Sediment    450   \n",
       "4   deep marine sediment [ENVO:00002113]    185   \n",
       "..                                   ...    ...   \n",
       "70                             Sequences      1   \n",
       "71                       Marine sediment      1   \n",
       "72                         Lake Sediment      1   \n",
       "73                  Environmental_sample      1   \n",
       "74                             Crude Oil      1   \n",
       "\n",
       "                                repeated  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "70                             Sequences   \n",
       "71                       Marine sediment   \n",
       "72                         Lake Sediment   \n",
       "73                  Environmental_sample   \n",
       "74                             Crude Oil   \n",
       "\n",
       "                                splitted  part_count  seq  \\\n",
       "0                                    mud         1.0  1.0   \n",
       "1                        marine sediment         1.0  1.0   \n",
       "2                               sediment         1.0  1.0   \n",
       "3                               Sediment         1.0  1.0   \n",
       "4   deep marine sediment [ENVO:00002113]         1.0  1.0   \n",
       "..                                   ...         ...  ...   \n",
       "70                             Sequences         1.0  1.0   \n",
       "71                       Marine sediment         1.0  1.0   \n",
       "72                         Lake Sediment         1.0  1.0   \n",
       "73                  Environmental_sample         1.0  1.0   \n",
       "74                             Crude Oil         1.0  1.0   \n",
       "\n",
       "                               term_tidy  \\\n",
       "0                                    mud   \n",
       "1                        marine sediment   \n",
       "2                               sediment   \n",
       "3                               Sediment   \n",
       "4   deep marine sediment [ENVO:00002113]   \n",
       "..                                   ...   \n",
       "70                             Sequences   \n",
       "71                       Marine sediment   \n",
       "72                         Lake Sediment   \n",
       "73                  Environmental_sample   \n",
       "74                             Crude Oil   \n",
       "\n",
       "                                  string        extract  \\\n",
       "0                                    mud                  \n",
       "1                        marine sediment                  \n",
       "2                               sediment                  \n",
       "3                               Sediment                  \n",
       "4   deep marine sediment [ENVO:00002113]  ENVO:00002113   \n",
       "..                                   ...            ...   \n",
       "70                             Sequences                  \n",
       "71                       Marine sediment                  \n",
       "72                         Lake Sediment                  \n",
       "73                  Environmental_sample                  \n",
       "74                             Crude Oil                  \n",
       "\n",
       "         remaining_string  ...           lab_from_id lfi_cosine  \\\n",
       "0                     mud  ...                              1.0   \n",
       "1         marine sediment  ...                              1.0   \n",
       "2                sediment  ...                              1.0   \n",
       "3                Sediment  ...                              1.0   \n",
       "4   deep marine sediment   ...  deep marine sediment        0.0   \n",
       "..                    ...  ...                   ...        ...   \n",
       "70              Sequences  ...                              1.0   \n",
       "71        Marine sediment  ...                              1.0   \n",
       "72          Lake Sediment  ...                              1.0   \n",
       "73   Environmental_sample  ...                              1.0   \n",
       "74              Crude Oil  ...                              1.0   \n",
       "\n",
       "   term_id_from_lab                 value onto_prefix rank   consensus_id  \\\n",
       "0     ENVO:01000001                   mud        envo  0.0  ENVO:01000001   \n",
       "1     ENVO:03000033       marine sediment        envo  0.0  ENVO:03000033   \n",
       "2     ENVO:00002007              sediment        envo  0.0  ENVO:00002007   \n",
       "3     ENVO:00002007              sediment        envo  0.0  ENVO:00002007   \n",
       "4     ENVO:00002113  deep marine sediment        envo  0.0  ENVO:00002113   \n",
       "..              ...                   ...         ...  ...            ...   \n",
       "70              NaN                   NaN              NaN            NaN   \n",
       "71    ENVO:03000033       marine sediment        envo  0.0  ENVO:03000033   \n",
       "72    ENVO:00000546         lake sediment        envo  0.0  ENVO:00000546   \n",
       "73              NaN                   NaN              NaN            NaN   \n",
       "74              NaN                   NaN              NaN            NaN   \n",
       "\n",
       "           consensus_lab id_or_lab_ok                   assembled_consensus  \n",
       "0                    mud         True                   mud [ENVO:01000001]  \n",
       "1        marine sediment         True       marine sediment [ENVO:03000033]  \n",
       "2               sediment         True              sediment [ENVO:00002007]  \n",
       "3               sediment         True              sediment [ENVO:00002007]  \n",
       "4   deep marine sediment         True  deep marine sediment [ENVO:00002113]  \n",
       "..                   ...          ...                                   ...  \n",
       "70                   NaN        False                                   NaN  \n",
       "71       marine sediment         True       marine sediment [ENVO:03000033]  \n",
       "72         lake sediment         True         lake sediment [ENVO:00000546]  \n",
       "73                   NaN        False                                   NaN  \n",
       "74                   NaN        False                                   NaN  \n",
       "\n",
       "[75 rows x 22 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13d44c8d",
   "metadata": {},
   "source": [
    "## For which Biosample annotations were no easy mappings found?\n",
    "How many Biosamples uses those annotations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c29338b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>106</td>\n",
       "      <td>contaminated sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>72</td>\n",
       "      <td>sediments</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>36</td>\n",
       "      <td>enrichment culture</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20</td>\n",
       "      <td>cryoconite sediments</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>freshwater lake sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>12</td>\n",
       "      <td>genomic dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>11</td>\n",
       "      <td>microbial mats</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>10</td>\n",
       "      <td>sediment sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>wet sediments</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>clay layer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>6</td>\n",
       "      <td>sea mud</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>5</td>\n",
       "      <td>sediment depth 1 2cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>5</td>\n",
       "      <td>sediment depth 0 1cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>4</td>\n",
       "      <td>salt crust</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4</td>\n",
       "      <td>mats</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4</td>\n",
       "      <td>wet sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>3</td>\n",
       "      <td>dry sediments</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>3</td>\n",
       "      <td>dy sediments</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>3</td>\n",
       "      <td>thermophilic anaerobic sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>3</td>\n",
       "      <td>mesophilic anaerobic sludge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2</td>\n",
       "      <td>solid waste</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>sequences</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>sediment2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>surface sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1</td>\n",
       "      <td>stainless steel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>sediment3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1</td>\n",
       "      <td>anode sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>sediment1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1</td>\n",
       "      <td>sediment at spring source</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1</td>\n",
       "      <td>sediemnt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1</td>\n",
       "      <td>chimney sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1</td>\n",
       "      <td>environmental sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1</td>\n",
       "      <td>deep sea surface sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1</td>\n",
       "      <td>crude oil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1</td>\n",
       "      <td>clay loam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1</td>\n",
       "      <td>salt pan crust</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    count               remaining_tidied\n",
       "0     106          contaminated sediment\n",
       "1      72                      sediments\n",
       "2      36             enrichment culture\n",
       "3      20           cryoconite sediments\n",
       "4      16       freshwater lake sediment\n",
       "5      12                    genomic dna\n",
       "6      11                 microbial mats\n",
       "7      10                sediment sample\n",
       "8       9                  wet sediments\n",
       "9       6                     clay layer\n",
       "10      6                        sea mud\n",
       "11      5           sediment depth 1 2cm\n",
       "12      5           sediment depth 0 1cm\n",
       "13      4                     salt crust\n",
       "14      4                           mats\n",
       "15      4                   wet sediment\n",
       "16      3                  dry sediments\n",
       "17      3                   dy sediments\n",
       "18      3  thermophilic anaerobic sludge\n",
       "19      3    mesophilic anaerobic sludge\n",
       "20      2                    solid waste\n",
       "21      1                      sequences\n",
       "22      1                      sediment2\n",
       "23      1               surface sediment\n",
       "24      1                stainless steel\n",
       "25      1                      sediment3\n",
       "26      1                 anode sediment\n",
       "27      1                      sediment1\n",
       "28      1      sediment at spring source\n",
       "29      1                       sediemnt\n",
       "30      1                 chimney sample\n",
       "31      1           environmental sample\n",
       "32      1      deep sea surface sediment\n",
       "33      1                      crude oil\n",
       "34      1                      clay loam\n",
       "35      1                 salt pan crust"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flag = ~mapping_candidates.id_or_lab_ok\n",
    "needs_search = mapping_candidates.loc[flag]\n",
    "\n",
    "needs_search = needs_search[[\"remaining_tidied\", \"count\"]]\n",
    "\n",
    "sum_by_needed = needs_search.groupby(\"remaining_tidied\")[\"count\"].sum()\n",
    "\n",
    "sum_by_needed = sum_by_needed.to_frame()\n",
    "sum_by_needed[\"remaining_tidied\"] = sum_by_needed.index\n",
    "\n",
    "sum_by_needed = sum_by_needed.sort_values(\"count\", ascending=False)\n",
    "sum_by_needed.reset_index(drop=True)\n",
    "\n",
    "# sum_by_needed.to_sql(\"sum_by_needed\", biosample_cnx, if_exists=\"replace\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "91fec97a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sum_by_needed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53bcffc5",
   "metadata": {},
   "source": [
    "## Extract the tidied strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "02a43f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_raw_list = list(sum_by_needed[\"remaining_tidied\"])\n",
    "ebs_raw_list.sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bd91eb2",
   "metadata": {},
   "source": [
    "## Submit those tidied strings to a search engine\n",
    "\n",
    "Specifically OLS search. This takes roughly one second per unique post-tidied submission\n",
    "\n",
    "_Turn logging back on to show status?_\n",
    "_Print the count and pre- and post- datestamps_\n",
    "`datetime.now().replace(microsecond=0).isoformat()` ?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9aecf3bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-07-04T19:10:17\n",
      "36\n",
      "0.6\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now().replace(microsecond=0).isoformat())\n",
    "print(len(sum_by_needed.index))\n",
    "print((len(sum_by_needed.index)) / 60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "08e39629",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    ebs_raw_list,\n",
    "    bad_chars=chars_to_whiteout,\n",
    "    cat_name=biosample_col_to_map,\n",
    "    ontoprefix=additional_preferred_ontologies_str,\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=my_string_dist_arg,\n",
    ")\n",
    "# ebs_search_res.to_sql(\n",
    "#     \"ols_search_results\", biosample_cnx, if_exists=\"replace\", index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae3d29dc",
   "metadata": {},
   "source": [
    "## Filter out the best of the acceptable mappings\n",
    "From a string distance perspective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b695a649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>crude oil</td>\n",
       "      <td>crude oil</td>\n",
       "      <td>crude oil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002984</td>\n",
       "      <td>petroleum</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002984</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>enrichment culture</td>\n",
       "      <td>enrichment culture</td>\n",
       "      <td>enrichment culture</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCBITaxon:1566338</td>\n",
       "      <td>enrichment culture</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1566338</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>1</td>\n",
       "      <td>0.039</td>\n",
       "      <td>ENVO:01000008</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000008</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>surface sediment</td>\n",
       "      <td>surface sediment</td>\n",
       "      <td>surface sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01001192</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001192</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       category                 raw               query                name  \\\n",
       "43   env_medium           crude oil           crude oil           crude oil   \n",
       "86   env_medium  enrichment culture  enrichment culture  enrichment culture   \n",
       "118  env_medium      microbial mats      microbial mats       microbial mat   \n",
       "257  env_medium    surface sediment    surface sediment    surface sediment   \n",
       "\n",
       "     string_dist_rank  string_dist             obo_id               label  \\\n",
       "43                  1        0.000      ENVO:00002984           petroleum   \n",
       "86                  1        0.000  NCBITaxon:1566338  enrichment culture   \n",
       "118                 1        0.039      ENVO:01000008       microbial mat   \n",
       "257                 1        0.000      ENVO:01001192    sediment surface   \n",
       "\n",
       "     search_rank ontology_prefix                scope        type  \\\n",
       "43             5            ENVO  has_related_synonym  annotation   \n",
       "86             1       NCBITAXON                label       label   \n",
       "118            1            ENVO                label       label   \n",
       "257            1            ENVO              synonym     synonym   \n",
       "\n",
       "                                                  iri ontology_name  \n",
       "43       http://purl.obolibrary.org/obo/ENVO_00002984          envo  \n",
       "86   http://purl.obolibrary.org/obo/NCBITaxon_1566338     ncbitaxon  \n",
       "118      http://purl.obolibrary.org/obo/ENVO_01000008          envo  \n",
       "257      http://purl.obolibrary.org/obo/ENVO_01001192          envo  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    ebs_search_res, max_string_dist=my_max_string_dist\n",
    ")\n",
    "my_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43aa5018",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Filter out the submissions with no acceptable matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "39e98bf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>anoxic sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>0.259</td>\n",
       "      <td>ENVO:00002045</td>\n",
       "      <td>anaerobic sediment</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002045</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>sediment</td>\n",
       "      <td>2</td>\n",
       "      <td>0.266</td>\n",
       "      <td>ENVO:00002007</td>\n",
       "      <td>sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002007</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>3</td>\n",
       "      <td>0.333</td>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000033</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>marine sediments</td>\n",
       "      <td>4</td>\n",
       "      <td>0.355</td>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000033</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>anode sediment</td>\n",
       "      <td>biogenous sediment</td>\n",
       "      <td>5</td>\n",
       "      <td>0.364</td>\n",
       "      <td>ENVO:01000082</td>\n",
       "      <td>biogenous sediment</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000082</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>2</td>\n",
       "      <td>0.383</td>\n",
       "      <td>ENVO:03000033</td>\n",
       "      <td>marine sediment</td>\n",
       "      <td>4</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_03000033</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>wet</td>\n",
       "      <td>3</td>\n",
       "      <td>0.592</td>\n",
       "      <td>PATO:0001823</td>\n",
       "      <td>wet</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0001823</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>wet microburst</td>\n",
       "      <td>4</td>\n",
       "      <td>0.760</td>\n",
       "      <td>ENVO:01001756</td>\n",
       "      <td>wet microburst</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001756</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>wet macroburst</td>\n",
       "      <td>5</td>\n",
       "      <td>0.760</td>\n",
       "      <td>ENVO:01001753</td>\n",
       "      <td>wet macroburst</td>\n",
       "      <td>3</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001753</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>wet sediments</td>\n",
       "      <td>wet pitfall trap</td>\n",
       "      <td>6</td>\n",
       "      <td>0.776</td>\n",
       "      <td>ENVO:01000978</td>\n",
       "      <td>wet pitfall trap</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000978</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>250 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       category             raw           query                name  \\\n",
       "2    env_medium  anode sediment  anode sediment     anoxic sediment   \n",
       "0    env_medium  anode sediment  anode sediment            sediment   \n",
       "3    env_medium  anode sediment  anode sediment     marine sediment   \n",
       "4    env_medium  anode sediment  anode sediment    marine sediments   \n",
       "6    env_medium  anode sediment  anode sediment  biogenous sediment   \n",
       "..          ...             ...             ...                 ...   \n",
       "279  env_medium   wet sediments   wet sediments     marine sediment   \n",
       "276  env_medium   wet sediments   wet sediments                 wet   \n",
       "277  env_medium   wet sediments   wet sediments      wet microburst   \n",
       "278  env_medium   wet sediments   wet sediments      wet macroburst   \n",
       "281  env_medium   wet sediments   wet sediments    wet pitfall trap   \n",
       "\n",
       "     string_dist_rank  string_dist         obo_id               label  \\\n",
       "2                   1        0.259  ENVO:00002045  anaerobic sediment   \n",
       "0                   2        0.266  ENVO:00002007            sediment   \n",
       "3                   3        0.333  ENVO:03000033     marine sediment   \n",
       "4                   4        0.355  ENVO:03000033     marine sediment   \n",
       "6                   5        0.364  ENVO:01000082  biogenous sediment   \n",
       "..                ...          ...            ...                 ...   \n",
       "279                 2        0.383  ENVO:03000033     marine sediment   \n",
       "276                 3        0.592   PATO:0001823                 wet   \n",
       "277                 4        0.760  ENVO:01001756      wet microburst   \n",
       "278                 5        0.760  ENVO:01001753      wet macroburst   \n",
       "281                 6        0.776  ENVO:01000978    wet pitfall trap   \n",
       "\n",
       "     search_rank ontology_prefix    scope     type  \\\n",
       "2              2            ENVO  synonym  synonym   \n",
       "0              1            ENVO    label    label   \n",
       "3              3            ENVO    label    label   \n",
       "4              3            ENVO  synonym  synonym   \n",
       "6              5            ENVO    label    label   \n",
       "..           ...             ...      ...      ...   \n",
       "279            4            ENVO    label    label   \n",
       "276            1            ENVO    label    label   \n",
       "277            2            ENVO    label    label   \n",
       "278            3            ENVO    label    label   \n",
       "281            5            ENVO    label    label   \n",
       "\n",
       "                                              iri ontology_name  \n",
       "2    http://purl.obolibrary.org/obo/ENVO_00002045          envo  \n",
       "0    http://purl.obolibrary.org/obo/ENVO_00002007          envo  \n",
       "3    http://purl.obolibrary.org/obo/ENVO_03000033          envo  \n",
       "4    http://purl.obolibrary.org/obo/ENVO_03000033          envo  \n",
       "6    http://purl.obolibrary.org/obo/ENVO_01000082          envo  \n",
       "..                                            ...           ...  \n",
       "279  http://purl.obolibrary.org/obo/ENVO_03000033          envo  \n",
       "276   http://purl.obolibrary.org/obo/PATO_0001823          envo  \n",
       "277  http://purl.obolibrary.org/obo/ENVO_01001756          envo  \n",
       "278  http://purl.obolibrary.org/obo/ENVO_01001753          envo  \n",
       "281  http://purl.obolibrary.org/obo/ENVO_01000978          envo  \n",
       "\n",
       "[250 rows x 14 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_acceptable_mappings = scoped_mapping.get_no_acceptable_mappings(\n",
    "    ebs_search_res, my_best_acceptable\n",
    ")\n",
    "\n",
    "no_acceptable_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70be6244",
   "metadata": {},
   "source": [
    "## Try searching the failures against all ontologies in OLS\n",
    "what would it be like to use a local OLS instance? (latency-wise)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d5adfca5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['anode sediment',\n",
       " 'chimney sample',\n",
       " 'clay layer',\n",
       " 'clay loam',\n",
       " 'contaminated sediment',\n",
       " 'cryoconite sediments',\n",
       " 'deep sea surface sediment',\n",
       " 'dry sediments',\n",
       " 'dy sediments',\n",
       " 'environmental sample',\n",
       " 'freshwater lake sediment',\n",
       " 'genomic dna',\n",
       " 'mats',\n",
       " 'mesophilic anaerobic sludge',\n",
       " 'salt crust',\n",
       " 'salt pan crust',\n",
       " 'sea mud',\n",
       " 'sediemnt',\n",
       " 'sediment at spring source',\n",
       " 'sediment depth 0 1cm',\n",
       " 'sediment depth 1 2cm',\n",
       " 'sediment sample',\n",
       " 'sediment1',\n",
       " 'sediment2',\n",
       " 'sediment3',\n",
       " 'sediments',\n",
       " 'sequences',\n",
       " 'solid waste',\n",
       " 'stainless steel',\n",
       " 'thermophilic anaerobic sludge',\n",
       " 'wet sediment',\n",
       " 'wet sediments']"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "still_unmapped = list(set(list(no_acceptable_mappings[\"raw\"])))\n",
    "still_unmapped.sort()\n",
    "\n",
    "print(len(still_unmapped))\n",
    "\n",
    "still_unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "93462447",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-07-04T21:43:47\n",
      "36\n",
      "0.6\n"
     ]
    }
   ],
   "source": [
    "print(datetime.now().replace(microsecond=0).isoformat())\n",
    "print(len(sum_by_needed.index))\n",
    "print((len(sum_by_needed.index)) / 60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ac5c4f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "salvage_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    still_unmapped,\n",
    "    bad_chars=\"._-\",\n",
    "    cat_name=\"salvage\",\n",
    "    ontoprefix=\"\",\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=2,\n",
    ")\n",
    "\n",
    "# salvage_search_res.to_sql(\n",
    "#     \"salvage_search_res\", biosample_cnx, if_exists=\"replace\", index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "453b8356",
   "metadata": {},
   "source": [
    "## We appear to be at a point of diminishing returns\n",
    "At the very least, will require some review"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d44fe45b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>salvage</td>\n",
       "      <td>clay layer</td>\n",
       "      <td>clay layer</td>\n",
       "      <td>clay</td>\n",
       "      <td>1</td>\n",
       "      <td>0.199</td>\n",
       "      <td>ENVO:00002982</td>\n",
       "      <td>clay</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002982</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>salvage</td>\n",
       "      <td>contaminated sediment</td>\n",
       "      <td>contaminated sediment</td>\n",
       "      <td>contaminated sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:00002114</td>\n",
       "      <td>contaminated sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002114</td>\n",
       "      <td>micro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>salvage</td>\n",
       "      <td>deep sea surface sediment</td>\n",
       "      <td>deep sea surface sediment</td>\n",
       "      <td>surface sediment</td>\n",
       "      <td>1</td>\n",
       "      <td>0.178</td>\n",
       "      <td>ENVO:01001192</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001192</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>salvage</td>\n",
       "      <td>environmental sample</td>\n",
       "      <td>environmental sample</td>\n",
       "      <td>Cichlidae environmental sample</td>\n",
       "      <td>1</td>\n",
       "      <td>0.177</td>\n",
       "      <td>NCBITaxon:2497813</td>\n",
       "      <td>Cichlidae environmental sample</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2497813</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>salvage</td>\n",
       "      <td>genomic dna</td>\n",
       "      <td>genomic dna</td>\n",
       "      <td>Genomic DNA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C95940</td>\n",
       "      <td>Genomic DNA</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C95940</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>salvage</td>\n",
       "      <td>mats</td>\n",
       "      <td>mats</td>\n",
       "      <td>mats</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mats</td>\n",
       "      <td>1</td>\n",
       "      <td>CCO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://identifiers.org/ncbigene/42634</td>\n",
       "      <td>cco</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>salvage</td>\n",
       "      <td>mesophilic anaerobic sludge</td>\n",
       "      <td>mesophilic anaerobic sludge</td>\n",
       "      <td>anaerobic sludge</td>\n",
       "      <td>1</td>\n",
       "      <td>0.199</td>\n",
       "      <td>ENVO:00002129</td>\n",
       "      <td>anaerobic sludge</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002129</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>salvage</td>\n",
       "      <td>sequences</td>\n",
       "      <td>sequences</td>\n",
       "      <td>sequences</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>data:2044</td>\n",
       "      <td>Sequence</td>\n",
       "      <td>1</td>\n",
       "      <td>EDAM</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://edamontology.org/data_2044</td>\n",
       "      <td>edam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>salvage</td>\n",
       "      <td>solid waste</td>\n",
       "      <td>solid waste</td>\n",
       "      <td>Solid Waste</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>OMIT:0028125</td>\n",
       "      <td>Solid Waste</td>\n",
       "      <td>1</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0028125</td>\n",
       "      <td>omit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>salvage</td>\n",
       "      <td>stainless steel</td>\n",
       "      <td>stainless steel</td>\n",
       "      <td>Stainless Steel</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>OMIT:0014071</td>\n",
       "      <td>Stainless Steel</td>\n",
       "      <td>1</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0014071</td>\n",
       "      <td>omit</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    category                          raw                        query  \\\n",
       "38   salvage                   clay layer                   clay layer   \n",
       "88   salvage        contaminated sediment        contaminated sediment   \n",
       "107  salvage    deep sea surface sediment    deep sea surface sediment   \n",
       "148  salvage         environmental sample         environmental sample   \n",
       "162  salvage                  genomic dna                  genomic dna   \n",
       "175  salvage                         mats                         mats   \n",
       "192  salvage  mesophilic anaerobic sludge  mesophilic anaerobic sludge   \n",
       "418  salvage                    sequences                    sequences   \n",
       "431  salvage                  solid waste                  solid waste   \n",
       "439  salvage              stainless steel              stainless steel   \n",
       "\n",
       "                               name  string_dist_rank  string_dist  \\\n",
       "38                             clay                 1        0.199   \n",
       "88            contaminated sediment                 1        0.000   \n",
       "107                surface sediment                 1        0.178   \n",
       "148  Cichlidae environmental sample                 1        0.177   \n",
       "162                     Genomic DNA                 1        0.000   \n",
       "175                            mats                 1        0.000   \n",
       "192                anaerobic sludge                 1        0.199   \n",
       "418                       sequences                 1        0.000   \n",
       "431                     Solid Waste                 1        0.000   \n",
       "439                 Stainless Steel                 1        0.000   \n",
       "\n",
       "                obo_id                           label  search_rank  \\\n",
       "38       ENVO:00002982                            clay            1   \n",
       "88       ENVO:00002114           contaminated sediment            1   \n",
       "107      ENVO:01001192                sediment surface            2   \n",
       "148  NCBITaxon:2497813  Cichlidae environmental sample            1   \n",
       "162        NCIT:C95940                     Genomic DNA            1   \n",
       "175                NaN                            mats            1   \n",
       "192      ENVO:00002129                anaerobic sludge            1   \n",
       "418          data:2044                        Sequence            1   \n",
       "431       OMIT:0028125                     Solid Waste            1   \n",
       "439       OMIT:0014071                 Stainless Steel            1   \n",
       "\n",
       "    ontology_prefix    scope     type  \\\n",
       "38             ENVO    label    label   \n",
       "88            MICRO    label    label   \n",
       "107            ENVO  synonym  synonym   \n",
       "148       NCBITAXON    label    label   \n",
       "162            NCIT    label    label   \n",
       "175             CCO    label    label   \n",
       "192            ENVO    label    label   \n",
       "418            EDAM  synonym  synonym   \n",
       "431            OMIT    label    label   \n",
       "439            OMIT    label    label   \n",
       "\n",
       "                                                  iri ontology_name  \n",
       "38       http://purl.obolibrary.org/obo/ENVO_00002982          envo  \n",
       "88       http://purl.obolibrary.org/obo/ENVO_00002114         micro  \n",
       "107      http://purl.obolibrary.org/obo/ENVO_01001192          envo  \n",
       "148  http://purl.obolibrary.org/obo/NCBITaxon_2497813     ncbitaxon  \n",
       "162        http://purl.obolibrary.org/obo/NCIT_C95940          ncit  \n",
       "175             http://identifiers.org/ncbigene/42634           cco  \n",
       "192      http://purl.obolibrary.org/obo/ENVO_00002129          envo  \n",
       "418                 http://edamontology.org/data_2044          edam  \n",
       "431       http://purl.obolibrary.org/obo/OMIT_0028125          omit  \n",
       "439       http://purl.obolibrary.org/obo/OMIT_0014071          omit  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "salvage_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    salvage_search_res, max_string_dist=salvage_max_dist\n",
    ")\n",
    "salvage_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6bd9e5e",
   "metadata": {},
   "source": [
    "## Save the mappings for review!\n",
    "\n",
    "_Note that the merge on ID or label mappings are not included here_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a2b88c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_best_acceptable[include_col] = first_pass_include_val\n",
    "salvage_best_acceptable[include_col] = salvage_include_val\n",
    "best_and_salvage = pd.concat([my_best_acceptable, salvage_best_acceptable])\n",
    "\n",
    "best_and_salvage[\"ols_curation_notes\"] = \"\"\n",
    "\n",
    "# best_and_salvage.to_csv(ols_review_file, sep=ols_review_seperator, index=False)\n",
    "\n",
    "best_and_salvage = best_and_salvage[\n",
    "    [\n",
    "        \"category\",\n",
    "        \"raw\",\n",
    "        \"query\",\n",
    "        \"name\",\n",
    "        \"obo_id\",\n",
    "        \"label\",\n",
    "        \"ontology_prefix\",\n",
    "        \"include\",\n",
    "        \"ols_curation_notes\",\n",
    "        \"scope\",\n",
    "        \"type\",\n",
    "        \"iri\",\n",
    "        \"ontology_name\",\n",
    "        \"string_dist_rank\",\n",
    "        \"string_dist\",\n",
    "        \"search_rank\",\n",
    "    ]\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d90a12de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>include</th>\n",
       "      <th>ols_curation_notes</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>search_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>salvage</td>\n",
       "      <td>mats</td>\n",
       "      <td>mats</td>\n",
       "      <td>mats</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mats</td>\n",
       "      <td>CCO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://identifiers.org/ncbigene/42634</td>\n",
       "      <td>cco</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>salvage</td>\n",
       "      <td>sequences</td>\n",
       "      <td>sequences</td>\n",
       "      <td>sequences</td>\n",
       "      <td>data:2044</td>\n",
       "      <td>Sequence</td>\n",
       "      <td>EDAM</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://edamontology.org/data_2044</td>\n",
       "      <td>edam</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>salvage</td>\n",
       "      <td>clay layer</td>\n",
       "      <td>clay layer</td>\n",
       "      <td>clay</td>\n",
       "      <td>ENVO:00002982</td>\n",
       "      <td>clay</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002982</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.199</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>salvage</td>\n",
       "      <td>deep sea surface sediment</td>\n",
       "      <td>deep sea surface sediment</td>\n",
       "      <td>surface sediment</td>\n",
       "      <td>ENVO:01001192</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001192</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.178</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>salvage</td>\n",
       "      <td>mesophilic anaerobic sludge</td>\n",
       "      <td>mesophilic anaerobic sludge</td>\n",
       "      <td>anaerobic sludge</td>\n",
       "      <td>ENVO:00002129</td>\n",
       "      <td>anaerobic sludge</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002129</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.199</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>salvage</td>\n",
       "      <td>contaminated sediment</td>\n",
       "      <td>contaminated sediment</td>\n",
       "      <td>contaminated sediment</td>\n",
       "      <td>ENVO:00002114</td>\n",
       "      <td>contaminated sediment</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002114</td>\n",
       "      <td>micro</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>salvage</td>\n",
       "      <td>environmental sample</td>\n",
       "      <td>environmental sample</td>\n",
       "      <td>Cichlidae environmental sample</td>\n",
       "      <td>NCBITaxon:2497813</td>\n",
       "      <td>Cichlidae environmental sample</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_2497813</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>1</td>\n",
       "      <td>0.177</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>salvage</td>\n",
       "      <td>genomic dna</td>\n",
       "      <td>genomic dna</td>\n",
       "      <td>Genomic DNA</td>\n",
       "      <td>NCIT:C95940</td>\n",
       "      <td>Genomic DNA</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C95940</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>salvage</td>\n",
       "      <td>solid waste</td>\n",
       "      <td>solid waste</td>\n",
       "      <td>Solid Waste</td>\n",
       "      <td>OMIT:0028125</td>\n",
       "      <td>Solid Waste</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0028125</td>\n",
       "      <td>omit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>salvage</td>\n",
       "      <td>stainless steel</td>\n",
       "      <td>stainless steel</td>\n",
       "      <td>Stainless Steel</td>\n",
       "      <td>OMIT:0014071</td>\n",
       "      <td>Stainless Steel</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0014071</td>\n",
       "      <td>omit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>crude oil</td>\n",
       "      <td>crude oil</td>\n",
       "      <td>crude oil</td>\n",
       "      <td>ENVO:00002984</td>\n",
       "      <td>petroleum</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>has_related_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00002984</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mats</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>ENVO:01000008</td>\n",
       "      <td>microbial mat</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000008</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.039</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>surface sediment</td>\n",
       "      <td>surface sediment</td>\n",
       "      <td>surface sediment</td>\n",
       "      <td>ENVO:01001192</td>\n",
       "      <td>sediment surface</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01001192</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>env_medium</td>\n",
       "      <td>enrichment culture</td>\n",
       "      <td>enrichment culture</td>\n",
       "      <td>enrichment culture</td>\n",
       "      <td>NCBITaxon:1566338</td>\n",
       "      <td>enrichment culture</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>True</td>\n",
       "      <td></td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1566338</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       category                          raw                        query  \\\n",
       "175     salvage                         mats                         mats   \n",
       "418     salvage                    sequences                    sequences   \n",
       "38      salvage                   clay layer                   clay layer   \n",
       "107     salvage    deep sea surface sediment    deep sea surface sediment   \n",
       "192     salvage  mesophilic anaerobic sludge  mesophilic anaerobic sludge   \n",
       "88      salvage        contaminated sediment        contaminated sediment   \n",
       "148     salvage         environmental sample         environmental sample   \n",
       "162     salvage                  genomic dna                  genomic dna   \n",
       "431     salvage                  solid waste                  solid waste   \n",
       "439     salvage              stainless steel              stainless steel   \n",
       "43   env_medium                    crude oil                    crude oil   \n",
       "118  env_medium               microbial mats               microbial mats   \n",
       "257  env_medium             surface sediment             surface sediment   \n",
       "86   env_medium           enrichment culture           enrichment culture   \n",
       "\n",
       "                               name             obo_id  \\\n",
       "175                            mats                NaN   \n",
       "418                       sequences          data:2044   \n",
       "38                             clay      ENVO:00002982   \n",
       "107                surface sediment      ENVO:01001192   \n",
       "192                anaerobic sludge      ENVO:00002129   \n",
       "88            contaminated sediment      ENVO:00002114   \n",
       "148  Cichlidae environmental sample  NCBITaxon:2497813   \n",
       "162                     Genomic DNA        NCIT:C95940   \n",
       "431                     Solid Waste       OMIT:0028125   \n",
       "439                 Stainless Steel       OMIT:0014071   \n",
       "43                        crude oil      ENVO:00002984   \n",
       "118                   microbial mat      ENVO:01000008   \n",
       "257                surface sediment      ENVO:01001192   \n",
       "86               enrichment culture  NCBITaxon:1566338   \n",
       "\n",
       "                              label ontology_prefix  include  \\\n",
       "175                            mats             CCO    False   \n",
       "418                        Sequence            EDAM    False   \n",
       "38                             clay            ENVO    False   \n",
       "107                sediment surface            ENVO    False   \n",
       "192                anaerobic sludge            ENVO    False   \n",
       "88            contaminated sediment           MICRO    False   \n",
       "148  Cichlidae environmental sample       NCBITAXON    False   \n",
       "162                     Genomic DNA            NCIT    False   \n",
       "431                     Solid Waste            OMIT    False   \n",
       "439                 Stainless Steel            OMIT    False   \n",
       "43                        petroleum            ENVO     True   \n",
       "118                   microbial mat            ENVO     True   \n",
       "257                sediment surface            ENVO     True   \n",
       "86               enrichment culture       NCBITAXON     True   \n",
       "\n",
       "    ols_curation_notes                scope        type  \\\n",
       "175                                   label       label   \n",
       "418                                 synonym     synonym   \n",
       "38                                    label       label   \n",
       "107                                 synonym     synonym   \n",
       "192                                   label       label   \n",
       "88                                    label       label   \n",
       "148                                   label       label   \n",
       "162                                   label       label   \n",
       "431                                   label       label   \n",
       "439                                   label       label   \n",
       "43                      has_related_synonym  annotation   \n",
       "118                                   label       label   \n",
       "257                                 synonym     synonym   \n",
       "86                                    label       label   \n",
       "\n",
       "                                                  iri ontology_name  \\\n",
       "175             http://identifiers.org/ncbigene/42634           cco   \n",
       "418                 http://edamontology.org/data_2044          edam   \n",
       "38       http://purl.obolibrary.org/obo/ENVO_00002982          envo   \n",
       "107      http://purl.obolibrary.org/obo/ENVO_01001192          envo   \n",
       "192      http://purl.obolibrary.org/obo/ENVO_00002129          envo   \n",
       "88       http://purl.obolibrary.org/obo/ENVO_00002114         micro   \n",
       "148  http://purl.obolibrary.org/obo/NCBITaxon_2497813     ncbitaxon   \n",
       "162        http://purl.obolibrary.org/obo/NCIT_C95940          ncit   \n",
       "431       http://purl.obolibrary.org/obo/OMIT_0028125          omit   \n",
       "439       http://purl.obolibrary.org/obo/OMIT_0014071          omit   \n",
       "43       http://purl.obolibrary.org/obo/ENVO_00002984          envo   \n",
       "118      http://purl.obolibrary.org/obo/ENVO_01000008          envo   \n",
       "257      http://purl.obolibrary.org/obo/ENVO_01001192          envo   \n",
       "86   http://purl.obolibrary.org/obo/NCBITaxon_1566338     ncbitaxon   \n",
       "\n",
       "     string_dist_rank  string_dist  search_rank  \n",
       "175                 1        0.000            1  \n",
       "418                 1        0.000            1  \n",
       "38                  1        0.199            1  \n",
       "107                 1        0.178            2  \n",
       "192                 1        0.199            1  \n",
       "88                  1        0.000            1  \n",
       "148                 1        0.177            1  \n",
       "162                 1        0.000            1  \n",
       "431                 1        0.000            1  \n",
       "439                 1        0.000            1  \n",
       "43                  1        0.000            5  \n",
       "118                 1        0.039            1  \n",
       "257                 1        0.000            1  \n",
       "86                  1        0.000            1  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_and_salvage = best_and_salvage.sort_values(\n",
    "    [\"include\", \"ontology_prefix\"], ascending=(True, True)\n",
    ")\n",
    "best_and_salvage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "131b2343",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_and_salvage.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0fd18d2",
   "metadata": {},
   "source": [
    "## print mapped column and scoping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "3d8b64ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env_package_normalization.EnvPackage\n",
      "sediment\n",
      "env_medium\n"
     ]
    }
   ],
   "source": [
    "print(scoping_col)\n",
    "print(scoping_value)\n",
    "print(biosample_col_to_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "4f4a3988",
   "metadata": {},
   "outputs": [
    {
     "ename": "SystemExit",
     "evalue": "Don't skip reviewing and saving!",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m Don't skip reviewing and saving!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/scoped-mapping/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3449: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "raise SystemExit(\"Don't skip reviewing and saving!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a331a6",
   "metadata": {},
   "source": [
    "## Do some review in a spreadsheet application!\n",
    "Specifically, columns `include` and `ols_curation_notes`\n",
    "\n",
    "Then Save\n",
    "\n",
    "**Add counts back in to motivate manual lookups if necessary**\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be6da72c",
   "metadata": {},
   "source": [
    "## Now read the reviewed spreadsheet back in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c94cf3b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# curated = pd.read_csv(ols_review_file, sep=ols_review_seperator)\n",
    "# curated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49389e4e",
   "metadata": {},
   "source": [
    "## OR copy and paste"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd84b014",
   "metadata": {},
   "outputs": [],
   "source": [
    "curated = pd.read_clipboard()\n",
    "curated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ea200e4",
   "metadata": {},
   "source": [
    "## Filter on `include`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1473206f",
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = curated[\"include\"]\n",
    "curated = curated.loc[flag]\n",
    "\n",
    "curated.columns = \"ols_\" + curated.columns\n",
    "# curated.to_sql(\"curated\", biosample_cnx, if_exists=\"replace\", index=False)\n",
    "# curated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac6219e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_search_merged = mapping_candidates.merge(\n",
    "    curated, how=\"outer\", left_on=\"remaining_tidied\", right_on=\"ols_raw\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2c242b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = (\n",
    "    ~merge_search_merged[\"id_or_lab_ok\"]\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].eq(\"\")\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].isna()\n",
    ")\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_obo_id\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_id\"] = replacement\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_label\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_lab\"] = replacement\n",
    "\n",
    "replacement = (\n",
    "    merge_search_merged.loc[flag, \"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + merge_search_merged.loc[flag, \"consensus_id\"]\n",
    "    + \"]\"\n",
    ")\n",
    "\n",
    "merge_search_merged.loc[flag, \"assembled_consensus\"] = replacement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b333c111",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge_search_merged.to_sql(\n",
    "#     \"merge_search_merged\", biosample_cnx, if_exists=\"replace\", index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e772701",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "summary = merge_search_merged[[biosample_col_to_map, \"count\", \"assembled_consensus\"]]\n",
    "summary = summary.sort_values(\"count\", ascending=False)\n",
    "# summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38acac10",
   "metadata": {},
   "outputs": [],
   "source": [
    "frame_for_sssom = merge_search_merged\n",
    "\n",
    "na_flag = frame_for_sssom[\"splitted\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "na_flag = frame_for_sssom[\"consensus_id\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "raw_queries = list(frame_for_sssom[\"splitted\"])\n",
    "urlencodeds = list(map(urllib.parse.quote, raw_queries))\n",
    "urlencodeds = [sssom_subject_prefix + \":\" + item for item in urlencodeds]\n",
    "\n",
    "results_rows = len(frame_for_sssom.index)\n",
    "iso8601_stamp = datetime.now().replace(microsecond=0).isoformat()\n",
    "\n",
    "sssom_frame = {\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"subject_label\": frame_for_sssom[\"splitted\"],\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"object_id\": frame_for_sssom[\"consensus_id\"],\n",
    "    \"object_label\": frame_for_sssom[\"consensus_lab\"],\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    \"mapping_date\": [iso8601_stamp] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "    \"subject_id\": urlencodeds,\n",
    "}\n",
    "\n",
    "sssom_frame = pd.DataFrame(sssom_frame)\n",
    "\n",
    "sssom_frame.to_csv(sssom_file, sep=ols_review_seperator, index=False)\n",
    "\n",
    "# sssom_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ad82349",
   "metadata": {},
   "source": [
    "# From here down requires refactoring and generalization \n",
    "\n",
    "That may even be true for some of the blocks above this!\n",
    "\n",
    "## Join together pipe-delimited env_broad_scale mappings \n",
    "_Any other constraints? If part_count is > 1, do we need to check for empty or null env_broad_scale, repeated or splitted?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32ad2b25",
   "metadata": {},
   "outputs": [],
   "source": [
    "multi_part_flag = merge_search_merged[\"part_count\"] > 1\n",
    "multi_part_flag.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d686f6ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "multi_part_frame = merge_search_merged.loc[multi_part_flag]\n",
    "\n",
    "multi_part_frame = multi_part_frame[\n",
    "    [\n",
    "        biosample_col_to_map,\n",
    "        \"count\",\n",
    "        \"repeated\",\n",
    "        \"splitted\",\n",
    "        \"part_count\",\n",
    "        \"seq\",\n",
    "        \"consensus_id\",\n",
    "        \"consensus_lab\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "multi_part_frame = multi_part_frame.sort_values(\n",
    "    [biosample_col_to_map, \"seq\"], ascending=(True, True)\n",
    ")\n",
    "\n",
    "multi_part_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f147c6b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "part_count = multi_part_frame.groupby([biosample_col_to_map], sort=False)[\n",
    "    \"part_count\"\n",
    "].max()\n",
    "part_count.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5db86116",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_max = multi_part_frame.groupby([biosample_col_to_map], sort=False)[\"seq\"].max()\n",
    "seq_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4ccf78a",
   "metadata": {},
   "outputs": [],
   "source": [
    "parts_check = pd.DataFrame(dict(part_count=part_count, seq_max=seq_max)).reset_index()\n",
    "\n",
    "all_parts_flag = parts_check[\"seq_max\"] == parts_check[\"part_count\"]\n",
    "ebs_with_all_parts = parts_check.loc[all_parts_flag, biosample_col_to_map]\n",
    "ebs_with_all_parts\n",
    "\n",
    "all_parts_flag = multi_part_frame[biosample_col_to_map].isin(list(ebs_with_all_parts))\n",
    "all_parts_frame = multi_part_frame.loc[all_parts_flag]\n",
    "all_parts_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0cbd251",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_pipesep_ebs = list(set(list(all_parts_frame[biosample_col_to_map])))\n",
    "unique_pipesep_ebs.sort()\n",
    "repipe_dict_list = []\n",
    "for one_pipesep in unique_pipesep_ebs:\n",
    "    flag = all_parts_frame[biosample_col_to_map] == one_pipesep\n",
    "    temp_frame = all_parts_frame.loc[flag]\n",
    "    ids_list = list(temp_frame[\"consensus_id\"])\n",
    "    lab_list = list(temp_frame[\"consensus_lab\"])\n",
    "    s = \"|\"\n",
    "    ids_repipe = s.join(ids_list)\n",
    "    lab_repipe = s.join(lab_list)\n",
    "\n",
    "    repipe_dict = {\n",
    "        biosample_col_to_map: one_pipesep,\n",
    "        \"consensus_id\": ids_repipe,\n",
    "        \"consensus_lab\": lab_repipe,\n",
    "    }\n",
    "    repipe_dict_list.append(repipe_dict)\n",
    "repipe_frame = pd.DataFrame(repipe_dict_list)\n",
    "repipe_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e8d54b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sssom_min_cols = sssom_frame[[\"subject_label\", \"object_id\", \"object_label\"]]\n",
    "\n",
    "sssom_min_cols.columns = [biosample_col_to_map, \"consensus_id\", \"consensus_lab\"]\n",
    "\n",
    "# sssom_min_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25424849",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "repipe_rows = len(repipe_frame.index)\n",
    "if repipe_rows > 0:\n",
    "    direct_and_repipe = pd.concat([sssom_min_cols, repipe_frame])\n",
    "else:\n",
    "    direct_and_repipe = sssom_min_cols\n",
    "# direct_and_repipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ab3aaa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "direct_and_repipe.to_sql(\n",
    "    \"direct_and_repipe\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "922c09b3",
   "metadata": {},
   "source": [
    "## refactor this even more?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8de27d49",
   "metadata": {},
   "outputs": [],
   "source": [
    "q = (\n",
    "    \"\"\"\n",
    "select\n",
    "\tb.id, b.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\",\n",
    "\tdar.consensus_id ,\n",
    "\tdar.consensus_lab\n",
    "from\n",
    "\tdirect_and_repipe dar\n",
    "left join biosample b on\n",
    "\tb.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\" = dar.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\"\n",
    "inner join biosample_sample_taxon_summary stic on\n",
    "\tb.taxonomy_id = stic.biosample_taxid\n",
    "join env_package_normalization on\n",
    "\tb.env_package = env_package_normalization.env_package\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"'\"\n",
    ")\n",
    "\n",
    "[per_biosample_scoped_mapping_results, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a863e532",
   "metadata": {},
   "outputs": [],
   "source": [
    "per_biosample_scoped_mapping_results[\"notes\"] = scoping_col + \"=\" + scoping_value\n",
    "\n",
    "table_name = \"repaired_\" + biosample_col_to_map\n",
    "\n",
    "per_biosample_scoped_mapping_results.to_sql(\n",
    "    table_name, biosample_cnx, if_exists=\"append\", index=False\n",
    ")\n",
    "\n",
    "per_biosample_scoped_mapping_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7e01f17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f16f8f8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
