{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "621bbcd9",
   "metadata": {},
   "source": [
    "be really clear about what's going into repaired tables: \n",
    "- repaired\n",
    "- already OK\n",
    "- unrepairable\n",
    "\n",
    "Only ~ 215k repaired `env_package`s ?!\n",
    "- yeah, there just aren't that many annotated in the biosample metadata\n",
    "\n",
    "\n",
    "- some of this code should probably be moved back into the package\n",
    "    - splitting and repiping \n",
    "\n",
    "- read in previous SSSOMs and don't re-map? or at least double check?\n",
    "\n",
    "- don't forget to consult \"model\" field in biosample table too\n",
    "    - for scoping, in addition to en_package and taxonomy_id?\n",
    "\n",
    "- what to do about hits against obsolete classes... does OLS provide a mechanism to avoid?\n",
    "- be careful with GAZ matches\n",
    "- avoid IDless, OMIT, NCIT, MICRO, BTO\n",
    "\n",
    "- is it worth trimming and concatenating \"x1\", \"x2\", \"x3\", where x is a mappable term? numbers may be low.\n",
    "    - example = XXX\n",
    "\n",
    "- reconcile \"x biome\" and \"x\" mappings\n",
    "- take note of non-local salvage hits\n",
    "    - like XXX\n",
    "    - small/large (freshwater) lake?\n",
    "    - include native labels in the review frame, in addition to local?\n",
    "    - should use rdftab, not additional OLS calls\n",
    "    \n",
    "\n",
    "- switch from hard-coded `env_package_overrides` to something like SSSOM?\n",
    "- mappings can come from merging strategies, first pass OLS or salvage OLS\n",
    "- reflect those methods in SSSOM\n",
    "- add string distance/confidence\n",
    "    - multiple columns for multiple methods?\n",
    "\n",
    "review these SSSOM columns\n",
    "\n",
    "```\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "```\n",
    "\n",
    "- are there samples with a host_taxid and a (sample) taxonomy_id from \"unclassified\"?\n",
    "- start at uncultured or metagenome taxonomy_id?\n",
    "\n",
    "- what scoping on or mapping of  \"gold\" or \"path\"? WON'T FIND THAT IN NCBI BIOSAMPLE\n",
    "\n",
    "\n",
    "What to save where? \n",
    "- new columns in biosample sqlite? \n",
    "- delimited file dumps\n",
    "- move content from original column and insert the new values into that original column?\n",
    "- pandas clipboard integration handy for review step\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fac2962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import requests as requests\n",
    "import sqlite3\n",
    "import string\n",
    "import urllib\n",
    "import yaml\n",
    "\n",
    "import scoped_mapping\n",
    "\n",
    "from datetime import datetime\n",
    "from pkg_resources import get_distribution, DistributionNotFound\n",
    "from strsimpy.cosine import Cosine\n",
    "from xml.etree import ElementTree\n",
    "from tdda import rexpy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3300da64",
   "metadata": {},
   "source": [
    "# Settings required from user:\n",
    "\n",
    "See repo README for notes on setting up SQLite databases of OBO ontologies with semantic-sql, relation-graph and rdftab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5684af0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://www.ncbi.nlm.nih.gov/biosample/docs/packages/?format=xml\n",
    "# see also https://www.ncbi.nlm.nih.gov/biosample/docs/packages/\n",
    "biosample_packages_file = \"../target/biosample_packages.xml\"\n",
    "\n",
    "# from ftp://ftp.ncbi.nlm.nih.gov//biosample/biosample_set.xml.gz\n",
    "# via harmonized_table.db.gz\n",
    "# in https://drive.google.com/drive/u/0/folders/1eL0v0stoduahjDpoDJIk3z2pJBAU4b2Y\n",
    "biosample_sqlite_file = \"../target/harmonized_table.db\"\n",
    "ncbitaxon_sqlite_file = \"../semantic-sql/db/ncbitaxon.db\"\n",
    "\n",
    "\n",
    "# where do we require a single ontology and where can we use multiple?\n",
    "primary_onto_prefix = \"ENVO\"\n",
    "# 'gaz',\n",
    "# need to rank results\n",
    "additional_preferred_ontologies = [\"po\", \"ncbitaxon\"]\n",
    "\n",
    "chars_to_whiteout = \"._-\"\n",
    "my_query_fields = \"\"  # OLS weighted default\n",
    "my_row_req = 3\n",
    "my_string_dist_arg = 2\n",
    "my_max_string_dist = 0.1\n",
    "salvage_max_dist = 0.2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2d7f8bf",
   "metadata": {},
   "source": [
    "### What Biosample field should be mapped to ontology classes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b79c519",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_col_to_map = \"env_local_scale\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87e3b6f1",
   "metadata": {},
   "source": [
    "### How should the mapping effort be scoped?\n",
    "Additional taxonomic filters may be applied below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ec160085",
   "metadata": {},
   "outputs": [],
   "source": [
    "scoping_col = \"env_package_normalization.EnvPackage\"\n",
    "scoping_value = \"plant-associated\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62a568c1",
   "metadata": {},
   "source": [
    "### Settings for manual review of OLS search-based results\n",
    "Results based on merging Biosample annotations to ontology classes by label or embedded ID are not exported for review at this time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "20cdb474",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ols_review_file = \"../local/ols_review.tsv\"\n",
    "ols_review_seperator = \"\\t\"\n",
    "\n",
    "strategy_col = \"strategy\"\n",
    "include_col = \"include\"\n",
    "\n",
    "first_pass_include_val = True\n",
    "# first_pass_strategy_val = \"env_braod_scale vs envo and gaz @ 0.1\"\n",
    "salvage_include_val = False\n",
    "# salvage_strategy_val    = \"env_braod_scale vs all of ols @ 0.2\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7edfbe0",
   "metadata": {},
   "source": [
    "### Settings for SSSOM output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9d71fa5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sssom_subject_prefix = biosample_col_to_map + \"_\" + scoping_col + \"_\" + scoping_value\n",
    "sssom_file = sssom_subject_prefix + \"_SSSOM.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c8336b96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# per_biosample_mapping_results_file = \"per_biosample_scoped_mapping_results.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f21778e4",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c4494ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "biosample_cnx = sqlite3.connect(biosample_sqlite_file)\n",
    "ncbitaxon_cnx = sqlite3.connect(ncbitaxon_sqlite_file)\n",
    "primary_ontology_sqlite_file = (\n",
    "    \"../semantic-sql/db/\" + primary_onto_prefix.lower() + \".db\"\n",
    ")\n",
    "primary_ontology_cnx = sqlite3.connect(primary_ontology_sqlite_file)\n",
    "additional_preferred_ontologies.insert(0, primary_onto_prefix.lower())\n",
    "additional_preferred_ontologies_str = \",\".join(additional_preferred_ontologies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "48241248",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ontology</th>\n",
       "      <th>id_pattern</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFO</td>\n",
       "      <td>BFO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CARO</td>\n",
       "      <td>CARO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEBI</td>\n",
       "      <td>CHEBI:\\d{4,6}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENVO</td>\n",
       "      <td>ENVO:\\d{7,8}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAO</td>\n",
       "      <td>FAO:0000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FOODON</td>\n",
       "      <td>FOODON:\\d{8}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO</td>\n",
       "      <td>GO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>IAO</td>\n",
       "      <td>IAO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NCBITaxon</td>\n",
       "      <td>NCBITaxon:\\d+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OBI</td>\n",
       "      <td>OBI:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PATO</td>\n",
       "      <td>PATO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PCO</td>\n",
       "      <td>PCO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PO</td>\n",
       "      <td>PO:\\d{7}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RO</td>\n",
       "      <td>RO:0002577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>UBERON</td>\n",
       "      <td>UBERON:\\d{7}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ontology     id_pattern\n",
       "0         BFO      BFO:\\d{7}\n",
       "1        CARO     CARO:\\d{7}\n",
       "2       CHEBI  CHEBI:\\d{4,6}\n",
       "3        ENVO   ENVO:\\d{7,8}\n",
       "4         FAO    FAO:0000001\n",
       "5      FOODON   FOODON:\\d{8}\n",
       "6          GO       GO:\\d{7}\n",
       "7         IAO      IAO:\\d{7}\n",
       "8   NCBITaxon  NCBITaxon:\\d+\n",
       "9         OBI      OBI:\\d{7}\n",
       "10       PATO     PATO:\\d{7}\n",
       "11        PCO      PCO:\\d{7}\n",
       "12         PO       PO:\\d{7}\n",
       "13         RO     RO:0002577\n",
       "14     UBERON   UBERON:\\d{7}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"select * from id_patterns\"\n",
    "[id_patterns_frame, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "id_patterns_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6cea64bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BFO': 'BFO:\\\\d{7}',\n",
       " 'CARO': 'CARO:\\\\d{7}',\n",
       " 'CHEBI': 'CHEBI:\\\\d{4,6}',\n",
       " 'ENVO': 'ENVO:\\\\d{7,8}',\n",
       " 'FAO': 'FAO:0000001',\n",
       " 'FOODON': 'FOODON:\\\\d{8}',\n",
       " 'GO': 'GO:\\\\d{7}',\n",
       " 'IAO': 'IAO:\\\\d{7}',\n",
       " 'NCBITaxon': 'NCBITaxon:\\\\d+',\n",
       " 'OBI': 'OBI:\\\\d{7}',\n",
       " 'PATO': 'PATO:\\\\d{7}',\n",
       " 'PCO': 'PCO:\\\\d{7}',\n",
       " 'PO': 'PO:\\\\d{7}',\n",
       " 'RO': 'RO:0002577',\n",
       " 'UBERON': 'UBERON:\\\\d{7}'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_patterns = dict(zip(id_patterns_frame.ontology, id_patterns_frame.id_pattern))\n",
    "id_patterns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db3fd267",
   "metadata": {},
   "source": [
    "- env_package = env_package annotation from NCBI Biosample file XXX\n",
    "- count = number of biosamples using that env_package annotation\n",
    "- lhs = checklist info\n",
    "- rhs = potential package info\n",
    "- extract = potential OBO ID from rhs column (currently harcoded and only looking for ENVO IDs)\n",
    "- remaining_string = rhs/string, with potential OBO IDs removed\n",
    "- remaining_tidied = remaining_string with case, whitespace and punctuation normailzastion\n",
    "- rt_override = some remaining_tidied values can be replaced according to env_package_overrides\n",
    "- EnvPackage = corresponding de-normalized value from package_dictionary\n",
    "- is_canonical = false when EnvPackage is NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3766cbe7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_package</th>\n",
       "      <th>count</th>\n",
       "      <th>lhs</th>\n",
       "      <th>rhs</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>rt_override</th>\n",
       "      <th>EnvPackage</th>\n",
       "      <th>is_canonical</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14083847</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>host-associated</td>\n",
       "      <td>49254</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td></td>\n",
       "      <td>host-associated</td>\n",
       "      <td>host associated</td>\n",
       "      <td>host associated</td>\n",
       "      <td>host-associated</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human-gut</td>\n",
       "      <td>47921</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td></td>\n",
       "      <td>human-gut</td>\n",
       "      <td>human gut</td>\n",
       "      <td>human gut</td>\n",
       "      <td>human-gut</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>water</td>\n",
       "      <td>16367</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>water</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human-skin</td>\n",
       "      <td>13706</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td></td>\n",
       "      <td>human-skin</td>\n",
       "      <td>human skin</td>\n",
       "      <td>human skin</td>\n",
       "      <td>human-skin</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>env_package</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td></td>\n",
       "      <td>env_package</td>\n",
       "      <td>env package</td>\n",
       "      <td>env package</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td></td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>gut microbiome</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td></td>\n",
       "      <td>marine sediment (ENVO:00002113)</td>\n",
       "      <td>marine sediment envo 00002113</td>\n",
       "      <td>marine sediment envo 00002113</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>saliva</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td></td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "      <td>saliva</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td></td>\n",
       "      <td>sea water, [ENVO:00002149]</td>\n",
       "      <td>sea water envo 00002149</td>\n",
       "      <td>sea water envo 00002149</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        env_package     count lhs  \\\n",
       "0                              None  14083847       \n",
       "1                   host-associated     49254       \n",
       "2                         human-gut     47921       \n",
       "3                             water     16367       \n",
       "4                        human-skin     13706       \n",
       "..                              ...       ...  ..   \n",
       "87                      env_package         1       \n",
       "88                   gut microbiome         1       \n",
       "89  marine sediment (ENVO:00002113)         1       \n",
       "90                           saliva         1       \n",
       "91       sea water, [ENVO:00002149]         1       \n",
       "\n",
       "                                rhs extract                 remaining_string  \\\n",
       "0                              None                                            \n",
       "1                   host-associated                          host-associated   \n",
       "2                         human-gut                                human-gut   \n",
       "3                             water                                    water   \n",
       "4                        human-skin                               human-skin   \n",
       "..                              ...     ...                              ...   \n",
       "87                      env_package                              env_package   \n",
       "88                   gut microbiome                           gut microbiome   \n",
       "89  marine sediment (ENVO:00002113)          marine sediment (ENVO:00002113)   \n",
       "90                           saliva                                   saliva   \n",
       "91       sea water, [ENVO:00002149]               sea water, [ENVO:00002149]   \n",
       "\n",
       "                 remaining_tidied                    rt_override  \\\n",
       "0                                                                  \n",
       "1                 host associated                host associated   \n",
       "2                       human gut                      human gut   \n",
       "3                           water                          water   \n",
       "4                      human skin                     human skin   \n",
       "..                            ...                            ...   \n",
       "87                    env package                    env package   \n",
       "88                 gut microbiome                 gut microbiome   \n",
       "89  marine sediment envo 00002113  marine sediment envo 00002113   \n",
       "90                         saliva                         saliva   \n",
       "91        sea water envo 00002149        sea water envo 00002149   \n",
       "\n",
       "         EnvPackage  is_canonical  \n",
       "0                               1  \n",
       "1   host-associated             1  \n",
       "2         human-gut             1  \n",
       "3             water             1  \n",
       "4        human-skin             1  \n",
       "..              ...           ...  \n",
       "87             None             0  \n",
       "88             None             0  \n",
       "89             None             0  \n",
       "90             None             0  \n",
       "91             None             0  \n",
       "\n",
       "[92 rows x 10 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"select * from env_package_normalization\"\n",
    "[env_package_normalization, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx\n",
    ")\n",
    "\n",
    "env_package_normalization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2363d41",
   "metadata": {},
   "source": [
    "## Get a table of MIxS annotations to be mapped to ontology classes.\n",
    "\n",
    "Explicitly scope based on normalized package data. These values were set at the top of this notebook.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "91dcca09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env_local_scale\n",
      "env_package_normalization.EnvPackage\n",
      "plant-associated\n"
     ]
    }
   ],
   "source": [
    "print(biosample_col_to_map)\n",
    "print(scoping_col)\n",
    "print(scoping_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80727cfd",
   "metadata": {},
   "source": [
    "**In this case, the scoping includes an inner join requirement for 'unclassified entities'**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7e777c8a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count\n",
       "0              Root    558\n",
       "1        rice field    310\n",
       "2              Agar    270\n",
       "3     lettuce field    220\n",
       "4     Two-weeks old    219\n",
       "..              ...    ...\n",
       "121              12      1\n",
       "122              11      1\n",
       "123              10      1\n",
       "124               1      1\n",
       "125            None      1\n",
       "\n",
       "[126 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = (\n",
    "    \"select \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\", count(*) as count\n",
    "from\n",
    "    biosample b\n",
    "join env_package_normalization on\n",
    "    b.env_package = env_package_normalization.env_package\n",
    "inner join biosample_sample_taxon_summary stic on\n",
    "    b.taxonomy_id = stic.biosample_taxid\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"' group by \"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\"\n",
    "order by\n",
    "    count(*) desc\"\"\"\n",
    ")\n",
    "\n",
    "[mapping_candidates, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86f43bb4",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## The Biosample format allows for pipe-delimited environmental package lists. \n",
    "\n",
    "Separate those out into their components.\n",
    "\n",
    "----\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "45e1859f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         repeated       splitted  part_count  seq\n",
       "0            Root           Root           1    1\n",
       "0      rice field     rice field           1    1\n",
       "0            Agar           Agar           1    1\n",
       "0   lettuce field  lettuce field           1    1\n",
       "0   Two-weeks old  Two-weeks old           1    1\n",
       "..            ...            ...         ...  ...\n",
       "0              12             12           1    1\n",
       "0              11             11           1    1\n",
       "0              10             10           1    1\n",
       "0               1              1           1    1\n",
       "0                                          1    1\n",
       "\n",
       "[126 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_frames = []\n",
    "for row in mapping_candidates.itertuples(index=True, name=\"PandaZ\"):\n",
    "    split_check = row[1]\n",
    "    if split_check is None:\n",
    "        split_check = \"\"\n",
    "    splitted = pd.Series(split_check.split(\"|\"))\n",
    "    splitted_count = len(splitted)\n",
    "    repeated = [split_check] * splitted_count\n",
    "    repeated = pd.Series(repeated)\n",
    "    as_frame = pd.DataFrame(dict(repeated=repeated, splitted=splitted)).reset_index()\n",
    "    seq_list = list(range(1, splitted_count + 1))\n",
    "    as_frame[\"part_count\"] = splitted_count\n",
    "    as_frame[\"seq\"] = seq_list\n",
    "    multi_frames.append(as_frame)\n",
    "concat_frame = pd.concat(multi_frames)\n",
    "concat_frame = concat_frame[[\"repeated\", \"splitted\", \"part_count\", \"seq\"]]\n",
    "concat_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c67a272e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>219</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count       repeated       splitted  part_count  seq\n",
       "0              Root    558           Root           Root         1.0  1.0\n",
       "1        rice field    310     rice field     rice field         1.0  1.0\n",
       "2              Agar    270           Agar           Agar         1.0  1.0\n",
       "3     lettuce field    220  lettuce field  lettuce field         1.0  1.0\n",
       "4     Two-weeks old    219  Two-weeks old  Two-weeks old         1.0  1.0\n",
       "..              ...    ...            ...            ...         ...  ...\n",
       "121              12      1             12             12         1.0  1.0\n",
       "122              11      1             11             11         1.0  1.0\n",
       "123              10      1             10             10         1.0  1.0\n",
       "124               1      1              1              1         1.0  1.0\n",
       "125            None      1            NaN            NaN         NaN  NaN\n",
       "\n",
       "[126 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    concat_frame, left_on=biosample_col_to_map, right_on=\"repeated\", how=\"left\"\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3af421ac",
   "metadata": {},
   "source": [
    "## Normalize a few different ways `OBO` IDs have been entered\n",
    "In the Biosample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ed3a89fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rice field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Agar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>lettuce field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>219</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Two-weeks old</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count       repeated       splitted  part_count  seq  \\\n",
       "0              Root    558           Root           Root         1.0  1.0   \n",
       "1        rice field    310     rice field     rice field         1.0  1.0   \n",
       "2              Agar    270           Agar           Agar         1.0  1.0   \n",
       "3     lettuce field    220  lettuce field  lettuce field         1.0  1.0   \n",
       "4     Two-weeks old    219  Two-weeks old  Two-weeks old         1.0  1.0   \n",
       "..              ...    ...            ...            ...         ...  ...   \n",
       "121              12      1             12             12         1.0  1.0   \n",
       "122              11      1             11             11         1.0  1.0   \n",
       "123              10      1             10             10         1.0  1.0   \n",
       "124               1      1              1              1         1.0  1.0   \n",
       "125            None      1            NaN            NaN         NaN  NaN   \n",
       "\n",
       "         term_tidy  \n",
       "0             Root  \n",
       "1       rice field  \n",
       "2             Agar  \n",
       "3    lettuce field  \n",
       "4    Two-weeks old  \n",
       "..             ...  \n",
       "121             12  \n",
       "122             11  \n",
       "123             10  \n",
       "124              1  \n",
       "125            NaN  \n",
       "\n",
       "[126 rows x 7 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this helps by standardizing what could be an ID\n",
    "# but if the RHS is text, then envo should just be removed\n",
    "mapping_candidates[\"term_tidy\"] = mapping_candidates.splitted.str.replace(\n",
    "    primary_onto_prefix.lower() + \"[:_ ]\",\n",
    "    primary_onto_prefix + \":\",\n",
    "    regex=True,\n",
    "    case=False,\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cc22a1a",
   "metadata": {},
   "source": [
    "# Now try to extract ontology terms that are already present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e423fcb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td></td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td></td>\n",
       "      <td>Agar</td>\n",
       "      <td>agar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td></td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>219</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td></td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>two weeks old</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count       repeated       splitted  part_count  seq  \\\n",
       "0              Root    558           Root           Root         1.0  1.0   \n",
       "1        rice field    310     rice field     rice field         1.0  1.0   \n",
       "2              Agar    270           Agar           Agar         1.0  1.0   \n",
       "3     lettuce field    220  lettuce field  lettuce field         1.0  1.0   \n",
       "4     Two-weeks old    219  Two-weeks old  Two-weeks old         1.0  1.0   \n",
       "..              ...    ...            ...            ...         ...  ...   \n",
       "121              12      1             12             12         1.0  1.0   \n",
       "122              11      1             11             11         1.0  1.0   \n",
       "123              10      1             10             10         1.0  1.0   \n",
       "124               1      1              1              1         1.0  1.0   \n",
       "125            None      1            NaN            NaN         NaN  NaN   \n",
       "\n",
       "         term_tidy         string extract remaining_string remaining_tidied  \n",
       "0             Root           Root                     Root             root  \n",
       "1       rice field     rice field               rice field       rice field  \n",
       "2             Agar           Agar                     Agar             agar  \n",
       "3    lettuce field  lettuce field            lettuce field    lettuce field  \n",
       "4    Two-weeks old  Two-weeks old            Two-weeks old    two weeks old  \n",
       "..             ...            ...     ...              ...              ...  \n",
       "121             12             12                       12               12  \n",
       "122             11             11                       11               11  \n",
       "123             10             10                       10               10  \n",
       "124              1              1                        1                1  \n",
       "125            NaN                                                           \n",
       "\n",
       "[126 rows x 11 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_series_decomposition = scoped_mapping.decompose_series(\n",
    "    mapping_candidates[\"term_tidy\"], id_patterns[primary_onto_prefix]\n",
    ")\n",
    "\n",
    "# candidate_series_decomposition\n",
    "\n",
    "mapping_candidates = pd.concat(\n",
    "    [mapping_candidates, candidate_series_decomposition], axis=1\n",
    ")\n",
    "\n",
    "mapping_candidates[\n",
    "    \"remaining_tidied\"\n",
    "] = mapping_candidates.remaining_tidied.str.replace(\n",
    "    primary_onto_prefix.lower() + \"[:_ ]\", \"\", regex=True, case=False\n",
    ")\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccd038ff",
   "metadata": {},
   "source": [
    "## Join the extracted IDs with their labels\n",
    "\n",
    "_Start by connecting to the rdftab database from which the terms and label-like annotations will be obtained?_"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db5d19a9",
   "metadata": {},
   "source": [
    "## extracting the labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3bcd7e47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BFO:0000001</td>\n",
       "      <td>entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BFO:0000002</td>\n",
       "      <td>continuant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BFO:0000003</td>\n",
       "      <td>occurrent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BFO:0000004</td>\n",
       "      <td>independent continuant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BFO:0000006</td>\n",
       "      <td>spatial region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8350</th>\n",
       "      <td>PO:0030132</td>\n",
       "      <td>cincinnus inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8351</th>\n",
       "      <td>PO:0030133</td>\n",
       "      <td>compound cincinnus inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8352</th>\n",
       "      <td>PO:0030134</td>\n",
       "      <td>bostryx inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8353</th>\n",
       "      <td>PO:0030135</td>\n",
       "      <td>drepanium inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8354</th>\n",
       "      <td>PO:0030136</td>\n",
       "      <td>compound drepanium inflorescence</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8355 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          subject                             value\n",
       "0     BFO:0000001                            entity\n",
       "1     BFO:0000002                        continuant\n",
       "2     BFO:0000003                         occurrent\n",
       "3     BFO:0000004            independent continuant\n",
       "4     BFO:0000006                    spatial region\n",
       "...           ...                               ...\n",
       "8350   PO:0030132           cincinnus inflorescence\n",
       "8351   PO:0030133  compound cincinnus inflorescence\n",
       "8352   PO:0030134             bostryx inflorescence\n",
       "8353   PO:0030135           drepanium inflorescence\n",
       "8354   PO:0030136  compound drepanium inflorescence\n",
       "\n",
       "[8355 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = \"select stanza as subject, value from ids_labs_selected_ontolgies\"\n",
    "\n",
    "[onto_labels, query_duration] = scoped_mapping.timed_query(q, biosample_cnx)\n",
    "\n",
    "onto_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "840b922f",
   "metadata": {},
   "source": [
    "## and merging "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d82daf6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_candidates.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8543ed0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td></td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td></td>\n",
       "      <td>Agar</td>\n",
       "      <td>agar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td></td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>219</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td></td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>two weeks old</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count       repeated       splitted  part_count  seq  \\\n",
       "0              Root    558           Root           Root         1.0  1.0   \n",
       "1        rice field    310     rice field     rice field         1.0  1.0   \n",
       "2              Agar    270           Agar           Agar         1.0  1.0   \n",
       "3     lettuce field    220  lettuce field  lettuce field         1.0  1.0   \n",
       "4     Two-weeks old    219  Two-weeks old  Two-weeks old         1.0  1.0   \n",
       "..              ...    ...            ...            ...         ...  ...   \n",
       "121              12      1             12             12         1.0  1.0   \n",
       "122              11      1             11             11         1.0  1.0   \n",
       "123              10      1             10             10         1.0  1.0   \n",
       "124               1      1              1              1         1.0  1.0   \n",
       "125            None      1            NaN            NaN         NaN  NaN   \n",
       "\n",
       "         term_tidy         string extract remaining_string remaining_tidied  \\\n",
       "0             Root           Root                     Root             root   \n",
       "1       rice field     rice field               rice field       rice field   \n",
       "2             Agar           Agar                     Agar             agar   \n",
       "3    lettuce field  lettuce field            lettuce field    lettuce field   \n",
       "4    Two-weeks old  Two-weeks old            Two-weeks old    two weeks old   \n",
       "..             ...            ...     ...              ...              ...   \n",
       "121             12             12                       12               12   \n",
       "122             11             11                       11               11   \n",
       "123             10             10                       10               10   \n",
       "124              1              1                        1                1   \n",
       "125            NaN                                                            \n",
       "\n",
       "    subject value  \n",
       "0       NaN   NaN  \n",
       "1       NaN   NaN  \n",
       "2       NaN   NaN  \n",
       "3       NaN   NaN  \n",
       "4       NaN   NaN  \n",
       "..      ...   ...  \n",
       "121     NaN   NaN  \n",
       "122     NaN   NaN  \n",
       "123     NaN   NaN  \n",
       "124     NaN   NaN  \n",
       "125     NaN   NaN  \n",
       "\n",
       "[126 rows x 13 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"extract\", right_on=\"subject\", how=\"left\"\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64b3baef",
   "metadata": {},
   "source": [
    "## Use cosine string distance to see if the labels match closely enough\n",
    "\n",
    "I.e. the labels claimed by the Biosample data set and the labels asserted in the ontology. if they're close enough, consider the assigned ID legit\n",
    "\n",
    "\n",
    "_How close is close enough?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "abd0a035",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>subject</th>\n",
       "      <th>value</th>\n",
       "      <th>cosine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td></td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td></td>\n",
       "      <td>Agar</td>\n",
       "      <td>agar</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td></td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>219</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td></td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>two weeks old</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>126 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count       repeated       splitted  part_count  seq  \\\n",
       "0              Root    558           Root           Root         1.0  1.0   \n",
       "1        rice field    310     rice field     rice field         1.0  1.0   \n",
       "2              Agar    270           Agar           Agar         1.0  1.0   \n",
       "3     lettuce field    220  lettuce field  lettuce field         1.0  1.0   \n",
       "4     Two-weeks old    219  Two-weeks old  Two-weeks old         1.0  1.0   \n",
       "..              ...    ...            ...            ...         ...  ...   \n",
       "121              12      1             12             12         1.0  1.0   \n",
       "122              11      1             11             11         1.0  1.0   \n",
       "123              10      1             10             10         1.0  1.0   \n",
       "124               1      1              1              1         1.0  1.0   \n",
       "125            None      1            NaN            NaN         NaN  NaN   \n",
       "\n",
       "         term_tidy         string extract remaining_string remaining_tidied  \\\n",
       "0             Root           Root                     Root             root   \n",
       "1       rice field     rice field               rice field       rice field   \n",
       "2             Agar           Agar                     Agar             agar   \n",
       "3    lettuce field  lettuce field            lettuce field    lettuce field   \n",
       "4    Two-weeks old  Two-weeks old            Two-weeks old    two weeks old   \n",
       "..             ...            ...     ...              ...              ...   \n",
       "121             12             12                       12               12   \n",
       "122             11             11                       11               11   \n",
       "123             10             10                       10               10   \n",
       "124              1              1                        1                1   \n",
       "125            NaN                                                            \n",
       "\n",
       "    subject value  cosine  \n",
       "0       NaN           1.0  \n",
       "1       NaN           1.0  \n",
       "2       NaN           1.0  \n",
       "3       NaN           1.0  \n",
       "4       NaN           1.0  \n",
       "..      ...   ...     ...  \n",
       "121     NaN           1.0  \n",
       "122     NaN           1.0  \n",
       "123     NaN           1.0  \n",
       "124     NaN           1.0  \n",
       "125     NaN           0.0  \n",
       "\n",
       "[126 rows x 14 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_cosine_obj = Cosine(my_string_dist_arg)\n",
    "mapping_candidates[\"value\"] = mapping_candidates[\"value\"].fillna(\"\")\n",
    "mapping_candidates[\"remaining_tidied\"] = mapping_candidates[\"remaining_tidied\"].fillna(\n",
    "    \"\"\n",
    ")\n",
    "mapping_candidates[\"cosine\"] = mapping_candidates.apply(\n",
    "    lambda my_row: my_cosine_obj.distance(\n",
    "        my_row[\"remaining_tidied\"].lower(), my_row[\"value\"].lower()\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2358a89",
   "metadata": {},
   "source": [
    "**Previously, we did a reality check on the claimed IDs and labels. If a label is claimed without any ID, that could still be a path to an ontology term.**\n",
    "\n",
    "We'll be doing some merging, so make sure column names are meaningful and unique\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "092aced0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon:1</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td></td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td></td>\n",
       "      <td>Agar</td>\n",
       "      <td>agar</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count    repeated    splitted  part_count  seq  \\\n",
       "0              Root    558        Root        Root         1.0  1.0   \n",
       "1              Root    558        Root        Root         1.0  1.0   \n",
       "2              Root    558        Root        Root         1.0  1.0   \n",
       "3        rice field    310  rice field  rice field         1.0  1.0   \n",
       "4              Agar    270        Agar        Agar         1.0  1.0   \n",
       "..              ...    ...         ...         ...         ...  ...   \n",
       "123              12      1          12          12         1.0  1.0   \n",
       "124              11      1          11          11         1.0  1.0   \n",
       "125              10      1          10          10         1.0  1.0   \n",
       "126               1      1           1           1         1.0  1.0   \n",
       "127            None      1         NaN         NaN         NaN  NaN   \n",
       "\n",
       "      term_tidy      string extract remaining_string remaining_tidied term_id  \\\n",
       "0          Root        Root                     Root             root     NaN   \n",
       "1          Root        Root                     Root             root     NaN   \n",
       "2          Root        Root                     Root             root     NaN   \n",
       "3    rice field  rice field               rice field       rice field     NaN   \n",
       "4          Agar        Agar                     Agar             agar     NaN   \n",
       "..          ...         ...     ...              ...              ...     ...   \n",
       "123          12          12                       12               12     NaN   \n",
       "124          11          11                       11               11     NaN   \n",
       "125          10          10                       10               10     NaN   \n",
       "126           1           1                        1                1     NaN   \n",
       "127         NaN                                                           NaN   \n",
       "\n",
       "    lab_from_id  lfi_cosine term_id_from_lab       value  \n",
       "0                       1.0      NCBITaxon:1        root  \n",
       "1                       1.0       PO:0009005        root  \n",
       "2                       1.0       PO:0009005        root  \n",
       "3                       1.0    ENVO:00000296  rice field  \n",
       "4                       1.0              NaN         NaN  \n",
       "..          ...         ...              ...         ...  \n",
       "123                     1.0              NaN         NaN  \n",
       "124                     1.0              NaN         NaN  \n",
       "125                     1.0              NaN         NaN  \n",
       "126                     1.0              NaN         NaN  \n",
       "127                     0.0              NaN         NaN  \n",
       "\n",
       "[128 rows x 16 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates.columns = [\n",
    "    biosample_col_to_map,\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"term_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "]\n",
    "mapping_candidates = mapping_candidates.merge(\n",
    "    onto_labels, left_on=\"remaining_tidied\", right_on=\"value\", how=\"left\"\n",
    ")\n",
    "\n",
    "mapping_candidates.columns = [\n",
    "    biosample_col_to_map,\n",
    "    \"count\",\n",
    "    \"repeated\",\n",
    "    \"splitted\",\n",
    "    \"part_count\",\n",
    "    \"seq\",\n",
    "    \"term_tidy\",\n",
    "    \"string\",\n",
    "    \"extract\",\n",
    "    \"remaining_string\",\n",
    "    \"remaining_tidied\",\n",
    "    \"term_id\",\n",
    "    \"lab_from_id\",\n",
    "    \"lfi_cosine\",\n",
    "    \"term_id_from_lab\",\n",
    "    \"value\",\n",
    "]\n",
    "\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e968917d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ncbitaxon',\n",
       " 'po',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'envo',\n",
       " 'envo',\n",
       " '',\n",
       " 'envo',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'envo',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.drop_duplicates()\n",
    "\n",
    "mc_tifl_split = mapping_candidates[\"term_id_from_lab\"].str.split(\":\")\n",
    "\n",
    "import math\n",
    "\n",
    "prefix_when_possible = []\n",
    "for one_split in mc_tifl_split:\n",
    "    if type(one_split) == list:\n",
    "        os_first = one_split[0]\n",
    "        prefix_when_possible.append(os_first)\n",
    "    else:\n",
    "        prefix_when_possible.append(\"\")\n",
    "\n",
    "prefix_when_possible = [each_string.lower() for each_string in prefix_when_possible]\n",
    "\n",
    "prefix_when_possible"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c10ced03",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/scoped-mapping/sm_venv/lib/python3.9/site-packages/pandas/core/indexing.py:1597: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = value\n",
      "/Users/MAM/Documents/gitrepos/scoped-mapping/sm_venv/lib/python3.9/site-packages/pandas/core/indexing.py:1676: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_column(ilocs[0], value, pi)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>onto_prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon:1</td>\n",
       "      <td>root</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "      <td>po</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td></td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td></td>\n",
       "      <td>Agar</td>\n",
       "      <td>agar</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td></td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count       repeated       splitted  part_count  seq  \\\n",
       "0              Root    558           Root           Root         1.0  1.0   \n",
       "1              Root    558           Root           Root         1.0  1.0   \n",
       "3        rice field    310     rice field     rice field         1.0  1.0   \n",
       "4              Agar    270           Agar           Agar         1.0  1.0   \n",
       "5     lettuce field    220  lettuce field  lettuce field         1.0  1.0   \n",
       "..              ...    ...            ...            ...         ...  ...   \n",
       "123              12      1             12             12         1.0  1.0   \n",
       "124              11      1             11             11         1.0  1.0   \n",
       "125              10      1             10             10         1.0  1.0   \n",
       "126               1      1              1              1         1.0  1.0   \n",
       "127            None      1            NaN            NaN         NaN  NaN   \n",
       "\n",
       "         term_tidy         string extract remaining_string remaining_tidied  \\\n",
       "0             Root           Root                     Root             root   \n",
       "1             Root           Root                     Root             root   \n",
       "3       rice field     rice field               rice field       rice field   \n",
       "4             Agar           Agar                     Agar             agar   \n",
       "5    lettuce field  lettuce field            lettuce field    lettuce field   \n",
       "..             ...            ...     ...              ...              ...   \n",
       "123             12             12                       12               12   \n",
       "124             11             11                       11               11   \n",
       "125             10             10                       10               10   \n",
       "126              1              1                        1                1   \n",
       "127            NaN                                                            \n",
       "\n",
       "    term_id lab_from_id  lfi_cosine term_id_from_lab       value onto_prefix  \n",
       "0       NaN                     1.0      NCBITaxon:1        root   ncbitaxon  \n",
       "1       NaN                     1.0       PO:0009005        root          po  \n",
       "3       NaN                     1.0    ENVO:00000296  rice field        envo  \n",
       "4       NaN                     1.0              NaN         NaN              \n",
       "5       NaN                     1.0              NaN         NaN              \n",
       "..      ...         ...         ...              ...         ...         ...  \n",
       "123     NaN                     1.0              NaN         NaN              \n",
       "124     NaN                     1.0              NaN         NaN              \n",
       "125     NaN                     1.0              NaN         NaN              \n",
       "126     NaN                     1.0              NaN         NaN              \n",
       "127     NaN                     0.0              NaN         NaN              \n",
       "\n",
       "[127 rows x 17 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if \"onto_prefix\" in mapping_candidates.columns:\n",
    "    mapping_candidates = mapping_candidates.drop(\"onto_prefix\", 1)\n",
    "\n",
    "mapping_candidates.loc[:, \"onto_prefix\"] = pd.Series(prefix_when_possible).values\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9720bade",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>onto_prefix</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>envo</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>po</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  onto_prefix  rank\n",
       "0        envo     0\n",
       "1          po     1\n",
       "2   ncbitaxon     2"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ranks = list(range(len(additional_preferred_ontologies)))\n",
    "\n",
    "d = {\"onto_prefix\": additional_preferred_ontologies, \"rank\": ranks}\n",
    "onto_rank_frame = pd.DataFrame(d)\n",
    "\n",
    "onto_rank_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "536a6f9c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>onto_prefix</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon:1</td>\n",
       "      <td>root</td>\n",
       "      <td>ncbitaxon</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "      <td>po</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td></td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td></td>\n",
       "      <td>Agar</td>\n",
       "      <td>agar</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td></td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count       repeated       splitted  part_count  seq  \\\n",
       "0              Root    558           Root           Root         1.0  1.0   \n",
       "1              Root    558           Root           Root         1.0  1.0   \n",
       "2        rice field    310     rice field     rice field         1.0  1.0   \n",
       "3              Agar    270           Agar           Agar         1.0  1.0   \n",
       "4     lettuce field    220  lettuce field  lettuce field         1.0  1.0   \n",
       "..              ...    ...            ...            ...         ...  ...   \n",
       "122              12      1             12             12         1.0  1.0   \n",
       "123              11      1             11             11         1.0  1.0   \n",
       "124              10      1             10             10         1.0  1.0   \n",
       "125               1      1              1              1         1.0  1.0   \n",
       "126            None      1            NaN            NaN         NaN  NaN   \n",
       "\n",
       "         term_tidy         string extract remaining_string remaining_tidied  \\\n",
       "0             Root           Root                     Root             root   \n",
       "1             Root           Root                     Root             root   \n",
       "2       rice field     rice field               rice field       rice field   \n",
       "3             Agar           Agar                     Agar             agar   \n",
       "4    lettuce field  lettuce field            lettuce field    lettuce field   \n",
       "..             ...            ...     ...              ...              ...   \n",
       "122             12             12                       12               12   \n",
       "123             11             11                       11               11   \n",
       "124             10             10                       10               10   \n",
       "125              1              1                        1                1   \n",
       "126            NaN                                                            \n",
       "\n",
       "    term_id lab_from_id  lfi_cosine term_id_from_lab       value onto_prefix  \\\n",
       "0       NaN                     1.0      NCBITaxon:1        root   ncbitaxon   \n",
       "1       NaN                     1.0       PO:0009005        root          po   \n",
       "2       NaN                     1.0    ENVO:00000296  rice field        envo   \n",
       "3       NaN                     1.0              NaN         NaN               \n",
       "4       NaN                     1.0              NaN         NaN               \n",
       "..      ...         ...         ...              ...         ...         ...   \n",
       "122     NaN                     1.0              NaN         NaN               \n",
       "123     NaN                     1.0              NaN         NaN               \n",
       "124     NaN                     1.0              NaN         NaN               \n",
       "125     NaN                     1.0              NaN         NaN               \n",
       "126     NaN                     0.0              NaN         NaN               \n",
       "\n",
       "     rank  \n",
       "0     2.0  \n",
       "1     1.0  \n",
       "2     0.0  \n",
       "3     NaN  \n",
       "4     NaN  \n",
       "..    ...  \n",
       "122   NaN  \n",
       "123   NaN  \n",
       "124   NaN  \n",
       "125   NaN  \n",
       "126   NaN  \n",
       "\n",
       "[127 rows x 18 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(onto_rank_frame, on=\"onto_prefix\", how='left')\n",
    "\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "46ce3f1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>rice field</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>soils used to inoculate seedlings</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>stream</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>sugar beet tissue</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>vineyard</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>125 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       env_local_scale  rank\n",
       "0                                    1   NaN\n",
       "1                                   10   NaN\n",
       "2                                   11   NaN\n",
       "3                                   12   NaN\n",
       "4                                   13   NaN\n",
       "..                                 ...   ...\n",
       "120                         rice field   0.0\n",
       "121  soils used to inoculate seedlings   NaN\n",
       "122                             stream   0.0\n",
       "123                  sugar beet tissue   NaN\n",
       "124                           vineyard   0.0\n",
       "\n",
       "[125 rows x 2 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_maps = mapping_candidates.groupby(biosample_col_to_map)['rank'].min().reset_index()\n",
    "best_maps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2667e541",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th>term_id</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>onto_prefix</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>root</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "      <td>po</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td></td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td></td>\n",
       "      <td>Agar</td>\n",
       "      <td>agar</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td></td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>219</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td></td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>two weeks old</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>125 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count       repeated       splitted  part_count  seq  \\\n",
       "0              Root    558           Root           Root         1.0  1.0   \n",
       "1        rice field    310     rice field     rice field         1.0  1.0   \n",
       "2              Agar    270           Agar           Agar         1.0  1.0   \n",
       "3     lettuce field    220  lettuce field  lettuce field         1.0  1.0   \n",
       "4     Two-weeks old    219  Two-weeks old  Two-weeks old         1.0  1.0   \n",
       "..              ...    ...            ...            ...         ...  ...   \n",
       "120              13      1             13             13         1.0  1.0   \n",
       "121              12      1             12             12         1.0  1.0   \n",
       "122              11      1             11             11         1.0  1.0   \n",
       "123              10      1             10             10         1.0  1.0   \n",
       "124               1      1              1              1         1.0  1.0   \n",
       "\n",
       "         term_tidy         string extract remaining_string remaining_tidied  \\\n",
       "0             Root           Root                     Root             root   \n",
       "1       rice field     rice field               rice field       rice field   \n",
       "2             Agar           Agar                     Agar             agar   \n",
       "3    lettuce field  lettuce field            lettuce field    lettuce field   \n",
       "4    Two-weeks old  Two-weeks old            Two-weeks old    two weeks old   \n",
       "..             ...            ...     ...              ...              ...   \n",
       "120             13             13                       13               13   \n",
       "121             12             12                       12               12   \n",
       "122             11             11                       11               11   \n",
       "123             10             10                       10               10   \n",
       "124              1              1                        1                1   \n",
       "\n",
       "    term_id lab_from_id  lfi_cosine term_id_from_lab       value onto_prefix  \\\n",
       "0       NaN                     1.0       PO:0009005        root          po   \n",
       "1       NaN                     1.0    ENVO:00000296  rice field        envo   \n",
       "2       NaN                     1.0              NaN         NaN               \n",
       "3       NaN                     1.0              NaN         NaN               \n",
       "4       NaN                     1.0              NaN         NaN               \n",
       "..      ...         ...         ...              ...         ...         ...   \n",
       "120     NaN                     1.0              NaN         NaN               \n",
       "121     NaN                     1.0              NaN         NaN               \n",
       "122     NaN                     1.0              NaN         NaN               \n",
       "123     NaN                     1.0              NaN         NaN               \n",
       "124     NaN                     1.0              NaN         NaN               \n",
       "\n",
       "     rank  \n",
       "0     1.0  \n",
       "1     0.0  \n",
       "2     NaN  \n",
       "3     NaN  \n",
       "4     NaN  \n",
       "..    ...  \n",
       "120   NaN  \n",
       "121   NaN  \n",
       "122   NaN  \n",
       "123   NaN  \n",
       "124   NaN  \n",
       "\n",
       "[125 rows x 18 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates = mapping_candidates.merge(best_maps, on=[biosample_col_to_map, \"rank\"])\n",
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41082d99",
   "metadata": {},
   "source": [
    "## Find consensus term IDs and labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4f5e3fdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_candidates[\"consensus_id\"] = mapping_candidates[\"term_id_from_lab\"]\n",
    "mapping_candidates[\"consensus_lab\"] = mapping_candidates[\"value\"]\n",
    "\n",
    "flag = mapping_candidates[\"consensus_id\"].isnull() & (\n",
    "    (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates.lfi_cosine.le(my_max_string_dist)\n",
    "    )\n",
    "    | (\n",
    "        ~mapping_candidates[\"term_id\"].isnull()\n",
    "        & mapping_candidates[\"remaining_tidied\"].eq(\"\")\n",
    "    )\n",
    ")\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"term_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_id\"] = replacements\n",
    "\n",
    "replacements = mapping_candidates.loc[flag, \"lab_from_id\"]\n",
    "mapping_candidates.loc[flag, \"consensus_lab\"] = replacements\n",
    "\n",
    "flag = mapping_candidates.consensus_id.isna()\n",
    "antiflag = ~flag\n",
    "mapping_candidates[\"id_or_lab_ok\"] = antiflag\n",
    "\n",
    "mapping_candidates[\"assembled_consensus\"] = (\n",
    "    mapping_candidates[\"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + mapping_candidates[\"consensus_id\"]\n",
    "    + \"]\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6af120f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>term_tidy</th>\n",
       "      <th>string</th>\n",
       "      <th>extract</th>\n",
       "      <th>remaining_string</th>\n",
       "      <th>...</th>\n",
       "      <th>lab_from_id</th>\n",
       "      <th>lfi_cosine</th>\n",
       "      <th>term_id_from_lab</th>\n",
       "      <th>value</th>\n",
       "      <th>onto_prefix</th>\n",
       "      <th>rank</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "      <th>id_or_lab_ok</th>\n",
       "      <th>assembled_consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Root</td>\n",
       "      <td>Root</td>\n",
       "      <td></td>\n",
       "      <td>Root</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "      <td>po</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "      <td>True</td>\n",
       "      <td>root [PO:0009005]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rice field</td>\n",
       "      <td>rice field</td>\n",
       "      <td></td>\n",
       "      <td>rice field</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "      <td>envo</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "      <td>True</td>\n",
       "      <td>rice field [ENVO:00000296]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td></td>\n",
       "      <td>Agar</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>lettuce field</td>\n",
       "      <td></td>\n",
       "      <td>lettuce field</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>219</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td></td>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>125 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    env_local_scale  count       repeated       splitted  part_count  seq  \\\n",
       "0              Root    558           Root           Root         1.0  1.0   \n",
       "1        rice field    310     rice field     rice field         1.0  1.0   \n",
       "2              Agar    270           Agar           Agar         1.0  1.0   \n",
       "3     lettuce field    220  lettuce field  lettuce field         1.0  1.0   \n",
       "4     Two-weeks old    219  Two-weeks old  Two-weeks old         1.0  1.0   \n",
       "..              ...    ...            ...            ...         ...  ...   \n",
       "120              13      1             13             13         1.0  1.0   \n",
       "121              12      1             12             12         1.0  1.0   \n",
       "122              11      1             11             11         1.0  1.0   \n",
       "123              10      1             10             10         1.0  1.0   \n",
       "124               1      1              1              1         1.0  1.0   \n",
       "\n",
       "         term_tidy         string extract remaining_string  ... lab_from_id  \\\n",
       "0             Root           Root                     Root  ...               \n",
       "1       rice field     rice field               rice field  ...               \n",
       "2             Agar           Agar                     Agar  ...               \n",
       "3    lettuce field  lettuce field            lettuce field  ...               \n",
       "4    Two-weeks old  Two-weeks old            Two-weeks old  ...               \n",
       "..             ...            ...     ...              ...  ...         ...   \n",
       "120             13             13                       13  ...               \n",
       "121             12             12                       12  ...               \n",
       "122             11             11                       11  ...               \n",
       "123             10             10                       10  ...               \n",
       "124              1              1                        1  ...               \n",
       "\n",
       "    lfi_cosine term_id_from_lab       value onto_prefix rank   consensus_id  \\\n",
       "0          1.0       PO:0009005        root          po  1.0     PO:0009005   \n",
       "1          1.0    ENVO:00000296  rice field        envo  0.0  ENVO:00000296   \n",
       "2          1.0              NaN         NaN              NaN            NaN   \n",
       "3          1.0              NaN         NaN              NaN            NaN   \n",
       "4          1.0              NaN         NaN              NaN            NaN   \n",
       "..         ...              ...         ...         ...  ...            ...   \n",
       "120        1.0              NaN         NaN              NaN            NaN   \n",
       "121        1.0              NaN         NaN              NaN            NaN   \n",
       "122        1.0              NaN         NaN              NaN            NaN   \n",
       "123        1.0              NaN         NaN              NaN            NaN   \n",
       "124        1.0              NaN         NaN              NaN            NaN   \n",
       "\n",
       "     consensus_lab id_or_lab_ok         assembled_consensus  \n",
       "0             root         True           root [PO:0009005]  \n",
       "1       rice field         True  rice field [ENVO:00000296]  \n",
       "2              NaN        False                         NaN  \n",
       "3              NaN        False                         NaN  \n",
       "4              NaN        False                         NaN  \n",
       "..             ...          ...                         ...  \n",
       "120            NaN        False                         NaN  \n",
       "121            NaN        False                         NaN  \n",
       "122            NaN        False                         NaN  \n",
       "123            NaN        False                         NaN  \n",
       "124            NaN        False                         NaN  \n",
       "\n",
       "[125 rows x 22 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13d44c8d",
   "metadata": {},
   "source": [
    "## For which Biosample annotations were no easy mappings found?\n",
    "How many Biosamples uses those annotations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c29338b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>270</td>\n",
       "      <td>agar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>219</td>\n",
       "      <td>two weeks old</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>207</td>\n",
       "      <td>plant associated habitat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>198</td>\n",
       "      <td>beech forest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>1</td>\n",
       "      <td>yacon fermentation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     count          remaining_tidied\n",
       "0      270                      agar\n",
       "1      220             lettuce field\n",
       "2      219             two weeks old\n",
       "3      207  plant associated habitat\n",
       "4      198              beech forest\n",
       "..     ...                       ...\n",
       "106      1                        38\n",
       "107      1                        37\n",
       "108      1                        36\n",
       "109      1                        35\n",
       "110      1        yacon fermentation\n",
       "\n",
       "[111 rows x 2 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flag = ~mapping_candidates.id_or_lab_ok\n",
    "needs_search = mapping_candidates.loc[flag]\n",
    "\n",
    "needs_search = needs_search[[\"remaining_tidied\", \"count\"]]\n",
    "\n",
    "sum_by_needed = needs_search.groupby(\"remaining_tidied\")[\"count\"].sum()\n",
    "\n",
    "sum_by_needed = sum_by_needed.to_frame()\n",
    "sum_by_needed[\"remaining_tidied\"] = sum_by_needed.index\n",
    "\n",
    "sum_by_needed = sum_by_needed.sort_values(\"count\", ascending=False)\n",
    "sum_by_needed.reset_index(drop=True)\n",
    "\n",
    "# sum_by_needed.to_sql(\"sum_by_needed\", biosample_cnx, if_exists=\"replace\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "91fec97a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>remaining_tidied</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>remaining_tidied</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>agar</th>\n",
       "      <td>270</td>\n",
       "      <td>agar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lettuce field</th>\n",
       "      <td>220</td>\n",
       "      <td>lettuce field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>two weeks old</th>\n",
       "      <td>219</td>\n",
       "      <td>two weeks old</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plant associated habitat</th>\n",
       "      <td>207</td>\n",
       "      <td>plant associated habitat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>beech forest</th>\n",
       "      <td>198</td>\n",
       "      <td>beech forest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yacon fermentation</th>\n",
       "      <td>1</td>\n",
       "      <td>yacon fermentation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          count          remaining_tidied\n",
       "remaining_tidied                                         \n",
       "agar                        270                      agar\n",
       "lettuce field               220             lettuce field\n",
       "two weeks old               219             two weeks old\n",
       "plant associated habitat    207  plant associated habitat\n",
       "beech forest                198              beech forest\n",
       "...                         ...                       ...\n",
       "38                            1                        38\n",
       "37                            1                        37\n",
       "36                            1                        36\n",
       "35                            1                        35\n",
       "yacon fermentation            1        yacon fermentation\n",
       "\n",
       "[111 rows x 2 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_by_needed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53bcffc5",
   "metadata": {},
   "source": [
    "## Extract the tidied strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "02a43f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_raw_list = list(sum_by_needed[\"remaining_tidied\"])\n",
    "ebs_raw_list.sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bd91eb2",
   "metadata": {},
   "source": [
    "## Submit those tidied strings to a search engine\n",
    "\n",
    "Specifically OLS search. This takes roughly one second per unique post-tidied submission\n",
    "\n",
    "_Turn logging back on to show status?_\n",
    "_Print the count and pre- and post- datestamps_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "08e39629",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebs_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    ebs_raw_list,\n",
    "    bad_chars=chars_to_whiteout,\n",
    "    cat_name=biosample_col_to_map,\n",
    "    ontoprefix=additional_preferred_ontologies_str,\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=my_string_dist_arg,\n",
    ")\n",
    "# ebs_search_res.to_sql(\n",
    "#     \"ols_search_results\", biosample_cnx, if_exists=\"replace\", index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae3d29dc",
   "metadata": {},
   "source": [
    "## Filter out the best of the acceptable mappings\n",
    "From a string distance perspective"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b695a649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:01000433</td>\n",
       "      <td>needleleaf forest</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000433</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1282</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>pool</td>\n",
       "      <td>pool</td>\n",
       "      <td>pool</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENVO:01000966</td>\n",
       "      <td>private swimming pool</td>\n",
       "      <td>2</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>has_broad_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000966</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             category                raw              query  \\\n",
       "428   env_local_scale  coniferous forest  coniferous forest   \n",
       "1282  env_local_scale               pool               pool   \n",
       "\n",
       "                   name  string_dist_rank  string_dist         obo_id  \\\n",
       "428   coniferous forest                 1          0.0  ENVO:01000433   \n",
       "1282               pool                 1          0.0  ENVO:01000966   \n",
       "\n",
       "                      label  search_rank ontology_prefix              scope  \\\n",
       "428       needleleaf forest            1            ENVO            synonym   \n",
       "1282  private swimming pool            2            ENVO  has_broad_synonym   \n",
       "\n",
       "            type                                           iri ontology_name  \n",
       "428      synonym  http://purl.obolibrary.org/obo/ENVO_01000433          envo  \n",
       "1282  annotation  http://purl.obolibrary.org/obo/ENVO_01000966          envo  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    ebs_search_res, max_string_dist=my_max_string_dist\n",
    ")\n",
    "my_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43aa5018",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Filter out the submissions with no acceptable matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "39e98bf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Halomonas sp. 1-1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NCBITaxon:314588</td>\n",
       "      <td>Halomonas sp. 1-1</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_314588</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>bacterium ST36-1-1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NCBITaxon:1705821</td>\n",
       "      <td>bacterium ST36-1-1</td>\n",
       "      <td>2</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1705821</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Vibrio sp. 1-1</td>\n",
       "      <td>3</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NCBITaxon:640334</td>\n",
       "      <td>Vibrio sp. 1-1</td>\n",
       "      <td>3</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_640334</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Streptomyces sp. 1-1</td>\n",
       "      <td>4</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NCBITaxon:1119062</td>\n",
       "      <td>Streptomyces sp. 1-1</td>\n",
       "      <td>4</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1119062</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Ralstonia sp. 1-1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.000</td>\n",
       "      <td>NCBITaxon:269146</td>\n",
       "      <td>Ralstonia sp. 1-1</td>\n",
       "      <td>5</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_269146</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1589</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>fermentation metagenome</td>\n",
       "      <td>1</td>\n",
       "      <td>0.311</td>\n",
       "      <td>NCBITaxon:1326787</td>\n",
       "      <td>fermentation metagenome</td>\n",
       "      <td>3</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1326787</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1590</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>food fermentation metagenome</td>\n",
       "      <td>2</td>\n",
       "      <td>0.321</td>\n",
       "      <td>NCBITaxon:1154581</td>\n",
       "      <td>food fermentation metagenome</td>\n",
       "      <td>4</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1154581</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1587</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>Yacon virus A</td>\n",
       "      <td>3</td>\n",
       "      <td>0.603</td>\n",
       "      <td>NCBITaxon:1868472</td>\n",
       "      <td>Yacon virus A</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1868472</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1588</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>Yacon necrotic mottle virus</td>\n",
       "      <td>4</td>\n",
       "      <td>0.697</td>\n",
       "      <td>NCBITaxon:1561150</td>\n",
       "      <td>Yacon necrotic mottle virus</td>\n",
       "      <td>2</td>\n",
       "      <td>NCBITAXON</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_1561150</td>\n",
       "      <td>ncbitaxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1591</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>vinegar food product</td>\n",
       "      <td>5</td>\n",
       "      <td>0.950</td>\n",
       "      <td>FOODON:00001073</td>\n",
       "      <td>vinegar food product</td>\n",
       "      <td>5</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/FOODON_00001073</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1569 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             category                 raw               query  \\\n",
       "0     env_local_scale                   1                   1   \n",
       "1     env_local_scale                   1                   1   \n",
       "2     env_local_scale                   1                   1   \n",
       "3     env_local_scale                   1                   1   \n",
       "4     env_local_scale                   1                   1   \n",
       "...               ...                 ...                 ...   \n",
       "1589  env_local_scale  yacon fermentation  yacon fermentation   \n",
       "1590  env_local_scale  yacon fermentation  yacon fermentation   \n",
       "1587  env_local_scale  yacon fermentation  yacon fermentation   \n",
       "1588  env_local_scale  yacon fermentation  yacon fermentation   \n",
       "1591  env_local_scale  yacon fermentation  yacon fermentation   \n",
       "\n",
       "                              name  string_dist_rank  string_dist  \\\n",
       "0                Halomonas sp. 1-1                 1        1.000   \n",
       "1               bacterium ST36-1-1                 2        1.000   \n",
       "2                   Vibrio sp. 1-1                 3        1.000   \n",
       "3             Streptomyces sp. 1-1                 4        1.000   \n",
       "4                Ralstonia sp. 1-1                 5        1.000   \n",
       "...                            ...               ...          ...   \n",
       "1589       fermentation metagenome                 1        0.311   \n",
       "1590  food fermentation metagenome                 2        0.321   \n",
       "1587                 Yacon virus A                 3        0.603   \n",
       "1588   Yacon necrotic mottle virus                 4        0.697   \n",
       "1591          vinegar food product                 5        0.950   \n",
       "\n",
       "                 obo_id                         label  search_rank  \\\n",
       "0      NCBITaxon:314588             Halomonas sp. 1-1            1   \n",
       "1     NCBITaxon:1705821            bacterium ST36-1-1            2   \n",
       "2      NCBITaxon:640334                Vibrio sp. 1-1            3   \n",
       "3     NCBITaxon:1119062          Streptomyces sp. 1-1            4   \n",
       "4      NCBITaxon:269146             Ralstonia sp. 1-1            5   \n",
       "...                 ...                           ...          ...   \n",
       "1589  NCBITaxon:1326787       fermentation metagenome            3   \n",
       "1590  NCBITaxon:1154581  food fermentation metagenome            4   \n",
       "1587  NCBITaxon:1868472                 Yacon virus A            1   \n",
       "1588  NCBITaxon:1561150   Yacon necrotic mottle virus            2   \n",
       "1591    FOODON:00001073          vinegar food product            5   \n",
       "\n",
       "     ontology_prefix  scope   type  \\\n",
       "0          NCBITAXON  label  label   \n",
       "1          NCBITAXON  label  label   \n",
       "2          NCBITAXON  label  label   \n",
       "3          NCBITAXON  label  label   \n",
       "4          NCBITAXON  label  label   \n",
       "...              ...    ...    ...   \n",
       "1589       NCBITAXON  label  label   \n",
       "1590       NCBITAXON  label  label   \n",
       "1587       NCBITAXON  label  label   \n",
       "1588       NCBITAXON  label  label   \n",
       "1591            ENVO  label  label   \n",
       "\n",
       "                                                   iri ontology_name  \n",
       "0      http://purl.obolibrary.org/obo/NCBITaxon_314588     ncbitaxon  \n",
       "1     http://purl.obolibrary.org/obo/NCBITaxon_1705821     ncbitaxon  \n",
       "2      http://purl.obolibrary.org/obo/NCBITaxon_640334     ncbitaxon  \n",
       "3     http://purl.obolibrary.org/obo/NCBITaxon_1119062     ncbitaxon  \n",
       "4      http://purl.obolibrary.org/obo/NCBITaxon_269146     ncbitaxon  \n",
       "...                                                ...           ...  \n",
       "1589  http://purl.obolibrary.org/obo/NCBITaxon_1326787     ncbitaxon  \n",
       "1590  http://purl.obolibrary.org/obo/NCBITaxon_1154581     ncbitaxon  \n",
       "1587  http://purl.obolibrary.org/obo/NCBITaxon_1868472     ncbitaxon  \n",
       "1588  http://purl.obolibrary.org/obo/NCBITaxon_1561150     ncbitaxon  \n",
       "1591    http://purl.obolibrary.org/obo/FOODON_00001073          envo  \n",
       "\n",
       "[1569 rows x 14 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_acceptable_mappings = scoped_mapping.get_no_acceptable_mappings(\n",
    "    ebs_search_res, my_best_acceptable\n",
    ")\n",
    "\n",
    "no_acceptable_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70be6244",
   "metadata": {},
   "source": [
    "## Try searching the failures against all ontologies in OLS\n",
    "what would it be like to use a local OLS instance? (latency-wise)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d5adfca5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "109\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['1',\n",
       " '10',\n",
       " '11',\n",
       " '12',\n",
       " '13',\n",
       " '14',\n",
       " '15',\n",
       " '16',\n",
       " '17',\n",
       " '18',\n",
       " '19',\n",
       " '2',\n",
       " '20',\n",
       " '21',\n",
       " '22',\n",
       " '23',\n",
       " '24',\n",
       " '25',\n",
       " '26',\n",
       " '27',\n",
       " '28',\n",
       " '29',\n",
       " '3',\n",
       " '30',\n",
       " '31',\n",
       " '32',\n",
       " '33',\n",
       " '34',\n",
       " '35',\n",
       " '36',\n",
       " '37',\n",
       " '38',\n",
       " '39',\n",
       " '4',\n",
       " '40',\n",
       " '41',\n",
       " '42',\n",
       " '43',\n",
       " '44',\n",
       " '5',\n",
       " '6',\n",
       " '7',\n",
       " '8',\n",
       " '9',\n",
       " 'agar',\n",
       " 'agricultural field or laboratory growth chamber',\n",
       " 'agrigulutural field',\n",
       " 'animal associated habitat',\n",
       " 'beech forest',\n",
       " 'blank',\n",
       " 'cedar glade',\n",
       " 'characterization of the microbial community present at vineyard',\n",
       " 'chlorotic flecks on leaves',\n",
       " 'crop',\n",
       " 'crop fields',\n",
       " 'disturbed field',\n",
       " 'dna from roots of tree 14 13',\n",
       " 'dna from roots of tree 14 14',\n",
       " 'dna from roots of tree 14 15',\n",
       " 'dna from roots of tree 16 10',\n",
       " 'dna from roots of tree 16 11',\n",
       " 'dna from roots of tree 16 12',\n",
       " 'dna from roots of tree 18 7',\n",
       " 'dna from roots of tree 20 2',\n",
       " 'dna from roots of tree 20 3',\n",
       " 'dna from roots of tree 20 6',\n",
       " 'dna from roots of tree 24 8',\n",
       " 'experimental pecan orchard',\n",
       " 'experimental pond',\n",
       " 'forest',\n",
       " 'genome sequence',\n",
       " 'glasshouse',\n",
       " 'greenhouse',\n",
       " 'growth chamber',\n",
       " 'highlands',\n",
       " 'hilly area',\n",
       " 'inoculum',\n",
       " 'leaves without infection',\n",
       " 'lettuce field',\n",
       " 'lowlands',\n",
       " 'missing',\n",
       " 'mountainous forest',\n",
       " 'no tillage soybean crop',\n",
       " 'normal',\n",
       " 'oak tree',\n",
       " 'och',\n",
       " 'organic feature',\n",
       " 'oulu',\n",
       " 'paratistic on pyrus communis',\n",
       " 'plant',\n",
       " 'plant associated habitat',\n",
       " 'plant roots',\n",
       " 'pool of 10 plant tissue',\n",
       " 'pustules on leaves',\n",
       " 'rice canopy',\n",
       " 'rna from roots of tree 14 13',\n",
       " 'rna from roots of tree 14 15',\n",
       " 'rna from roots of tree 16 10',\n",
       " 'rna from roots of tree 16 12',\n",
       " 'rna from roots of tree 20 3',\n",
       " 'root associated',\n",
       " 'soils used to inoculate seedlings',\n",
       " 'sugar beet tissue',\n",
       " 't0agar',\n",
       " 'temperate continental monsoon climate',\n",
       " 'teosinte field',\n",
       " 'two weeks old',\n",
       " 'urediniospore tissue',\n",
       " 'yacon fermentation']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "still_unmapped = list(set(list(no_acceptable_mappings[\"raw\"])))\n",
    "still_unmapped.sort()\n",
    "\n",
    "print(len(still_unmapped))\n",
    "\n",
    "still_unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ac5c4f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "salvage_search_res = scoped_mapping.search_get_annotations_wrapper(\n",
    "    still_unmapped,\n",
    "    bad_chars=\"._-\",\n",
    "    cat_name=\"salvage\",\n",
    "    ontoprefix=\"\",\n",
    "    query_fields=\"\",\n",
    "    rr=5,\n",
    "    string_dist_arg=2,\n",
    ")\n",
    "\n",
    "# salvage_search_res.to_sql(\n",
    "#     \"salvage_search_res\", biosample_cnx, if_exists=\"replace\", index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "453b8356",
   "metadata": {},
   "source": [
    "## We appear to be at a point of diminishing returns\n",
    "At the very least, will require some review"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d44fe45b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>search_rank</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>salvage</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C179011</td>\n",
       "      <td>COVID-19 Disease Severity 1</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179011</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>salvage</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C105718</td>\n",
       "      <td>Karnofsky Performance Status 10</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C105718</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>salvage</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C113425</td>\n",
       "      <td>Eleven</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113425</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>salvage</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C113429</td>\n",
       "      <td>Fifteen</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113429</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>salvage</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C14338</td>\n",
       "      <td>Human Papillomavirus Type 16</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14338</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>salvage</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C113431</td>\n",
       "      <td>Seventeen</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113431</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>salvage</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C14377</td>\n",
       "      <td>Human Papillomavirus Type 18</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14377</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>salvage</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C105801</td>\n",
       "      <td>Nineteen</td>\n",
       "      <td>4</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C105801</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>salvage</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C172969</td>\n",
       "      <td>Worry Score 2</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172969</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>salvage</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C105716</td>\n",
       "      <td>Karnofsky Performance Status 20</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C105716</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>salvage</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C107434</td>\n",
       "      <td>Twenty One</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C107434</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>salvage</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C102995</td>\n",
       "      <td>Human Papillomavirus-26</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102995</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>salvage</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C126888</td>\n",
       "      <td>Twenty Eight</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C126888</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>salvage</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C172976</td>\n",
       "      <td>Fear Score 3</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172976</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>salvage</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C70541</td>\n",
       "      <td>Lansky Performance Status 30</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C70541</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>salvage</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C102904</td>\n",
       "      <td>Human Papillomavirus-31</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102904</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>salvage</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C102996</td>\n",
       "      <td>Human Papillomavirus-33</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102996</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>salvage</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C102997</td>\n",
       "      <td>Human Papillomavirus-35</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102997</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>salvage</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C102998</td>\n",
       "      <td>Human Papillomavirus-39</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102998</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>salvage</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C172971</td>\n",
       "      <td>Worry Score 4</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172971</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>salvage</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C105714</td>\n",
       "      <td>Karnofsky Performance Status 40</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C105714</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>salvage</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C178273</td>\n",
       "      <td>Forty One</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C178273</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>513</th>\n",
       "      <td>salvage</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C175676</td>\n",
       "      <td>Forty Two</td>\n",
       "      <td>4</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C175676</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>salvage</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C178274</td>\n",
       "      <td>Forty Four</td>\n",
       "      <td>4</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C178274</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>554</th>\n",
       "      <td>salvage</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C172966</td>\n",
       "      <td>Upset Score 5</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172966</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>salvage</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C179016</td>\n",
       "      <td>COVID-19 Disease Severity 6</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179016</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>609</th>\n",
       "      <td>salvage</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C179017</td>\n",
       "      <td>COVID-19 Disease Severity 7</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>None</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179017</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>salvage</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C66839</td>\n",
       "      <td>Eight</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C66839</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>salvage</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C136557</td>\n",
       "      <td>Neck Pain Score 9</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C136557</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>salvage</td>\n",
       "      <td>agar</td>\n",
       "      <td>agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>OMIT:0001790</td>\n",
       "      <td>Agar</td>\n",
       "      <td>1</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0001790</td>\n",
       "      <td>omit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>849</th>\n",
       "      <td>salvage</td>\n",
       "      <td>beech forest</td>\n",
       "      <td>beech forest</td>\n",
       "      <td>beech forest soil</td>\n",
       "      <td>1</td>\n",
       "      <td>0.171</td>\n",
       "      <td>ENVO:00005770</td>\n",
       "      <td>beech forest soil</td>\n",
       "      <td>1</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005770</td>\n",
       "      <td>envo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>salvage</td>\n",
       "      <td>blank</td>\n",
       "      <td>blank</td>\n",
       "      <td>blank</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>AFRL:0000259</td>\n",
       "      <td>blank role</td>\n",
       "      <td>1</td>\n",
       "      <td>AFO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.allotrope.org/ontologies/role#AFRL...</td>\n",
       "      <td>afo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>915</th>\n",
       "      <td>salvage</td>\n",
       "      <td>crop</td>\n",
       "      <td>crop</td>\n",
       "      <td>Crop</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C111162</td>\n",
       "      <td>Crop</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C111162</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1344</th>\n",
       "      <td>salvage</td>\n",
       "      <td>forest</td>\n",
       "      <td>forest</td>\n",
       "      <td>Forest</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00454366</td>\n",
       "      <td>Forest</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00454366</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1384</th>\n",
       "      <td>salvage</td>\n",
       "      <td>genome sequence</td>\n",
       "      <td>genome sequence</td>\n",
       "      <td>Genome Assembly Sequence</td>\n",
       "      <td>1</td>\n",
       "      <td>0.182</td>\n",
       "      <td>NCIT:C73517</td>\n",
       "      <td>Genome Assembly Sequence</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C73517</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>salvage</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>1</td>\n",
       "      <td>AGRO</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/AGRO_00000363</td>\n",
       "      <td>agro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1409</th>\n",
       "      <td>salvage</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>1</td>\n",
       "      <td>AGRO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/AGRO_00000363</td>\n",
       "      <td>agro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1426</th>\n",
       "      <td>salvage</td>\n",
       "      <td>growth chamber</td>\n",
       "      <td>growth chamber</td>\n",
       "      <td>growth chamber study</td>\n",
       "      <td>1</td>\n",
       "      <td>0.173</td>\n",
       "      <td>PECO:0007269</td>\n",
       "      <td>growth chamber study</td>\n",
       "      <td>1</td>\n",
       "      <td>PECO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PECO_0007269</td>\n",
       "      <td>peco</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>salvage</td>\n",
       "      <td>highlands</td>\n",
       "      <td>highlands</td>\n",
       "      <td>Highlands</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00246323</td>\n",
       "      <td>Highlands</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00246323</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1443</th>\n",
       "      <td>salvage</td>\n",
       "      <td>inoculum</td>\n",
       "      <td>inoculum</td>\n",
       "      <td>inoculum</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ARO:3004417</td>\n",
       "      <td>inoculum</td>\n",
       "      <td>1</td>\n",
       "      <td>ARO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ARO_3004417</td>\n",
       "      <td>aro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1469</th>\n",
       "      <td>salvage</td>\n",
       "      <td>missing</td>\n",
       "      <td>missing</td>\n",
       "      <td>Missing</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C54031</td>\n",
       "      <td>Missing</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C54031</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1498</th>\n",
       "      <td>salvage</td>\n",
       "      <td>normal</td>\n",
       "      <td>normal</td>\n",
       "      <td>Normal</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C14165</td>\n",
       "      <td>Normal</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14165</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1507</th>\n",
       "      <td>salvage</td>\n",
       "      <td>oak tree</td>\n",
       "      <td>oak tree</td>\n",
       "      <td>Oak Tree</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00454476</td>\n",
       "      <td>Oak Tree</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00454476</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1539</th>\n",
       "      <td>salvage</td>\n",
       "      <td>organic feature</td>\n",
       "      <td>organic feature</td>\n",
       "      <td>organic feature</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>ENVO:01000159</td>\n",
       "      <td>organic feature</td>\n",
       "      <td>1</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000159</td>\n",
       "      <td>micro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1549</th>\n",
       "      <td>salvage</td>\n",
       "      <td>oulu</td>\n",
       "      <td>oulu</td>\n",
       "      <td>Oulu</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>GAZ:00010238</td>\n",
       "      <td>Oulu</td>\n",
       "      <td>1</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00010238</td>\n",
       "      <td>gaz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1668</th>\n",
       "      <td>salvage</td>\n",
       "      <td>plant</td>\n",
       "      <td>plant</td>\n",
       "      <td>Plant</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NCIT:C14258</td>\n",
       "      <td>Plant</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14258</td>\n",
       "      <td>ncit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1694</th>\n",
       "      <td>salvage</td>\n",
       "      <td>plant roots</td>\n",
       "      <td>plant roots</td>\n",
       "      <td>Plant Roots</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>OMIT:0018707</td>\n",
       "      <td>Plant Roots</td>\n",
       "      <td>1</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0018707</td>\n",
       "      <td>omit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1869</th>\n",
       "      <td>salvage</td>\n",
       "      <td>urediniospore tissue</td>\n",
       "      <td>urediniospore tissue</td>\n",
       "      <td>urediniospore</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>FAO:0000026</td>\n",
       "      <td>urediniospore</td>\n",
       "      <td>1</td>\n",
       "      <td>FAO</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/FAO_0000026</td>\n",
       "      <td>fao</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1896</th>\n",
       "      <td>salvage</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>Fermentation</td>\n",
       "      <td>1</td>\n",
       "      <td>0.170</td>\n",
       "      <td>OMIT:0006501</td>\n",
       "      <td>Fermentation</td>\n",
       "      <td>4</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0006501</td>\n",
       "      <td>omit</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     category                   raw                 query  \\\n",
       "13    salvage                     1                     1   \n",
       "29    salvage                    10                    10   \n",
       "51    salvage                    11                    11   \n",
       "78    salvage                    15                    15   \n",
       "98    salvage                    16                    16   \n",
       "107   salvage                    17                    17   \n",
       "112   salvage                    18                    18   \n",
       "132   salvage                    19                    19   \n",
       "158   salvage                     2                     2   \n",
       "163   salvage                    20                    20   \n",
       "181   salvage                    21                    21   \n",
       "240   salvage                    26                    26   \n",
       "266   salvage                    28                    28   \n",
       "305   salvage                     3                     3   \n",
       "326   salvage                    30                    30   \n",
       "333   salvage                    31                    31   \n",
       "359   salvage                    33                    33   \n",
       "397   salvage                    35                    35   \n",
       "447   salvage                    39                    39   \n",
       "468   salvage                     4                     4   \n",
       "484   salvage                    40                    40   \n",
       "503   salvage                    41                    41   \n",
       "513   salvage                    42                    42   \n",
       "536   salvage                    44                    44   \n",
       "554   salvage                     5                     5   \n",
       "581   salvage                     6                     6   \n",
       "609   salvage                     7                     7   \n",
       "625   salvage                     8                     8   \n",
       "646   salvage                     9                     9   \n",
       "665   salvage                  agar                  agar   \n",
       "849   salvage          beech forest          beech forest   \n",
       "863   salvage                 blank                 blank   \n",
       "915   salvage                  crop                  crop   \n",
       "1344  salvage                forest                forest   \n",
       "1384  salvage       genome sequence       genome sequence   \n",
       "1393  salvage            glasshouse            glasshouse   \n",
       "1409  salvage            greenhouse            greenhouse   \n",
       "1426  salvage        growth chamber        growth chamber   \n",
       "1432  salvage             highlands             highlands   \n",
       "1443  salvage              inoculum              inoculum   \n",
       "1469  salvage               missing               missing   \n",
       "1498  salvage                normal                normal   \n",
       "1507  salvage              oak tree              oak tree   \n",
       "1539  salvage       organic feature       organic feature   \n",
       "1549  salvage                  oulu                  oulu   \n",
       "1668  salvage                 plant                 plant   \n",
       "1694  salvage           plant roots           plant roots   \n",
       "1869  salvage  urediniospore tissue  urediniospore tissue   \n",
       "1896  salvage    yacon fermentation    yacon fermentation   \n",
       "\n",
       "                          name  string_dist_rank  string_dist         obo_id  \\\n",
       "13                           1                 1        0.000   NCIT:C179011   \n",
       "29                          10                 1        0.000   NCIT:C105718   \n",
       "51                          11                 1        0.000   NCIT:C113425   \n",
       "78                          15                 1        0.000   NCIT:C113429   \n",
       "98                          16                 1        0.000    NCIT:C14338   \n",
       "107                         17                 1        0.000   NCIT:C113431   \n",
       "112                         18                 1        0.000    NCIT:C14377   \n",
       "132                         19                 1        0.000   NCIT:C105801   \n",
       "158                          2                 1        0.000   NCIT:C172969   \n",
       "163                         20                 1        0.000   NCIT:C105716   \n",
       "181                         21                 1        0.000   NCIT:C107434   \n",
       "240                         26                 1        0.000   NCIT:C102995   \n",
       "266                         28                 1        0.000   NCIT:C126888   \n",
       "305                          3                 1        0.000   NCIT:C172976   \n",
       "326                         30                 1        0.000    NCIT:C70541   \n",
       "333                         31                 1        0.000   NCIT:C102904   \n",
       "359                         33                 1        0.000   NCIT:C102996   \n",
       "397                         35                 1        0.000   NCIT:C102997   \n",
       "447                         39                 1        0.000   NCIT:C102998   \n",
       "468                          4                 1        0.000   NCIT:C172971   \n",
       "484                         40                 1        0.000   NCIT:C105714   \n",
       "503                         41                 1        0.000   NCIT:C178273   \n",
       "513                         42                 1        0.000   NCIT:C175676   \n",
       "536                         44                 1        0.000   NCIT:C178274   \n",
       "554                          5                 1        0.000   NCIT:C172966   \n",
       "581                          6                 1        0.000   NCIT:C179016   \n",
       "609                          7                 1        0.000   NCIT:C179017   \n",
       "625                          8                 1        0.000    NCIT:C66839   \n",
       "646                          9                 1        0.000   NCIT:C136557   \n",
       "665                       Agar                 1        0.000   OMIT:0001790   \n",
       "849          beech forest soil                 1        0.171  ENVO:00005770   \n",
       "863                      blank                 1        0.000   AFRL:0000259   \n",
       "915                       Crop                 1        0.000   NCIT:C111162   \n",
       "1344                    Forest                 1        0.000   GAZ:00454366   \n",
       "1384  Genome Assembly Sequence                 1        0.182    NCIT:C73517   \n",
       "1393                glasshouse                 1        0.000  AGRO:00000363   \n",
       "1409                greenhouse                 1        0.000  AGRO:00000363   \n",
       "1426      growth chamber study                 1        0.173   PECO:0007269   \n",
       "1432                 Highlands                 1        0.000   GAZ:00246323   \n",
       "1443                  inoculum                 1        0.000    ARO:3004417   \n",
       "1469                   Missing                 1        0.000    NCIT:C54031   \n",
       "1498                    Normal                 1        0.000    NCIT:C14165   \n",
       "1507                  Oak Tree                 1        0.000   GAZ:00454476   \n",
       "1539           organic feature                 1        0.000  ENVO:01000159   \n",
       "1549                      Oulu                 1        0.000   GAZ:00010238   \n",
       "1668                     Plant                 1        0.000    NCIT:C14258   \n",
       "1694               Plant Roots                 1        0.000   OMIT:0018707   \n",
       "1869             urediniospore                 1        0.184    FAO:0000026   \n",
       "1896              Fermentation                 1        0.170   OMIT:0006501   \n",
       "\n",
       "                                label  search_rank ontology_prefix  \\\n",
       "13        COVID-19 Disease Severity 1            3            NCIT   \n",
       "29    Karnofsky Performance Status 10            1            NCIT   \n",
       "51                             Eleven            1            NCIT   \n",
       "78                            Fifteen            5            NCIT   \n",
       "98       Human Papillomavirus Type 16            5            NCIT   \n",
       "107                         Seventeen            5            NCIT   \n",
       "112      Human Papillomavirus Type 18            1            NCIT   \n",
       "132                          Nineteen            4            NCIT   \n",
       "158                     Worry Score 2            5            NCIT   \n",
       "163   Karnofsky Performance Status 20            1            NCIT   \n",
       "181                        Twenty One            5            NCIT   \n",
       "240           Human Papillomavirus-26            1            NCIT   \n",
       "266                      Twenty Eight            3            NCIT   \n",
       "305                      Fear Score 3            5            NCIT   \n",
       "326      Lansky Performance Status 30            5            NCIT   \n",
       "333           Human Papillomavirus-31            1            NCIT   \n",
       "359           Human Papillomavirus-33            1            NCIT   \n",
       "397           Human Papillomavirus-35            1            NCIT   \n",
       "447           Human Papillomavirus-39            1            NCIT   \n",
       "468                     Worry Score 4            3            NCIT   \n",
       "484   Karnofsky Performance Status 40            1            NCIT   \n",
       "503                         Forty One            5            NCIT   \n",
       "513                         Forty Two            4            NCIT   \n",
       "536                        Forty Four            4            NCIT   \n",
       "554                     Upset Score 5            3            NCIT   \n",
       "581       COVID-19 Disease Severity 6            3            NCIT   \n",
       "609       COVID-19 Disease Severity 7            3            NCIT   \n",
       "625                             Eight            1            NCIT   \n",
       "646                 Neck Pain Score 9            1            NCIT   \n",
       "665                              Agar            1            OMIT   \n",
       "849                 beech forest soil            1            ENVO   \n",
       "863                        blank role            1             AFO   \n",
       "915                              Crop            1            NCIT   \n",
       "1344                           Forest            1             GAZ   \n",
       "1384         Genome Assembly Sequence            3            NCIT   \n",
       "1393                       greenhouse            1            AGRO   \n",
       "1409                       greenhouse            1            AGRO   \n",
       "1426             growth chamber study            1            PECO   \n",
       "1432                        Highlands            1             GAZ   \n",
       "1443                         inoculum            1             ARO   \n",
       "1469                          Missing            1            NCIT   \n",
       "1498                           Normal            1            NCIT   \n",
       "1507                         Oak Tree            1             GAZ   \n",
       "1539                  organic feature            1           MICRO   \n",
       "1549                             Oulu            1             GAZ   \n",
       "1668                            Plant            1            NCIT   \n",
       "1694                      Plant Roots            1            OMIT   \n",
       "1869                    urediniospore            1             FAO   \n",
       "1896                     Fermentation            4            OMIT   \n",
       "\n",
       "                scope     type  \\\n",
       "13    hasExactSynonym     None   \n",
       "29            synonym  synonym   \n",
       "51            synonym  synonym   \n",
       "78            synonym  synonym   \n",
       "98    hasExactSynonym     None   \n",
       "107           synonym  synonym   \n",
       "112           synonym  synonym   \n",
       "132           synonym  synonym   \n",
       "158           synonym  synonym   \n",
       "163           synonym  synonym   \n",
       "181           synonym  synonym   \n",
       "240   hasExactSynonym     None   \n",
       "266           synonym  synonym   \n",
       "305           synonym  synonym   \n",
       "326   hasExactSynonym     None   \n",
       "333           synonym  synonym   \n",
       "359           synonym  synonym   \n",
       "397   hasExactSynonym     None   \n",
       "447   hasExactSynonym     None   \n",
       "468           synonym  synonym   \n",
       "484           synonym  synonym   \n",
       "503           synonym  synonym   \n",
       "513           synonym  synonym   \n",
       "536           synonym  synonym   \n",
       "554           synonym  synonym   \n",
       "581   hasExactSynonym     None   \n",
       "609   hasExactSynonym     None   \n",
       "625           synonym  synonym   \n",
       "646           synonym  synonym   \n",
       "665             label    label   \n",
       "849             label    label   \n",
       "863           synonym  synonym   \n",
       "915             label    label   \n",
       "1344            label    label   \n",
       "1384            label    label   \n",
       "1393          synonym  synonym   \n",
       "1409            label    label   \n",
       "1426            label    label   \n",
       "1432            label    label   \n",
       "1443            label    label   \n",
       "1469            label    label   \n",
       "1498            label    label   \n",
       "1507            label    label   \n",
       "1539            label    label   \n",
       "1549            label    label   \n",
       "1668            label    label   \n",
       "1694            label    label   \n",
       "1869            label    label   \n",
       "1896            label    label   \n",
       "\n",
       "                                                    iri ontology_name  \n",
       "13          http://purl.obolibrary.org/obo/NCIT_C179011          ncit  \n",
       "29          http://purl.obolibrary.org/obo/NCIT_C105718          ncit  \n",
       "51          http://purl.obolibrary.org/obo/NCIT_C113425          ncit  \n",
       "78          http://purl.obolibrary.org/obo/NCIT_C113429          ncit  \n",
       "98           http://purl.obolibrary.org/obo/NCIT_C14338          ncit  \n",
       "107         http://purl.obolibrary.org/obo/NCIT_C113431          ncit  \n",
       "112          http://purl.obolibrary.org/obo/NCIT_C14377          ncit  \n",
       "132         http://purl.obolibrary.org/obo/NCIT_C105801          ncit  \n",
       "158         http://purl.obolibrary.org/obo/NCIT_C172969          ncit  \n",
       "163         http://purl.obolibrary.org/obo/NCIT_C105716          ncit  \n",
       "181         http://purl.obolibrary.org/obo/NCIT_C107434          ncit  \n",
       "240         http://purl.obolibrary.org/obo/NCIT_C102995          ncit  \n",
       "266         http://purl.obolibrary.org/obo/NCIT_C126888          ncit  \n",
       "305         http://purl.obolibrary.org/obo/NCIT_C172976          ncit  \n",
       "326          http://purl.obolibrary.org/obo/NCIT_C70541          ncit  \n",
       "333         http://purl.obolibrary.org/obo/NCIT_C102904          ncit  \n",
       "359         http://purl.obolibrary.org/obo/NCIT_C102996          ncit  \n",
       "397         http://purl.obolibrary.org/obo/NCIT_C102997          ncit  \n",
       "447         http://purl.obolibrary.org/obo/NCIT_C102998          ncit  \n",
       "468         http://purl.obolibrary.org/obo/NCIT_C172971          ncit  \n",
       "484         http://purl.obolibrary.org/obo/NCIT_C105714          ncit  \n",
       "503         http://purl.obolibrary.org/obo/NCIT_C178273          ncit  \n",
       "513         http://purl.obolibrary.org/obo/NCIT_C175676          ncit  \n",
       "536         http://purl.obolibrary.org/obo/NCIT_C178274          ncit  \n",
       "554         http://purl.obolibrary.org/obo/NCIT_C172966          ncit  \n",
       "581         http://purl.obolibrary.org/obo/NCIT_C179016          ncit  \n",
       "609         http://purl.obolibrary.org/obo/NCIT_C179017          ncit  \n",
       "625          http://purl.obolibrary.org/obo/NCIT_C66839          ncit  \n",
       "646         http://purl.obolibrary.org/obo/NCIT_C136557          ncit  \n",
       "665         http://purl.obolibrary.org/obo/OMIT_0001790          omit  \n",
       "849        http://purl.obolibrary.org/obo/ENVO_00005770          envo  \n",
       "863   http://purl.allotrope.org/ontologies/role#AFRL...           afo  \n",
       "915         http://purl.obolibrary.org/obo/NCIT_C111162          ncit  \n",
       "1344        http://purl.obolibrary.org/obo/GAZ_00454366           gaz  \n",
       "1384         http://purl.obolibrary.org/obo/NCIT_C73517          ncit  \n",
       "1393       http://purl.obolibrary.org/obo/AGRO_00000363          agro  \n",
       "1409       http://purl.obolibrary.org/obo/AGRO_00000363          agro  \n",
       "1426        http://purl.obolibrary.org/obo/PECO_0007269          peco  \n",
       "1432        http://purl.obolibrary.org/obo/GAZ_00246323           gaz  \n",
       "1443         http://purl.obolibrary.org/obo/ARO_3004417           aro  \n",
       "1469         http://purl.obolibrary.org/obo/NCIT_C54031          ncit  \n",
       "1498         http://purl.obolibrary.org/obo/NCIT_C14165          ncit  \n",
       "1507        http://purl.obolibrary.org/obo/GAZ_00454476           gaz  \n",
       "1539       http://purl.obolibrary.org/obo/ENVO_01000159         micro  \n",
       "1549        http://purl.obolibrary.org/obo/GAZ_00010238           gaz  \n",
       "1668         http://purl.obolibrary.org/obo/NCIT_C14258          ncit  \n",
       "1694        http://purl.obolibrary.org/obo/OMIT_0018707          omit  \n",
       "1869         http://purl.obolibrary.org/obo/FAO_0000026           fao  \n",
       "1896        http://purl.obolibrary.org/obo/OMIT_0006501          omit  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "salvage_best_acceptable = scoped_mapping.get_best_acceptable(\n",
    "    salvage_search_res, max_string_dist=salvage_max_dist\n",
    ")\n",
    "salvage_best_acceptable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6bd9e5e",
   "metadata": {},
   "source": [
    "## Save the mappings for review!\n",
    "\n",
    "_Note that the merge on ID or label mappings are not included here_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "a2b88c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_best_acceptable[include_col] = first_pass_include_val\n",
    "salvage_best_acceptable[include_col] = salvage_include_val\n",
    "best_and_salvage = pd.concat([my_best_acceptable, salvage_best_acceptable])\n",
    "\n",
    "best_and_salvage[\"ols_curation_notes\"] = \"\"\n",
    "\n",
    "# best_and_salvage.to_csv(ols_review_file, sep=ols_review_seperator, index=False)\n",
    "\n",
    "best_and_salvage = best_and_salvage[\n",
    "    [\n",
    "        \"category\",\n",
    "        \"raw\",\n",
    "        \"query\",\n",
    "        \"name\",\n",
    "        \"obo_id\",\n",
    "        \"label\",\n",
    "        \"ontology_prefix\",\n",
    "        \"include\",\n",
    "        \"ols_curation_notes\",\n",
    "        \"scope\",\n",
    "        \"type\",\n",
    "        \"iri\",\n",
    "        \"ontology_name\",\n",
    "        \"string_dist_rank\",\n",
    "        \"string_dist\",\n",
    "        \"search_rank\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "best_and_salvage = best_and_salvage.sort_values(['include','ontology_prefix'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "131b2343",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_and_salvage.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f4a3988",
   "metadata": {},
   "outputs": [],
   "source": [
    "raise SystemExit(\"Don't skip reviewing and saving!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a331a6",
   "metadata": {},
   "source": [
    "## Do some review in a spreadsheet application!\n",
    "Specifically, columns `include` and `ols_curation_notes`\n",
    "\n",
    "Then Save\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be6da72c",
   "metadata": {},
   "source": [
    "## Now read the reviewed spreadsheet back in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c94cf3b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# curated = pd.read_csv(ols_review_file, sep=ols_review_seperator)\n",
    "# curated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49389e4e",
   "metadata": {},
   "source": [
    "## OR copy and paste"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "bd84b014",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>raw</th>\n",
       "      <th>query</th>\n",
       "      <th>name</th>\n",
       "      <th>obo_id</th>\n",
       "      <th>label</th>\n",
       "      <th>ontology_prefix</th>\n",
       "      <th>include</th>\n",
       "      <th>ols_curation_notes</th>\n",
       "      <th>scope</th>\n",
       "      <th>type</th>\n",
       "      <th>iri</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>string_dist_rank</th>\n",
       "      <th>string_dist</th>\n",
       "      <th>search_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>salvage</td>\n",
       "      <td>blank</td>\n",
       "      <td>blank</td>\n",
       "      <td>blank</td>\n",
       "      <td>AFRL:0000259</td>\n",
       "      <td>blank role</td>\n",
       "      <td>AFO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.allotrope.org/ontologies/role#AFRL...</td>\n",
       "      <td>afo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>salvage</td>\n",
       "      <td>inoculum</td>\n",
       "      <td>inoculum</td>\n",
       "      <td>inoculum</td>\n",
       "      <td>ARO:3004417</td>\n",
       "      <td>inoculum</td>\n",
       "      <td>ARO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ARO_3004417</td>\n",
       "      <td>aro</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>salvage</td>\n",
       "      <td>beech forest</td>\n",
       "      <td>beech forest</td>\n",
       "      <td>beech forest soil</td>\n",
       "      <td>ENVO:00005770</td>\n",
       "      <td>beech forest soil</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_00005770</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.171</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>salvage</td>\n",
       "      <td>urediniospore tissue</td>\n",
       "      <td>urediniospore tissue</td>\n",
       "      <td>urediniospore</td>\n",
       "      <td>FAO:0000026</td>\n",
       "      <td>urediniospore</td>\n",
       "      <td>FAO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/FAO_0000026</td>\n",
       "      <td>fao</td>\n",
       "      <td>1</td>\n",
       "      <td>0.184</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>salvage</td>\n",
       "      <td>forest</td>\n",
       "      <td>forest</td>\n",
       "      <td>Forest</td>\n",
       "      <td>GAZ:00454366</td>\n",
       "      <td>Forest</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00454366</td>\n",
       "      <td>gaz</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>salvage</td>\n",
       "      <td>highlands</td>\n",
       "      <td>highlands</td>\n",
       "      <td>Highlands</td>\n",
       "      <td>GAZ:00246323</td>\n",
       "      <td>Highlands</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00246323</td>\n",
       "      <td>gaz</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>salvage</td>\n",
       "      <td>oak tree</td>\n",
       "      <td>oak tree</td>\n",
       "      <td>Oak Tree</td>\n",
       "      <td>GAZ:00454476</td>\n",
       "      <td>Oak Tree</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00454476</td>\n",
       "      <td>gaz</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>salvage</td>\n",
       "      <td>oulu</td>\n",
       "      <td>oulu</td>\n",
       "      <td>Oulu</td>\n",
       "      <td>GAZ:00010238</td>\n",
       "      <td>Oulu</td>\n",
       "      <td>GAZ</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GAZ_00010238</td>\n",
       "      <td>gaz</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>salvage</td>\n",
       "      <td>organic feature</td>\n",
       "      <td>organic feature</td>\n",
       "      <td>organic feature</td>\n",
       "      <td>ENVO:01000159</td>\n",
       "      <td>organic feature</td>\n",
       "      <td>MICRO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000159</td>\n",
       "      <td>micro</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>salvage</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NCIT:C179011</td>\n",
       "      <td>COVID-19 Disease Severity 1</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179011</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>salvage</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>NCIT:C105718</td>\n",
       "      <td>Karnofsky Performance Status 10</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C105718</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>salvage</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>NCIT:C113425</td>\n",
       "      <td>Eleven</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113425</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>salvage</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>NCIT:C113429</td>\n",
       "      <td>Fifteen</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113429</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>salvage</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>NCIT:C14338</td>\n",
       "      <td>Human Papillomavirus Type 16</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14338</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>salvage</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>NCIT:C113431</td>\n",
       "      <td>Seventeen</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C113431</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>salvage</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>NCIT:C14377</td>\n",
       "      <td>Human Papillomavirus Type 18</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14377</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>salvage</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>NCIT:C105801</td>\n",
       "      <td>Nineteen</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C105801</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>salvage</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NCIT:C172969</td>\n",
       "      <td>Worry Score 2</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172969</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>salvage</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>NCIT:C105716</td>\n",
       "      <td>Karnofsky Performance Status 20</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C105716</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>salvage</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>NCIT:C107434</td>\n",
       "      <td>Twenty One</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C107434</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>salvage</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>NCIT:C102995</td>\n",
       "      <td>Human Papillomavirus-26</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102995</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>salvage</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>NCIT:C126888</td>\n",
       "      <td>Twenty Eight</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C126888</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>salvage</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>NCIT:C172976</td>\n",
       "      <td>Fear Score 3</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172976</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>salvage</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>NCIT:C70541</td>\n",
       "      <td>Lansky Performance Status 30</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C70541</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>salvage</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>NCIT:C102904</td>\n",
       "      <td>Human Papillomavirus-31</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102904</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>salvage</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>NCIT:C102996</td>\n",
       "      <td>Human Papillomavirus-33</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102996</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>salvage</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>NCIT:C102997</td>\n",
       "      <td>Human Papillomavirus-35</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102997</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>salvage</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>NCIT:C102998</td>\n",
       "      <td>Human Papillomavirus-39</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C102998</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>salvage</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NCIT:C172971</td>\n",
       "      <td>Worry Score 4</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172971</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>salvage</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>NCIT:C105714</td>\n",
       "      <td>Karnofsky Performance Status 40</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C105714</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>salvage</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>NCIT:C178273</td>\n",
       "      <td>Forty One</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C178273</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>salvage</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>NCIT:C175676</td>\n",
       "      <td>Forty Two</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C175676</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>salvage</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>NCIT:C178274</td>\n",
       "      <td>Forty Four</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C178274</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>salvage</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>NCIT:C172966</td>\n",
       "      <td>Upset Score 5</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C172966</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>salvage</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>NCIT:C179016</td>\n",
       "      <td>COVID-19 Disease Severity 6</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179016</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>salvage</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>NCIT:C179017</td>\n",
       "      <td>COVID-19 Disease Severity 7</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hasExactSynonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C179017</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>salvage</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>NCIT:C66839</td>\n",
       "      <td>Eight</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C66839</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>salvage</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>NCIT:C136557</td>\n",
       "      <td>Neck Pain Score 9</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C136557</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>salvage</td>\n",
       "      <td>crop</td>\n",
       "      <td>crop</td>\n",
       "      <td>Crop</td>\n",
       "      <td>NCIT:C111162</td>\n",
       "      <td>Crop</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C111162</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>salvage</td>\n",
       "      <td>genome sequence</td>\n",
       "      <td>genome sequence</td>\n",
       "      <td>Genome Assembly Sequence</td>\n",
       "      <td>NCIT:C73517</td>\n",
       "      <td>Genome Assembly Sequence</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C73517</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.182</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>salvage</td>\n",
       "      <td>missing</td>\n",
       "      <td>missing</td>\n",
       "      <td>Missing</td>\n",
       "      <td>NCIT:C54031</td>\n",
       "      <td>Missing</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C54031</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>salvage</td>\n",
       "      <td>normal</td>\n",
       "      <td>normal</td>\n",
       "      <td>Normal</td>\n",
       "      <td>NCIT:C14165</td>\n",
       "      <td>Normal</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14165</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>salvage</td>\n",
       "      <td>plant</td>\n",
       "      <td>plant</td>\n",
       "      <td>Plant</td>\n",
       "      <td>NCIT:C14258</td>\n",
       "      <td>Plant</td>\n",
       "      <td>NCIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCIT_C14258</td>\n",
       "      <td>ncit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>salvage</td>\n",
       "      <td>agar</td>\n",
       "      <td>agar</td>\n",
       "      <td>Agar</td>\n",
       "      <td>OMIT:0001790</td>\n",
       "      <td>Agar</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0001790</td>\n",
       "      <td>omit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>salvage</td>\n",
       "      <td>plant roots</td>\n",
       "      <td>plant roots</td>\n",
       "      <td>Plant Roots</td>\n",
       "      <td>OMIT:0018707</td>\n",
       "      <td>Plant Roots</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0018707</td>\n",
       "      <td>omit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>salvage</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>yacon fermentation</td>\n",
       "      <td>Fermentation</td>\n",
       "      <td>OMIT:0006501</td>\n",
       "      <td>Fermentation</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/OMIT_0006501</td>\n",
       "      <td>omit</td>\n",
       "      <td>1</td>\n",
       "      <td>0.170</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>salvage</td>\n",
       "      <td>growth chamber</td>\n",
       "      <td>growth chamber</td>\n",
       "      <td>growth chamber study</td>\n",
       "      <td>PECO:0007269</td>\n",
       "      <td>growth chamber study</td>\n",
       "      <td>PECO</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PECO_0007269</td>\n",
       "      <td>peco</td>\n",
       "      <td>1</td>\n",
       "      <td>0.173</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>salvage</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>AGRO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/AGRO_00000363</td>\n",
       "      <td>agro</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>salvage</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>AGRO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/AGRO_00000363</td>\n",
       "      <td>agro</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>ENVO:01000433</td>\n",
       "      <td>needleleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000433</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>pool</td>\n",
       "      <td>pool</td>\n",
       "      <td>pool</td>\n",
       "      <td>ENVO:01000966</td>\n",
       "      <td>private swimming pool</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>has_broad_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000966</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           category                   raw                 query  \\\n",
       "0           salvage                 blank                 blank   \n",
       "1           salvage              inoculum              inoculum   \n",
       "2           salvage          beech forest          beech forest   \n",
       "3           salvage  urediniospore tissue  urediniospore tissue   \n",
       "4           salvage                forest                forest   \n",
       "5           salvage             highlands             highlands   \n",
       "6           salvage              oak tree              oak tree   \n",
       "7           salvage                  oulu                  oulu   \n",
       "8           salvage       organic feature       organic feature   \n",
       "9           salvage                     1                     1   \n",
       "10          salvage                    10                    10   \n",
       "11          salvage                    11                    11   \n",
       "12          salvage                    15                    15   \n",
       "13          salvage                    16                    16   \n",
       "14          salvage                    17                    17   \n",
       "15          salvage                    18                    18   \n",
       "16          salvage                    19                    19   \n",
       "17          salvage                     2                     2   \n",
       "18          salvage                    20                    20   \n",
       "19          salvage                    21                    21   \n",
       "20          salvage                    26                    26   \n",
       "21          salvage                    28                    28   \n",
       "22          salvage                     3                     3   \n",
       "23          salvage                    30                    30   \n",
       "24          salvage                    31                    31   \n",
       "25          salvage                    33                    33   \n",
       "26          salvage                    35                    35   \n",
       "27          salvage                    39                    39   \n",
       "28          salvage                     4                     4   \n",
       "29          salvage                    40                    40   \n",
       "30          salvage                    41                    41   \n",
       "31          salvage                    42                    42   \n",
       "32          salvage                    44                    44   \n",
       "33          salvage                     5                     5   \n",
       "34          salvage                     6                     6   \n",
       "35          salvage                     7                     7   \n",
       "36          salvage                     8                     8   \n",
       "37          salvage                     9                     9   \n",
       "38          salvage                  crop                  crop   \n",
       "39          salvage       genome sequence       genome sequence   \n",
       "40          salvage               missing               missing   \n",
       "41          salvage                normal                normal   \n",
       "42          salvage                 plant                 plant   \n",
       "43          salvage                  agar                  agar   \n",
       "44          salvage           plant roots           plant roots   \n",
       "45          salvage    yacon fermentation    yacon fermentation   \n",
       "46          salvage        growth chamber        growth chamber   \n",
       "47          salvage            glasshouse            glasshouse   \n",
       "48          salvage            greenhouse            greenhouse   \n",
       "49  env_local_scale     coniferous forest     coniferous forest   \n",
       "50  env_local_scale                  pool                  pool   \n",
       "\n",
       "                        name         obo_id                            label  \\\n",
       "0                      blank   AFRL:0000259                       blank role   \n",
       "1                   inoculum    ARO:3004417                         inoculum   \n",
       "2          beech forest soil  ENVO:00005770                beech forest soil   \n",
       "3              urediniospore    FAO:0000026                    urediniospore   \n",
       "4                     Forest   GAZ:00454366                           Forest   \n",
       "5                  Highlands   GAZ:00246323                        Highlands   \n",
       "6                   Oak Tree   GAZ:00454476                         Oak Tree   \n",
       "7                       Oulu   GAZ:00010238                             Oulu   \n",
       "8            organic feature  ENVO:01000159                  organic feature   \n",
       "9                          1   NCIT:C179011      COVID-19 Disease Severity 1   \n",
       "10                        10   NCIT:C105718  Karnofsky Performance Status 10   \n",
       "11                        11   NCIT:C113425                           Eleven   \n",
       "12                        15   NCIT:C113429                          Fifteen   \n",
       "13                        16    NCIT:C14338     Human Papillomavirus Type 16   \n",
       "14                        17   NCIT:C113431                        Seventeen   \n",
       "15                        18    NCIT:C14377     Human Papillomavirus Type 18   \n",
       "16                        19   NCIT:C105801                         Nineteen   \n",
       "17                         2   NCIT:C172969                    Worry Score 2   \n",
       "18                        20   NCIT:C105716  Karnofsky Performance Status 20   \n",
       "19                        21   NCIT:C107434                       Twenty One   \n",
       "20                        26   NCIT:C102995          Human Papillomavirus-26   \n",
       "21                        28   NCIT:C126888                     Twenty Eight   \n",
       "22                         3   NCIT:C172976                     Fear Score 3   \n",
       "23                        30    NCIT:C70541     Lansky Performance Status 30   \n",
       "24                        31   NCIT:C102904          Human Papillomavirus-31   \n",
       "25                        33   NCIT:C102996          Human Papillomavirus-33   \n",
       "26                        35   NCIT:C102997          Human Papillomavirus-35   \n",
       "27                        39   NCIT:C102998          Human Papillomavirus-39   \n",
       "28                         4   NCIT:C172971                    Worry Score 4   \n",
       "29                        40   NCIT:C105714  Karnofsky Performance Status 40   \n",
       "30                        41   NCIT:C178273                        Forty One   \n",
       "31                        42   NCIT:C175676                        Forty Two   \n",
       "32                        44   NCIT:C178274                       Forty Four   \n",
       "33                         5   NCIT:C172966                    Upset Score 5   \n",
       "34                         6   NCIT:C179016      COVID-19 Disease Severity 6   \n",
       "35                         7   NCIT:C179017      COVID-19 Disease Severity 7   \n",
       "36                         8    NCIT:C66839                            Eight   \n",
       "37                         9   NCIT:C136557                Neck Pain Score 9   \n",
       "38                      Crop   NCIT:C111162                             Crop   \n",
       "39  Genome Assembly Sequence    NCIT:C73517         Genome Assembly Sequence   \n",
       "40                   Missing    NCIT:C54031                          Missing   \n",
       "41                    Normal    NCIT:C14165                           Normal   \n",
       "42                     Plant    NCIT:C14258                            Plant   \n",
       "43                      Agar   OMIT:0001790                             Agar   \n",
       "44               Plant Roots   OMIT:0018707                      Plant Roots   \n",
       "45              Fermentation   OMIT:0006501                     Fermentation   \n",
       "46      growth chamber study   PECO:0007269             growth chamber study   \n",
       "47                glasshouse  AGRO:00000363                       greenhouse   \n",
       "48                greenhouse  AGRO:00000363                       greenhouse   \n",
       "49         coniferous forest  ENVO:01000433                needleleaf forest   \n",
       "50                      pool  ENVO:01000966            private swimming pool   \n",
       "\n",
       "   ontology_prefix  include  ols_curation_notes              scope  \\\n",
       "0              AFO    False                 NaN            synonym   \n",
       "1              ARO    False                 NaN              label   \n",
       "2             ENVO    False                 NaN              label   \n",
       "3              FAO    False                 NaN              label   \n",
       "4              GAZ    False                 NaN              label   \n",
       "5              GAZ    False                 NaN              label   \n",
       "6              GAZ    False                 NaN              label   \n",
       "7              GAZ    False                 NaN              label   \n",
       "8            MICRO    False                 NaN              label   \n",
       "9             NCIT    False                 NaN    hasExactSynonym   \n",
       "10            NCIT    False                 NaN            synonym   \n",
       "11            NCIT    False                 NaN            synonym   \n",
       "12            NCIT    False                 NaN            synonym   \n",
       "13            NCIT    False                 NaN    hasExactSynonym   \n",
       "14            NCIT    False                 NaN            synonym   \n",
       "15            NCIT    False                 NaN            synonym   \n",
       "16            NCIT    False                 NaN            synonym   \n",
       "17            NCIT    False                 NaN            synonym   \n",
       "18            NCIT    False                 NaN            synonym   \n",
       "19            NCIT    False                 NaN            synonym   \n",
       "20            NCIT    False                 NaN    hasExactSynonym   \n",
       "21            NCIT    False                 NaN            synonym   \n",
       "22            NCIT    False                 NaN            synonym   \n",
       "23            NCIT    False                 NaN    hasExactSynonym   \n",
       "24            NCIT    False                 NaN            synonym   \n",
       "25            NCIT    False                 NaN            synonym   \n",
       "26            NCIT    False                 NaN    hasExactSynonym   \n",
       "27            NCIT    False                 NaN    hasExactSynonym   \n",
       "28            NCIT    False                 NaN            synonym   \n",
       "29            NCIT    False                 NaN            synonym   \n",
       "30            NCIT    False                 NaN            synonym   \n",
       "31            NCIT    False                 NaN            synonym   \n",
       "32            NCIT    False                 NaN            synonym   \n",
       "33            NCIT    False                 NaN            synonym   \n",
       "34            NCIT    False                 NaN    hasExactSynonym   \n",
       "35            NCIT    False                 NaN    hasExactSynonym   \n",
       "36            NCIT    False                 NaN            synonym   \n",
       "37            NCIT    False                 NaN            synonym   \n",
       "38            NCIT    False                 NaN              label   \n",
       "39            NCIT    False                 NaN              label   \n",
       "40            NCIT    False                 NaN              label   \n",
       "41            NCIT    False                 NaN              label   \n",
       "42            NCIT    False                 NaN              label   \n",
       "43            OMIT    False                 NaN              label   \n",
       "44            OMIT    False                 NaN              label   \n",
       "45            OMIT    False                 NaN              label   \n",
       "46            PECO    False                 NaN              label   \n",
       "47            AGRO     True                 NaN            synonym   \n",
       "48            AGRO     True                 NaN              label   \n",
       "49            ENVO     True                 NaN            synonym   \n",
       "50            ENVO     True                 NaN  has_broad_synonym   \n",
       "\n",
       "          type                                                iri  \\\n",
       "0      synonym  http://purl.allotrope.org/ontologies/role#AFRL...   \n",
       "1        label         http://purl.obolibrary.org/obo/ARO_3004417   \n",
       "2        label       http://purl.obolibrary.org/obo/ENVO_00005770   \n",
       "3        label         http://purl.obolibrary.org/obo/FAO_0000026   \n",
       "4        label        http://purl.obolibrary.org/obo/GAZ_00454366   \n",
       "5        label        http://purl.obolibrary.org/obo/GAZ_00246323   \n",
       "6        label        http://purl.obolibrary.org/obo/GAZ_00454476   \n",
       "7        label        http://purl.obolibrary.org/obo/GAZ_00010238   \n",
       "8        label       http://purl.obolibrary.org/obo/ENVO_01000159   \n",
       "9          NaN        http://purl.obolibrary.org/obo/NCIT_C179011   \n",
       "10     synonym        http://purl.obolibrary.org/obo/NCIT_C105718   \n",
       "11     synonym        http://purl.obolibrary.org/obo/NCIT_C113425   \n",
       "12     synonym        http://purl.obolibrary.org/obo/NCIT_C113429   \n",
       "13         NaN         http://purl.obolibrary.org/obo/NCIT_C14338   \n",
       "14     synonym        http://purl.obolibrary.org/obo/NCIT_C113431   \n",
       "15     synonym         http://purl.obolibrary.org/obo/NCIT_C14377   \n",
       "16     synonym        http://purl.obolibrary.org/obo/NCIT_C105801   \n",
       "17     synonym        http://purl.obolibrary.org/obo/NCIT_C172969   \n",
       "18     synonym        http://purl.obolibrary.org/obo/NCIT_C105716   \n",
       "19     synonym        http://purl.obolibrary.org/obo/NCIT_C107434   \n",
       "20         NaN        http://purl.obolibrary.org/obo/NCIT_C102995   \n",
       "21     synonym        http://purl.obolibrary.org/obo/NCIT_C126888   \n",
       "22     synonym        http://purl.obolibrary.org/obo/NCIT_C172976   \n",
       "23         NaN         http://purl.obolibrary.org/obo/NCIT_C70541   \n",
       "24     synonym        http://purl.obolibrary.org/obo/NCIT_C102904   \n",
       "25     synonym        http://purl.obolibrary.org/obo/NCIT_C102996   \n",
       "26         NaN        http://purl.obolibrary.org/obo/NCIT_C102997   \n",
       "27         NaN        http://purl.obolibrary.org/obo/NCIT_C102998   \n",
       "28     synonym        http://purl.obolibrary.org/obo/NCIT_C172971   \n",
       "29     synonym        http://purl.obolibrary.org/obo/NCIT_C105714   \n",
       "30     synonym        http://purl.obolibrary.org/obo/NCIT_C178273   \n",
       "31     synonym        http://purl.obolibrary.org/obo/NCIT_C175676   \n",
       "32     synonym        http://purl.obolibrary.org/obo/NCIT_C178274   \n",
       "33     synonym        http://purl.obolibrary.org/obo/NCIT_C172966   \n",
       "34         NaN        http://purl.obolibrary.org/obo/NCIT_C179016   \n",
       "35         NaN        http://purl.obolibrary.org/obo/NCIT_C179017   \n",
       "36     synonym         http://purl.obolibrary.org/obo/NCIT_C66839   \n",
       "37     synonym        http://purl.obolibrary.org/obo/NCIT_C136557   \n",
       "38       label        http://purl.obolibrary.org/obo/NCIT_C111162   \n",
       "39       label         http://purl.obolibrary.org/obo/NCIT_C73517   \n",
       "40       label         http://purl.obolibrary.org/obo/NCIT_C54031   \n",
       "41       label         http://purl.obolibrary.org/obo/NCIT_C14165   \n",
       "42       label         http://purl.obolibrary.org/obo/NCIT_C14258   \n",
       "43       label        http://purl.obolibrary.org/obo/OMIT_0001790   \n",
       "44       label        http://purl.obolibrary.org/obo/OMIT_0018707   \n",
       "45       label        http://purl.obolibrary.org/obo/OMIT_0006501   \n",
       "46       label        http://purl.obolibrary.org/obo/PECO_0007269   \n",
       "47     synonym       http://purl.obolibrary.org/obo/AGRO_00000363   \n",
       "48       label       http://purl.obolibrary.org/obo/AGRO_00000363   \n",
       "49     synonym       http://purl.obolibrary.org/obo/ENVO_01000433   \n",
       "50  annotation       http://purl.obolibrary.org/obo/ENVO_01000966   \n",
       "\n",
       "   ontology_name  string_dist_rank  string_dist  search_rank  \n",
       "0            afo                 1        0.000            1  \n",
       "1            aro                 1        0.000            1  \n",
       "2           envo                 1        0.171            1  \n",
       "3            fao                 1        0.184            1  \n",
       "4            gaz                 1        0.000            1  \n",
       "5            gaz                 1        0.000            1  \n",
       "6            gaz                 1        0.000            1  \n",
       "7            gaz                 1        0.000            1  \n",
       "8          micro                 1        0.000            1  \n",
       "9           ncit                 1        0.000            3  \n",
       "10          ncit                 1        0.000            1  \n",
       "11          ncit                 1        0.000            1  \n",
       "12          ncit                 1        0.000            5  \n",
       "13          ncit                 1        0.000            5  \n",
       "14          ncit                 1        0.000            5  \n",
       "15          ncit                 1        0.000            1  \n",
       "16          ncit                 1        0.000            4  \n",
       "17          ncit                 1        0.000            5  \n",
       "18          ncit                 1        0.000            1  \n",
       "19          ncit                 1        0.000            5  \n",
       "20          ncit                 1        0.000            1  \n",
       "21          ncit                 1        0.000            3  \n",
       "22          ncit                 1        0.000            5  \n",
       "23          ncit                 1        0.000            5  \n",
       "24          ncit                 1        0.000            1  \n",
       "25          ncit                 1        0.000            1  \n",
       "26          ncit                 1        0.000            1  \n",
       "27          ncit                 1        0.000            1  \n",
       "28          ncit                 1        0.000            3  \n",
       "29          ncit                 1        0.000            1  \n",
       "30          ncit                 1        0.000            5  \n",
       "31          ncit                 1        0.000            4  \n",
       "32          ncit                 1        0.000            4  \n",
       "33          ncit                 1        0.000            3  \n",
       "34          ncit                 1        0.000            3  \n",
       "35          ncit                 1        0.000            3  \n",
       "36          ncit                 1        0.000            1  \n",
       "37          ncit                 1        0.000            1  \n",
       "38          ncit                 1        0.000            1  \n",
       "39          ncit                 1        0.182            3  \n",
       "40          ncit                 1        0.000            1  \n",
       "41          ncit                 1        0.000            1  \n",
       "42          ncit                 1        0.000            1  \n",
       "43          omit                 1        0.000            1  \n",
       "44          omit                 1        0.000            1  \n",
       "45          omit                 1        0.170            4  \n",
       "46          peco                 1        0.173            1  \n",
       "47          agro                 1        0.000            1  \n",
       "48          agro                 1        0.000            1  \n",
       "49          envo                 1        0.000            1  \n",
       "50          envo                 1        0.000            2  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curated = pd.read_clipboard()\n",
    "curated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ea200e4",
   "metadata": {},
   "source": [
    "## Filter on `include`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "1473206f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ols_category</th>\n",
       "      <th>ols_raw</th>\n",
       "      <th>ols_query</th>\n",
       "      <th>ols_name</th>\n",
       "      <th>ols_obo_id</th>\n",
       "      <th>ols_label</th>\n",
       "      <th>ols_ontology_prefix</th>\n",
       "      <th>ols_include</th>\n",
       "      <th>ols_ols_curation_notes</th>\n",
       "      <th>ols_scope</th>\n",
       "      <th>ols_type</th>\n",
       "      <th>ols_iri</th>\n",
       "      <th>ols_ontology_name</th>\n",
       "      <th>ols_string_dist_rank</th>\n",
       "      <th>ols_string_dist</th>\n",
       "      <th>ols_search_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>salvage</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>AGRO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/AGRO_00000363</td>\n",
       "      <td>agro</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>salvage</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>AGRO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>label</td>\n",
       "      <td>label</td>\n",
       "      <td>http://purl.obolibrary.org/obo/AGRO_00000363</td>\n",
       "      <td>agro</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>ENVO:01000433</td>\n",
       "      <td>needleleaf forest</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>synonym</td>\n",
       "      <td>synonym</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000433</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>pool</td>\n",
       "      <td>pool</td>\n",
       "      <td>pool</td>\n",
       "      <td>ENVO:01000966</td>\n",
       "      <td>private swimming pool</td>\n",
       "      <td>ENVO</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>has_broad_synonym</td>\n",
       "      <td>annotation</td>\n",
       "      <td>http://purl.obolibrary.org/obo/ENVO_01000966</td>\n",
       "      <td>envo</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ols_category            ols_raw          ols_query           ols_name  \\\n",
       "47          salvage         glasshouse         glasshouse         glasshouse   \n",
       "48          salvage         greenhouse         greenhouse         greenhouse   \n",
       "49  env_local_scale  coniferous forest  coniferous forest  coniferous forest   \n",
       "50  env_local_scale               pool               pool               pool   \n",
       "\n",
       "       ols_obo_id              ols_label ols_ontology_prefix  ols_include  \\\n",
       "47  AGRO:00000363             greenhouse                AGRO         True   \n",
       "48  AGRO:00000363             greenhouse                AGRO         True   \n",
       "49  ENVO:01000433      needleleaf forest                ENVO         True   \n",
       "50  ENVO:01000966  private swimming pool                ENVO         True   \n",
       "\n",
       "    ols_ols_curation_notes          ols_scope    ols_type  \\\n",
       "47                     NaN            synonym     synonym   \n",
       "48                     NaN              label       label   \n",
       "49                     NaN            synonym     synonym   \n",
       "50                     NaN  has_broad_synonym  annotation   \n",
       "\n",
       "                                         ols_iri ols_ontology_name  \\\n",
       "47  http://purl.obolibrary.org/obo/AGRO_00000363              agro   \n",
       "48  http://purl.obolibrary.org/obo/AGRO_00000363              agro   \n",
       "49  http://purl.obolibrary.org/obo/ENVO_01000433              envo   \n",
       "50  http://purl.obolibrary.org/obo/ENVO_01000966              envo   \n",
       "\n",
       "    ols_string_dist_rank  ols_string_dist  ols_search_rank  \n",
       "47                     1              0.0                1  \n",
       "48                     1              0.0                1  \n",
       "49                     1              0.0                1  \n",
       "50                     1              0.0                2  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flag = curated[\"include\"]\n",
    "curated = curated.loc[flag]\n",
    "\n",
    "curated.columns = \"ols_\" + curated.columns\n",
    "# curated.to_sql(\"curated\", biosample_cnx, if_exists=\"replace\", index=False)\n",
    "curated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ac6219e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_search_merged = mapping_candidates.merge(\n",
    "    curated, how=\"outer\", left_on=\"remaining_tidied\", right_on=\"ols_raw\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "d2c242b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "flag = (\n",
    "    ~merge_search_merged[\"id_or_lab_ok\"]\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].eq(\"\")\n",
    "    & ~merge_search_merged[\"ols_obo_id\"].isna()\n",
    ")\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_obo_id\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_id\"] = replacement\n",
    "\n",
    "replacement = merge_search_merged.loc[flag, \"ols_label\"]\n",
    "\n",
    "merge_search_merged.loc[flag, \"consensus_lab\"] = replacement\n",
    "\n",
    "replacement = (\n",
    "    merge_search_merged.loc[flag, \"consensus_lab\"]\n",
    "    + \" [\"\n",
    "    + merge_search_merged.loc[flag, \"consensus_id\"]\n",
    "    + \"]\"\n",
    ")\n",
    "\n",
    "merge_search_merged.loc[flag, \"assembled_consensus\"] = replacement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b333c111",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge_search_merged.to_sql(\n",
    "#     \"merge_search_merged\", biosample_cnx, if_exists=\"replace\", index=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "5e772701",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>assembled_consensus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>558</td>\n",
       "      <td>root [PO:0009005]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>310</td>\n",
       "      <td>rice field [ENVO:00000296]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agar</td>\n",
       "      <td>270</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lettuce field</td>\n",
       "      <td>220</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Two-weeks old</td>\n",
       "      <td>219</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>DNA from roots of tree 24_8</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>Highlands</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Lowlands</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Mountain</td>\n",
       "      <td>1</td>\n",
       "      <td>mountain [ENVO:00000081]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>125 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 env_local_scale  count         assembled_consensus\n",
       "0                           Root    558           root [PO:0009005]\n",
       "1                     rice field    310  rice field [ENVO:00000296]\n",
       "2                           Agar    270                         NaN\n",
       "3                  lettuce field    220                         NaN\n",
       "4                  Two-weeks old    219                         NaN\n",
       "..                           ...    ...                         ...\n",
       "69   DNA from roots of tree 24_8      1                         NaN\n",
       "68                     Highlands      1                         NaN\n",
       "67                      Lowlands      1                         NaN\n",
       "66                      Mountain      1    mountain [ENVO:00000081]\n",
       "124                            1      1                         NaN\n",
       "\n",
       "[125 rows x 3 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary = merge_search_merged[[biosample_col_to_map, \"count\", \"assembled_consensus\"]]\n",
    "summary = summary.sort_values(\"count\", ascending=False)\n",
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "38acac10",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_category</th>\n",
       "      <th>subject_label</th>\n",
       "      <th>predicate_id</th>\n",
       "      <th>object_id</th>\n",
       "      <th>object_label</th>\n",
       "      <th>match_type</th>\n",
       "      <th>creator_id</th>\n",
       "      <th>mapping_date</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>Root</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>rice field</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>garden</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000011</td>\n",
       "      <td>garden</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000433</td>\n",
       "      <td>needleleaf forest</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>prairie</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000260</td>\n",
       "      <td>prairie</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>field</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>vineyard</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>Vineyard</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>stream</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000023</td>\n",
       "      <td>stream</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>pool</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:01000966</td>\n",
       "      <td>private swimming pool</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>Rhizosphere</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>research facility</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000469</td>\n",
       "      <td>research facility</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>bulk soil</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00005802</td>\n",
       "      <td>bulk soil</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>glasshouse</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>Plantation</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000117</td>\n",
       "      <td>plantation</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>env_local_scale</td>\n",
       "      <td>Mountain</td>\n",
       "      <td>skos:relatedMatch</td>\n",
       "      <td>ENVO:00000081</td>\n",
       "      <td>mountain</td>\n",
       "      <td>Lexical</td>\n",
       "      <td>https://github.com/turbomam/scoped-mapping</td>\n",
       "      <td>2021-06-11T18:33:50</td>\n",
       "      <td>env_local_scale_env_package_normalization.EnvP...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   subject_category      subject_label       predicate_id      object_id  \\\n",
       "0   env_local_scale               Root  skos:relatedMatch     PO:0009005   \n",
       "1   env_local_scale         rice field  skos:relatedMatch  ENVO:00000296   \n",
       "7   env_local_scale             garden  skos:relatedMatch  ENVO:00000011   \n",
       "8   env_local_scale  coniferous forest  skos:relatedMatch  ENVO:01000433   \n",
       "11  env_local_scale            prairie  skos:relatedMatch  ENVO:00000260   \n",
       "12  env_local_scale              field  skos:relatedMatch  ENVO:01000352   \n",
       "14  env_local_scale           vineyard  skos:relatedMatch  ENVO:00000116   \n",
       "15  env_local_scale           Vineyard  skos:relatedMatch  ENVO:00000116   \n",
       "16  env_local_scale             stream  skos:relatedMatch  ENVO:00000023   \n",
       "17  env_local_scale               pool  skos:relatedMatch  ENVO:01000966   \n",
       "20  env_local_scale        rhizosphere  skos:relatedMatch  ENVO:00005801   \n",
       "21  env_local_scale        Rhizosphere  skos:relatedMatch  ENVO:00005801   \n",
       "28  env_local_scale  research facility  skos:relatedMatch  ENVO:00000469   \n",
       "31  env_local_scale          bulk soil  skos:relatedMatch  ENVO:00005802   \n",
       "34  env_local_scale         greenhouse  skos:relatedMatch  AGRO:00000363   \n",
       "37  env_local_scale         glasshouse  skos:relatedMatch  AGRO:00000363   \n",
       "62  env_local_scale         Plantation  skos:relatedMatch  ENVO:00000117   \n",
       "66  env_local_scale           Mountain  skos:relatedMatch  ENVO:00000081   \n",
       "\n",
       "             object_label match_type  \\\n",
       "0                    root    Lexical   \n",
       "1              rice field    Lexical   \n",
       "7                  garden    Lexical   \n",
       "8       needleleaf forest    Lexical   \n",
       "11                prairie    Lexical   \n",
       "12                  field    Lexical   \n",
       "14               vineyard    Lexical   \n",
       "15               vineyard    Lexical   \n",
       "16                 stream    Lexical   \n",
       "17  private swimming pool    Lexical   \n",
       "20            rhizosphere    Lexical   \n",
       "21            rhizosphere    Lexical   \n",
       "28      research facility    Lexical   \n",
       "31              bulk soil    Lexical   \n",
       "34             greenhouse    Lexical   \n",
       "37             greenhouse    Lexical   \n",
       "62             plantation    Lexical   \n",
       "66               mountain    Lexical   \n",
       "\n",
       "                                    creator_id         mapping_date  \\\n",
       "0   https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "1   https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "7   https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "8   https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "11  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "12  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "14  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "15  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "16  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "17  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "20  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "21  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "28  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "31  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "34  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "37  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "62  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "66  https://github.com/turbomam/scoped-mapping  2021-06-11T18:33:50   \n",
       "\n",
       "                                           subject_id  \n",
       "0   env_local_scale_env_package_normalization.EnvP...  \n",
       "1   env_local_scale_env_package_normalization.EnvP...  \n",
       "7   env_local_scale_env_package_normalization.EnvP...  \n",
       "8   env_local_scale_env_package_normalization.EnvP...  \n",
       "11  env_local_scale_env_package_normalization.EnvP...  \n",
       "12  env_local_scale_env_package_normalization.EnvP...  \n",
       "14  env_local_scale_env_package_normalization.EnvP...  \n",
       "15  env_local_scale_env_package_normalization.EnvP...  \n",
       "16  env_local_scale_env_package_normalization.EnvP...  \n",
       "17  env_local_scale_env_package_normalization.EnvP...  \n",
       "20  env_local_scale_env_package_normalization.EnvP...  \n",
       "21  env_local_scale_env_package_normalization.EnvP...  \n",
       "28  env_local_scale_env_package_normalization.EnvP...  \n",
       "31  env_local_scale_env_package_normalization.EnvP...  \n",
       "34  env_local_scale_env_package_normalization.EnvP...  \n",
       "37  env_local_scale_env_package_normalization.EnvP...  \n",
       "62  env_local_scale_env_package_normalization.EnvP...  \n",
       "66  env_local_scale_env_package_normalization.EnvP...  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frame_for_sssom = merge_search_merged\n",
    "\n",
    "na_flag = frame_for_sssom[\"splitted\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "na_flag = frame_for_sssom[\"consensus_id\"].isna()\n",
    "frame_for_sssom = frame_for_sssom[~na_flag]\n",
    "\n",
    "raw_queries = list(frame_for_sssom[\"splitted\"])\n",
    "urlencodeds = list(map(urllib.parse.quote, raw_queries))\n",
    "urlencodeds = [sssom_subject_prefix + \":\" + item for item in urlencodeds]\n",
    "\n",
    "results_rows = len(frame_for_sssom.index)\n",
    "iso8601_stamp = datetime.now().replace(microsecond=0).isoformat()\n",
    "\n",
    "sssom_frame = {\n",
    "    \"subject_category\": [biosample_col_to_map] * results_rows,\n",
    "    \"subject_label\": frame_for_sssom[\"splitted\"],\n",
    "    \"predicate_id\": [\"skos:relatedMatch\"] * results_rows,\n",
    "    \"object_id\": frame_for_sssom[\"consensus_id\"],\n",
    "    \"object_label\": frame_for_sssom[\"consensus_lab\"],\n",
    "    \"match_type\": [\"Lexical\"] * results_rows,\n",
    "    \"creator_id\": [\"https://github.com/turbomam/scoped-mapping\"] * results_rows,\n",
    "    #     \"mapping_tool\": [\"https://www.ebi.ac.uk/ols/docs/api\"] * results_rows,\n",
    "    \"mapping_date\": [iso8601_stamp] * results_rows,\n",
    "    #     \"confidence\": 1 - frame_for_sssom[\"string_dist\"],\n",
    "    \"subject_id\": urlencodeds,\n",
    "}\n",
    "\n",
    "sssom_frame = pd.DataFrame(sssom_frame)\n",
    "\n",
    "sssom_frame.to_csv(sssom_file, sep=ols_review_seperator, index=False)\n",
    "\n",
    "sssom_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ad82349",
   "metadata": {},
   "source": [
    "# From here down requires refactoring and generalization \n",
    "\n",
    "That may even be true for some of the blocks above this!\n",
    "\n",
    "## Join together pipe-delimited env_broad_scale mappings \n",
    "_Any other constraints? If part_count is > 1, do we need to check for empty or null env_broad_scale, repeated or splitted?_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "32ad2b25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    125\n",
       "Name: part_count, dtype: int64"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_part_flag = merge_search_merged[\"part_count\"] > 1\n",
    "multi_part_flag.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "d686f6ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [env_local_scale, count, repeated, splitted, part_count, seq, consensus_id, consensus_lab]\n",
       "Index: []"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_part_frame = merge_search_merged.loc[multi_part_flag]\n",
    "\n",
    "multi_part_frame = multi_part_frame[\n",
    "    [\n",
    "        biosample_col_to_map,\n",
    "        \"count\",\n",
    "        \"repeated\",\n",
    "        \"splitted\",\n",
    "        \"part_count\",\n",
    "        \"seq\",\n",
    "        \"consensus_id\",\n",
    "        \"consensus_lab\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "multi_part_frame = multi_part_frame.sort_values(\n",
    "    [biosample_col_to_map, \"seq\"], ascending=(True, True)\n",
    ")\n",
    "\n",
    "multi_part_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "f147c6b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: part_count, dtype: int64)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "part_count = multi_part_frame.groupby([biosample_col_to_map], sort=False)[\n",
    "    \"part_count\"\n",
    "].max()\n",
    "part_count.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "5db86116",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: seq, dtype: float64)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_max = multi_part_frame.groupby([biosample_col_to_map], sort=False)[\"seq\"].max()\n",
    "seq_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "c4ccf78a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>count</th>\n",
       "      <th>repeated</th>\n",
       "      <th>splitted</th>\n",
       "      <th>part_count</th>\n",
       "      <th>seq</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [env_local_scale, count, repeated, splitted, part_count, seq, consensus_id, consensus_lab]\n",
       "Index: []"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parts_check = pd.DataFrame(dict(part_count=part_count, seq_max=seq_max)).reset_index()\n",
    "\n",
    "all_parts_flag = parts_check[\"seq_max\"] == parts_check[\"part_count\"]\n",
    "ebs_with_all_parts = parts_check.loc[all_parts_flag, biosample_col_to_map]\n",
    "ebs_with_all_parts\n",
    "\n",
    "all_parts_flag = multi_part_frame[biosample_col_to_map].isin(list(ebs_with_all_parts))\n",
    "all_parts_frame = multi_part_frame.loc[all_parts_flag]\n",
    "all_parts_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f0cbd251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_pipesep_ebs = list(set(list(all_parts_frame[biosample_col_to_map])))\n",
    "unique_pipesep_ebs.sort()\n",
    "repipe_dict_list = []\n",
    "for one_pipesep in unique_pipesep_ebs:\n",
    "    flag = all_parts_frame[biosample_col_to_map] == one_pipesep\n",
    "    temp_frame = all_parts_frame.loc[flag]\n",
    "    ids_list = list(temp_frame[\"consensus_id\"])\n",
    "    lab_list = list(temp_frame[\"consensus_lab\"])\n",
    "    s = \"|\"\n",
    "    ids_repipe = s.join(ids_list)\n",
    "    lab_repipe = s.join(lab_list)\n",
    "\n",
    "    repipe_dict = {\n",
    "        biosample_col_to_map: one_pipesep,\n",
    "        \"consensus_id\": ids_repipe,\n",
    "        \"consensus_lab\": lab_repipe,\n",
    "    }\n",
    "    repipe_dict_list.append(repipe_dict)\n",
    "repipe_frame = pd.DataFrame(repipe_dict_list)\n",
    "repipe_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "0e8d54b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>garden</td>\n",
       "      <td>ENVO:00000011</td>\n",
       "      <td>garden</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>ENVO:01000433</td>\n",
       "      <td>needleleaf forest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>prairie</td>\n",
       "      <td>ENVO:00000260</td>\n",
       "      <td>prairie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>field</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>vineyard</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Vineyard</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>stream</td>\n",
       "      <td>ENVO:00000023</td>\n",
       "      <td>stream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>pool</td>\n",
       "      <td>ENVO:01000966</td>\n",
       "      <td>private swimming pool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>research facility</td>\n",
       "      <td>ENVO:00000469</td>\n",
       "      <td>research facility</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>bulk soil</td>\n",
       "      <td>ENVO:00005802</td>\n",
       "      <td>bulk soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>greenhouse</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>glasshouse</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Plantation</td>\n",
       "      <td>ENVO:00000117</td>\n",
       "      <td>plantation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Mountain</td>\n",
       "      <td>ENVO:00000081</td>\n",
       "      <td>mountain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_local_scale   consensus_id          consensus_lab\n",
       "0                Root     PO:0009005                   root\n",
       "1          rice field  ENVO:00000296             rice field\n",
       "7              garden  ENVO:00000011                 garden\n",
       "8   coniferous forest  ENVO:01000433      needleleaf forest\n",
       "11            prairie  ENVO:00000260                prairie\n",
       "12              field  ENVO:01000352                  field\n",
       "14           vineyard  ENVO:00000116               vineyard\n",
       "15           Vineyard  ENVO:00000116               vineyard\n",
       "16             stream  ENVO:00000023                 stream\n",
       "17               pool  ENVO:01000966  private swimming pool\n",
       "20        rhizosphere  ENVO:00005801            rhizosphere\n",
       "21        Rhizosphere  ENVO:00005801            rhizosphere\n",
       "28  research facility  ENVO:00000469      research facility\n",
       "31          bulk soil  ENVO:00005802              bulk soil\n",
       "34         greenhouse  AGRO:00000363             greenhouse\n",
       "37         glasshouse  AGRO:00000363             greenhouse\n",
       "62         Plantation  ENVO:00000117             plantation\n",
       "66           Mountain  ENVO:00000081               mountain"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sssom_min_cols = sssom_frame[[\"subject_label\", \"object_id\", \"object_label\"]]\n",
    "\n",
    "sssom_min_cols.columns = [biosample_col_to_map, \"consensus_id\", \"consensus_lab\"]\n",
    "\n",
    "sssom_min_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "25424849",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Root</td>\n",
       "      <td>PO:0009005</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rice field</td>\n",
       "      <td>ENVO:00000296</td>\n",
       "      <td>rice field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>garden</td>\n",
       "      <td>ENVO:00000011</td>\n",
       "      <td>garden</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>coniferous forest</td>\n",
       "      <td>ENVO:01000433</td>\n",
       "      <td>needleleaf forest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>prairie</td>\n",
       "      <td>ENVO:00000260</td>\n",
       "      <td>prairie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>field</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>vineyard</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Vineyard</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>stream</td>\n",
       "      <td>ENVO:00000023</td>\n",
       "      <td>stream</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>pool</td>\n",
       "      <td>ENVO:01000966</td>\n",
       "      <td>private swimming pool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>research facility</td>\n",
       "      <td>ENVO:00000469</td>\n",
       "      <td>research facility</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>bulk soil</td>\n",
       "      <td>ENVO:00005802</td>\n",
       "      <td>bulk soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>greenhouse</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>glasshouse</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Plantation</td>\n",
       "      <td>ENVO:00000117</td>\n",
       "      <td>plantation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Mountain</td>\n",
       "      <td>ENVO:00000081</td>\n",
       "      <td>mountain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      env_local_scale   consensus_id          consensus_lab\n",
       "0                Root     PO:0009005                   root\n",
       "1          rice field  ENVO:00000296             rice field\n",
       "7              garden  ENVO:00000011                 garden\n",
       "8   coniferous forest  ENVO:01000433      needleleaf forest\n",
       "11            prairie  ENVO:00000260                prairie\n",
       "12              field  ENVO:01000352                  field\n",
       "14           vineyard  ENVO:00000116               vineyard\n",
       "15           Vineyard  ENVO:00000116               vineyard\n",
       "16             stream  ENVO:00000023                 stream\n",
       "17               pool  ENVO:01000966  private swimming pool\n",
       "20        rhizosphere  ENVO:00005801            rhizosphere\n",
       "21        Rhizosphere  ENVO:00005801            rhizosphere\n",
       "28  research facility  ENVO:00000469      research facility\n",
       "31          bulk soil  ENVO:00005802              bulk soil\n",
       "34         greenhouse  AGRO:00000363             greenhouse\n",
       "37         glasshouse  AGRO:00000363             greenhouse\n",
       "62         Plantation  ENVO:00000117             plantation\n",
       "66           Mountain  ENVO:00000081               mountain"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repipe_rows = len(repipe_frame.index)\n",
    "if repipe_rows > 0:\n",
    "    direct_and_repipe = pd.concat([temp, repipe_frame])\n",
    "else:\n",
    "    direct_and_repipe = sssom_min_cols\n",
    "direct_and_repipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "5ab3aaa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "direct_and_repipe.to_sql(\n",
    "    \"direct_and_repipe\", biosample_cnx, if_exists=\"replace\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "922c09b3",
   "metadata": {},
   "source": [
    "## refactor this even more?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "8de27d49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-06-11 18:33:51.065664\n",
      "2021-06-11 18:33:51.210968\n",
      "0:00:00.145304\n"
     ]
    }
   ],
   "source": [
    "q = (\n",
    "    \"\"\"\n",
    "select\n",
    "\tb.id, b.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\",\n",
    "\tdar.consensus_id ,\n",
    "\tdar.consensus_lab\n",
    "from\n",
    "\tdirect_and_repipe dar\n",
    "left join biosample b on\n",
    "\tb.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\" = dar.\"\"\"\n",
    "    + biosample_col_to_map\n",
    "    + \"\"\"\n",
    "inner join biosample_sample_taxon_summary stic on\n",
    "\tb.taxonomy_id = stic.biosample_taxid\n",
    "join env_package_normalization on\n",
    "\tb.env_package = env_package_normalization.env_package\n",
    "where \"\"\"\n",
    "    + scoping_col\n",
    "    + \" = '\"\n",
    "    + scoping_value\n",
    "    + \"'\"\n",
    ")\n",
    "\n",
    "[per_biosample_scoped_mapping_results, query_duration] = scoped_mapping.timed_query(\n",
    "    q, biosample_cnx, print_timing=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "a863e532",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>env_local_scale</th>\n",
       "      <th>consensus_id</th>\n",
       "      <th>consensus_lab</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BIOSAMPLE:SAMEA1695842</td>\n",
       "      <td>field</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BIOSAMPLE:SAMEA1695808</td>\n",
       "      <td>field</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIOSAMPLE:SAMEA1695787</td>\n",
       "      <td>field</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BIOSAMPLE:SAMEA1695839</td>\n",
       "      <td>field</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BIOSAMPLE:SAMEA1695833</td>\n",
       "      <td>field</td>\n",
       "      <td>ENVO:01000352</td>\n",
       "      <td>field</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1457</th>\n",
       "      <td>BIOSAMPLE:SAMN02628361</td>\n",
       "      <td>Vineyard</td>\n",
       "      <td>ENVO:00000116</td>\n",
       "      <td>vineyard</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1458</th>\n",
       "      <td>BIOSAMPLE:SAMN02640034</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>ENVO:00005801</td>\n",
       "      <td>rhizosphere</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>BIOSAMPLE:SAMN02641154</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>AGRO:00000363</td>\n",
       "      <td>greenhouse</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1460</th>\n",
       "      <td>BIOSAMPLE:SAMN02664684</td>\n",
       "      <td>Mountain</td>\n",
       "      <td>ENVO:00000081</td>\n",
       "      <td>mountain</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1461</th>\n",
       "      <td>BIOSAMPLE:SAMN02709059</td>\n",
       "      <td>Plantation</td>\n",
       "      <td>ENVO:00000117</td>\n",
       "      <td>plantation</td>\n",
       "      <td>env_package_normalization.EnvPackage=plant-ass...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1462 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          id env_local_scale   consensus_id consensus_lab  \\\n",
       "0     BIOSAMPLE:SAMEA1695842           field  ENVO:01000352         field   \n",
       "1     BIOSAMPLE:SAMEA1695808           field  ENVO:01000352         field   \n",
       "2     BIOSAMPLE:SAMEA1695787           field  ENVO:01000352         field   \n",
       "3     BIOSAMPLE:SAMEA1695839           field  ENVO:01000352         field   \n",
       "4     BIOSAMPLE:SAMEA1695833           field  ENVO:01000352         field   \n",
       "...                      ...             ...            ...           ...   \n",
       "1457  BIOSAMPLE:SAMN02628361        Vineyard  ENVO:00000116      vineyard   \n",
       "1458  BIOSAMPLE:SAMN02640034     rhizosphere  ENVO:00005801   rhizosphere   \n",
       "1459  BIOSAMPLE:SAMN02641154      greenhouse  AGRO:00000363    greenhouse   \n",
       "1460  BIOSAMPLE:SAMN02664684        Mountain  ENVO:00000081      mountain   \n",
       "1461  BIOSAMPLE:SAMN02709059      Plantation  ENVO:00000117    plantation   \n",
       "\n",
       "                                                  notes  \n",
       "0     env_package_normalization.EnvPackage=plant-ass...  \n",
       "1     env_package_normalization.EnvPackage=plant-ass...  \n",
       "2     env_package_normalization.EnvPackage=plant-ass...  \n",
       "3     env_package_normalization.EnvPackage=plant-ass...  \n",
       "4     env_package_normalization.EnvPackage=plant-ass...  \n",
       "...                                                 ...  \n",
       "1457  env_package_normalization.EnvPackage=plant-ass...  \n",
       "1458  env_package_normalization.EnvPackage=plant-ass...  \n",
       "1459  env_package_normalization.EnvPackage=plant-ass...  \n",
       "1460  env_package_normalization.EnvPackage=plant-ass...  \n",
       "1461  env_package_normalization.EnvPackage=plant-ass...  \n",
       "\n",
       "[1462 rows x 5 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "per_biosample_scoped_mapping_results[\"notes\"] = scoping_col + \"=\" + scoping_value\n",
    "\n",
    "table_name = \"repaired_\" + biosample_col_to_map\n",
    "\n",
    "per_biosample_scoped_mapping_results.to_sql(\n",
    "    table_name, biosample_cnx, if_exists=\"append\", index=False\n",
    ")\n",
    "\n",
    "per_biosample_scoped_mapping_results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
